---
title: "R Notebook"
output: html_notebook
editor_options: 
  chunk_output_type: console
---
##################################################
# AIM 2 Analysis
##################################################
#ref https://www.nicholas-ollberding.com/post/introduction-to-the-statistical-analysis-of-microbiome-data-in-r/

Prepare Environment
# set dir
```{r}
git_base="~/../../Volumes/sevillas2-1/git/gmu/dissertation/"
data_base="~/../../Volumes/data/gmu/"
```

Load Libraries
```{r}
source(paste0(git_base,"sources/qiime2R.R")) #github https://github.com/jbisanz/qiime2R
source(paste0(git_base,"sources/qzq_to_phyloseq.R")) #github https://github.com/jbisanz/qiime2R
source(paste0(git_base,"sources/ggrare.R")) #github library: https://rdrr.io/github/gauravsk/ranacapa/

library(phyloseq)
library(biomformat)
library(yaml)
library(stringr)
library(microbiome)
library(dplyr)
library(Rmisc)
library(ggpubr)
library(ggplot2)
library(dplyr)
library(vegan)
library(agricolae)
library(DESeq2)
library(tidyr)
library(purrr)
library(tibble)
library(ALDEx2)
library(rms)
library(themetagenomics)
library(wordcloud)
library(tm)
```

#Read in config file
```{r}
#system
system_opt = "m"

#read in config
config_file = read_yaml(paste0(git_base,"aim2/analysis/config.yaml"))

#set dirs
git_dir=str_replace(config_file$git_dir[[system_opt]],'git_base/',git_base)
data_dir=str_replace(config_file$data_dir[[system_opt]],'data_base/',data_base)
analysis_dir=str_replace(config_file$analysis_dir[[system_opt]],'data_base/',data_base)
ref_dir=str_replace(config_file$ref_dir[[system_opt]],'data_base/',data_base)
manifest_dir=str_replace(config_file$manifest_dir[[system_opt]],'git_base/',git_base)

#set manifests
manifest_file=paste(manifest_dir, config_file$clean_manifest,sep="")
```

#Create phyloseq object
```{r}
#read metadata
metadata<-read.table(manifest_file,sep='\t', header=T, row.names=1, comment="")

prep_data<-function(){
  #Read OTUS
  otus<-read_qza(paste0("4_tab.qza"))
  
  #Read taxonomy reference file
  taxonomy<-read_qza(paste0("scikit_silva.qza"))
  
  #convert tax to matrix for phyloseq input
  GenData <- function(x){
      tax_list = unlist(strsplit(as.character(x),";"))
      
      for (i in (length(tax_list)+1):17){
        tax_list = append(tax_list,"")
      }
      
      return(tax_list)
  }
  
  taxonomy_list <- as.list(taxonomy$data$Taxon)
  taxonomy_table <- do.call("rbind", lapply(taxonomy_list, GenData))
  taxonomy_table <- taxonomy_table[,c(1:7)]
  colnames(taxonomy_table) <- c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species")
  taxonomy_table <- as.data.frame(taxonomy_table)
  taxonomy_table <-taxonomy_table %>% 
      mutate_all(funs(str_replace(., "D_.__", "")))
  taxonomy_table <-taxonomy_table %>% 
      mutate_all(funs(str_replace(., "]", "")))
  taxonomy_table <-taxonomy_table %>% 
      mutate_all(funs(str_replace(., "-", "_")))
  taxonomy_table <-taxonomy_table %>% 
      mutate_all(funs(str_replace(., " ", "")))
    
  rownames(taxonomy_table)<-taxonomy$data$Feature.ID
  tax_convert<-as.matrix(taxonomy_table)
  
  #create phyloseq object
  phy_obj<-phyloseq(otu_table(otus$data, taxa_are_rows = T), tax_table(tax_convert), sample_data(metadata))
  return(phy_obj)
}
phy_t1 = prep_data()
```

sample counts
```{r}
count_samples<-function(df.in){
  #samples by type
  for (typeid in unique(df.in$SampleType)){
    for (dietid in unique(df.in$Diet)){
      for (timeid in unique(df.in$Timepoint)){
        print(paste0(typeid,",",dietid,",",timeid,":",nrow(subset(df.in,SampleType==typeid & 
                                                              Diet==dietid & 
                                                              Timepoint==timeid))))
      }
    }
  }
}

#total samples sequenced
print(paste0("Before seq: ", nrow(metadata)))
count_samples(metadata)

#### AFTER SEQ
#total samples sequenced
print(paste0("After seq: ",nrow(sample_data(phy_t1))))
count_samples(data.frame(sample_data(phy_t1)))

#determine num of reads by Sample.Desc
plot_read_depth<-function(phy.in){
  sampledf = data.frame()
  
  #for each sample type
  for(typeid in unique(sample_data(phy.in)$SampleType)){
  
    #create list of ids for ext.kit
    sample_list = rownames(subset(sample_data(phy.in),SampleType==typeid))
    
    for (sampleid in sample_list){
      sampledf[nrow(sampledf)+1,"Sample.Type"] = typeid
      sampledf[nrow(sampledf),"Timepoint"] = as.character(sample_data(phy.in)[sampleid,"Timepoint"])
      sampledf[nrow(sampledf),"reads"] = sample_sums(phy.in)[sampleid][[1]]
      sampledf[nrow(sampledf),"sample"] = sampleid
    }
  }
  sampledf$reads = as.numeric(sampledf$reads)
  print(paste0("Mean of all: ",mean(sampledf$reads)))
  sumdf = summarySE(sampledf, measurevar="reads", groupvars=c("Sample.Type"))
  print(sumdf)
  
  #plot
  p1 = ggplot(sumdf, aes(x=Sample.Type, y=reads, fill=Sample.Type)) + 
          geom_bar(position=position_dodge(), stat="identity",
                 colour="black", # Use black outlines,
                 size=.3) +      # Thinner lines
        geom_errorbar(aes(ymin=reads-se, ymax=reads+se),
                      size=.3,    # Thinner lines
                      width=.2,
                      position=position_dodge(.9)) +
        xlab("Sample Type") +
        ylab("Average Read Length") +
        scale_fill_hue(name="Sample.Type") + # Legend label, use darker colors
        ggtitle(paste0("By Sample Type")) +
        theme_bw()
  print(p1)
    
  sumdf = summarySE(sampledf, measurevar="reads", groupvars=c("Timepoint"))
  print(sumdf)
  
  #plot
  p2 = ggplot(sumdf, aes(x=Timepoint, y=reads, fill=Timepoint)) + 
          geom_bar(position=position_dodge(), stat="identity",
                 colour="black", # Use black outlines,
                 size=.3) +      # Thinner lines
        geom_errorbar(aes(ymin=reads-se, ymax=reads+se),
                      size=.3,    # Thinner lines
                      width=.2,
                      position=position_dodge(.9)) +
        xlab("Timepoint") +
        ylab("") +
        scale_fill_hue(name="Timepoint") + # Legend label, use darker colors
        ggtitle(paste0("By Timepoint")) +
        theme_bw()
  print(p2)
    
  file_name = paste0(analysis_dir,"readlength_time_type.png")
  pf = ggarrange(p1, p2,
            labels = c("A", "B"),
            ncol = 2, nrow = 1)
  
  pf = annotate_figure(pf, top = text_grob("Average Number of Reads", 
                face = "bold", size = 14))
  print(pf)
  ggsave(file_name,pf)
}
plot_read_depth(phy_t1)
summarize_phyloseq(phy_t1)

```

#prevelance values
```{r}
# Compute prevalence of each feature, store as data.frame
prevdf = apply(X = otu_table(phy_t1),
               MARGIN = ifelse(taxa_are_rows(phy_t1), yes = 1, no = 2),
               FUN = function(x){sum(x > 0)})
# Add taxonomy and total read counts to this data.frame
prevdf = data.frame(Prevalence = prevdf,
                    TotalAbundance = taxa_sums(phy_t1),
                    tax_table(phy_t1))

plyr::ddply(prevdf, "Kingdom", function(df1){cbind(mean(df1$Prevalence),sum(df1$Prevalence))})
plyr::ddply(prevdf, "Phylum", function(df1){cbind(mean(df1$Prevalence),sum(df1$Prevalence))})

```

#abundances
```{r}
#Figure 3A https://pubs.acs.org/doi/pdf/10.1021/acs.est.8b02301
top_abun_plots<-function(phy.in,type.in){
  #melt data to relative abundance
  phy_tmp <- phy.in %>%
      tax_glom(taxrank = "Phylum") %>%                     # agglomerate at tax level
      transform_sample_counts(function(x) {x/sum(x)} ) %>% # Transform to rel. abundance
      psmelt() 
  phy_tmp = phy_tmp[order(phy_tmp$Abundance),]
  
  #plot
  phylum_colors <- c(
    "#CBD588", "#5F7FC7", "orange","#DA5724", "#508578", "#CD9BCD",
     "#AD6F3B", "#673770","#D14285", "#652926", "#C84248", 
    "#8569D5", "#5E738F","#D1A33D", "#8A7C64", "#599861",  "#CBD588", "#5F7FC7", "orange","#DA5724", "#508578", "#CD9BCD",
     "#AD6F3B", "#673770","#D14285", "#652926", "#C84248", 
    "#8569D5", "#5E738F","#D1A33D", "#8A7C64", "#599861"
  )
  
  addSmallLegend <- function(myPlot, pointSize = 1, textSize = 6, spaceLegend = 0.1) {
      myPlot +
          guides(shape = guide_legend(override.aes = list(size = pointSize)),
                 color = guide_legend(override.aes = list(size = pointSize))) +
          theme(legend.title = element_text(size = textSize), 
                legend.text  = element_text(size = textSize),
                legend.key.size = unit(spaceLegend, "lines"))
  }
  
  #print bar plot of relative abundance
  p = ggplot(phy_tmp, aes(x = Sample, y = Abundance, fill = Phylum)) + 
      theme(legend.position = "bottom", legend.key.width = unit(0.2,"cm"),
            axis.text.x = element_text(angle = 90, hjust = 1), 
            legend.text = element_text(size = 8))  +
            geom_bar(stat = "identity") +
            xlab(type.in) +
            ylab("Relative Abundance (%)") +
            ggtitle("Relative Abundance of Phylum: By Diet") +
            labs(fill = "Phylum")
  p1 = addSmallLegend(p)
  print(p1)
  
  file_name = paste0("relabun_",type.in)
  ggsave(paste0(analysis_dir,file_name,".jpg"),plot=p1)
  
  
  #select top OTUS
  TopNOTUs <- names(sort(taxa_sums(phy.in), TRUE)[1:50])
  #phy_pr <- prune_taxa(TopNOTUs,phy.in)

  #melt data to phylum
  phy_tmp <- phy_t1 %>%
      tax_glom(taxrank = "Genus") %>% # agglomerate at tax level
      psmelt() 
  phy_tmp = phy_tmp[with(phy_tmp,order(-Abundance)),]
  phy_tmp = subset(phy_tmp,OTU %in% TopNOTUs)
  
  #print plot of top abundance OTU
  p = ggplot(phy_tmp, aes(x = Timepoint, y = Abundance, color=Diet)) + 
      theme(legend.position = "bottom", legend.key.width = unit(0.2,"cm"),
            axis.text.x = element_text(angle = 90, hjust = 1), 
            legend.text = element_text(size = 8))  +
            geom_boxplot() +
            xlab("Timepoint") +
            ylab("Abundnance") +
            ggtitle("Abundance of Top 50 OTU's\nPhylum: By Diet, Timepoint") +
            labs(fill = "Timpepoint")+
    facet_wrap(~Phylum, scales = "free_y")
  p1 = addSmallLegend(p)
  print(p1)
  
  file_name = paste0("abun_",type.in)
  ggsave(paste0(analysis_dir,file_name,".jpg"),plot=p1)
  
}
top_abun_plots(merge_samples(phy_t1,"Diet"),"diet")

sig_test_abun<-function(phy.in,tax.in){

  #melt OTUS  
  phy_tmp <- phy.in %>%
      tax_glom(taxrank = tax.in) %>%                     # agglomerate at tax level
      psmelt() 
  phy_tmp = phy_tmp[order(phy_tmp$Abundance),]

  #Subset groups
  prepare_df<-function(type.in){
    df_otu <- subset(phy_tmp, Diet == type.in)

    #merge to phylum
    df_otu = subset(aggregate(Abundance~get(tax.in)+Sample, df_otu, sum),Abundance>0)
    colnames(df_otu)[1]=tax.in
    
    #format df
    df.out=data.frame()
    for(rowid in rownames(df_otu)){
      df.out[df_otu[rowid,"Sample"],df_otu[rowid,tax.in]]=df_otu[rowid,"Abundance"]
    }
    
    df.out[is.na(df.out)]=0
    return(df.out)
  }
  
  control_otu = prepare_df("Rice Bran")
  cf_otu = prepare_df("Apple Slices")
 
  #if there are different cols, add them to df
  if(length(setdiff(colnames(control_otu),colnames(cf_otu)))>0){
    cf_otu[,setdiff(colnames(control_otu),colnames(cf_otu))]=0
  }
  if(length(setdiff(colnames(cf_otu),colnames(control_otu)))>0){
    control_otu[,setdiff(colnames(cf_otu),colnames(control_otu))]=0
  }
  
  #HMP test
  group_data <- list(control_otu, cf_otu)
  (xdc <- HMP::Xdc.sevsample(group_data))           
}
sig_test_abun(phy_t1,"Phylum")
sig_test_abun(phy_t1,"Genus")

create_heatmap_diet<-function(phy.in,tax.in){
  #melt OTUS  
  phy_tmp <- phy.in %>%
      tax_glom(taxrank = tax.in) # agglomerate at tax 
  
  tmp_df = data.frame(sample_data(subset_samples(phy_tmp, Diet=="Rice Bran")))
  tmp_df = tmp_df[order(tmp_df$Timepoint),]
  
  #plot heatmap
  tmp_df = data.frame(sample_data(subset_samples(phy_tmp, Diet=="Rice Bran")))
  sample_list = rownames(tmp_df[order(tmp_df$Timepoint),])
  
  tmp_df = data.frame(tax_table(subset_samples(phy_t1, Diet=="Rice Bran")))
  taxa_list = rownames(tmp[order(tmp$Phylum,decreasing=TRUE),])
  p1= plot_heatmap(subset_samples(phy_tmp, Diet=="Rice Bran"),
                   "NMDS", "bray", "Timepoint", tax.in, low="#66CCFF", high="#000033",
                   na.value="white", sample.order=sample_list, taxa.order = taxa_list) + ggtitle("Rice Bran")
  
  #plot second heatmap
  tmp_df = data.frame(sample_data(subset_samples(phy_tmp, Diet=="Apple Slices")))
  sample_list = rownames(tmp_df[order(tmp_df$Timepoint),])
  
  tmp_df = data.frame(tax_table(subset_samples(phy_t1, Diet=="Apple Slices")))
  taxa_list = rownames(tmp[order(tmp$Phylum,decreasing=TRUE),])
  p2= plot_heatmap(subset_samples(phy_tmp, Diet=="Apple Slices"),
                   "NMDS", "bray", "Timepoint", tax.in, low="#66CCFF", high="#000033", 
                   na.value="white", ample.order=sample_list, taxa.order = taxa_list) + ggtitle("Apple Slices")

  #merge heatmaps
  file_name = paste0(analysis_dir,"heatplot_diet_",tax.in,".png")
  p = ggarrange(p1,p2,
            labels = c("A", "C"),
            ncol = 2, nrow = 1)
  p = annotate_figure(p, top = text_grob(paste0("Absolute Abundance, ",tax.in," by Diet"), 
                face = "bold", size = 14))
  print(p)
  ggsave(file_name,p)
}
create_heatmap_diet(phy_t1,"Phylum")
create_heatmap_diet(phy_t1,"Genus")
```

alpha table
```{r}
##################################################
# richness summary
##################################################
richness_variables<-function(phy.in){
  #calculate alpha at binf level by sample type
  compare_sampletype = data.frame()
  sampledf = data.frame()
    
  # Calculate richness
  df_ob = phyloseq::estimate_richness(phy.in, measures = "Observed")
  df_ch = phyloseq::estimate_richness(phy.in, measures = "Chao1")
  df_inv = phyloseq::estimate_richness(phy.in, measures = "InvSimpson")
  
  # Store inv distances within type
  for (rowid in rownames(df_ob)){
    lookupval = gsub("[.]","-",rowid)
    compare_sampletype[nrow(compare_sampletype)+1,"type"] = as.character(sample_data(phy.in)[lookupval,"SampleType"])
    compare_sampletype[nrow(compare_sampletype),"timepoint"] = as.character(sample_data(phy.in)[lookupval,"Timepoint"])
    compare_sampletype[nrow(compare_sampletype),"diet"] = as.character(sample_data(phy.in)[lookupval,"Diet"])
    compare_sampletype[nrow(compare_sampletype),"INV"] = df_inv[lookupval,"InvSimpson"]  
    compare_sampletype[nrow(compare_sampletype),"CH"] = df_ch[rowid,"Chao1"]  
    compare_sampletype[nrow(compare_sampletype),"OBS"] = df_ob[rowid,"Observed"]  
  }
  
  #merge at group level
  for(typeid in sort(unique(compare_sampletype$type))){
    for (timeid in sort(unique(compare_sampletype$timepoint))){
      sub_df = subset(compare_sampletype, type==typeid & timepoint==timeid)
      sampledf[nrow(sampledf)+1,"Sample Type"] = typeid
      sampledf[nrow(sampledf),"Timepoint"] = timeid
      sampledf[nrow(sampledf),"Inverse Simpson"] = paste0(round(mean(sub_df$INV),2)," (", round(sd(sub_df$INV),2),")")
      sampledf[nrow(sampledf),"Chao1"] = paste0(round(mean(sub_df$CH),2)," (", round(sd(sub_df$CH),2),")")
      sampledf[nrow(sampledf),"Observed Species"] = paste0(round(mean(sub_df$OBS),2)," (", round(sd(sub_df$OBS),2),")")
    }
  }
  
  print(sampledf)
  
  file_save = paste0(analysis_dir,'richness.csv')
  write.csv(sampledf,file_save)
}
richness_variables(phy_t1)

richness_plots <- function(phy.in){
  #define alpha
  sample_df = data.frame(sample_data(phy.in))
  sample_df$alpha = diversity(otu_table(phy.in),
                              MARGIN = 2, #samples are in cols, not rows
                              index = "invsimpson")
  #check distribution
  print(shapiro.test(sample_df$alpha))
  
  #run anova, tukey HSD, plot
  ## diet
  anova_result <- aov(alpha ~ Diet, sample_df)
  print(summary(anova_result))
  tukey_result <- HSD.test(anova_result, "Diet", group = TRUE)
  print(tukey_result)
  group_data <- tukey_result$groups[order(rownames(tukey_result$groups)),]

  p1 = ggplot(sample_df, aes(x = Diet, y = alpha)) +
    geom_text(data = data.frame(),
              aes(x = rownames(group_data), y = max(sample_df$alpha) + 1, label = group_data$groups),
              col = 'black',
              size = 10) +
    geom_boxplot() +
    ggtitle("Diet") +
    xlab("") +
    ylab("Alpha diversity index")
  ggsave(paste0(paste0(analysis_dir,"alpha_diversity_diet.png")),p1)
  
  ## timepoint
  anova_result <- aov(alpha ~ Timepoint, sample_df)
  print(summary(anova_result))
  tukey_result <- HSD.test(anova_result, "Timepoint", group = TRUE)
  print(tukey_result)
  group_data <- tukey_result$groups[order(rownames(tukey_result$groups)),]
  p2 = ggplot(sample_df, aes(x = Timepoint, y = alpha)) +
    geom_text(data = data.frame(),
              aes(x = rownames(group_data), y = max(sample_df$alpha) + 1, label = group_data$groups),
              col = 'black',
              size = 10) +
    geom_boxplot() +
    ggtitle("Timepoint") +
    xlab("") +
    ylab("")
    ggsave(paste0(paste0(analysis_dir,"alpha_diversity_time.png")),p2)
  
  ##sampletype
  anova_result <- aov(alpha ~ SampleType, sample_df)
  print(summary(anova_result))
  tukey_result <- HSD.test(anova_result, "SampleType", group = TRUE)
  print(tukey_result)
  group_data <- tukey_result$groups[order(rownames(tukey_result$groups)),]
  p3 = ggplot(sample_df, aes(x = SampleType, y = alpha)) +
    geom_text(data = data.frame(),
              aes(x = rownames(group_data), y = max(sample_df$alpha) + 1, label = group_data$groups),
              col = 'black',
              size = 10) +
    geom_boxplot() +
    ggtitle("SampleType") +
    xlab("") +
    ylab("Alpha diversity index")
    ggsave(paste0(paste0(analysis_dir,"alpha_diversity_type.png")),p3)

  
  file_name = paste0(analysis_dir,"alpha_diversity.png")
  p = ggarrange(p1,p2,p3,
            labels = c("A", "B","C"),
            ncol = 2, nrow = 2)
  p = annotate_figure(p, top = text_grob("Alpha Diversity, Inverse Simpson Diversity", 
                face = "bold", size = 14))
  print(p)
  ggsave(file_name,p)
  
  p1 = plot_richness(phy_t1,measures = c("Observed","Simpson","Chao1"),color="SampleType",x="Diet",shape="Timepoint")
  file_name = paste0(analysis_dir,"alpha_diversity_merged.png")
  p = ggarrange(p1,
            ncol = 1, nrow = 1)
  p = annotate_figure(p, top = text_grob("Alpha Diversity", 
                face = "bold", size = 14))
  print(p)
  ggsave(file_name,p)
}
richness_plots(phy_t1)
#samples are not normally distributed Shapiro-Wilk
```

Bray curtis
```{r}
create_ord_plots <-function(){
  #https://microbiome.github.io/tutorials/Ordination.html
  ordination <- ordinate(phy_t1,"NMDS", "bray")
  p1 = plot_ordination(phy_t1, ordination,  
                       color="Diet", title="Diet") + theme_bw() + 
    stat_ellipse(geom = "polygon", type="norm", alpha=0.4)
  
  p2 = plot_ordination(phy_t1, ordination,  
                       color="Timepoint", title="Timepoint") + theme_bw() + 
    stat_ellipse(geom = "polygon", type="norm", alpha=0.4, aes(fill=Diet))
  
  p3 = plot_ordination(phy_t1, ordination,  
                       color="SampleType", title="Sample Type") + theme_bw() + 
    stat_ellipse(geom = "polygon", type="norm", alpha=0.4, aes(fill=Diet))
  
  file_name = paste0(analysis_dir,"NMDS_bray.png")
  p = ggarrange(p1,p2,p3,
                labels = c("A", "B", "C"),
                ncol = 2, nrow = 2)
  p = annotate_figure(p, top = text_grob("NMDS-Bray plots", 
                  face = "bold", size = 14))
  print(p)
  ggsave(file_name,p)
}
create_ord_plots()
```

permanova analysis
```{r}
#https://www.nicholas-ollberding.com/post/introduction-to-the-statistical-analysis-of-microbiome-data-in-r/
create_atich_plots<-function(phy.in,compare.in){
  #Generate distance matrix
  clr_dist_matrix <- phyloseq::distance(phy.in, method = "euclidean") 
  
  #ADONIS test
  print(vegan::adonis(clr_dist_matrix ~ phyloseq::sample_data(phy.in)$Diet))
  
  #Dispersion test and plot
  dispr <- vegan::betadisper(clr_dist_matrix, phyloseq::sample_data(phy.in)$Diet)
  print(permutest(dispr))

  file_name = paste0(analysis_dir,"ordination_",compare.in,".png")
  jpeg(file_name)
  par(mfrow=c(1,2),oma=c(1.5,2,3,1))    # set the plotting area into a 1*2 array
  plot(dispr, main="",sub="")
  boxplot(dispr, main = "", xlab = "",xaxt="n")
    axis(1, labels = FALSE)
    text(x = 1:length(sort(unique(dispr$group))),
     ## Move labels to just below bottom of chart.
     y = par("usr")[3] - 5,
     ## Use names from the data list.
     labels = paste0(sort(unique(dispr$group)),"   "),
     ## Change the clipping region.
     xpd = NA,
     ## Rotate the labels by 35 degrees.
     srt = 90,
     ## Adjust the labels to almost 100% right-justified.
     adj = 0.965,
     ## Increase label size.
     cex = 1)
  mtext("Ordination Centroids and Dispersion Labeled:\nAitchison Distance", line=0, side=3, outer=TRUE, cex=1.5)
  dev.off()
}
create_atich_plots(subset_samples(phy_t1,Timepoint=="Baseline" | Timepoint=="Day12" & Diet=="Apple Slices"),"bto12")

#https://microbiome.github.io/tutorials/PERMANOVA.html
#For Diet
adonis2(t(otu_table(phy_t1)) ~ Diet, 
        data=data.frame(sample_data(phy_t1)), permutations=99,
        method = "bray")
anova(betadisper(vegdist(t(otu_table(phy_t1))), data.frame(sample_data(phy_t1))$Diet))

#For Timepoints
adonis2(t(otu_table(phy_t1)) ~ Timepoint, 
        data=data.frame(sample_data(phy_t1)), permutations=99,
        method = "bray")
anova(betadisper(vegdist(t(otu_table(phy_t1))), data.frame(sample_data(phy_t1))$Timepoint))
```

Diff analysis
```{r}
#https://informatics.fas.harvard.edu/differential-expression-with-deseq2.html
#https://joey711.github.io/phyloseq-extensions/DESeq2.html
#https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1003531 - use un-rare data
#https://bioconductor.org/packages/release/bioc/vignettes/ALDEx2/inst/doc/ALDEx2_vignette.html

cal_wilcox<-function(phy.in,tax.in){
  #merge to family
  phy_tmp <- phy.in %>%
      tax_glom(taxrank = tax.in) # agglomerate at tax 
  phyloseq::taxa_names(phy_tmp) <- phyloseq::tax_table(phy_tmp)[, tax.in]

  #Generate data.frame with OTUs and metadata
  ps_wilcox <- data.frame(t(data.frame(phyloseq::otu_table(phy_tmp))))
  ps_wilcox$Diet <- phyloseq::sample_data(phy_tmp)$Diet
  
  #Define functions to pass to map
  wilcox_model <- function(df){
    wilcox.test(abund ~ Diet, data = df)
  }
  wilcox_pval <- function(df){
    wilcox.test(abund ~ Diet, data = df)$p.value
  }

  #Create nested data frames by OTU and loop over each using map 
  wilcox_results <- ps_wilcox %>%
    gather(key = OTU, value = abund, -Diet) %>%
    group_by(OTU) %>%
    nest() %>%
    mutate(wilcox_test = map(data, wilcox_model),
           p_value = map(data, wilcox_pval))                       
  #Show results
  print(head(as.data.frame(wilcox_results)))
  print(wilcox_results$wilcox_test[[1]])
  
  #unnest
  wilcox_results <- wilcox_results %>%
    dplyr::select(OTU, p_value) %>%
    unnest()
  
  #Computing FDR corrected p-values
  wilcox_results <- wilcox_results %>%
    arrange(p_value) %>%
    mutate(BH_FDR = p.adjust(p_value, "BH")) %>%
    filter(BH_FDR < 0.05) %>%
    dplyr::select(OTU, p_value, BH_FDR, everything())
  
  print(data.frame(wilcox_results))
}
cal_wilcox(phy_t1,"Phylum")
cal_wilcox(subset_samples(phy_t1,Timepoint=="Baseline" | Timepoint=="Day12" & Diet=="Apple Slices"),"Phylum")

#aldex2
cal_alex<-function(phy.in,tax.in){
  #Run ALDEx2
  aldex2_da <- ALDEx2::aldex(data.frame(phyloseq::otu_table(phy.in)), 
                             phyloseq::sample_data(phy.in)$Diet, test="t", effect = TRUE, denom="all")
  
  file_name = paste0(analysis_dir,"aldex_diet.png")
  jpeg(file_name)
  par(mfrow=c(1,2),oma=c(1.5,2,3,1))    # set the plotting area into a 1*2 array
  aldex.plot(aldex2_da, type="MA", test="welch", xlab="Log-ratio abundance", ylab="Difference")
  aldex.plot(aldex2_da, type="MW", test="welch", xlab="Dispersion", ylab="Difference")
  mtext("ALDEx2 stratified by diet", line=0, side=3, outer=TRUE, cex=1.5)
  dev.off()
  
  #Adding taxonomic labels
  taxa_info <- data.frame(tax_table(phy.in))
  taxa_info <- taxa_info %>% rownames_to_column(var = "OTU")
  
  sig_aldex2 <- aldex2_da %>%
    rownames_to_column(var = "OTU") %>%
    filter(wi.eBH < 0.05) %>%
    arrange(effect, wi.eBH) %>%
    dplyr::select(OTU, diff.btw, diff.win, effect, wi.ep, wi.eBH)
  sig_aldex2 <- left_join(sig_aldex2, taxa_info)
  
  print(sig_aldex2)
}
cal_alex(phy_t1,"Phylum")

#deseq2 
DiffAnalysis<-function(phy_in,de_in,subtitle,file_name){
  #conver to deseq
  phy_deseq = de_in

  #run deseq
  diagdds = DESeq(phy_deseq, test="Wald", sfType="poscounts")

  #run stats
  res = results(diagdds, cooksCutoff = FALSE)
  alpha = 0.05
  sigtab = res[which(res$padj < alpha), ]
  result_df = data.frame(res)

  #run correction
  result_df$BH = p.adjust(result_df$pvalue, method = "BH", n = length(result_df$pvalue))
  sig_df = subset(result_df,BH<0.05)

  #add tax info
  sigtab = cbind(as(sig_df, "data.frame"), as(tax_table(phy_in)[rownames(sig_df), ], "matrix"))
  sigtab = sigtab[with(sigtab, order(log2FoldChange)),]
  
  #remove D_6__, add "higher level"
  sigtab$Genus = gsub("D_5__", "", sigtab[,"Genus"])
  sigtab$Genus[sigtab$Genus==""] = "HigherTaxonomicLevel"

  #rownames
  sigtab$OTU = rownames(sigtab)
  rownames(sigtab)=make.names(sigtab$Genus,unique=TRUE)

  #plot
    if(nrow(sigtab)>0){
      p = ggplot(data = sigtab,
           aes(x = reorder(rownames(sigtab),log2FoldChange), y = log2FoldChange,
               fill = rownames(sigtab))) +
          geom_bar(stat = "identity")+
          coord_flip()+
          labs(x = "Genus", y = "log2 Fold Change",
               title = paste("Differential Abundance Analysis: \n",subtitle))+
          theme_minimal()+
          guides(fill = FALSE)
      
      print(p)
      ggsave(paste(analysis_dir,file_name,".jpg",sep=""),plot=p)
    } else{
      print(paste("There are no sig genes for",subtitle))
    }
  return(sigtab)
}
SigTables<-function(sig_in,val_in){
  sig_out = sig_in
  sig_out$Genus = rownames(sig_out)
  rownames(sig_out) = NULL
  sig_out$val = val_in
  return(sig_out[,c("Genus","val","BH")])
}
run_analysis<-function(phy.in){
  sig_complete = data.frame()
  #Diet
  sigtab = DiffAnalysis(phy.in,phyloseq_to_deseq2(phy.in, ~ Diet), "Apple Slices to Rice Bran","diffanal_diet")
  sig_complete = rbind(sig_complete,SigTables(sigtab,"DIET"))
  
  #Timepoints
  phy_tmp = subset_samples(phy.in,Timepoint %in% c("Baseline","Day05"))
  sigtab = DiffAnalysis(phy_tmp, phyloseq_to_deseq2(phy_tmp, ~ Timepoint), "Timepoint Baseline to Day 5","diffanal_tp_b5")
  sig_complete = rbind(sig_complete,SigTables(sigtab,"Bto5"))
  
  phy_tmp = subset_samples(phy.in,Timepoint %in% c("Baseline","Day08"))
  sigtab = DiffAnalysis(phy_tmp, phyloseq_to_deseq2(phy_tmp, ~ Timepoint), "Timepoint Baseline to Day 8","diffanal_tp_b8")
  sig_complete = rbind(sig_complete,SigTables(sigtab,"Bto8"))
  
  phy_tmp = subset_samples(phy.in,Timepoint %in% c("Baseline","Day12"))
  sigtab = DiffAnalysis(phy_tmp, phyloseq_to_deseq2(phy_tmp, ~ Timepoint), "Timepoint Baseline to Day 12","diffanal_tp_b12")
  sig_complete = rbind(sig_complete,SigTables(sigtab,"Bto12"))
  
  phy_tmp = subset_samples(phy.in,Timepoint %in% c("Day05","Day08"))
  sigtab = DiffAnalysis(phy_tmp, phyloseq_to_deseq2(phy_tmp, ~ Timepoint), "Timepoint Day 5 to Day 8","diffanal_tp_58")
  sig_complete = rbind(sig_complete,SigTables(sigtab,"5to8"))
  
  phy_tmp = subset_samples(phy.in,Timepoint %in% c("Day08","Day12"))
  sigtab = DiffAnalysis(phy_tmp, phyloseq_to_deseq2(phy_tmp, ~ Timepoint), "Timepoint Day 8 to Day 12","diffanal_tp_812")
  sig_complete = rbind(sig_complete,SigTables(sigtab,"8to12"))
  
  sig_final = data.frame()
  for (rowid in rownames(sig_complete)){
    sig_final[sig_complete[rowid,"Genus"],sig_complete[rowid,"val"]]=sig_complete[rowid,"BH"]
  }

  return(sig_final)
}

sig_table = run_analysis(phy_t1)
for (colid in colnames(sig_table)){print(colid); print(nrow(subset(sig_table,sig_table[,colid]>0)))}
write.csv(sig_table,paste0(analysis_dir,"diff_sig.csv"))
```

#prediction analysis
```{r}
#rms
#https://www.nicholas-ollberding.com/post/introduction-to-the-statistical-analysis-of-microbiome-data-in-r/
cal_rms<-function(phy.in,sub.id){
  ord_clr <- phyloseq::ordinate(phy.in, "RDA")
  
  #print variance explained
  print(sapply(ord_clr$CA$eig[1:5], function(x) x / sum(ord_clr$CA$eig)))
  
  #Generate data.frame
  clr_pcs <- data.frame(
    "pc1" = ord_clr$CA$u[,1],
    "pc2" = ord_clr$CA$u[,2],
    "pc3" = ord_clr$CA$u[,3],
    "Status" = phyloseq::sample_data(phy.in)$Diet
  )
  clr_pcs$Status_num <- ifelse(clr_pcs$Status == "Rice Bran", 0, 1)

  #Specify a datadist object (for rms)
  options(datadist = rms::datadist(clr_pcs))
  
  #Plot the unconditional associations
  a <- ggplot(clr_pcs, aes(x = pc1, y = Status_num)) +
    Hmisc::histSpikeg(Status_num ~ pc1, lowess = TRUE, data = clr_pcs) +
    labs(x = "\nPC1", y = "Pr(Apple Slices)\n")
  b <- ggplot(clr_pcs, aes(x = pc2, y = Status_num)) +
    Hmisc::histSpikeg(Status_num ~ pc2, lowess = TRUE, data = clr_pcs) +
    labs(x = "\nPC2", y = "Pr(Apple Slices)\n")
  c <- ggplot(clr_pcs, aes(x = pc3, y = Status_num)) +
    Hmisc::histSpikeg(Status_num ~ pc3, lowess = TRUE, data = clr_pcs) +
    labs(x = "\nPC3", y = "Pr(Apple Slices)\n")
  p = cowplot::plot_grid(a, b, c, nrow = 2, ncol = 2, scale = .9, labels = "AUTO")

  file_name = paste0(analysis_dir,"rms_apple",sub.id,".png")
  pf = ggarrange(p,
            ncol = 1, nrow = 1)
  
  pf = annotate_figure(pf, 
                       top = text_grob(paste0("RMS Prediction of Apple Slices\n",sub.id), face = "bold", size = 14))
  print(pf)
  ggsave(file_name,pf)
  
  #Fit full model with splines (3 knots each)
  m1 <- rms::lrm(Status_num ~ rcs(pc1, 3) + rcs(pc2, 3) + rcs(pc3, 3), data = clr_pcs, x = TRUE, y = TRUE)
  
  #Grid search for penalties
  pentrace(m1, list(simple = c(0, 1, 2), nonlinear = c(0, 100, 200)))
  pen_m1 <- update(m1, penalty = list(simple = 1, nonlinear = 200))
  print(pen_m1)
  ggplot(Predict(pen_m1))

  #Obtain optimism corrected estimates
  (val <- rms::validate(pen_m1))

  #Compute corrected c-statistic
  (c_opt_corr <- 0.5 * (val[1, 5] + 1))
  cal <- rms::calibrate(pen_m1, B = 200)
  
  file_name = paste0(analysis_dir,"rms_diet",sub.id,".png")
  jpeg(file_name)
  par(mfrow=c(1,1),oma=c(1.5,2,3,1))    # set the plotting area into a 1*2 array
  plot(cal)
  mtext(paste0("Actual Probability compared to Predicted Probability for Apple Slices\n",sub.id))
  dev.off()
  
}
cal_rms(phy_t1,"allsamples")
cal_rms(subset_samples(phy_t1,Timepoint=="Baseline" | Timepoint=="Day12" & Diet=="Apple Slices"),"subsetsamples")

#sebal
#https://www.nicholas-ollberding.com/post/introduction-to-the-statistical-analysis-of-microbiome-data-in-r/
#https://htmlpreview.github.io/?https://github.com/malucalle/selbal/blob/master/vignettes/vignette.html
cal_sebal<-function(phy.in,compare.in,tax.in){
  #merge to family
  phy_tmp <- phy.in %>%
      tax_glom(taxrank = tax.in) # agglomerate at tax 
  phyloseq::taxa_names(phy_tmp) <- phyloseq::tax_table(phy_tmp)[, tax.in]
  

  #create and filter df
  phy_df = data.frame(t(data.frame(phyloseq::otu_table(phy_tmp))))
  phy_df = phy_df[,colSums(phy_df[,1:(ncol(phy_df)-1)]) > 200]

  #Run selbal
  cv_sebal <- selbal::selbal.cv(x = phy_df[1:(ncol(phy_df)-1)], 
                                y = as.factor(phyloseq::sample_data(phy_tmp)$Diet), 
                             n.fold = 5, n.iter = 1) 
  
  #plot/print results
  p1 = cv_sebal$accuracy.nvar
  p2 = cv_sebal$var.barplot + ggtitle("")
  file_name = paste0(analysis_dir,"selbal_AUC_bar.png")
  p = ggarrange(p1,p2,
                labels = c("A", "B"),
                ncol = 1, nrow = 2)
  p = annotate_figure(p, top = text_grob("SELBAL Accuracy and Cross Validation", 
                  face = "bold", size = 14))
  print(p)
  ggsave(file_name,p)
  
  #plot/print results
  file_name = paste0(analysis_dir,"selbal_prediction.png")
  jpeg(file_name)
  plot.new()
  plot(cv_sebal$global.plot)
  mtext("Prediction Results using Sebal\n All samples for Diet", line=0, side=3, outer=TRUE, cex=1.5)
  dev.off()
  
  file_name = paste0(analysis_dir,"selbal_table.png")
  jpeg(file_name)
  plot.new()
  plot.tab(cv_sebal$cv.tab)
  mtext("Prediction Results using Sebal\n All samples for Diet", line=0, side=3, outer=TRUE, cex=1.5)
  dev.off()
}
cal_sebal(phy_t1,"diet","Phylum")
```

#Tax4Fun installation
```{r}
#http://tax4fun.gobics.de/RPackage/Readme_Tax4Fun.pdf

#download and install
#https://cran.r-project.org/src/contrib/Archive/qiimer/
#unzip
#mv to /Library/Frameworks/R.framework/Versions/4.0/Resources/library
#on terminal: R CMD INSTALL /Library/Frameworks/R.framework/Versions/4.0/Resources/library/qiimer
#in R: library(qiimer)

#RJSONIO from biocmanager

#https://cran.r-project.org/src/contrib/Archive/biom/
#repeat steps above

#http://tax4fun.gobics.de/
#repeat steps above
```

#Tax4Fun Analysis
```{r}
#https://rdrr.io/cran/themetagenomics/f/vignettes/functional_prediction.Rmd
#https://cran.r-project.org/web/packages/themetagenomics/vignettes/thematic_structure.html

#download ref
tmp <- tempdir()
download_ref(tmp,reference='silva_ko',overwrite=FALSE)

create_wordcloud<-function(df.in,file.save){
  str.list = df.in
  
  remove.list = c("SUBCELLULAR LOCATION","[[:punct:]]",
                                      "ECO[0-9]*[|][A-Za-z]*[0-9]",
                                      "Note=","ECO[0-9]*","[0-9]*",
                                      " and "," [tT]he "," in ","PubMed", " with ")
  for (items in remove.list){
    str.list=str_remove_all(str.list,items)
  }
  list.final = stripWhitespace(str.list)
  dtm <- TermDocumentMatrix(Corpus(VectorSource(list.final)))
  v <- sort(rowSums(as.matrix(dtm)),decreasing=TRUE)
  d <- data.frame(word = names(v),freq=v)

  #save wordcloud and histogram counts
  #png(file=paste(analysis_dir,"word_",file.save,".jpg",sep=""))
  
  file_name = paste0(analysis_dir,"word_cloud_bray.png")
  jpeg(file_name)
  par(mfrow=c(1,2),oma=c(0,0,4,0))
  par(cex=.7)
  wordcloud(list.final,colors=brewer.pal(8,"Dark2"),max.words = 25)
  par(fig=c(.5,1,.2,.7), new=TRUE)
  barplot(d[1:10,]$freq, las = 2, names.arg = d[1:10,]$word,
          col = "light blue", ylab = "Word frequencies")
  mtext("Sub-Cellular Distribution of \nTop Pathway Descriptions",outer=TRUE, cex=2)
  dev.off()
}
run_tax4fun<-function(phy.in,tax.in){
  #merge
  phy_tmp <- phy.in %>%
      tax_glom(taxrank = tax.in) # agglomerate at tax 
  phyloseq::taxa_names(phy_tmp) <- phyloseq::tax_table(phy_tmp)[, tax.in]
  
  #create otu df
  phy_df = data.frame(t(data.frame(phyloseq::otu_table(phy_tmp))))
  
  #run analysis
  t4f_output <- t4f(phy_df,rows_are_taxa=FALSE,
                   tax_table=data.frame(tax_table(phy_tmp)),
                   reference_path=tmp,type='uproc',short=TRUE,
                   cn_normalize=TRUE,sample_normalize=TRUE,drop=TRUE)
  #top ko
  top_ko = names(sort(colMeans(t4f_output$fxn_table),decreasing = TRUE)[1:10])

  #create list of top ko's
  list_ko_des  = vector()
  list_ko_path = vector()
  i = 1
  for (ko in top_ko){
    list_ko_des[i] = (t4f_output$fxn_meta$KEGG_Description[ko])[[1]]
    list_ko_path[i] =  (t4f_output$fxn_meta$KEGG_Pathways[ko])[[1]][[1]][3]
    i = i + 1
  }
  #output
  print("Top 10 KO Descriptions")
  print(list_ko_des)
  print("Top 10 KO Pathways")
  print(list_ko_path)
  
  #generate wordcloud
  create_wordcloud(list_ko_path,file.save)
  
  #repeat analysis
  CLEAN <- prepare_data(otu_table=phy_df,
                      rows_are_taxa=FALSE,
                      tax_table=data.frame(tax_table(phy_tmp)),
                      metadata=data.frame(sample_data(phy_tmp)),
                      formula= ~Diet,
                      refs="Rice Bran",
                      cn_normalize=FALSE,
                      drop=TRUE)
  TOPICS <- find_topics(CLEAN,K=15)
  TOPIC_EFFECTS<- est(TOPICS)
  vis(TOPIC_EFFECTS,type="continuous")
  FUNCTIONS <- predict(TOPICS,reference='silva_ko',reference_path=tmp,cn_normalize=TRUE,sample_normalize=FALSE,scalar=25)
  FUNCTION_EFFECTS <- est(FUNCTIONS,level=3,iters=300,
                          chains=2,cores=2,
                          return_summary=TRUE,
                          prior=c('normal','normal','normal'))
  FUNCTION_EFFECTS2 <- resume(FUNCTION_EFFECTS,init_type='last',iters=600,chains=2,cores=2)
  
  vis(FUNCTION_EFFECTS2,est(TOPICS),ui_level=.95)
}

run_tax4fun(phy_t1,"Phylum")
```

