{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h1> Overview <h1>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h2> Project Set-up <h2>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "pkg_variables <- function(status){\n",
    "    pkg_list = c(\"ape\",\"biomformat\",\"corrplot\",\"cowplot\",\"cluster\",\n",
    "             \"data.table\",\"expss\",\n",
    "             \"GGally\",\"Hmisc\",\"ggpubr\",\"ggnetwork\",\"ggplot2\",\"grid\",\n",
    "             \"gridExtra\",\n",
    "             \"otuSummary\",\n",
    "             \"plyr\",\"RColorBrewer\",\n",
    "             \"scales\",\"tibble\",\"vegan\",\"yaml\",\"dplyr\",\"tidyverse\",\"phyloseq\")\n",
    "\n",
    "    for (pkg in pkg_list){\n",
    "        print(pkg)\n",
    "        #Detaching\n",
    "        if(status==\"detach\"){\n",
    "            print(\"detaching\")\n",
    "            \n",
    "            #If the package has been loaded detach\n",
    "            if(pkg %in% (.packages())){\n",
    "                pkg_name = paste(\"package\", pkg, sep = \":\")\n",
    "                detach(pkg_name, unload = TRUE, character.only = TRUE)   \n",
    "            }\n",
    "        }\n",
    "        else{\n",
    "            lapply(pkg, require, character.only = TRUE)\n",
    "        }\n",
    "    }\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"package 'extrafont' was built under R version 3.6.2\"\n",
      "Registering fonts with R\n",
      "\n",
      "Agency FB already registered with windowsFonts().\n",
      "\n",
      "Algerian already registered with windowsFonts().\n",
      "\n",
      "Arial Black already registered with windowsFonts().\n",
      "\n",
      "Arial already registered with windowsFonts().\n",
      "\n",
      "Arial Narrow already registered with windowsFonts().\n",
      "\n",
      "Arial Rounded MT Bold already registered with windowsFonts().\n",
      "\n",
      "Bahnschrift already registered with windowsFonts().\n",
      "\n",
      "Baskerville Old Face already registered with windowsFonts().\n",
      "\n",
      "Bauhaus 93 already registered with windowsFonts().\n",
      "\n",
      "Bell MT already registered with windowsFonts().\n",
      "\n",
      "Berlin Sans FB already registered with windowsFonts().\n",
      "\n",
      "Berlin Sans FB Demi already registered with windowsFonts().\n",
      "\n",
      "Bernard MT Condensed already registered with windowsFonts().\n",
      "\n",
      "Blackadder ITC already registered with windowsFonts().\n",
      "\n",
      "Bodoni MT already registered with windowsFonts().\n",
      "\n",
      "Bodoni MT Black already registered with windowsFonts().\n",
      "\n",
      "Bodoni MT Condensed already registered with windowsFonts().\n",
      "\n",
      "Bodoni MT Poster Compressed already registered with windowsFonts().\n",
      "\n",
      "Book Antiqua already registered with windowsFonts().\n",
      "\n",
      "Bookman Old Style already registered with windowsFonts().\n",
      "\n",
      "Bookshelf Symbol 7 already registered with windowsFonts().\n",
      "\n",
      "Bradley Hand ITC already registered with windowsFonts().\n",
      "\n",
      "Britannic Bold already registered with windowsFonts().\n",
      "\n",
      "Broadway already registered with windowsFonts().\n",
      "\n",
      "Brush Script MT already registered with windowsFonts().\n",
      "\n",
      "Calibri already registered with windowsFonts().\n",
      "\n",
      "Calibri Light already registered with windowsFonts().\n",
      "\n",
      "Californian FB already registered with windowsFonts().\n",
      "\n",
      "Calisto MT already registered with windowsFonts().\n",
      "\n",
      "Cambria already registered with windowsFonts().\n",
      "\n",
      "Candara already registered with windowsFonts().\n",
      "\n",
      "Candara Light already registered with windowsFonts().\n",
      "\n",
      "Castellar already registered with windowsFonts().\n",
      "\n",
      "Centaur already registered with windowsFonts().\n",
      "\n",
      "Century already registered with windowsFonts().\n",
      "\n",
      "Century Gothic already registered with windowsFonts().\n",
      "\n",
      "Century Schoolbook already registered with windowsFonts().\n",
      "\n",
      "Chiller already registered with windowsFonts().\n",
      "\n",
      "Colonna MT already registered with windowsFonts().\n",
      "\n",
      "Comic Sans MS already registered with windowsFonts().\n",
      "\n",
      "Consolas already registered with windowsFonts().\n",
      "\n",
      "Constantia already registered with windowsFonts().\n",
      "\n",
      "Cooper Black already registered with windowsFonts().\n",
      "\n",
      "Copperplate Gothic Bold already registered with windowsFonts().\n",
      "\n",
      "Copperplate Gothic Light already registered with windowsFonts().\n",
      "\n",
      "Corbel already registered with windowsFonts().\n",
      "\n",
      "Corbel Light already registered with windowsFonts().\n",
      "\n",
      "Courier New already registered with windowsFonts().\n",
      "\n",
      "Curlz MT already registered with windowsFonts().\n",
      "\n",
      "Dubai already registered with windowsFonts().\n",
      "\n",
      "Dubai Light already registered with windowsFonts().\n",
      "\n",
      "Dubai Medium already registered with windowsFonts().\n",
      "\n",
      "Ebrima already registered with windowsFonts().\n",
      "\n",
      "Edwardian Script ITC already registered with windowsFonts().\n",
      "\n",
      "Elephant already registered with windowsFonts().\n",
      "\n",
      "Engravers MT already registered with windowsFonts().\n",
      "\n",
      "Eras Bold ITC already registered with windowsFonts().\n",
      "\n",
      "Eras Demi ITC already registered with windowsFonts().\n",
      "\n",
      "Eras Light ITC already registered with windowsFonts().\n",
      "\n",
      "Eras Medium ITC already registered with windowsFonts().\n",
      "\n",
      "Felix Titling already registered with windowsFonts().\n",
      "\n",
      "Footlight MT Light already registered with windowsFonts().\n",
      "\n",
      "Forte already registered with windowsFonts().\n",
      "\n",
      "Franklin Gothic Book already registered with windowsFonts().\n",
      "\n",
      "Franklin Gothic Demi already registered with windowsFonts().\n",
      "\n",
      "Franklin Gothic Demi Cond already registered with windowsFonts().\n",
      "\n",
      "Franklin Gothic Heavy already registered with windowsFonts().\n",
      "\n",
      "Franklin Gothic Medium already registered with windowsFonts().\n",
      "\n",
      "Franklin Gothic Medium Cond already registered with windowsFonts().\n",
      "\n",
      "Freestyle Script already registered with windowsFonts().\n",
      "\n",
      "French Script MT already registered with windowsFonts().\n",
      "\n",
      "Gabriola already registered with windowsFonts().\n",
      "\n",
      "Gadugi already registered with windowsFonts().\n",
      "\n",
      "Garamond already registered with windowsFonts().\n",
      "\n",
      "Georgia already registered with windowsFonts().\n",
      "\n",
      "Gigi already registered with windowsFonts().\n",
      "\n",
      "Gill Sans Ultra Bold already registered with windowsFonts().\n",
      "\n",
      "Gill Sans Ultra Bold Condensed already registered with windowsFonts().\n",
      "\n",
      "Gill Sans MT already registered with windowsFonts().\n",
      "\n",
      "Gill Sans MT Condensed already registered with windowsFonts().\n",
      "\n",
      "Gill Sans MT Ext Condensed Bold already registered with windowsFonts().\n",
      "\n",
      "Gloucester MT Extra Condensed already registered with windowsFonts().\n",
      "\n",
      "Goudy Old Style already registered with windowsFonts().\n",
      "\n",
      "Goudy Stout already registered with windowsFonts().\n",
      "\n",
      "HP Simplified already registered with windowsFonts().\n",
      "\n",
      "HP Simplified Light already registered with windowsFonts().\n",
      "\n",
      "Haettenschweiler already registered with windowsFonts().\n",
      "\n",
      "Harlow Solid Italic already registered with windowsFonts().\n",
      "\n",
      "Harrington already registered with windowsFonts().\n",
      "\n",
      "High Tower Text already registered with windowsFonts().\n",
      "\n",
      "HoloLens MDL2 Assets already registered with windowsFonts().\n",
      "\n",
      "Impact already registered with windowsFonts().\n",
      "\n",
      "Imprint MT Shadow already registered with windowsFonts().\n",
      "\n",
      "Informal Roman already registered with windowsFonts().\n",
      "\n",
      "Ink Free already registered with windowsFonts().\n",
      "\n",
      "Javanese Text already registered with windowsFonts().\n",
      "\n",
      "Jokerman already registered with windowsFonts().\n",
      "\n",
      "Juice ITC already registered with windowsFonts().\n",
      "\n",
      "Kristen ITC already registered with windowsFonts().\n",
      "\n",
      "Kunstler Script already registered with windowsFonts().\n",
      "\n",
      "Wide Latin already registered with windowsFonts().\n",
      "\n",
      "Leelawadee UI already registered with windowsFonts().\n",
      "\n",
      "Leelawadee UI Semilight already registered with windowsFonts().\n",
      "\n",
      "Lucida Bright already registered with windowsFonts().\n",
      "\n",
      "Lucida Calligraphy already registered with windowsFonts().\n",
      "\n",
      "Lucida Console already registered with windowsFonts().\n",
      "\n",
      "Lucida Fax already registered with windowsFonts().\n",
      "\n",
      "Lucida Handwriting already registered with windowsFonts().\n",
      "\n",
      "Lucida Sans already registered with windowsFonts().\n",
      "\n",
      "Lucida Sans Typewriter already registered with windowsFonts().\n",
      "\n",
      "Lucida Sans Unicode already registered with windowsFonts().\n",
      "\n",
      "MS Outlook already registered with windowsFonts().\n",
      "\n",
      "MS Reference Sans Serif already registered with windowsFonts().\n",
      "\n",
      "MS Reference Specialty already registered with windowsFonts().\n",
      "\n",
      "MV Boli already registered with windowsFonts().\n",
      "\n",
      "Magneto already registered with windowsFonts().\n",
      "\n",
      "Maiandra GD already registered with windowsFonts().\n",
      "\n",
      "Malgun Gothic already registered with windowsFonts().\n",
      "\n",
      "Malgun Gothic Semilight already registered with windowsFonts().\n",
      "\n",
      "Marlett already registered with windowsFonts().\n",
      "\n",
      "Matura MT Script Capitals already registered with windowsFonts().\n",
      "\n",
      "Microsoft Himalaya already registered with windowsFonts().\n",
      "\n",
      "Microsoft Yi Baiti already registered with windowsFonts().\n",
      "\n",
      "Microsoft New Tai Lue already registered with windowsFonts().\n",
      "\n",
      "Microsoft PhagsPa already registered with windowsFonts().\n",
      "\n",
      "Microsoft Sans Serif already registered with windowsFonts().\n",
      "\n",
      "Microsoft Tai Le already registered with windowsFonts().\n",
      "\n",
      "Mistral already registered with windowsFonts().\n",
      "\n",
      "Modern No. 20 already registered with windowsFonts().\n",
      "\n",
      "Mongolian Baiti already registered with windowsFonts().\n",
      "\n",
      "Monotype Corsiva already registered with windowsFonts().\n",
      "\n",
      "Myanmar Text already registered with windowsFonts().\n",
      "\n",
      "Niagara Engraved already registered with windowsFonts().\n",
      "\n",
      "Niagara Solid already registered with windowsFonts().\n",
      "\n",
      "Nirmala UI already registered with windowsFonts().\n",
      "\n",
      "Nirmala UI Semilight already registered with windowsFonts().\n",
      "\n",
      "OCR A Extended already registered with windowsFonts().\n",
      "\n",
      "Old English Text MT already registered with windowsFonts().\n",
      "\n",
      "Onyx already registered with windowsFonts().\n",
      "\n",
      "Palace Script MT already registered with windowsFonts().\n",
      "\n",
      "Palatino Linotype already registered with windowsFonts().\n",
      "\n",
      "Papyrus already registered with windowsFonts().\n",
      "\n",
      "Parchment already registered with windowsFonts().\n",
      "\n",
      "Perpetua already registered with windowsFonts().\n",
      "\n",
      "Perpetua Titling MT already registered with windowsFonts().\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Playbill already registered with windowsFonts().\n",
      "\n",
      "Poor Richard already registered with windowsFonts().\n",
      "\n",
      "Pristina already registered with windowsFonts().\n",
      "\n",
      "Rage Italic already registered with windowsFonts().\n",
      "\n",
      "Ravie already registered with windowsFonts().\n",
      "\n",
      "Rockwell already registered with windowsFonts().\n",
      "\n",
      "Rockwell Condensed already registered with windowsFonts().\n",
      "\n",
      "Rockwell Extra Bold already registered with windowsFonts().\n",
      "\n",
      "Script MT Bold already registered with windowsFonts().\n",
      "\n",
      "Segoe MDL2 Assets already registered with windowsFonts().\n",
      "\n",
      "Segoe Print already registered with windowsFonts().\n",
      "\n",
      "Segoe Script already registered with windowsFonts().\n",
      "\n",
      "Segoe UI already registered with windowsFonts().\n",
      "\n",
      "Segoe UI Light already registered with windowsFonts().\n",
      "\n",
      "Segoe UI Semibold already registered with windowsFonts().\n",
      "\n",
      "Segoe UI Semilight already registered with windowsFonts().\n",
      "\n",
      "Segoe UI Black already registered with windowsFonts().\n",
      "\n",
      "Segoe UI Emoji already registered with windowsFonts().\n",
      "\n",
      "Segoe UI Historic already registered with windowsFonts().\n",
      "\n",
      "Segoe UI Symbol already registered with windowsFonts().\n",
      "\n",
      "Showcard Gothic already registered with windowsFonts().\n",
      "\n",
      "SimSun-ExtB already registered with windowsFonts().\n",
      "\n",
      "Snap ITC already registered with windowsFonts().\n",
      "\n",
      "Stencil already registered with windowsFonts().\n",
      "\n",
      "Sylfaen already registered with windowsFonts().\n",
      "\n",
      "Symbol already registered with windowsFonts().\n",
      "\n",
      "Tahoma already registered with windowsFonts().\n",
      "\n",
      "Tempus Sans ITC already registered with windowsFonts().\n",
      "\n",
      "Times New Roman already registered with windowsFonts().\n",
      "\n",
      "Trebuchet MS already registered with windowsFonts().\n",
      "\n",
      "Tw Cen MT already registered with windowsFonts().\n",
      "\n",
      "Tw Cen MT Condensed already registered with windowsFonts().\n",
      "\n",
      "Tw Cen MT Condensed Extra Bold already registered with windowsFonts().\n",
      "\n",
      "Verdana already registered with windowsFonts().\n",
      "\n",
      "Viner Hand ITC already registered with windowsFonts().\n",
      "\n",
      "Vivaldi already registered with windowsFonts().\n",
      "\n",
      "Vladimir Script already registered with windowsFonts().\n",
      "\n",
      "Webdings already registered with windowsFonts().\n",
      "\n",
      "Wingdings already registered with windowsFonts().\n",
      "\n",
      "Wingdings 2 already registered with windowsFonts().\n",
      "\n",
      "Wingdings 3 already registered with windowsFonts().\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"ape\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: ape\n",
      "\n",
      "Warning message:\n",
      "\"package 'ape' was built under R version 3.6.3\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"biomformat\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: biomformat\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"corrplot\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: corrplot\n",
      "\n",
      "Warning message:\n",
      "\"package 'corrplot' was built under R version 3.6.3\"\n",
      "corrplot 0.84 loaded\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"cowplot\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: cowplot\n",
      "\n",
      "Warning message:\n",
      "\"package 'cowplot' was built under R version 3.6.3\"\n",
      "\n",
      "********************************************************\n",
      "\n",
      "Note: As of version 1.0.0, cowplot does not change the\n",
      "\n",
      "  default ggplot2 theme anymore. To recover the previous\n",
      "\n",
      "  behavior, execute:\n",
      "  theme_set(theme_cowplot())\n",
      "\n",
      "********************************************************\n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"cluster\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: cluster\n",
      "\n",
      "Warning message:\n",
      "\"package 'cluster' was built under R version 3.6.3\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"data.table\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: data.table\n",
      "\n",
      "Warning message:\n",
      "\"package 'data.table' was built under R version 3.6.3\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"expss\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: expss\n",
      "\n",
      "Warning message:\n",
      "\"package 'expss' was built under R version 3.6.3\"\n",
      "\n",
      "Attaching package: 'expss'\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:data.table':\n",
      "\n",
      "    copy, like\n",
      "\n",
      "\n",
      "The following object is masked from 'package:ape':\n",
      "\n",
      "    where\n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"GGally\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: GGally\n",
      "\n",
      "Warning message:\n",
      "\"package 'GGally' was built under R version 3.6.3\"\n",
      "Loading required package: ggplot2\n",
      "\n",
      "Warning message:\n",
      "\"package 'ggplot2' was built under R version 3.6.3\"\n",
      "\n",
      "Attaching package: 'ggplot2'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:expss':\n",
      "\n",
      "    vars\n",
      "\n",
      "\n",
      "Registered S3 method overwritten by 'GGally':\n",
      "  method from   \n",
      "  +.gg   ggplot2\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Hmisc\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: Hmisc\n",
      "\n",
      "Warning message:\n",
      "\"package 'Hmisc' was built under R version 3.6.3\"\n",
      "Loading required package: lattice\n",
      "\n",
      "Warning message:\n",
      "\"package 'lattice' was built under R version 3.6.3\"\n",
      "Loading required package: survival\n",
      "\n",
      "Warning message:\n",
      "\"package 'survival' was built under R version 3.6.3\"\n",
      "Loading required package: Formula\n",
      "\n",
      "Registered S3 methods overwritten by 'Hmisc':\n",
      "  method                 from \n",
      "  [.labelled             expss\n",
      "  print.labelled         expss\n",
      "  as.data.frame.labelled expss\n",
      "\n",
      "\n",
      "Attaching package: 'Hmisc'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:ape':\n",
      "\n",
      "    zoom\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:base':\n",
      "\n",
      "    format.pval, units\n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"ggpubr\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: ggpubr\n",
      "\n",
      "Warning message:\n",
      "\"package 'ggpubr' was built under R version 3.6.3\"\n",
      "Loading required package: magrittr\n",
      "\n",
      "\n",
      "Attaching package: 'magrittr'\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:expss':\n",
      "\n",
      "    and, equals, or\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: 'ggpubr'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:expss':\n",
      "\n",
      "    compare_means\n",
      "\n",
      "\n",
      "The following object is masked from 'package:cowplot':\n",
      "\n",
      "    get_legend\n",
      "\n",
      "\n",
      "The following object is masked from 'package:ape':\n",
      "\n",
      "    rotate\n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"ggnetwork\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: ggnetwork\n",
      "\n",
      "Warning message:\n",
      "\"package 'ggnetwork' was built under R version 3.6.3\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"ggplot2\"\n",
      "[1] \"grid\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: grid\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"gridExtra\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: gridExtra\n",
      "\n",
      "Warning message:\n",
      "\"package 'gridExtra' was built under R version 3.6.3\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"otuSummary\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: otuSummary\n",
      "\n",
      "Warning message:\n",
      "\"package 'otuSummary' was built under R version 3.6.3\"\n",
      "Loading required package: vegan\n",
      "\n",
      "Warning message:\n",
      "\"package 'vegan' was built under R version 3.6.3\"\n",
      "Loading required package: permute\n",
      "\n",
      "Warning message:\n",
      "\"package 'permute' was built under R version 3.6.3\"\n",
      "This is vegan 2.5-6\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"plyr\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: plyr\n",
      "\n",
      "Warning message:\n",
      "\"package 'plyr' was built under R version 3.6.3\"\n",
      "\n",
      "Attaching package: 'plyr'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:ggpubr':\n",
      "\n",
      "    mutate\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:Hmisc':\n",
      "\n",
      "    is.discrete, summarize\n",
      "\n",
      "\n",
      "The following object is masked from 'package:expss':\n",
      "\n",
      "    split_labels\n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"RColorBrewer\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: RColorBrewer\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"scales\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: scales\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"tibble\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: tibble\n",
      "\n",
      "Warning message:\n",
      "\"package 'tibble' was built under R version 3.6.3\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"vegan\"\n",
      "[1] \"yaml\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: yaml\n",
      "\n",
      "Warning message:\n",
      "\"package 'yaml' was built under R version 3.6.2\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"dplyr\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: dplyr\n",
      "\n",
      "Warning message:\n",
      "\"package 'dplyr' was built under R version 3.6.3\"\n",
      "\n",
      "Attaching package: 'dplyr'\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:plyr':\n",
      "\n",
      "    arrange, count, desc, failwith, id, mutate, rename, summarise,\n",
      "    summarize\n",
      "\n",
      "\n",
      "The following object is masked from 'package:gridExtra':\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:Hmisc':\n",
      "\n",
      "    src, summarize\n",
      "\n",
      "\n",
      "The following object is masked from 'package:GGally':\n",
      "\n",
      "    nasa\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:expss':\n",
      "\n",
      "    between, compute, contains, first, last, na_if, recode, vars\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:data.table':\n",
      "\n",
      "    between, first, last\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:stats':\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:base':\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"tidyverse\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: tidyverse\n",
      "\n",
      "Warning message:\n",
      "\"package 'tidyverse' was built under R version 3.6.3\"\n",
      "-- \u001b[1mAttaching packages\u001b[22m --------------------------------------- tidyverse 1.3.0 --\n",
      "\n",
      "\u001b[32m<U+221A>\u001b[39m \u001b[34mtidyr  \u001b[39m 1.0.2     \u001b[32m<U+221A>\u001b[39m \u001b[34mstringr\u001b[39m 1.4.0\n",
      "\u001b[32m<U+221A>\u001b[39m \u001b[34mreadr  \u001b[39m 1.3.1     \u001b[32m<U+221A>\u001b[39m \u001b[34mforcats\u001b[39m 0.5.0\n",
      "\u001b[32m<U+221A>\u001b[39m \u001b[34mpurrr  \u001b[39m 0.3.2     \n",
      "\n",
      "Warning message:\n",
      "\"package 'tidyr' was built under R version 3.6.3\"\n",
      "Warning message:\n",
      "\"package 'forcats' was built under R version 3.6.3\"\n",
      "-- \u001b[1mConflicts\u001b[22m ------------------------------------------ tidyverse_conflicts() --\n",
      "\u001b[31mx\u001b[39m \u001b[34mmagrittr\u001b[39m::\u001b[32mand()\u001b[39m     masks \u001b[34mexpss\u001b[39m::and()\n",
      "\u001b[31mx\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32marrange()\u001b[39m    masks \u001b[34mplyr\u001b[39m::arrange()\n",
      "\u001b[31mx\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mbetween()\u001b[39m    masks \u001b[34mexpss\u001b[39m::between(), \u001b[34mdata.table\u001b[39m::between()\n",
      "\u001b[31mx\u001b[39m \u001b[34mreadr\u001b[39m::\u001b[32mcol_factor()\u001b[39m masks \u001b[34mscales\u001b[39m::col_factor()\n",
      "\u001b[31mx\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mcombine()\u001b[39m    masks \u001b[34mgridExtra\u001b[39m::combine()\n",
      "\u001b[31mx\u001b[39m \u001b[34mpurrr\u001b[39m::\u001b[32mcompact()\u001b[39m    masks \u001b[34mplyr\u001b[39m::compact()\n",
      "\u001b[31mx\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mcompute()\u001b[39m    masks \u001b[34mexpss\u001b[39m::compute()\n",
      "\u001b[31mx\u001b[39m \u001b[34mtidyr\u001b[39m::\u001b[32mcontains()\u001b[39m   masks \u001b[34mdplyr\u001b[39m::contains(), \u001b[34mexpss\u001b[39m::contains()\n",
      "\u001b[31mx\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mcount()\u001b[39m      masks \u001b[34mplyr\u001b[39m::count()\n",
      "\u001b[31mx\u001b[39m \u001b[34mpurrr\u001b[39m::\u001b[32mdiscard()\u001b[39m    masks \u001b[34mscales\u001b[39m::discard()\n",
      "\u001b[31mx\u001b[39m \u001b[34mmagrittr\u001b[39m::\u001b[32mequals()\u001b[39m  masks \u001b[34mexpss\u001b[39m::equals()\n",
      "\u001b[31mx\u001b[39m \u001b[34mtidyr\u001b[39m::\u001b[32mextract()\u001b[39m    masks \u001b[34mmagrittr\u001b[39m::extract()\n",
      "\u001b[31mx\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mfailwith()\u001b[39m   masks \u001b[34mplyr\u001b[39m::failwith()\n",
      "\u001b[31mx\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mfilter()\u001b[39m     masks \u001b[34mstats\u001b[39m::filter()\n",
      "\u001b[31mx\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mfirst()\u001b[39m      masks \u001b[34mexpss\u001b[39m::first(), \u001b[34mdata.table\u001b[39m::first()\n",
      "\u001b[31mx\u001b[39m \u001b[34mstringr\u001b[39m::\u001b[32mfixed()\u001b[39m    masks \u001b[34mexpss\u001b[39m::fixed()\n",
      "\u001b[31mx\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mid()\u001b[39m         masks \u001b[34mplyr\u001b[39m::id()\n",
      "\u001b[31mx\u001b[39m \u001b[34mpurrr\u001b[39m::\u001b[32mkeep()\u001b[39m       masks \u001b[34mexpss\u001b[39m::keep()\n",
      "\u001b[31mx\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mlag()\u001b[39m        masks \u001b[34mstats\u001b[39m::lag()\n",
      "\u001b[31mx\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mlast()\u001b[39m       masks \u001b[34mexpss\u001b[39m::last(), \u001b[34mdata.table\u001b[39m::last()\n",
      "\u001b[31mx\u001b[39m \u001b[34mpurrr\u001b[39m::\u001b[32mmodify()\u001b[39m     masks \u001b[34mexpss\u001b[39m::modify()\n",
      "\u001b[31mx\u001b[39m \u001b[34mpurrr\u001b[39m::\u001b[32mmodify_if()\u001b[39m  masks \u001b[34mexpss\u001b[39m::modify_if()\n",
      "\u001b[31mx\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mmutate()\u001b[39m     masks \u001b[34mplyr\u001b[39m::mutate(), \u001b[34mggpubr\u001b[39m::mutate()\n",
      "\u001b[31mx\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mna_if()\u001b[39m      masks \u001b[34mexpss\u001b[39m::na_if()\n",
      "\u001b[31mx\u001b[39m \u001b[34mtidyr\u001b[39m::\u001b[32mnest()\u001b[39m       masks \u001b[34mexpss\u001b[39m::nest()\n",
      "\u001b[31mx\u001b[39m \u001b[34mmagrittr\u001b[39m::\u001b[32mor()\u001b[39m      masks \u001b[34mexpss\u001b[39m::or()\n",
      "\u001b[31mx\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mrecode()\u001b[39m     masks \u001b[34mexpss\u001b[39m::recode()\n",
      "\u001b[31mx\u001b[39m \u001b[34mstringr\u001b[39m::\u001b[32mregex()\u001b[39m    masks \u001b[34mexpss\u001b[39m::regex()\n",
      "\u001b[31mx\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mrename()\u001b[39m     masks \u001b[34mplyr\u001b[39m::rename()\n",
      "\u001b[31mx\u001b[39m \u001b[34mpurrr\u001b[39m::\u001b[32mset_names()\u001b[39m  masks \u001b[34mmagrittr\u001b[39m::set_names()\n",
      "\u001b[31mx\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32msrc()\u001b[39m        masks \u001b[34mHmisc\u001b[39m::src()\n",
      "\u001b[31mx\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32msummarise()\u001b[39m  masks \u001b[34mplyr\u001b[39m::summarise()\n",
      "\u001b[31mx\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32msummarize()\u001b[39m  masks \u001b[34mplyr\u001b[39m::summarize(), \u001b[34mHmisc\u001b[39m::summarize()\n",
      "\u001b[31mx\u001b[39m \u001b[34mpurrr\u001b[39m::\u001b[32mtranspose()\u001b[39m  masks \u001b[34mexpss\u001b[39m::transpose(), \u001b[34mdata.table\u001b[39m::transpose()\n",
      "\u001b[31mx\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mvars()\u001b[39m       masks \u001b[34mggplot2\u001b[39m::vars(), \u001b[34mexpss\u001b[39m::vars()\n",
      "\u001b[31mx\u001b[39m \u001b[34mpurrr\u001b[39m::\u001b[32mwhen()\u001b[39m       masks \u001b[34mexpss\u001b[39m::when()\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"phyloseq\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: phyloseq\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(extrafont)\n",
    "extrafont::loadfonts(device=\"win\")\n",
    "windowsFonts(Times=windowsFont(\"TT Times New Roman\"))\n",
    "\n",
    "#pkg_variables(status=\"detach\")\n",
    "pkg_variables(status=\"attach\")\n",
    "\n",
    "source(\"sources/ggrare.R\") #github library: https://rdrr.io/github/gauravsk/ranacapa/\n",
    "source(\"sources/miseqR.R\")\n",
    "source(\"sources/qiime2R.R\") #github https://github.com/jbisanz/qiime2R\n",
    "source('sources/abundances.R') #github https://github.com/microbiome/microbiome\n",
    "source('sources/ordination.R') #github https://github.com/umerijaz/microbiomeSeq\n",
    "source('sources/plot.ordination.R') #github https://github.com/umerijaz/microbiomeSeq\n",
    "source('sources/plot_anova_diversity.R') #github https://github.com/umerijaz/microbiomeSeq\n",
    "source('sources/alpha_div.R')\n",
    "source('sources/perform_anova.R')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "run_param=\"dceg\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "proj_dir='/Users/slsevilla/Google Drive/MyDocuments_Current/Education/George Mason University/Dissertation/Aim3/'\n",
    "data_dir=paste(proj_dir,'output/data/',sep=\"\")\n",
    "img_dir=paste(proj_dir,'output/analysis/img/',sep=\"\")\n",
    "stats_dir=paste(proj_dir,'output/analysis/stats/',sep=\"\")\n",
    "man_dir = paste(proj_dir,'manifest/',sep=\"\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h3> Metadata input <h3>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata<-read.table(man_dir + 'manifest.txt',sep='\\t', header=T, row.names=1, comment=\"\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h2> Create PhyloSeq Object<h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h3> Functions <h3>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "phylo_create <- function(tax_in,tab_in){\n",
    "    #Read OTUS\n",
    "    otus<-read_qza(tab_in)\n",
    "\n",
    "    #Read Greengenes taxonomy file\n",
    "    taxonomy<-read_qza(tax_in)\n",
    "\n",
    "    #Create taxonomy table\n",
    "    tax_table = data.frame()\n",
    "    col = 1\n",
    "    row = 1\n",
    "\n",
    "    for (i in taxonomy$data$Taxon){\n",
    "        tax_list = as.list(strsplit(as.character(i),\";\"))\n",
    "        for (i in 1:lengths(tax_list)){\n",
    "            tax_table[row,col] = tax_list[[1]][i]\n",
    "            col = col + 1\n",
    "        }\n",
    "        col = 1\n",
    "        row = row + 1\n",
    "    }\n",
    "    rownames(tax_table)<-taxonomy$data$Feature.ID\n",
    "    colnames(tax_table)<-c(\"Domain\", \"Phylum\", \"Class\", \"Order\", \"Family\", \"Genus\", \"Species\")\n",
    "    tax_convert<-as.matrix(tax_table)  \n",
    "    \n",
    "    #Create phylo object\n",
    "    phy_obj<-phyloseq(otu_table(otus$data, taxa_are_rows = T), tax_table(tax_convert), sample_data(metadata))\n",
    "    print(phy_obj) \n",
    "    return(phy_obj)\n",
    "}   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "phy_subset_ext <- function(phy_in){\n",
    "    #glom to tax level, subset for only seq controls\n",
    "    phy_obj = tax_glom(phy_in,taxrank=\"Genus\") #only merges genera that have the same upper taxonomy\n",
    "    phy_obj = subset_samples(phy_obj,Sample.Type==\"Ext_Control\" & Sample.Des!=\"Ext_Blank\")\n",
    "    return(phy_obj)\n",
    "}\n",
    "\n",
    "phy_subset_blank <- function(phy_in){\n",
    "    #glom to tax level, subset for only seq controls\n",
    "    phy_obj = tax_glom(phy_in,taxrank=\"Genus\") #only merges genera that have the same upper taxonomy\n",
    "    phy_obj = subset_samples(phy_obj,Sample.Des==\"Ext_Blank\")\n",
    "    return(phy_obj)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "otu_extract <- function(phy_in){\n",
    "    # Extract abundance matrix from the phyloseq object\n",
    "    OTU1 = as(otu_table(phy_in), \"matrix\")\n",
    "    if(taxa_are_rows(phy_in)){OTU1 <- t(OTU1)} # transpose\n",
    "    OTUdf = as.data.frame(OTU1)# Coerce to data.frame\n",
    "    return(OTUdf)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "tax_extract <- function(phy_in){\n",
    "    #extract taxonomy information into dataframe\n",
    "    TAXdf = tax_table(phy_in)\n",
    "    return(TAXdf)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "otu_tax_clean <- function(OTU_in,TAX_in){\n",
    "    #Remove zero cols\n",
    "    OTUdf_clean = OTU_in[, which(colSums(OTU_in) != 0)]\n",
    "\n",
    "    #rename col from features to genus\n",
    "    for (i in colnames(OTU_in)){\n",
    "        newname = paste(as.character(TAX_in[i,\"Family\"]),as.character(TAX_in[i,\"Genus\"]),sep=\"\")\n",
    "        \n",
    "        #standardize gg naming\n",
    "        newname = trimws(newname) #remove first space in gg naming\n",
    "        newname = sub(\" g__\",\"g__\",newname)\n",
    "        newname = sub(\"\\\\[\",\"\",newname)\n",
    "        newname = sub(\"\\\\]\",\"\",newname)\n",
    "        \n",
    "        if(is.na(newname)){\n",
    "            next;\n",
    "        } else{\n",
    "            names(OTUdf_clean)[names(OTUdf_clean) == i] <- newname\n",
    "        }\n",
    "    }\n",
    "    \n",
    "    #add metadata col\n",
    "    for (a in rownames(OTUdf_clean)){\n",
    "        OTUdf_clean[a,\"Subject.ID\"] = metadata[a,\"Subject.ID\"]\n",
    "        OTUdf_clean[a,\"Ext.Kit\"] = metadata[a,\"Ext.Kit\"]\n",
    "        OTUdf_clean[a,\"Homo.Status\"] = metadata[a,\"Homo.Status\"]\n",
    "    }\n",
    "    OTUdf_clean = data.frame(OTUdf_clean,stringsAsFactors=FALSE)\n",
    "    \n",
    "    return(OTUdf_clean)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h3> Code <h3>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "phyloseq-class experiment-level object\n",
      "otu_table()   OTU Table:         [ 1818 taxa and 351 samples ]\n",
      "sample_data() Sample Data:       [ 351 samples by 27 sample variables ]\n",
      "tax_table()   Taxonomy Table:    [ 1818 taxa by 7 taxonomic ranks ]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "`set.seed(1)` was used to initialize repeatable random subsampling.\n",
      "\n",
      "Please record this for your records so others can reproduce.\n",
      "\n",
      "Try `set.seed(1); .Random.seed` for the full vector\n",
      "\n",
      "...\n",
      "\n",
      "69 samples removedbecause they contained fewer reads than `sample.size`.\n",
      "\n",
      "Up to first five removed samples are: \n",
      "\n",
      "\n",
      "NTC-PC04925-H-12SC249366-PC04925-D-01SC249382-PC04925-F-01SC249406SC249416-PC04925-A-03\n",
      "\n",
      "...\n",
      "\n",
      "612OTUs were removed because they are no longer \n",
      "present in any sample after random subsampling\n",
      "\n",
      "\n",
      "...\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#Create the phyloseq object\n",
    "tax_file = paste(data_dir,'taxonomic_classification/classify-sklearn_silva-132-99-nb-classifier.qza',sep=\"\")\n",
    "table_file = paste(data_dir,'denoising/feature_tables/merged_filtered.qza',sep=\"\")\n",
    "phy_obj = phylo_create(tax_file,table_file)  \n",
    "\n",
    "#Rarefy samples\n",
    "phy_r = rarefy_even_depth(phy_obj, rngseed=1, sample.size=20000, replace=F) \n",
    "\n",
    "#Create phy tree\n",
    "random_tree = rtree(ntaxa(phy_r), rooted=TRUE, tip.label=taxa_names(phy_r))\n",
    "phy_r_t = merge_phyloseq(phy_r, random_tree)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h2> TP,FP,FN <h2>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Create TP,FP,FN dataframe\n",
    "bench_compare <- function(OTUdf_in,CONTdf_in){\n",
    "    \n",
    "    F1df = OTUdf_in\n",
    "    F1df$Sample_names = rownames(F1df) #create col of sample_names\n",
    "    \n",
    "    #For each of the controls in the cleaned df\n",
    "    for (cont in unique(OTUdf_in$Subject.ID)){\n",
    "        temp=filter(CONTdf_in, get(cont) !=0) #create df of taxa related to control\n",
    "        \n",
    "        #Create taxa list, merging family and genus - this will help avoid id errors from taxonomic differences\n",
    "        #IE one taxa labeled at the genus level may be from family1 and the same taxa may also be from family2\n",
    "        taxa_list=list()\n",
    "        for (x in rownames(temp)){taxa_list = c(taxa_list,paste(temp[x,\"Family\"],temp[x,\"Genus\"],sep=\"\"))}  \n",
    "        \n",
    "        #Create list of samples that match the control\n",
    "        sample_list = filter(F1df, Subject.ID == cont)$Sample_names\n",
    "        \n",
    "        #If the data is not in the df (FN), then add the taxa as a new col\n",
    "        for (taxa in taxa_list){ \n",
    "            if(!(taxa %in% unique(colnames(F1df)))){\n",
    "                F1df[taxa[!(taxa %in% colnames(F1df))]] = 0\n",
    "            } \n",
    "        }\n",
    "\n",
    "        for (sample in sample_list){\n",
    "            for (col in unique(colnames(F1df))[unique(colnames(F1df)) != \"Subject.ID\" & unique(colnames(F1df)) != \"Sample_names\"\n",
    "                & unique(colnames(F1df)) != \"Ext.Kit\" & unique(colnames(F1df)) != \"Homo.Status\"]){ \n",
    "                #removes subjectID, sample_names cols from loop\n",
    "                value_check = F1df[sample,col]\n",
    "\n",
    "                if(col %in% taxa_list && value_check > 0){ #column matches expected taxa, has features\n",
    "                    F1df[sample,col] = \"TP\"\n",
    "                } else if (col %in% taxa_list && value_check == 0){ #column matches expected taxa, no features\n",
    "                    F1df[sample,col] = \"FN\"\n",
    "                } else if ((!col %in% taxa_list) && value_check > 0){ #column doesnt match expected taxa, has features\n",
    "                    F1df[sample,col] = \"FP\"\n",
    "                } else{ #column doesnt match expected taxa, has no features\n",
    "                    next\n",
    "                }\n",
    "            }\n",
    "        }\n",
    "    }\n",
    "    return(F1df)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Create counts of FP, FN, TP\n",
    "bench_counts <- function(OTUdf_in,F1_in,bmark_in,var_in){\n",
    "    F1df_counts = subset(OTUdf_in, select=c(\"Subject.ID\"))\n",
    "\n",
    "    for (sample in rownames(F1df)){\n",
    "        FN = rowSums(F1_in==\"FN\")[sample][[1]]\n",
    "        FP = rowSums(F1_in==\"FP\")[sample][[1]]\n",
    "        TP = rowSums(F1_in==\"TP\")[sample][[1]]\n",
    "\n",
    "        F1df_counts[sample,\"check\"]= FN + TP\n",
    "        F1df_counts[sample,\"FN\"] = FN\n",
    "        F1df_counts[sample,\"FP\"] = FP\n",
    "        F1df_counts[sample,\"TP\"] = TP\n",
    "        F1df_counts[sample,\"TAR\"] = TP / (TP + FP)\n",
    "        F1df_counts[sample,\"TDR\"] = TP / (TP + FN)\n",
    "\n",
    "        num = F1df_counts[sample,\"TAR\"]*F1df_counts[sample,\"TDR\"]\n",
    "        den = F1df_counts[sample,\"TAR\"]+F1df_counts[sample,\"TDR\"] \n",
    "        F1df_counts[sample,\"F1\"] = 2*(num/den)\n",
    "        F1df_counts[sample,\"Subject.ID\"] = OTUdf_in[sample,\"Subject.ID\"]\n",
    "        \n",
    "        F1df_counts[sample,\"bmark\"] = bmark_in\n",
    "        F1df_counts[sample,\"bmark2\"] = var_in\n",
    "    }\n",
    "    return(F1df_counts)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h3> Plotting <h3>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "plot_violin <- function(df_in,x_lab,y_ax,y_lab,col_in,y_lowerlim,y_upperlim){\n",
    "    if(y_lowerlim==\"\" | y_upperlim==\"\"){\n",
    "        p = ggplot(df_in, aes(x=get(x_lab), y=get(y_ax), color=get(col_in))) + \n",
    "            geom_violin(trim=FALSE) + \n",
    "            labs(color = col_in) +\n",
    "            ylab(y_lab) +\n",
    "            theme(axis.text.x = element_text(angle = 90, hjust = 1),\n",
    "                  axis.title.x = element_blank(), \n",
    "                 legend.position = \"none\") +\n",
    "            stat_summary(fun.data=mean_sdl, color=\"red\")\n",
    "    } else{\n",
    "        p = ggplot(df_in, aes(x=get(x_lab), y=get(y_ax), color=get(col_in))) + \n",
    "            geom_violin(trim=FALSE) + \n",
    "            labs(color = col_in) +\n",
    "             ylab(y_lab) +\n",
    "            theme(axis.text.x = element_text(angle = 90, hjust = 1),\n",
    "                  axis.title.x = element_blank(), \n",
    "                 legend.position = \"none\") +\n",
    "            ylim(y_lowerlim,y_upperlim) +\n",
    "            stat_summary(fun.data=mean_sdl, color=\"red\")\n",
    "    }    \n",
    "    return (p)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_f1_violin <- function(F1_in,x_list,y_list,id_in,title,ftype,y_lowerlim=\"\",y_upperlim=\"\"){\n",
    "    \n",
    "    p_1 = plot_violin(F1_in,id_in,x_list[1],y_list[1],id_in,y_lowerlim,y_upperlim)\n",
    "    p_2 = plot_violin(F1_in,id_in,x_list[2],y_list[2],id_in,y_lowerlim,y_upperlim)\n",
    "    p_3 = plot_violin(F1_in,id_in,x_list[3],y_list[3],id_in,y_lowerlim,y_upperlim)\n",
    "\n",
    "    #Create count violin plots\n",
    "    fig = ggarrange(p_1, p_2, p_3, \n",
    "            labels = c(\"A\", \"B\", \"C\"),\n",
    "            ncol = 2, nrow = 2)\n",
    "    \n",
    "    annotate_figure(fig, #https://rpkgs.datanovia.com/ggpubr/reference/annotate_figure.html\n",
    "                   top = text_grob(title, color = \"blue\", face = \"bold\", size = 14)\n",
    "    )\n",
    "    \n",
    "    ggsave(paste(img_dir,\"violin_f1_\",ftype,\"_\",title,\".jpg\",sep=\"\"))\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_violin_hmo <- function(df_in,x_lab,y_ax,y_lab,col_in,fac_in){\n",
    "    if(fac_in==2){\n",
    "        p = ggplot(df_in, aes(x=Subject.ID, y=get(y_ax), color=get(col_in))) + \n",
    "            facet_grid(Homo.Status~ get(x_lab))+\n",
    "            geom_violin(trim=FALSE) + \n",
    "            labs(color = col_in) +\n",
    "            theme(axis.text.x = element_text(angle = 90, hjust = 1),\n",
    "                  axis.title.x = element_blank(), \n",
    "                 legend.position = \"none\") + \n",
    "            ylab(y_lab) +\n",
    "            rotate_x_facet_text(angle = 90, align = 0.5) +\n",
    "            stat_summary(fun.data=mean_sdl, color=\"red\")\n",
    "    } else if(fac_in==3){\n",
    "        p = ggplot(df_in, aes(x=Homo.Status, y=get(y_ax), color=get(col_in))) + \n",
    "            facet_grid(Subject.ID ~get(x_lab))+\n",
    "            geom_violin(trim=FALSE) + \n",
    "            labs(color = col_in) +\n",
    "            theme(axis.text.x = element_text(angle = 90, hjust = 1),\n",
    "                  axis.title.x = element_blank(), \n",
    "                 legend.position = \"none\") + \n",
    "            ylab(y_lab) +\n",
    "            rotate_x_facet_text(angle = 90, align = 0.5) +\n",
    "            stat_summary(fun.data=mean_sdl, color=\"red\")    \n",
    "        \n",
    "    } else{\n",
    "        p = ggplot(df_in, aes(x=get(x_lab), y=get(y_ax), color=get(col_in))) + \n",
    "            facet_grid(Subject.ID~ .)+\n",
    "            geom_violin(trim=FALSE) + \n",
    "            labs(color = col_in) +\n",
    "            theme(axis.text.x = element_text(angle = 90, hjust = 1),\n",
    "                  axis.title.x = element_blank(), \n",
    "                 legend.position = \"none\") + \n",
    "            ylab(y_lab) +\n",
    "            rotate_x_facet_text(angle = 90, align = 0.5) +\n",
    "            stat_summary(fun.data=mean_sdl, color=\"red\")\n",
    "    }\n",
    "    return (p)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "rotate_x_facet_text <- function(angle = 45, align = 0, valign = 0.25) {\n",
    "    ggplot2::theme(strip.text.x = ggplot2::element_text(angle = angle,\n",
    "                                                        hjust = align,\n",
    "                                                        vjust = valign))\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_f1_violin_hmo <- function(F1_in,x_list,y_list,id_in,title,ftype,fac_in=2){\n",
    "    \n",
    "    p_1 = plot_violin_hmo(F1_in,id_in,x_list[1],y_list[1],id_in,fac_in)\n",
    "    annotate_figure(p_1, #https://rpkgs.datanovia.com/ggpubr/reference/annotate_figure.html\n",
    "                    top = text_grob(title, color = \"blue\", face = \"bold\", size = 14))\n",
    "    ggsave(paste(img_dir,\"violin_f1_\",ftype,\"_\",title,\"_\",x_list[1],\".jpg\",sep=\"\"))\n",
    "    \n",
    "    p_2 = plot_violin_hmo(F1_in,id_in,x_list[2],y_list[2],id_in,fac_in)\n",
    "    annotate_figure(p_2, #https://rpkgs.datanovia.com/ggpubr/reference/annotate_figure.html\n",
    "                    top = text_grob(title, color = \"blue\", face = \"bold\", size = 14))\n",
    "    ggsave(paste(img_dir,\"violin_f1_\",ftype,\"_\",title,\"_\",x_list[2],\".jpg\",sep=\"\"))\n",
    "    \n",
    "    \n",
    "    p_3 = plot_violin_hmo(F1_in,id_in,x_list[3],y_list[3],id_in,fac_in)\n",
    "    annotate_figure(p_3, #https://rpkgs.datanovia.com/ggpubr/reference/annotate_figure.html\n",
    "                    top = text_grob(title, color = \"blue\", face = \"bold\", size = 14))\n",
    "    ggsave(paste(img_dir,\"violin_f1_\",ftype,\"_\",title,\"_\",x_list[3],\".jpg\",sep=\"\"))\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h3> Relative Abunandance <h3>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "cal_relabun <- function(df_in){\n",
    "    temp = df_in\n",
    "    temp$Total = rowSums(df_in[,1:(ncol(df_in)-3)]) #increased from -1 to -3\n",
    "    col_list = colnames(temp[,1:(ncol(temp)-4)])\n",
    "    \n",
    "    for (row in rownames(temp)){\n",
    "        for (col in col_list){\n",
    "            val = temp[row,col]\n",
    "            if(val==0){\n",
    "                next\n",
    "            } else{\n",
    "                temp[row,col] = val/temp[row,\"Total\"]\n",
    "            }\n",
    "        }\n",
    "    }\n",
    "    return(temp)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h2> Expected to Observed Correlation <h2>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "cal_EO_kit <- function(df_in,contdf_in,bench_in,vari_in,title_in){\n",
    "    count=1\n",
    "    plot_list = list()\n",
    "    EOdf_merged = data.frame()\n",
    "    eo_out = data.frame()\n",
    "\n",
    "    #For each of the controls in the cleaned df\n",
    "    for (cont in unique(df_in$Subject.ID)){\n",
    "        \n",
    "        #Create df of taxa\n",
    "        taxdf = create_taxadf(cont,contdf_in)\n",
    "        taxa_list=taxdf$taxa.id #list of expected taxa for this control\n",
    "        \n",
    "        #Create list of samples that match the control\n",
    "        sample_list = filter(RAdf, Subject.ID == cont)$Sample_name\n",
    "        EOdf= data.frame()\n",
    "        \n",
    "        #for each expected taxa create df of sample name, taxa id, control, exp RA and obs RA\n",
    "        for (taxa in taxa_list){\n",
    "            for(sample in sample_list){\n",
    "                EOdf[nrow(EOdf)+1,\"Sample_name\"] = sample\n",
    "                EOdf[nrow(EOdf),\"taxa.id\"] = taxa\n",
    "                EOdf[nrow(EOdf),\"Subject.ID\"] = cont\n",
    "                EOdf[nrow(EOdf),\"exp\"] = taxdf[taxa,cont]\n",
    "                EOdf[nrow(EOdf),\"bench\"] = bench_in\n",
    "\n",
    "                if(is.null(df_in[sample,taxa])){ #if that sample doesn't have that taxa, obs is 0\n",
    "                     EOdf[nrow(EOdf),\"obs\"] = 0\n",
    "                } else{\n",
    "                    EOdf[nrow(EOdf),\"obs\"] = df_in[sample,taxa] #otherwise observed val saves\n",
    "                }\n",
    "            }\n",
    "        }\n",
    "        EOdf_merged = rbind(EOdf_merged,EOdf) #Merge dfs\n",
    "        \n",
    "        #Create scatter plots\n",
    "        plot_list[[count]] = plot_EO(EOdf,cont)\n",
    "        count=count+1\n",
    "        \n",
    "        #Calc stats\n",
    "        EOdf[is.na(EOdf)] = 0\n",
    "        print(bench_in)\n",
    "        temp = cal_EO_stattest(EOdf,cont,bench_in) #calc stats - answers Q is expected sig different from observed per cont?\n",
    "        eo_out = rbind(eo_out,temp) #merge df       \n",
    "    }\n",
    "    \n",
    "    #Plot controls\n",
    "    plot_length = as.numeric(length(plot_list))\n",
    "    if(plot_length==4){\n",
    "        fig = ggarrange(plot_list[[1]], plot_list[[2]], plot_list[[3]], plot_list[[4]],\n",
    "                            labels = c(\"A\", \"B\", \"C\", \"D\"), #edit number of controls\n",
    "                            ncol = 2, nrow = 3)  \n",
    "    } else if(plot_length==3){\n",
    "        fig = ggarrange(plot_list[[1]], plot_list[[2]], plot_list[[3]],\n",
    "                        labels = c(\"A\", \"B\", \"C\"), #edit number of controls\n",
    "                        ncol = 2, nrow = 3)\n",
    "    } else if(plot_length==2){\n",
    "        fig = ggarrange(plot_list[[1]], plot_list[[2]], \n",
    "                        labels = c(\"A\", \"B\"), #edit number of controls\n",
    "                        ncol = 2, nrow = 3)\n",
    "    } else{\n",
    "        fig = ggarrange(plot_list[[1]],\n",
    "                        labels = c(\"A\"), #edit number of controls\n",
    "                        ncol = 2, nrow = 3)\n",
    "\n",
    "    }\n",
    "    annotate_figure(fig, top = text_grob(title_in, color = \"blue\", face = \"bold\", size = 14))\n",
    "    ggsave(paste(img_dir,\"scatter_ext\",vari_in,\"_\",bench_in,\".jpg\",sep=\"\"))\n",
    "    \n",
    "    return(eo_out)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "cal_EO_stattest <- function(df_in,cont_in,bench_in){\n",
    "    my_data=subset(df_in,Subject.ID==cont_in)\n",
    "    df_out=data.frame()\n",
    "    \n",
    "    df_out[nrow(df_out)+1,\"bmark\"] = bench_in\n",
    "    df_out[nrow(df_out),\"cont\"] = cont_in\n",
    "    df_out[nrow(df_out),\"shapiro.p\"] = shapiro.test(my_data$obs)$p.value\n",
    "    df_out[nrow(df_out),\"shapiro.sig\"] = if(df_out[nrow(df_out),\"shapiro.p\"]<0.05){\"sig\"}else{\"ns\"}\n",
    "    df_out[nrow(df_out),\"spear.p\"] = cor.test(my_data$obs,my_data$exp,method = \"spearman\")$p.value\n",
    "    df_out[nrow(df_out),\"spear.rho\"] = cor.test(my_data$obs,my_data$exp,method = \"spearman\")$estimate[[1]]\n",
    "    df_out[nrow(df_out),\"spear.sig\"] = if(df_out[nrow(df_out),\"spear.p\"]<0.05){\"sig\"}else{\"ns\"}\n",
    "    df_out[nrow(df_out),\"pears.val\"] = cor.test(my_data$obs,my_data$exp,method = \"pearson\")$estimate[[1]]\n",
    "    df_out[nrow(df_out),\"pears.p\"] = cor.test(my_data$obs,my_data$exp,method = \"pearson\")$p.value\n",
    "    df_out[nrow(df_out),\"pears.sig\"] = if(df_out[nrow(df_out),\"pears.p\"]<0.05){\"sig\"}else{\"ns\"}\n",
    "    return(df_out)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "create_taxadf <- function(cont,contdf_in){\n",
    "    taxdf=filter(contdf_in, get(cont) !=0) #create df of taxa related to control\n",
    "    taxdf$taxa.id=paste(taxdf$Family,taxdf$Genus,sep=\"\") #create fam/genus col\n",
    "    taxdf = taxdf[,c(\"taxa.id\",cont),drop=FALSE] #remove all cols but the merged and control col\n",
    "    taxdf = aggregate(get(cont) ~ taxa.id, data = taxdf, sum) #collapse df for any dups    \n",
    "    names(taxdf)[2] = cont\n",
    "    rownames(taxdf) = taxdf$taxa.id\n",
    "    return(taxdf)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "cal_EO_cont <- function(df_in,contdf_in,taxdf_in,cont_in,title_in){\n",
    "    count=1\n",
    "    plot_list = list()\n",
    "    EOdf_merged = data.frame()\n",
    "    eo_out = data.frame()\n",
    "    \n",
    "    #Create list of samples that match the control\n",
    "    sample_list = filter(df_in, Subject.ID == cont_in)$Sample_name\n",
    "    EOdf= data.frame()\n",
    "    taxa_list=taxdf_in$taxa.id\n",
    "    \n",
    "    #for (kit in unique(df_in$Ext.Kit)){\n",
    "        for (taxa in taxa_list){\n",
    "            for(sample in sample_list){\n",
    "                EOdf[nrow(EOdf)+1,\"Sample_name\"] = sample\n",
    "                EOdf[nrow(EOdf),\"taxa.id\"] = taxa\n",
    "                EOdf[nrow(EOdf),\"Subject.ID\"] = cont\n",
    "                EOdf[nrow(EOdf),\"exp\"] = taxdf_in[taxa,cont]\n",
    "                EOdf[nrow(EOdf),\"bench\"] = df_in[sample,\"Ext.Kit\"]\n",
    "\n",
    "                if(is.null(df_in[sample,taxa])){\n",
    "                    next\n",
    "                } else{\n",
    "                    EOdf[nrow(EOdf),\"obs\"] = df_in[sample,taxa]\n",
    "                }\n",
    "            }\n",
    "        }\n",
    "    #}\n",
    "\n",
    "    for(kit in unique(EOdf$bench)){\n",
    "        EOdf_sub = subset(EOdf,bench==kit)\n",
    "        #Create scatter plots\n",
    "        if(nrow(EOdf_sub)>1){\n",
    "            plot_list[[count]] = plot_EO(EOdf_sub,cont_in,kit)\n",
    "        count=count+1\n",
    "        }\n",
    "    }\n",
    "    EOdf_merged = rbind(EOdf_merged,EOdf) #Merge dfs\n",
    "\n",
    "    #Calc stats\n",
    "    #EOdf[is.na(EOdf)] = 0\n",
    "    #temp = cal_EO_stattest(EOdf,cont,bench_in)\n",
    "    #eo_out = rbind(eo_out,temp) #merge df \n",
    "    plot_EO_save(plot_list,title_in,\"std\",cont_in)\n",
    "    \n",
    "    return(eo_out)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_EO <- function(df_in,cont_in,title=\"\"){\n",
    "    if(title==\"\"){\n",
    "        title=cont_in\n",
    "        my_data=subset(df_in,Subject.ID==cont_in)\n",
    "    } else{\n",
    "        my_data=df_in\n",
    "    }\n",
    "    \n",
    "    p = ggscatter(my_data, x = \"obs\", y = \"exp\", \n",
    "              add = \"reg.line\", conf.int = TRUE, \n",
    "              cor.coef = TRUE, cor.method = \"pearson\",\n",
    "              xlab = \"Expected Rel \\n Abunance (%)\", ylab = \"Observed Rel \\n Abun (%)\") +\n",
    "        ggtitle(paste(\"Scatter plot of Rel Abun:\\n\",title,sep=\"\"))\n",
    "    return(p)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_EO_save <- function(plot_list,title_in,vari_in,bench_in){\n",
    "    #Plot controls\n",
    "    plot_length = as.numeric(length(plot_list))\n",
    "    if(plot_length==7){\n",
    "        fig = ggarrange(plot_list[[1]], plot_list[[2]], plot_list[[3]], plot_list[[4]],plot_list[[5]],plot_list[[6]],plot_list[[7]],\n",
    "                            labels = c(\"A\", \"B\", \"C\", \"D\",\"E\",\"F\",\"G\"), #edit number of controls\n",
    "                            ncol = 2, nrow = 4)  \n",
    "    }else if(plot_length==6){\n",
    "        fig = ggarrange(plot_list[[1]], plot_list[[2]], plot_list[[3]], plot_list[[4]],plot_list[[5]],plot_list[[6]],\n",
    "                            labels = c(\"A\", \"B\", \"C\", \"D\",\"E\",\"F\"), #edit number of controls\n",
    "                            ncol = 2, nrow = 3)  \n",
    "    }else if(plot_length==5){\n",
    "        fig = ggarrange(plot_list[[1]], plot_list[[2]], plot_list[[3]], plot_list[[4]],plot_list[[5]],\n",
    "                            labels = c(\"A\", \"B\", \"C\", \"D\",\"E\"), #edit number of controls\n",
    "                            ncol = 2, nrow = 3)  \n",
    "    }else if(plot_length==4){\n",
    "        fig = ggarrange(plot_list[[1]], plot_list[[2]], plot_list[[3]], plot_list[[4]],\n",
    "                            labels = c(\"A\", \"B\", \"C\", \"D\"), #edit number of controls\n",
    "                            ncol = 2, nrow = 3)  \n",
    "    } else if(plot_length==3){\n",
    "        fig = ggarrange(plot_list[[1]], plot_list[[2]], plot_list[[3]],\n",
    "                        labels = c(\"A\", \"B\", \"C\"), #edit number of controls\n",
    "                        ncol = 2, nrow = 3)\n",
    "    } else if(plot_length==2){\n",
    "        fig = ggarrange(plot_list[[1]], plot_list[[2]], \n",
    "                        labels = c(\"A\", \"B\"), #edit number of controls\n",
    "                        ncol = 2, nrow = 3)\n",
    "    } else{\n",
    "        fig = ggarrange(plot_list[[1]],\n",
    "                        labels = c(\"A\"), #edit number of controls\n",
    "                        ncol = 2, nrow = 3)\n",
    "\n",
    "    }\n",
    "    annotate_figure(fig, top = text_grob(title_in, color = \"blue\", face = \"bold\", size = 14))\n",
    "    ggsave(paste(img_dir,\"scatter_ext\",vari_in,\"_\",bench_in,\".jpg\",sep=\"\"))\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h2>Dissimilarity <h2>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Create averaged relative abundance df\n",
    "create_av_RAdf <- function(RAdf_in,bench_in,var_in,status=\"std\"){\n",
    "    if(status==\"std\"){\n",
    "        tmp = aggregate(RAdf_in[,1:(ncol(RAdf_in)-5)], list(RAdf_in$Subject.ID), mean) #increase from 3 to 5\n",
    "        tmp$bmark=bench_in\n",
    "        tmp$bmark2 = var_in\n",
    "    } else{\n",
    "        tmp = aggregate(RAdf_in[,1:(ncol(RAdf_in)-5)], list(RAdf_in$Subject.ID,RAdf_in$Homo.Status), mean)\n",
    "        tmp$bmark=bench_in\n",
    "        tmp$bmark2 = var_in\n",
    "    }\n",
    "    return(tmp)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h3> Merged metrics <h3>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "#Add expected values to RA average dfs\n",
    "RAdf_addcontrols<-function(cont_in,df_in,hmo_in=\"std\"){\n",
    "    \n",
    "    tmp=get(df_in)\n",
    "    cont_list = unique(tmp$Group.1)\n",
    "    cont_leng = length(cont_list) #chnaged to Group.1\n",
    "    \n",
    "    for(i in 1:cont_leng){\n",
    "        cont = as.character(cont_list[i])\n",
    "        taxadf = create_taxadf(cont,get(cont_in))\n",
    "        tmp_count = nrow(tmp)+1\n",
    "        \n",
    "        if(hmo_in==\"hmo\"){\n",
    "            for(tx_row in 1:nrow(taxadf)){\n",
    "                tmp[tmp_count,\"Group.1\"]=cont\n",
    "                tmp[tmp_count,taxadf[tx_row,\"taxa.id\"]] = taxadf[tx_row,cont]\n",
    "                tmp[tmp_count,\"bmark\"]=\"expected\"\n",
    "                tmp[tmp_count,\"bmark2\"]=tmp[i,\"bmark2\"]\n",
    "            }\n",
    "        } else{\n",
    "            for(tx_row in 1:nrow(taxadf)){\n",
    "                tmp[tmp_count,\"Group.1\"]=cont\n",
    "                tmp[tmp_count,taxadf[tx_row,\"taxa.id\"]] = taxadf[tx_row,cont]\n",
    "                tmp[tmp_count,\"bmark\"]=\"expected\"\n",
    "                tmp[tmp_count,\"bmark2\"]=tmp[i,\"bmark2\"]\n",
    "                tmp[tmp_count,\"Group.2\"]=\"cnt\"\n",
    "            } \n",
    "        }\n",
    "        \n",
    "    }\n",
    "    tmp[is.na(tmp)] = 0 \n",
    "    return(tmp)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "calc_dist <- function(RA_in,CONTdf_in,hmo_in=\"std\"){\n",
    "    if(hmo_in==\"hmo\"){\n",
    "        for(cnt in 1:nrow(RA_in)){\n",
    "            if(is.na(RA_in[cnt,\"Group.2\"])){\n",
    "                new_name=paste(RA_in[cnt,\"Group.1\"],RA_in[cnt,\"bmark\"],RA_in[cnt,\"bmark2\"],sep=\"-\")\n",
    "                RA_in[cnt,\"new_name\"] = new_name\n",
    "            }else{\n",
    "                new_name = paste(RA_in[cnt,\"Group.1\"],RA_in[cnt,\"bmark\"],RA_in[cnt,\"Group.2\"],sep=\"-\")\n",
    "                RA_in[cnt,\"new_name\"] = new_name\n",
    "            }\n",
    "        }\n",
    "        rownames(RA_in)=RA_in$new_name\n",
    "        RA_in = subset(RA_in, select = -c(new_name))\n",
    "    } else{\n",
    "        rownames(RA_in)=paste(RA_in$Group.1,RA_in$bmark,RA_in$bmark2,sep=\"-\")\n",
    "    }\n",
    "    RA_in = subset(RA_in, select = -c(bmark2,Group.1,bmark,Group.2))\n",
    "\n",
    "    taxa_list<-unique(paste(CONTdf_in$Family,CONTdf_in$Genus,sep=\"\"))\n",
    "    df_clean = RA_in[,colnames(RA_in) %in% taxa_list] #remove cols not expected\n",
    "\n",
    "    df_out = vegdist(df_clean,method=\"bray\")   \n",
    "    return(df_out)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_dist_cluster <- function(df_in,ref_in){\n",
    "    fname = paste(img_dir,\"hclust_summary_bc_allgroups_\",ref_in,\".png\",sep=\"\")\n",
    "    png(fname,width = 900, height = 600)\n",
    "    p = plot(hclust(df_in), cex = .9, main = \"All Groups - Bray-Curtis Dissimilarity\",\n",
    "             xlab=\"Bioinformatic Variables\", sub=\"\")\n",
    "    dev.off()\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_dist_violin <- function(df_in,ref_in){\n",
    "    df_in = as.data.frame(as.matrix(df_in))\n",
    "    df_plot=data.frame()\n",
    "    \n",
    "    for (row in rownames(df_in)){\n",
    "        splits = strsplit(row,\"-\")\n",
    "        if(splits[[1]][2]==\"expected\"){\n",
    "            next\n",
    "        }else{\n",
    "            df_plot[nrow(df_plot)+1,\"dist\"]=df_in[row,paste(splits[[1]][1],\"expected\",\"std\",sep=\"-\")]\n",
    "            df_plot[nrow(df_plot),\"name\"]=paste(splits[[1]][2],splits[[1]][3],sep=\"-\")\n",
    "        }\n",
    "    }\n",
    "    \n",
    "    p = ggplot(df_plot, aes(x=name, y=dist, color=name)) + \n",
    "            geom_violin(trim=FALSE) + \n",
    "            labs(color = df_plot$name) +\n",
    "            theme(axis.text.x = element_text(angle = 90, hjust = 1),\n",
    "                  axis.title.x = element_blank(), \n",
    "                 legend.position = \"none\") + \n",
    "            xlab(\"Benchmark ID\") +\n",
    "            ylab(\"Bray-Curtis Dissimilarity\") +\n",
    "            stat_summary(fun.data=mean_sdl, color=\"red\")\n",
    "    \n",
    "    \n",
    "    fig = ggarrange(p, \n",
    "              #labels = c(\"A\"),\n",
    "              #top = as.character(title_in),\n",
    "              #common.legend = TRUE#,\n",
    "              ncol = 1, nrow = 1)\n",
    "    annotate_figure(fig, top = text_grob(\"Bray-Curtis Dissimilarity\", color = \"blue\", face = \"bold\", size = 14)\n",
    "    )\n",
    "    ggsave(paste(img_dir,\"violin_bc_summary_allgroups_\",ref_in,\".jpg\",sep=\"\"))\n",
    "    return(df_plot)\n",
    "\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_dist_violin_hmo <- function(df_in,ref_in){\n",
    "    df_in = as.data.frame(as.matrix(df_in))\n",
    "    df_plot=data.frame()\n",
    "    \n",
    "    for (row in rownames(df_in)){\n",
    "        splits = strsplit(row,\"-\")\n",
    "        if(splits[[1]][2]==\"expected\"){\n",
    "            next\n",
    "        }else{\n",
    "            df_plot[nrow(df_plot)+1,\"dist\"]=df_in[row,paste(splits[[1]][1],\"expected\",\"std\",sep=\"-\")]\n",
    "            df_plot[nrow(df_plot),\"name\"]=splits[[1]][1]\n",
    "            df_plot[nrow(df_plot),\"Ext.Kit\"]=splits[[1]][2]\n",
    "            df_plot[nrow(df_plot),\"Homo.Status\"]=splits[[1]][3]\n",
    "        }\n",
    "    }\n",
    "    \n",
    "    p = ggplot(df_plot, aes(x=name, y=dist, color=Homo.Status)) + \n",
    "            geom_violin(trim=FALSE) + \n",
    "            facet_grid(Homo.Status ~ Ext.Kit) +\n",
    "            labs(color = df_plot$name) +\n",
    "            rotate_x_facet_text(angle = 90, align = 0.5) +\n",
    "            theme(axis.text.x = element_text(angle = 90, hjust = 1),\n",
    "                  axis.title.x = element_blank(), \n",
    "                 legend.position = \"none\") + \n",
    "            xlab(\"Benchmark ID\") +\n",
    "            ylab(\"Bray-Curtis Dissimilarity\") +\n",
    "            stat_summary(fun.data=mean_sdl, color=\"red\")\n",
    "    \n",
    "    \n",
    "    fig = ggarrange(p, \n",
    "              ncol = 1, nrow = 1)\n",
    "    annotate_figure(fig, top = text_grob(\"Bray-Curtis Dissimilarity\", color = \"blue\", face = \"bold\", size = 14)\n",
    "    )\n",
    "    ggsave(paste(img_dir,\"violin_bc_summary_allgroups_\",ref_in,\".jpg\",sep=\"\"))\n",
    "    return(df_plot)\n",
    "\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_corrmatrix <- function(df_in,contdf_in,sort_in){\n",
    "    tmp = calc_dist(df_in[order(df_in[,sort_in]),],contdf_in)\n",
    "    tmp = as.matrix(tmp)\n",
    "    remove_list = list()\n",
    "    for (rows in rownames(tmp)){\n",
    "        if(strsplit(rows,\"-\")[[1]][2]==\"expected\"){\n",
    "            remove_list=append(remove_list,rows)\n",
    "        }\n",
    "    }\n",
    "    tmp_clean = tmp[!rownames(tmp) %in% remove_list,!colnames(tmp) %in% remove_list ]\n",
    "    \n",
    "    #cal sig\n",
    "    p.mat <- cor.mtest(tmp_clean)$p\n",
    "    \n",
    "    #image with stats\n",
    "    png(file = paste(img_dir,\"distmatrix_ext_sig_\",sort_in,\".jpg\",sep=\"\"))\n",
    "    corrplot(as.matrix(tmp_clean), type = \"upper\",col=brewer.pal(n=8, name=\"RdYlBu\"),\n",
    "        title = \"Correlation of  Dissimilarity (R): Ext Controls\",\n",
    "        tl.col = \"black\",\n",
    "        mar = c(0,0,1,0), number.cex = 0.5, number.digits = 2,\n",
    "        p.mat = p.mat, sig.level = 0.05, insig=\"blank\")\n",
    "    dev.off()\n",
    "    \n",
    "    #image without stats\n",
    "    png(file = paste(img_dir,\"distmatrix_ext_\",sort_in,\".jpg\",sep=\"\"))\n",
    "    \n",
    "    corrplot(as.matrix(tmp_clean),type=\"upper\",col=brewer.pal(n=8, name=\"RdYlBu\"),\n",
    "        title = \"Correlation of  Dissimilarity (R): Ext Controls\",\n",
    "        tl.col = \"black\",\n",
    "        mar = c(0,0,1,0), number.cex = 0.5, number.digits = 2)\n",
    "    dev.off()\n",
    "}\n",
    "#Ref https://cran.r-project.org/web/packages/corrplot/vignettes/corrplot-intro.html"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h2> Statistics <h2>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "tstat_paired_f1 <- function(f1_in,subgroup_in){\n",
    "    \n",
    "    df_out = data.frame()\n",
    "    \n",
    "    cal_list = c(\"F1\")\n",
    "    \n",
    "    for (cal in cal_list){\n",
    "    \n",
    "        comb_list = list()\n",
    "        for(kit in unique(f1_in[,subgroup_in])){\n",
    "            if(sum(f1_in[,subgroup_in] == kit)>1)\n",
    "                comb_list=append(comb_list,kit)\n",
    "        }\n",
    "        \n",
    "        if(length(comb_list)>1){\n",
    "            comb_list = combn(comb_list,2)\n",
    "            count=1\n",
    "        \n",
    "            for (i in 1:(length(comb_list)/2)){\n",
    "\n",
    "                tstatdf=data.frame()\n",
    "                var1 = as.character(comb_list[count])\n",
    "                var2 = as.character(comb_list[count+1])\n",
    "                count=count+2\n",
    "\n",
    "                df = f1_in %>%\n",
    "                    filter(f1_in[,subgroup_in] == var1 | f1_in[,subgroup_in] == var2) %>%\n",
    "                    select(subgroup_in, cal)\n",
    "\n",
    "                tmp = t.test(get(cal) ~ get(subgroup_in), data = df)\n",
    "                tstatdf[nrow(tstatdf)+1,\"level\"] = cal\n",
    "                tstatdf[nrow(tstatdf),\"comparison\"] = paste(var1,var2,sep=\"_\")\n",
    "                tstatdf[nrow(tstatdf),\"pval\"] = tmp$p.value\n",
    "                if(tmp$p.value<0.05){\n",
    "                    tstatdf[nrow(tstatdf),\"sig\"] = \"sig\"\n",
    "                } else{\n",
    "                    tstatdf[nrow(tstatdf),\"sig\"] = \"ns\"\n",
    "                }\n",
    "                df_out = bind_rows(df_out,tstatdf)\n",
    "            }        \n",
    "            \n",
    "        } else{\n",
    "            next;\n",
    "        }\n",
    "    }\n",
    "        \n",
    "    return(df_out)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "adonis_bc <- function(df_in){\n",
    "\n",
    "    df_out=data.frame()\n",
    "    \n",
    "    st = adonis(df_in[,1] ~ name,\n",
    "                data=df_in, permutations = 99, method=\"bray\")\n",
    "    df_out[nrow(df_out)+1,\"comparison\"] = \"bray-curtis\"\n",
    "    df_out[nrow(df_out),\"pval\"] = as.numeric(st$aov.tab$\"Pr(>F)\"[1])\n",
    "        \n",
    "    if(df_out[nrow(df_out),\"pval\"]<0.05){\n",
    "        df_out[nrow(df_out),\"sig\"] = \"sig\"\n",
    "    } else{\n",
    "        df_out[nrow(df_out),\"sig\"] = \"ns\"\n",
    "    }   \n",
    "    \n",
    "    dist_s = vegdist(subset(dissdf_s,select = -c(bmark,Group.1,bmark2)),na.rm=TRUE)\n",
    "\n",
    "        \n",
    "    df_out[nrow(df_out)+1,\"homo_s\"] = anova(betadisper(dist_s,dissdf_s$bmark))$\"Pr(>F)\"[1]\n",
    "    \n",
    "    return(df_out)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h3> Rel Abun Plots <h3>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "addSmallLegend <- function(myPlot, pointSize = 1, textSize = 6, spaceLegend = 0.1) {\n",
    "    myPlot +\n",
    "        guides(shape = guide_legend(override.aes = list(size = pointSize)),\n",
    "               color = guide_legend(override.aes = list(size = pointSize))) +\n",
    "        theme(legend.title = element_text(size = textSize), \n",
    "              legend.text  = element_text(size = textSize),\n",
    "              legend.key.size = unit(spaceLegend, \"lines\"))\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "phylum_colors <- c(\n",
    "  \"#CBD588\", \"#5F7FC7\", \"orange\",\"#DA5724\", \"#508578\", \"#CD9BCD\",\n",
    "   \"#AD6F3B\", \"#673770\",\"#D14285\", \"#652926\", \"#C84248\", \n",
    "  \"#8569D5\", \"#5E738F\",\"#D1A33D\", \"#8A7C64\", \"#599861\"\n",
    ")\n",
    "\n",
    "rel_abun_plot <- function(df_in,x_in,y_in,x_lab_in,level_in,title_in,file_name){\n",
    "    tmp = as.data.frame(unique(df_in[,c(\"Sample\",\"Ext.Kit\")]))\n",
    "    \n",
    "    p = ggplot(df_in, aes(x = get(x_in), y =get(y_in), fill = get(level_in))) + \n",
    "        theme(legend.position = \"bottom\", legend.key.width = unit(0.2,\"cm\"),\n",
    "            axis.text.x = element_text(angle = 90, hjust = 1), legend.text = element_text(size = 8)) +\n",
    "            facet_grid(Homo.Status~ .) +\n",
    "            scale_fill_manual(values = phylum_colors) +\n",
    "            geom_bar(stat = \"identity\") +\n",
    "            xlab(x_in) +\n",
    "            ylab(\"Relative Abundance \\n\") +\n",
    "            ggtitle(title_in) +\n",
    "            labs(fill = level_in)\n",
    "    addSmallLegend(p)\n",
    "    ggsave(paste(img_dir,\"\",file_name,\".jpg\",sep=\"\"))\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "adonis_rel <- function(df_in){\n",
    "    df_out=data.frame()\n",
    "    \n",
    "    ext_list = unique(df_in$bmark)[-length(unique(df_in$bmark))]\n",
    "    \n",
    "    #analysis of each kit, to determine the sig of a homo on the control distance\n",
    "    for (kit in ext_list){\n",
    "        df_kit = subset(df_in,bmark==kit) #df only with kit\n",
    "        cont_list = unique(df_kit$Group.1)\n",
    "        \n",
    "        for (cont in cont_list){\n",
    "            df_cont = subset(df_kit,Group.1==cont) #df only with kit, and control\n",
    "            df_cont = subset(df_cont,select=c(\"val\",\"Group.2\"))\n",
    "            homo_list = unique(df_cont$Group.2)\n",
    "            \n",
    "            if(length(homo_list)>1){ #if there is more than one control, comparison can be performed\n",
    "                \n",
    "                st = adonis(val ~ Group.2,\n",
    "                    data=df_cont, permutations = 99, method=\"bray\")\n",
    "                df_out[nrow(df_out)+1,\"stat\"] = \"bray-curtis\"\n",
    "                df_out[nrow(df_out),\"measure\"] = paste(kit,cont,sep=\"-\")\n",
    "                df_out[nrow(df_out),\"pval\"] = as.numeric(st$aov.tab$\"Pr(>F)\"[1])\n",
    "                \n",
    "                if(df_out[nrow(df_out),\"pval\"]>0.05 | is.na(df_out[nrow(df_out),\"pval\"])){\n",
    "                    df_out[nrow(df_out),\"sig\"] = \"ns\"\n",
    "                } else{\n",
    "                    df_out[nrow(df_out),\"sig\"] = \"sig\"\n",
    "                }\n",
    "            }\n",
    "        }\n",
    "    }\n",
    "    \n",
    "    #dist_s = vegdist(subset(dissdf_s,select = -c(bmark,Group.1,bmark2)),na.rm=TRUE)   \n",
    "    #df_out[nrow(df_out)+1,\"homo_s\"] = anova(betadisper(dist_s,dissdf_s$bmark))$\"Pr(>F)\"[1]\n",
    "    \n",
    "    return(df_out)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h3> Homog Stats <h3>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "adonis_bc_metadf <- function(df_in){\n",
    "    \n",
    "    #create metadata df related to distance names\n",
    "    metadf = as.data.frame(as.matrix(df_in))\n",
    "    \n",
    "    #Push out results\n",
    "    for (rows in rownames(metadf)){\n",
    "        splitname = strsplit(rows,\"-\")\n",
    "        metadf[rows,\"Subject.ID\"] = splitname[[1]][1]\n",
    "        metadf[rows,\"Ext.Kit\"] = splitname[[1]][2]\n",
    "        metadf[rows,\"Homo.Status\"] = splitname[[1]][3]\n",
    "        metadf[rows,\"Ext.Robotics\"] = as.character(metadf_clean[metadf[rows,\"Ext.Kit\"],\"Ext.Robotics\"])\n",
    "        metadf[rows,\"IR\"] = as.character(metadf_clean[metadf[rows,\"Ext.Kit\"],\"IR\"])\n",
    "        metadf[rows,\"Mag.Col\"] = as.character(metadf_clean[metadf[rows,\"Ext.Kit\"],\"Mag.Col\"])\n",
    "    }\n",
    "    metadf = subset(metadf,select=c(Subject.ID,Ext.Kit,Homo.Status,Ext.Robotics,IR,Mag.Col))\n",
    "    return(metadf)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "adonis_bc_stat <- function(cont_in,df_in,df_out,stat_in,strat_in=\"\",stat_label){\n",
    "    \n",
    "    #Create metadata table\n",
    "    metadf = adonis_bc_metadf(df_in)\n",
    "    \n",
    "    #Run adonis by control\n",
    "    if(strat_in!=\"\"){\n",
    "        st = adonis(df_in ~ Ext.Kit, data=metadf, permutations = 99, strata=metadf[,strat_in])\n",
    "    } else{\n",
    "        st = adonis(df_in ~ get(stat_in), data=metadf, permutations = 99)\n",
    "    }\n",
    "    #save data to df\n",
    "    df_out[nrow(df_out)+1,\"stat\"] = stat_label\n",
    "    df_out[nrow(df_out),\"measure\"] = cont_in\n",
    "    df_out[nrow(df_out),\"pval\"] = as.numeric(st$aov.tab$\"Pr(>F)\"[1])\n",
    "\n",
    "    if(df_out[nrow(df_out),\"pval\"]>0.05 | is.na(df_out[nrow(df_out),\"pval\"])){\n",
    "        df_out[nrow(df_out),\"sig\"] = \"ns\"\n",
    "    } else{\n",
    "        df_out[nrow(df_out),\"sig\"] = \"sig\"\n",
    "    }\n",
    "    return(df_out)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "adonis_bc_hmo <- function(df_in){\n",
    "    df_out=data.frame()\n",
    "\n",
    "    for (cont in unique(df_in$Group.1)){\n",
    "        df_sub = df_in\n",
    "\n",
    "        df_sub$Group.1 <- as.character(df_sub$Group.1) #cont name\n",
    "        df_sub$Group.2 <- as.character(df_sub$Group.2) #hmo method\n",
    "        df_sub[is.na(df_sub)] = \"std\" #will change NA in Group.2 to std \n",
    "\n",
    "        df_clean = subset(df_sub,Group.1==cont & bmark!=\"expected\") #remove expect control values\n",
    "\n",
    "        #Cal dissimilarity between kits stratified by homo method\n",
    "        df_cont = subset(df_clean)\n",
    "        tmp = calc_dist(df_cont,CONTdf_s,\"hmo\")\n",
    "        df_out = adonis_bc_stat(cont,tmp,df_out,\"Ext.Kit\",\"Homo.Status\",\"all kits,Std/varied hmo\")\n",
    "\n",
    "        #Cal dissimilarity between kits, standard method only\n",
    "        df_cont = subset(df_clean,Group.2==\"std\")\n",
    "        tmp = calc_dist(df_cont,CONTdf_s,\"hmo\")\n",
    "        df_out = adonis_bc_stat(cont,tmp,df_out,\"Ext.Kit\",stat_label=\"all kits,Std hmo\")\n",
    "        \n",
    "        #Cal dissimilarity between kits, standard method only, automated platform\n",
    "        df_cont = subset(df_clean,Group.2==\"std\" | Group.2==\"Altered\")\n",
    "        tmp = calc_dist(df_cont,CONTdf_s,\"hmo\")\n",
    "        df_out = adonis_bc_stat(cont,tmp,df_out,\"Ext.Robotics\", \"Ext.Kit\", stat_label=\"all kits,auto\")\n",
    "\n",
    "        #Cal dissimilarity between altered methods\n",
    "        for (kit in unique(RAdf_oe$bmark)[-length(unique(RAdf_oe$bmark))]){ #Dont want to include \"expected\" \n",
    "            df_kit = subset(df_clean,bmark==kit)\n",
    "\n",
    "            if(nrow(df_kit)>1){\n",
    "                \n",
    "                tmp = calc_dist(df_kit,CONTdf_s,\"hmo\")\n",
    "                df_out = adonis_bc_stat(cont,tmp,df_out,\"Homo.Status\",stat_label=paste(kit,\"-hmo\",sep=\"\"))\n",
    "            }\n",
    "        }\n",
    "        \n",
    "        #Cal dissimilarity between IR methods\n",
    "        df_cont = subset(df_clean)\n",
    "        tmp = calc_dist(df_cont,CONTdf_s,\"hmo\")\n",
    "        df_out = adonis_bc_stat(cont,tmp,df_out,\"Ext.Kit\",\"IR\",\"all kits,Std/varied hmo IR\")\n",
    "        \n",
    "        #Cal dissimilarity between Mag.Col methods\n",
    "        df_cont = subset(df_clean)\n",
    "        tmp = calc_dist(df_cont,CONTdf_s,\"hmo\")\n",
    "        df_out = adonis_bc_stat(cont,tmp,df_out,\"Ext.Kit\",\"Mag.Col\",\"all kits,Std/varied hmo MC\")\n",
    "        \n",
    "        df_sub = subset(RAdf_oe,bmark==kit)\n",
    "    }\n",
    "\n",
    "    return(df_out)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "cal_ttest_hmo <-function(df_in){\n",
    "    df_out = data.frame()\n",
    "    \n",
    "    #for each of the methods\n",
    "    for(kit in unique(df_in$bmark)){ #each \n",
    "        tmp = subset(df_in,bmark==kit)\n",
    "\n",
    "        #for each of the controls\n",
    "        for(cont in unique(tmp$Subject.ID)){\n",
    "            tmp2 = subset(tmp,Subject.ID==cont)\n",
    "\n",
    "            #if there is more than one hmo instance, perform t-test\n",
    "            if(sum(tmp2$bmark2==\"hmo\")>1){\n",
    "                df_out[nrow(df_out)+1,\"bmark\"] = kit\n",
    "                df_out[nrow(df_out),\"level\"] = \"F1\"\n",
    "                df_out[nrow(df_out),\"cont\"] = cont\n",
    "                df_out[nrow(df_out),\"pval\"] = t.test(F1~bmark2,tmp)$p.value\n",
    "            }\n",
    "        }\n",
    "    }\n",
    "    return(df_out)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h3> Alpha and Beta <h3>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Function for plotting alpha diversity\n",
    "alpha_meas <- function(phy_in,alpha_list,sub_in,col_in,title,file_name,facet_input=\"\"){\n",
    "    \n",
    "    \n",
    "    if(facet_input!=\"\"){\n",
    "        p = plot_richness(phy_in, sub_in, col_in, measures=alpha_list)\n",
    "        p_final = p + facet_wrap(~get(facet_input))\n",
    "    } else{ \n",
    "        p_final <- plot_richness(phy_in, sub_in, col_in, measures=alpha_list)\n",
    "    }\n",
    "    \n",
    "    fig = p_final + \n",
    "        geom_boxplot(data=p_final$data, aes(x=.data[[sub_in]], y=value, color=NULL), alpha=0.1) + \n",
    "        theme(legend.position=\"bottom\")\n",
    "    \n",
    "    annotate_figure(fig, \n",
    "                   top = text_grob(title, color = \"blue\", face = \"bold\", size = 14))\n",
    "    \n",
    "    ggsave(paste(img_dir,\"alpha_plot_\",file_name,\".jpg\",sep=\"\"))\n",
    "    \n",
    "}\n",
    "#https://rpkgs.datanovia.com/ggpubr/reference/annotate_figure.html"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_beta<-function(phyobj,type,dist_meas,colby,title_in,file_name,shape_in=\"\",facet_in=\"\"){\n",
    " \n",
    "    ordu = ordinate(phyobj, type, dist_meas, weighted=TRUE)\n",
    "    allGroupsColors<- c(\"grey0\")\n",
    "    allGroupsColors<-append(allGroupsColors,brewer.pal(n = 8, name = \"Dark2\"))\n",
    "    allGroupsColors<-append(allGroupsColors,brewer.pal(n = 8, name = \"Set2\"))\n",
    "    \n",
    "    if(shape_in==\"\"){\n",
    "        p<-plot_ordination(phyobj, ordu, color=colby,title=title_in)+ \n",
    "            theme(legend.position = \"bottom\") + \n",
    "            geom_point(size = 3) +\n",
    "            scale_color_manual(values = allGroupsColors) +\n",
    "            scale_shape_manual(values = c(1,2,3,4,5,6,7,8,9,10,11,12,13,15, 16, 17))\n",
    "    } else{\n",
    "        p<-plot_ordination(phyobj, ordu, color=colby,title=title_in,shape=shape_in)+ \n",
    "            theme(legend.position = \"bottom\") +\n",
    "            geom_point(size = 3) +\n",
    "            scale_color_manual(values = allGroupsColors) +\n",
    "            scale_shape_manual(values = c(1,2,3,4,5,6,7,8,9,10,11,12,13,15, 16, 17))\n",
    "    } \n",
    "    \n",
    "    if(facet_in==\"\"){\n",
    "        p_final = p \n",
    "    } else{\n",
    "        p_final = p + facet_wrap(~get(facet_in))\n",
    "    }\n",
    "    \n",
    "    addSmallLegend(p_final)\n",
    "\n",
    "    ggsave(paste(img_dir,\"\",file_name,\".tiff\",sep=\"\"))\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h2> Run Production Code <h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h3> Create objs <h3>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\n",
       "Ext_Control Seq_Control       Study \n",
       "        155          36         160 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "\n",
       "      A00       A01       A02       A03       M16       M22       M98 NTC_Blank \n",
       "        5         5         5         5         5        24        49         2 \n",
       "PCR_Blank     Water       Z00       Z05       Z06       Z10       Z11   human_1 \n",
       "        2        50        55         5         5        21         2       111 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "\n",
       "Ext_Control Seq_Control       Study \n",
       "        117          30         135 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "\n",
       "    A00     A01     A02     A03     M16     M22     M98   Water     Z00     Z05 \n",
       "      4       5       5       5       4      23      47      21      51       5 \n",
       "    Z06     Z10     Z11 human_1 \n",
       "      5      18       1      88 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#Summarize data\n",
    "temp = as.data.frame(sample_data(phy_obj))\n",
    "table(temp$Sample.Type)\n",
    "table(temp$Subject.ID)\n",
    "write.csv(ddply(temp, .(temp$Ext.Kit, temp$Subject.ID), nrow),paste(stats_dir,\"counts_pre.csv\",sep=\"\"))\n",
    "\n",
    "#Summarize data\n",
    "temp = as.data.frame(sample_data(phy_r))\n",
    "table(temp$Sample.Type)\n",
    "table(temp$Subject.ID)\n",
    "write.csv(ddply(temp, .(temp$Ext.Kit, temp$Subject.ID, temp$Homo.Status), nrow),paste(stats_dir,\"counts_post.csv\",sep=\"\"))\n",
    "write.csv(ddply(temp, .(temp$Ext.Kit, temp$Subject.ID), nrow),paste(stats_dir,\"counts_post_hmo.csv\",sep=\"\"))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h3> Extraction Controls <h3>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h4> Prepare Data <h4>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "54187.3238095238"
      ],
      "text/latex": [
       "54187.3238095238"
      ],
      "text/markdown": [
       "54187.3238095238"
      ],
      "text/plain": [
       "[1] 54187.32"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#Create glomed, subsetted obj\n",
    "phy_ext = phy_subset_ext(phy_obj)\n",
    "\n",
    "#Create OTUdf\n",
    "OTUdf = otu_extract(phy_ext)\n",
    "\n",
    "#Create TAXdf\n",
    "TAXdf = tax_extract(phy_ext)\n",
    "\n",
    "#Clean df's\n",
    "OTUdf_clean = otu_tax_clean(OTUdf,TAXdf)\n",
    "\n",
    "#Find averages reads\n",
    "mean(rowSums(OTUdf))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "write.csv(sample_data(phy_ext),\"save.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "#List of all variables to test\n",
    "test_list = unique(sample_data(phy_ext)$Ext.Kit)\n",
    "#test_list = test_list[7]\n",
    "#test_list = test_list[1]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h3> Standard Processing <h3>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Create a merged dfs for all variables\n",
    "mergedf_f1stats = data.frame()\n",
    "merged_RAstats = data.frame()\n",
    "dissdf_s = data.frame()\n",
    "RAdf_merged = data.frame()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"Removed 1 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 1 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 1 rows containing missing values (geom_segment).\"\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Warning message:\n",
      "\"Removed 1 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 1 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 2 rows containing non-finite values (stat_ydensity).\"\n",
      "Warning message:\n",
      "\"Removed 2 rows containing non-finite values (stat_summary).\"\n",
      "Warning message:\n",
      "\"Removed 1 rows containing missing values (geom_segment).\"\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 15\n",
      "[1] \"Z MagBead\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in cor.test.default(my_data$obs, my_data$exp, method = \"spearman\"):\n",
      "\"Cannot compute exact p-value with ties\"\n",
      "Warning message in cor.test.default(my_data$obs, my_data$exp, method = \"spearman\"):\n",
      "\"Cannot compute exact p-value with ties\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Z MagBead\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in cor.test.default(my_data$obs, my_data$exp, method = \"spearman\"):\n",
      "\"Cannot compute exact p-value with ties\"\n",
      "Warning message in cor.test.default(my_data$obs, my_data$exp, method = \"spearman\"):\n",
      "\"Cannot compute exact p-value with ties\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Z MagBead\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in cor.test.default(my_data$obs, my_data$exp, method = \"spearman\"):\n",
      "\"Cannot compute exact p-value with ties\"\n",
      "Warning message in cor.test.default(my_data$obs, my_data$exp, method = \"spearman\"):\n",
      "\"Cannot compute exact p-value with ties\"\n",
      "`geom_smooth()` using formula 'y ~ x'\n",
      "\n",
      "`geom_smooth()` using formula 'y ~ x'\n",
      "\n",
      "`geom_smooth()` using formula 'y ~ x'\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Warning message:\n",
      "\"Removed 1 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 1 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 1 rows containing missing values (geom_segment).\"\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Warning message:\n",
      "\"Removed 1 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 1 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 1 rows containing missing values (geom_segment).\"\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 14\n",
      "[1] \"Q PowerSoil\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in cor.test.default(my_data$obs, my_data$exp, method = \"spearman\"):\n",
      "\"Cannot compute exact p-value with ties\"\n",
      "Warning message in cor.test.default(my_data$obs, my_data$exp, method = \"spearman\"):\n",
      "\"Cannot compute exact p-value with ties\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Q PowerSoil\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in cor.test.default(my_data$obs, my_data$exp, method = \"spearman\"):\n",
      "\"Cannot compute exact p-value with ties\"\n",
      "Warning message in cor.test.default(my_data$obs, my_data$exp, method = \"spearman\"):\n",
      "\"Cannot compute exact p-value with ties\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Q PowerSoil\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in cor.test.default(my_data$obs, my_data$exp, method = \"spearman\"):\n",
      "\"Cannot compute exact p-value with ties\"\n",
      "Warning message in cor.test.default(my_data$obs, my_data$exp, method = \"spearman\"):\n",
      "\"Cannot compute exact p-value with ties\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Q PowerSoil\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in cor.test.default(my_data$obs, my_data$exp, method = \"spearman\"):\n",
      "\"Cannot compute exact p-value with ties\"\n",
      "Warning message in cor.test.default(my_data$obs, my_data$exp, method = \"spearman\"):\n",
      "\"Cannot compute exact p-value with ties\"\n",
      "`geom_smooth()` using formula 'y ~ x'\n",
      "\n",
      "`geom_smooth()` using formula 'y ~ x'\n",
      "\n",
      "`geom_smooth()` using formula 'y ~ x'\n",
      "\n",
      "`geom_smooth()` using formula 'y ~ x'\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Warning message:\n",
      "\"Removed 1 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 1 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 1 rows containing missing values (geom_segment).\"\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Warning message:\n",
      "\"Removed 1 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 1 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 1 rows containing missing values (geom_segment).\"\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 17\n",
      "[1] \"Q PowerMicrobiome\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in cor.test.default(my_data$obs, my_data$exp, method = \"spearman\"):\n",
      "\"Cannot compute exact p-value with ties\"\n",
      "Warning message in cor.test.default(my_data$obs, my_data$exp, method = \"spearman\"):\n",
      "\"Cannot compute exact p-value with ties\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Q PowerMicrobiome\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in cor.test.default(my_data$obs, my_data$exp, method = \"spearman\"):\n",
      "\"Cannot compute exact p-value with ties\"\n",
      "Warning message in cor.test.default(my_data$obs, my_data$exp, method = \"spearman\"):\n",
      "\"Cannot compute exact p-value with ties\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Q PowerMicrobiome\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "`geom_smooth()` using formula 'y ~ x'\n",
      "\n",
      "`geom_smooth()` using formula 'y ~ x'\n",
      "\n",
      "`geom_smooth()` using formula 'y ~ x'\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 11\n",
      "[1] \"Q QIAamp\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in cor.test.default(my_data$obs, my_data$exp, method = \"spearman\"):\n",
      "\"Cannot compute exact p-value with ties\"\n",
      "Warning message in cor.test.default(my_data$obs, my_data$exp, method = \"spearman\"):\n",
      "\"Cannot compute exact p-value with ties\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Q QIAamp\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in cor.test.default(my_data$obs, my_data$exp, method = \"spearman\"):\n",
      "\"Cannot compute exact p-value with ties\"\n",
      "Warning message in cor.test.default(my_data$obs, my_data$exp, method = \"spearman\"):\n",
      "\"Cannot compute exact p-value with ties\"\n",
      "`geom_smooth()` using formula 'y ~ x'\n",
      "\n",
      "`geom_smooth()` using formula 'y ~ x'\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Warning message:\n",
      "\"Removed 2 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 2 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 2 rows containing missing values (geom_segment).\"\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Warning message:\n",
      "\"Removed 2 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 2 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 2 rows containing missing values (geom_segment).\"\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 9\n",
      "[1] \"Q PowerSoil Pro\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in cor.test.default(my_data$obs, my_data$exp, method = \"spearman\"):\n",
      "\"Cannot compute exact p-value with ties\"\n",
      "Warning message in cor.test.default(my_data$obs, my_data$exp, method = \"spearman\"):\n",
      "\"Cannot compute exact p-value with ties\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Q PowerSoil Pro\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in cor.test.default(my_data$obs, my_data$exp, method = \"spearman\"):\n",
      "\"Cannot compute exact p-value with ties\"\n",
      "Warning message in cor.test.default(my_data$obs, my_data$exp, method = \"spearman\"):\n",
      "\"Cannot compute exact p-value with ties\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Q PowerSoil Pro\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in cor.test.default(my_data$obs, my_data$exp, method = \"spearman\"):\n",
      "\"Cannot compute exact p-value with ties\"\n",
      "Warning message in cor.test.default(my_data$obs, my_data$exp, method = \"spearman\"):\n",
      "\"Cannot compute exact p-value with ties\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Q PowerSoil Pro\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in cor.test.default(my_data$obs, my_data$exp, method = \"spearman\"):\n",
      "\"Cannot compute exact p-value with ties\"\n",
      "Warning message in cor.test.default(my_data$obs, my_data$exp, method = \"spearman\"):\n",
      "\"Cannot compute exact p-value with ties\"\n",
      "`geom_smooth()` using formula 'y ~ x'\n",
      "\n",
      "`geom_smooth()` using formula 'y ~ x'\n",
      "\n",
      "`geom_smooth()` using formula 'y ~ x'\n",
      "\n",
      "`geom_smooth()` using formula 'y ~ x'\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message:\n",
      "\"Removed 3 rows containing missing values (geom_segment).\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message:\n",
      "\"Removed 3 rows containing missing values (geom_segment).\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message:\n",
      "\"Removed 3 rows containing missing values (geom_segment).\"\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message:\n",
      "\"Removed 3 rows containing missing values (geom_segment).\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message:\n",
      "\"Removed 3 rows containing missing values (geom_segment).\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message:\n",
      "\"Removed 3 rows containing missing values (geom_segment).\"\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 3\n",
      "[1] \"T MagMax\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in cor.test.default(my_data$obs, my_data$exp, method = \"spearman\"):\n",
      "\"Cannot compute exact p-value with ties\"\n",
      "Warning message in cor.test.default(my_data$obs, my_data$exp, method = \"spearman\"):\n",
      "\"Cannot compute exact p-value with ties\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"T MagMax\"\n",
      "[1] \"T MagMax\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "`geom_smooth()` using formula 'y ~ x'\n",
      "\n",
      "`geom_smooth()` using formula 'y ~ x'\n",
      "\n",
      "`geom_smooth()` using formula 'y ~ x'\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 9\n",
      "[1] \"Q DSP Virus\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in cor.test.default(my_data$obs, my_data$exp, method = \"spearman\"):\n",
      "\"Cannot compute exact p-value with ties\"\n",
      "Warning message in cor.test.default(my_data$obs, my_data$exp, method = \"spearman\"):\n",
      "\"Cannot compute exact p-value with ties\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Q DSP Virus\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in cor.test.default(my_data$obs, my_data$exp, method = \"spearman\"):\n",
      "\"Cannot compute exact p-value with ties\"\n",
      "Warning message in cor.test.default(my_data$obs, my_data$exp, method = \"spearman\"):\n",
      "\"Cannot compute exact p-value with ties\"\n",
      "`geom_smooth()` using formula 'y ~ x'\n",
      "\n",
      "`geom_smooth()` using formula 'y ~ x'\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for (variable in test_list){\n",
    "    #Subset dataframe\n",
    "    OTUdf_clean2 = subset(OTUdf_clean,Ext.Kit==variable & Homo.Status==\"Standard\" & Subject.ID!=\"Ext_Blank\")\n",
    "    \n",
    "    #Create TP,FP,FN df\n",
    "    CONTdf_use = CONTdf_s\n",
    "    F1df = bench_compare(OTUdf_clean2, CONTdf_use)\n",
    "    \n",
    "    #Create TP,FP,FN counts    \n",
    "    bench_key = \"std\"\n",
    "    F1df_counts = bench_counts(OTUdf_clean2,F1df,variable, bench_key)#swapped variable and benchkey from original\n",
    "    temp = F1df_counts\n",
    "    temp$sample_name = rownames(temp)\n",
    "    rownames(temp) = c()\n",
    "    mergedf_f1stats = rbind(mergedf_f1stats,temp)\n",
    "    \n",
    "    #Create violin plots for each control - counts\n",
    "    bench_key = variable\n",
    "    x_list = c(\"FN\",\"TP\",\"FP\")\n",
    "    y_labs = c(\"False negative\",\"True positive\", \"False positive\")\n",
    "    plot_f1_violin(F1df_counts,x_list,y_labs,\"Subject.ID\",bench_key,\"counts_extstd\")\n",
    "    \n",
    "    #Create violin plots for each control - summary\n",
    "    bench_key = variable\n",
    "    x_list = c(\"TAR\",\"TDR\",\"F1\")\n",
    "    y_labs = c(\"Taxoonmic Accuracy \\n Rate\",\"Taxonomic Detection \\n Rate\", \"F-Measure\")\n",
    "    plot_f1_violin(F1df_counts,x_list,y_labs,\"Subject.ID\",bench_key,\"summary_extstd\",-1,3)\n",
    "    \n",
    "    #calculate relative abundance\n",
    "    RAdf = cal_relabun(OTUdf_clean2)\n",
    "    RAdf$Sample_name = rownames(RAdf)\n",
    "    RAdf_merged = rbind(RAdf_merged,RAdf)\n",
    "    \n",
    "    #Create scatter plots\n",
    "    #Determine expected to observed scatter plots, normality, spearman coeff \n",
    "    title = paste(\"Scatterplots of controls: \", variable,sep=\"\")\n",
    "    \n",
    "    print(nrow(RAdf))\n",
    "    RAdf_stats = cal_EO_kit(RAdf,CONTdf_use,variable,\"std\",title)\n",
    "    merged_RAstats = rbind(merged_RAstats,RAdf_stats)\n",
    "    \n",
    "    #average controls\n",
    "    bench_key = \"std\"\n",
    "    dissdf = create_av_RAdf(RAdf,variable,bench_key)\n",
    "    dissdf_s = rbind(dissdf_s,dissdf)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "write.csv(merged_RAstats,paste(stats_dir,\"RA_stats_kit_extstd.csv\",sep=\"\"))\n",
    "write.csv(mergedf_f1stats,paste(stats_dir,\"F1_vals_all_extstd.csv\",sep=\"\"))\n",
    "write.csv(aggregate(mergedf_f1stats[, 3:8], list(mergedf_f1stats$Subject.ID,mergedf_f1stats$bmark), mean),paste(stats_dir,\"f1_vals_mean_extstd.csv\",sep=\"\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h4> Benchmarked Assessment<h4>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Saving 6.67 x 6.67 in image\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#Create dist df of all methods\n",
    "Distdf_merged=data.frame()\n",
    "#Silva\n",
    "if(length(test_list)>1){\n",
    "    RAdf_av = RAdf_addcontrols(\"CONTdf_s\",\"dissdf_s\")\n",
    "    RAdf_clean = calc_dist(RAdf_av,CONTdf_s)\n",
    "    plot_dist_cluster(RAdf_clean,\"extstd\")\n",
    "    Distdf = plot_dist_violin(RAdf_clean,\"extstd\")\n",
    "} \n",
    "Distdf_merged = bind_rows(Distdf_merged,Distdf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A data.frame: 2  4</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>comparison</th><th scope=col>pval</th><th scope=col>sig</th><th scope=col>homo_s</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>bray-curtis</td><td>0.82</td><td>ns</td><td>       NA</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>NA         </td><td>  NA</td><td>NA</td><td>0.7571894</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2  4\n",
       "\\begin{tabular}{r|llll}\n",
       "  & comparison & pval & sig & homo\\_s\\\\\n",
       "  & <chr> & <dbl> & <chr> & <dbl>\\\\\n",
       "\\hline\n",
       "\t1 & bray-curtis & 0.82 & ns &        NA\\\\\n",
       "\t2 & NA          &   NA & NA & 0.7571894\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2  4\n",
       "\n",
       "| <!--/--> | comparison &lt;chr&gt; | pval &lt;dbl&gt; | sig &lt;chr&gt; | homo_s &lt;dbl&gt; |\n",
       "|---|---|---|---|---|\n",
       "| 1 | bray-curtis | 0.82 | ns |        NA |\n",
       "| 2 | NA          |   NA | NA | 0.7571894 |\n",
       "\n"
      ],
      "text/plain": [
       "  comparison  pval sig homo_s   \n",
       "1 bray-curtis 0.82 ns         NA\n",
       "2 NA            NA NA  0.7571894"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#is dist by method sig between all controls\n",
    "adonis_stats = adonis_bc(Distdf_merged)\n",
    "adonis_stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"Removed 1 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 1 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 2 rows containing non-finite values (stat_ydensity).\"\n",
      "Warning message:\n",
      "\"Removed 2 rows containing non-finite values (stat_summary).\"\n",
      "Warning message:\n",
      "\"Removed 1 rows containing missing values (geom_segment).\"\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Warning message:\n",
      "\"Removed 2 rows containing non-finite values (stat_ydensity).\"\n",
      "Warning message:\n",
      "\"Removed 2 rows containing non-finite values (stat_summary).\"\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Warning message:\n",
      "\"Removed 2 rows containing non-finite values (stat_ydensity).\"\n",
      "Warning message:\n",
      "\"Removed 2 rows containing non-finite values (stat_summary).\"\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message:\n",
      "\"Removed 4 rows containing missing values (geom_segment).\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message:\n",
      "\"Removed 4 rows containing missing values (geom_segment).\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message:\n",
      "\"Removed 4 rows containing missing values (geom_segment).\"\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Warning message:\n",
      "\"Removed 1 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 1 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 1 rows containing missing values (geom_segment).\"\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Warning message:\n",
      "\"Removed 1 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 1 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 1 rows containing missing values (geom_segment).\"\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Warning message:\n",
      "\"Removed 1 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 1 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 1 rows containing missing values (geom_segment).\"\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Warning message:\n",
      "\"Removed 2 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 2 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 2 rows containing missing values (geom_segment).\"\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#Overall violin plots\n",
    "x_list = c(\"TAR\",\"TDR\",\"F1\")\n",
    "y_labs = c(\"Taxonomic Accuracy \\n Rate\",\"Taxonomic Detection \\nRate\", \"F-Measure\")\n",
    "\n",
    "#Create violin plots for each control across all methods\n",
    "for(cont in unique(mergedf_f1stats$Subject.ID)){\n",
    "    subdf = subset(mergedf_f1stats, Subject.ID==cont)\n",
    "    \n",
    "    for (rows in rownames(subdf)){ #add IR and magnetic/column data\n",
    "        subdf[rows,\"Mag.Col\"] = metadata[subdf[rows,\"sample_name\"],\"Mag.Col\"]\n",
    "        subdf[rows,\"IR\"] = metadata[subdf[rows,\"sample_name\"],\"IR\"]\n",
    "    }\n",
    "    \n",
    "    plot_f1_violin(subdf,x_list,y_labs,\"bmark\",paste(\"Std Processing - \",cont,sep=\"\"),\"summary\",-1,3)\n",
    "    plot_f1_violin(subdf,x_list,y_labs,\"IR\",paste(\"Std Processing - \",cont,sep=\"\"),\"summary_ir\",-1,3)\n",
    "    plot_f1_violin(subdf,x_list,y_labs,\"Mag.Col\",paste(\"Std Processing - \",cont,sep=\"\"),\"summary_mc\",-1,3)\n",
    "\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Note: Using an external vector in selections is ambiguous.\n",
      "\u001b[34mi\u001b[39m Use `all_of(subgroup_in)` instead of `subgroup_in` to silence this message.\n",
      "\u001b[34mi\u001b[39m See <https://tidyselect.r-lib.org/reference/faq-external-vector.html>.\n",
      "\u001b[90mThis message is displayed once per session.\u001b[39m\n",
      "\n",
      "Note: Using an external vector in selections is ambiguous.\n",
      "\u001b[34mi\u001b[39m Use `all_of(cal)` instead of `cal` to silence this message.\n",
      "\u001b[34mi\u001b[39m See <https://tidyselect.r-lib.org/reference/faq-external-vector.html>.\n",
      "\u001b[90mThis message is displayed once per session.\u001b[39m\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#T-tests\n",
    "#Run by control - not all kits have the same control\n",
    "tstatdf_merged=data.frame()\n",
    "\n",
    "for (cont in unique(mergedf_f1stats$Subject.ID)){\n",
    "    subdf = subset(mergedf_f1stats,Subject.ID==cont)\n",
    "    tstatdf = tstat_paired_f1(subdf,\"bmark\")\n",
    "    if(nrow(tstatdf>1)){for(i in 1:nrow(tstatdf)){tstatdf[i,\"cont\"]=cont}}\n",
    "    tstatdf_merged=bind_rows(tstatdf_merged,tstatdf)  \n",
    "}\n",
    "write.csv(tstatdf_merged,paste(stats_dir,\"ttest_extstd.csv\",sep=\"\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "`geom_smooth()` using formula 'y ~ x'\n",
      "\n",
      "`geom_smooth()` using formula 'y ~ x'\n",
      "\n",
      "`geom_smooth()` using formula 'y ~ x'\n",
      "\n",
      "`geom_smooth()` using formula 'y ~ x'\n",
      "\n",
      "`geom_smooth()` using formula 'y ~ x'\n",
      "\n",
      "`geom_smooth()` using formula 'y ~ x'\n",
      "\n",
      "`geom_smooth()` using formula 'y ~ x'\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "`geom_smooth()` using formula 'y ~ x'\n",
      "\n",
      "Warning message:\n",
      "\"Removed 10 rows containing non-finite values (stat_smooth).\"\n",
      "Warning message:\n",
      "\"Removed 10 rows containing non-finite values (stat_cor).\"\n",
      "Warning message:\n",
      "\"Removed 10 rows containing missing values (geom_point).\"\n",
      "`geom_smooth()` using formula 'y ~ x'\n",
      "\n",
      "Warning message:\n",
      "\"Removed 10 rows containing non-finite values (stat_smooth).\"\n",
      "Warning message:\n",
      "\"Removed 10 rows containing non-finite values (stat_cor).\"\n",
      "Warning message:\n",
      "\"Removed 10 rows containing missing values (geom_point).\"\n",
      "`geom_smooth()` using formula 'y ~ x'\n",
      "\n",
      "Warning message:\n",
      "\"Removed 10 rows containing non-finite values (stat_smooth).\"\n",
      "Warning message:\n",
      "\"Removed 10 rows containing non-finite values (stat_cor).\"\n",
      "Warning message:\n",
      "\"Removed 10 rows containing missing values (geom_point).\"\n",
      "`geom_smooth()` using formula 'y ~ x'\n",
      "\n",
      "Warning message:\n",
      "\"Removed 10 rows containing non-finite values (stat_smooth).\"\n",
      "Warning message:\n",
      "\"Removed 10 rows containing non-finite values (stat_cor).\"\n",
      "Warning message:\n",
      "\"Removed 10 rows containing missing values (geom_point).\"\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "`geom_smooth()` using formula 'y ~ x'\n",
      "\n",
      "Warning message:\n",
      "\"Removed 20 rows containing non-finite values (stat_smooth).\"\n",
      "Warning message:\n",
      "\"Removed 20 rows containing non-finite values (stat_cor).\"\n",
      "Warning message:\n",
      "\"Removed 20 rows containing missing values (geom_point).\"\n",
      "`geom_smooth()` using formula 'y ~ x'\n",
      "\n",
      "Warning message:\n",
      "\"Removed 20 rows containing non-finite values (stat_smooth).\"\n",
      "Warning message:\n",
      "\"Removed 20 rows containing non-finite values (stat_cor).\"\n",
      "Warning message:\n",
      "\"Removed 20 rows containing missing values (geom_point).\"\n",
      "`geom_smooth()` using formula 'y ~ x'\n",
      "\n",
      "Warning message:\n",
      "\"Removed 35 rows containing non-finite values (stat_smooth).\"\n",
      "Warning message:\n",
      "\"Removed 35 rows containing non-finite values (stat_cor).\"\n",
      "Warning message:\n",
      "\"Removed 35 rows containing missing values (geom_point).\"\n",
      "`geom_smooth()` using formula 'y ~ x'\n",
      "\n",
      "Warning message:\n",
      "\"Removed 25 rows containing non-finite values (stat_smooth).\"\n",
      "Warning message:\n",
      "\"Removed 25 rows containing non-finite values (stat_cor).\"\n",
      "Warning message:\n",
      "\"Removed 25 rows containing missing values (geom_point).\"\n",
      "`geom_smooth()` using formula 'y ~ x'\n",
      "\n",
      "Warning message:\n",
      "\"Removed 5 rows containing non-finite values (stat_smooth).\"\n",
      "Warning message:\n",
      "\"Removed 5 rows containing non-finite values (stat_cor).\"\n",
      "Warning message:\n",
      "\"Removed 5 rows containing missing values (geom_point).\"\n",
      "`geom_smooth()` using formula 'y ~ x'\n",
      "\n",
      "Warning message:\n",
      "\"Removed 15 rows containing non-finite values (stat_smooth).\"\n",
      "Warning message:\n",
      "\"Removed 15 rows containing non-finite values (stat_cor).\"\n",
      "Warning message:\n",
      "\"Removed 15 rows containing missing values (geom_point).\"\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "`geom_smooth()` using formula 'y ~ x'\n",
      "\n",
      "`geom_smooth()` using formula 'y ~ x'\n",
      "\n",
      "`geom_smooth()` using formula 'y ~ x'\n",
      "\n",
      "`geom_smooth()` using formula 'y ~ x'\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for (cont in unique(RAdf_merged$Subject.ID)){\n",
    "    taxdf = create_taxadf(cont,CONTdf_s)\n",
    "    cal_EO_cont(RAdf_merged,CONTdf_s,taxdf,cont,paste(\"Std Processing - \",cont,sep=\"\"))\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#Create correlation matrices for extraction controls\n",
    "RAdf_oe_save = RAdf_av\n",
    "plot_corrmatrix(RAdf_av,CONTdf_s,\"Group.1\") #sort by kit\n",
    "plot_corrmatrix(RAdf_av,CONTdf_s,\"bmark\") #sort by control"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h3> Blank assessment <h3>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h3> Prepare Data <h3>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "phyloseq-class experiment-level object\n",
       "otu_table()   OTU Table:         [ 233 taxa and 21 samples ]\n",
       "sample_data() Sample Data:       [ 21 samples by 27 sample variables ]\n",
       "tax_table()   Taxonomy Table:    [ 233 taxa by 7 taxonomic ranks ]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#Create glomed, subsetted obj\n",
    "phy_blank = phy_subset_blank(phy_r)\n",
    "phy_blank"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A data.frame: 52  3</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>Ext.Kit</th><th scope=col>Subject.ID</th><th scope=col>Seq</th></tr>\n",
       "\t<tr><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>                 </td><td>A00      </td><td> 5</td></tr>\n",
       "\t<tr><td>                 </td><td>A01      </td><td> 5</td></tr>\n",
       "\t<tr><td>                 </td><td>A02      </td><td> 5</td></tr>\n",
       "\t<tr><td>                 </td><td>A03      </td><td> 5</td></tr>\n",
       "\t<tr><td>                 </td><td>NTC_Blank</td><td> 2</td></tr>\n",
       "\t<tr><td>                 </td><td>PCR_Blank</td><td> 2</td></tr>\n",
       "\t<tr><td>                 </td><td>Z05      </td><td> 5</td></tr>\n",
       "\t<tr><td>                 </td><td>Z06      </td><td> 5</td></tr>\n",
       "\t<tr><td>                 </td><td>Z11      </td><td> 2</td></tr>\n",
       "\t<tr><td>Q DSP Virus      </td><td>M22      </td><td> 3</td></tr>\n",
       "\t<tr><td>Q DSP Virus      </td><td>M98      </td><td>10</td></tr>\n",
       "\t<tr><td>Q DSP Virus      </td><td>Water    </td><td> 3</td></tr>\n",
       "\t<tr><td>Q DSP Virus      </td><td>Z00      </td><td> 6</td></tr>\n",
       "\t<tr><td>Q DSP Virus      </td><td>human_1  </td><td>11</td></tr>\n",
       "\t<tr><td>Q PowerMicrobiome</td><td>M22      </td><td> 7</td></tr>\n",
       "\t<tr><td>Q PowerMicrobiome</td><td>M98      </td><td>10</td></tr>\n",
       "\t<tr><td>Q PowerMicrobiome</td><td>Water    </td><td> 9</td></tr>\n",
       "\t<tr><td>Q PowerMicrobiome</td><td>Z00      </td><td>11</td></tr>\n",
       "\t<tr><td>Q PowerMicrobiome</td><td>Z10      </td><td> 2</td></tr>\n",
       "\t<tr><td>Q PowerMicrobiome</td><td>human_1  </td><td>34</td></tr>\n",
       "\t<tr><td>Q PowerSoil      </td><td>M16      </td><td> 2</td></tr>\n",
       "\t<tr><td>Q PowerSoil      </td><td>M22      </td><td> 4</td></tr>\n",
       "\t<tr><td>Q PowerSoil      </td><td>M98      </td><td>11</td></tr>\n",
       "\t<tr><td>Q PowerSoil      </td><td>Water    </td><td>12</td></tr>\n",
       "\t<tr><td>Q PowerSoil      </td><td>Z00      </td><td>11</td></tr>\n",
       "\t<tr><td>Q PowerSoil      </td><td>Z10      </td><td> 6</td></tr>\n",
       "\t<tr><td>Q PowerSoil      </td><td>human_1  </td><td>26</td></tr>\n",
       "\t<tr><td>Q PowerSoil Pro  </td><td>M16      </td><td> 1</td></tr>\n",
       "\t<tr><td>Q PowerSoil Pro  </td><td>M22      </td><td> 1</td></tr>\n",
       "\t<tr><td>Q PowerSoil Pro  </td><td>M98      </td><td> 2</td></tr>\n",
       "\t<tr><td>Q PowerSoil Pro  </td><td>Water    </td><td>10</td></tr>\n",
       "\t<tr><td>Q PowerSoil Pro  </td><td>Z00      </td><td> 5</td></tr>\n",
       "\t<tr><td>Q PowerSoil Pro  </td><td>Z10      </td><td> 8</td></tr>\n",
       "\t<tr><td>Q PowerSoil Pro  </td><td>human_1  </td><td>10</td></tr>\n",
       "\t<tr><td>Q QIAamp         </td><td>M22      </td><td> 5</td></tr>\n",
       "\t<tr><td>Q QIAamp         </td><td>M98      </td><td> 7</td></tr>\n",
       "\t<tr><td>Q QIAamp         </td><td>Water    </td><td> 2</td></tr>\n",
       "\t<tr><td>Q QIAamp         </td><td>Z00      </td><td> 6</td></tr>\n",
       "\t<tr><td>Q QIAamp         </td><td>human_1  </td><td>18</td></tr>\n",
       "\t<tr><td>T MagMax         </td><td>M16      </td><td> 1</td></tr>\n",
       "\t<tr><td>T MagMax         </td><td>M98      </td><td> 1</td></tr>\n",
       "\t<tr><td>T MagMax         </td><td>Water    </td><td> 3</td></tr>\n",
       "\t<tr><td>T MagMax         </td><td>Z00      </td><td> 3</td></tr>\n",
       "\t<tr><td>T MagMax         </td><td>Z10      </td><td> 3</td></tr>\n",
       "\t<tr><td>T MagMax         </td><td>human_1  </td><td> 2</td></tr>\n",
       "\t<tr><td>Z MagBead        </td><td>M16      </td><td> 1</td></tr>\n",
       "\t<tr><td>Z MagBead        </td><td>M22      </td><td> 4</td></tr>\n",
       "\t<tr><td>Z MagBead        </td><td>M98      </td><td> 8</td></tr>\n",
       "\t<tr><td>Z MagBead        </td><td>Water    </td><td>11</td></tr>\n",
       "\t<tr><td>Z MagBead        </td><td>Z00      </td><td>13</td></tr>\n",
       "\t<tr><td>Z MagBead        </td><td>Z10      </td><td> 2</td></tr>\n",
       "\t<tr><td>Z MagBead        </td><td>human_1  </td><td>10</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 52  3\n",
       "\\begin{tabular}{lll}\n",
       " Ext.Kit & Subject.ID & Seq\\\\\n",
       " <fct> & <fct> & <int>\\\\\n",
       "\\hline\n",
       "\t                   & A00       &  5\\\\\n",
       "\t                   & A01       &  5\\\\\n",
       "\t                   & A02       &  5\\\\\n",
       "\t                   & A03       &  5\\\\\n",
       "\t                   & NTC\\_Blank &  2\\\\\n",
       "\t                   & PCR\\_Blank &  2\\\\\n",
       "\t                   & Z05       &  5\\\\\n",
       "\t                   & Z06       &  5\\\\\n",
       "\t                   & Z11       &  2\\\\\n",
       "\t Q DSP Virus       & M22       &  3\\\\\n",
       "\t Q DSP Virus       & M98       & 10\\\\\n",
       "\t Q DSP Virus       & Water     &  3\\\\\n",
       "\t Q DSP Virus       & Z00       &  6\\\\\n",
       "\t Q DSP Virus       & human\\_1   & 11\\\\\n",
       "\t Q PowerMicrobiome & M22       &  7\\\\\n",
       "\t Q PowerMicrobiome & M98       & 10\\\\\n",
       "\t Q PowerMicrobiome & Water     &  9\\\\\n",
       "\t Q PowerMicrobiome & Z00       & 11\\\\\n",
       "\t Q PowerMicrobiome & Z10       &  2\\\\\n",
       "\t Q PowerMicrobiome & human\\_1   & 34\\\\\n",
       "\t Q PowerSoil       & M16       &  2\\\\\n",
       "\t Q PowerSoil       & M22       &  4\\\\\n",
       "\t Q PowerSoil       & M98       & 11\\\\\n",
       "\t Q PowerSoil       & Water     & 12\\\\\n",
       "\t Q PowerSoil       & Z00       & 11\\\\\n",
       "\t Q PowerSoil       & Z10       &  6\\\\\n",
       "\t Q PowerSoil       & human\\_1   & 26\\\\\n",
       "\t Q PowerSoil Pro   & M16       &  1\\\\\n",
       "\t Q PowerSoil Pro   & M22       &  1\\\\\n",
       "\t Q PowerSoil Pro   & M98       &  2\\\\\n",
       "\t Q PowerSoil Pro   & Water     & 10\\\\\n",
       "\t Q PowerSoil Pro   & Z00       &  5\\\\\n",
       "\t Q PowerSoil Pro   & Z10       &  8\\\\\n",
       "\t Q PowerSoil Pro   & human\\_1   & 10\\\\\n",
       "\t Q QIAamp          & M22       &  5\\\\\n",
       "\t Q QIAamp          & M98       &  7\\\\\n",
       "\t Q QIAamp          & Water     &  2\\\\\n",
       "\t Q QIAamp          & Z00       &  6\\\\\n",
       "\t Q QIAamp          & human\\_1   & 18\\\\\n",
       "\t T MagMax          & M16       &  1\\\\\n",
       "\t T MagMax          & M98       &  1\\\\\n",
       "\t T MagMax          & Water     &  3\\\\\n",
       "\t T MagMax          & Z00       &  3\\\\\n",
       "\t T MagMax          & Z10       &  3\\\\\n",
       "\t T MagMax          & human\\_1   &  2\\\\\n",
       "\t Z MagBead         & M16       &  1\\\\\n",
       "\t Z MagBead         & M22       &  4\\\\\n",
       "\t Z MagBead         & M98       &  8\\\\\n",
       "\t Z MagBead         & Water     & 11\\\\\n",
       "\t Z MagBead         & Z00       & 13\\\\\n",
       "\t Z MagBead         & Z10       &  2\\\\\n",
       "\t Z MagBead         & human\\_1   & 10\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 52  3\n",
       "\n",
       "| Ext.Kit &lt;fct&gt; | Subject.ID &lt;fct&gt; | Seq &lt;int&gt; |\n",
       "|---|---|---|\n",
       "| <!----> | A00       |  5 |\n",
       "| <!----> | A01       |  5 |\n",
       "| <!----> | A02       |  5 |\n",
       "| <!----> | A03       |  5 |\n",
       "| <!----> | NTC_Blank |  2 |\n",
       "| <!----> | PCR_Blank |  2 |\n",
       "| <!----> | Z05       |  5 |\n",
       "| <!----> | Z06       |  5 |\n",
       "| <!----> | Z11       |  2 |\n",
       "| Q DSP Virus       | M22       |  3 |\n",
       "| Q DSP Virus       | M98       | 10 |\n",
       "| Q DSP Virus       | Water     |  3 |\n",
       "| Q DSP Virus       | Z00       |  6 |\n",
       "| Q DSP Virus       | human_1   | 11 |\n",
       "| Q PowerMicrobiome | M22       |  7 |\n",
       "| Q PowerMicrobiome | M98       | 10 |\n",
       "| Q PowerMicrobiome | Water     |  9 |\n",
       "| Q PowerMicrobiome | Z00       | 11 |\n",
       "| Q PowerMicrobiome | Z10       |  2 |\n",
       "| Q PowerMicrobiome | human_1   | 34 |\n",
       "| Q PowerSoil       | M16       |  2 |\n",
       "| Q PowerSoil       | M22       |  4 |\n",
       "| Q PowerSoil       | M98       | 11 |\n",
       "| Q PowerSoil       | Water     | 12 |\n",
       "| Q PowerSoil       | Z00       | 11 |\n",
       "| Q PowerSoil       | Z10       |  6 |\n",
       "| Q PowerSoil       | human_1   | 26 |\n",
       "| Q PowerSoil Pro   | M16       |  1 |\n",
       "| Q PowerSoil Pro   | M22       |  1 |\n",
       "| Q PowerSoil Pro   | M98       |  2 |\n",
       "| Q PowerSoil Pro   | Water     | 10 |\n",
       "| Q PowerSoil Pro   | Z00       |  5 |\n",
       "| Q PowerSoil Pro   | Z10       |  8 |\n",
       "| Q PowerSoil Pro   | human_1   | 10 |\n",
       "| Q QIAamp          | M22       |  5 |\n",
       "| Q QIAamp          | M98       |  7 |\n",
       "| Q QIAamp          | Water     |  2 |\n",
       "| Q QIAamp          | Z00       |  6 |\n",
       "| Q QIAamp          | human_1   | 18 |\n",
       "| T MagMax          | M16       |  1 |\n",
       "| T MagMax          | M98       |  1 |\n",
       "| T MagMax          | Water     |  3 |\n",
       "| T MagMax          | Z00       |  3 |\n",
       "| T MagMax          | Z10       |  3 |\n",
       "| T MagMax          | human_1   |  2 |\n",
       "| Z MagBead         | M16       |  1 |\n",
       "| Z MagBead         | M22       |  4 |\n",
       "| Z MagBead         | M98       |  8 |\n",
       "| Z MagBead         | Water     | 11 |\n",
       "| Z MagBead         | Z00       | 13 |\n",
       "| Z MagBead         | Z10       |  2 |\n",
       "| Z MagBead         | human_1   | 10 |\n",
       "\n"
      ],
      "text/plain": [
       "   Ext.Kit           Subject.ID Seq\n",
       "1                    A00         5 \n",
       "2                    A01         5 \n",
       "3                    A02         5 \n",
       "4                    A03         5 \n",
       "5                    NTC_Blank   2 \n",
       "6                    PCR_Blank   2 \n",
       "7                    Z05         5 \n",
       "8                    Z06         5 \n",
       "9                    Z11         2 \n",
       "10 Q DSP Virus       M22         3 \n",
       "11 Q DSP Virus       M98        10 \n",
       "12 Q DSP Virus       Water       3 \n",
       "13 Q DSP Virus       Z00         6 \n",
       "14 Q DSP Virus       human_1    11 \n",
       "15 Q PowerMicrobiome M22         7 \n",
       "16 Q PowerMicrobiome M98        10 \n",
       "17 Q PowerMicrobiome Water       9 \n",
       "18 Q PowerMicrobiome Z00        11 \n",
       "19 Q PowerMicrobiome Z10         2 \n",
       "20 Q PowerMicrobiome human_1    34 \n",
       "21 Q PowerSoil       M16         2 \n",
       "22 Q PowerSoil       M22         4 \n",
       "23 Q PowerSoil       M98        11 \n",
       "24 Q PowerSoil       Water      12 \n",
       "25 Q PowerSoil       Z00        11 \n",
       "26 Q PowerSoil       Z10         6 \n",
       "27 Q PowerSoil       human_1    26 \n",
       "28 Q PowerSoil Pro   M16         1 \n",
       "29 Q PowerSoil Pro   M22         1 \n",
       "30 Q PowerSoil Pro   M98         2 \n",
       "31 Q PowerSoil Pro   Water      10 \n",
       "32 Q PowerSoil Pro   Z00         5 \n",
       "33 Q PowerSoil Pro   Z10         8 \n",
       "34 Q PowerSoil Pro   human_1    10 \n",
       "35 Q QIAamp          M22         5 \n",
       "36 Q QIAamp          M98         7 \n",
       "37 Q QIAamp          Water       2 \n",
       "38 Q QIAamp          Z00         6 \n",
       "39 Q QIAamp          human_1    18 \n",
       "40 T MagMax          M16         1 \n",
       "41 T MagMax          M98         1 \n",
       "42 T MagMax          Water       3 \n",
       "43 T MagMax          Z00         3 \n",
       "44 T MagMax          Z10         3 \n",
       "45 T MagMax          human_1     2 \n",
       "46 Z MagBead         M16         1 \n",
       "47 Z MagBead         M22         4 \n",
       "48 Z MagBead         M98         8 \n",
       "49 Z MagBead         Water      11 \n",
       "50 Z MagBead         Z00        13 \n",
       "51 Z MagBead         Z10         2 \n",
       "52 Z MagBead         human_1    10 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A data.frame: 5  3</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>Ext.Kit</th><th scope=col>Subject.ID</th><th scope=col>Passed QC</th></tr>\n",
       "\t<tr><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>Q PowerMicrobiome</td><td>Water</td><td>5</td></tr>\n",
       "\t<tr><td>Q PowerSoil      </td><td>Water</td><td>7</td></tr>\n",
       "\t<tr><td>Q QIAamp         </td><td>Water</td><td>1</td></tr>\n",
       "\t<tr><td>T MagMax         </td><td>Water</td><td>1</td></tr>\n",
       "\t<tr><td>Z MagBead        </td><td>Water</td><td>7</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 5  3\n",
       "\\begin{tabular}{lll}\n",
       " Ext.Kit & Subject.ID & Passed QC\\\\\n",
       " <fct> & <fct> & <int>\\\\\n",
       "\\hline\n",
       "\t Q PowerMicrobiome & Water & 5\\\\\n",
       "\t Q PowerSoil       & Water & 7\\\\\n",
       "\t Q QIAamp          & Water & 1\\\\\n",
       "\t T MagMax          & Water & 1\\\\\n",
       "\t Z MagBead         & Water & 7\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 5  3\n",
       "\n",
       "| Ext.Kit &lt;fct&gt; | Subject.ID &lt;fct&gt; | Passed QC &lt;int&gt; |\n",
       "|---|---|---|\n",
       "| Q PowerMicrobiome | Water | 5 |\n",
       "| Q PowerSoil       | Water | 7 |\n",
       "| Q QIAamp          | Water | 1 |\n",
       "| T MagMax          | Water | 1 |\n",
       "| Z MagBead         | Water | 7 |\n",
       "\n"
      ],
      "text/plain": [
       "  Ext.Kit           Subject.ID Passed QC\n",
       "1 Q PowerMicrobiome Water      5        \n",
       "2 Q PowerSoil       Water      7        \n",
       "3 Q QIAamp          Water      1        \n",
       "4 T MagMax          Water      1        \n",
       "5 Z MagBead         Water      7        "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#Determine counts for blanks included, and those that passed filtering\n",
    "tmp = ddply(sample_data(phy_obj), .(sample_data(phy_obj)$Ext.Kit, sample_data(phy_obj)$Subject.ID), nrow)\n",
    "names(tmp) <- c(\"Ext.Kit\", \"Subject.ID\", \"Seq\")\n",
    "tmp\n",
    "tmp = ddply(sample_data(phy_blank), .(sample_data(phy_blank)$Ext.Kit, sample_data(phy_blank)$Subject.ID), nrow)\n",
    "names(tmp) <- c(\"Ext.Kit\", \"Subject.ID\", \"Passed QC\")\n",
    "tmp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A data.frame: 6  36</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>OTU</th><th scope=col>Sample</th><th scope=col>Abundance</th><th scope=col>Source.PCR.Plate</th><th scope=col>Run.ID</th><th scope=col>Project.ID</th><th scope=col>CGR.Sample.ID</th><th scope=col>Sample.Type</th><th scope=col>Sample.Des</th><th scope=col>Subject.ID</th><th scope=col>...</th><th scope=col>Unassigned</th><th scope=col>frac_filtered</th><th scope=col>Mag.Col</th><th scope=col>IR</th><th scope=col>Domain</th><th scope=col>Phylum</th><th scope=col>Class</th><th scope=col>Order</th><th scope=col>Family</th><th scope=col>Genus</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>...</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>2512</th><td>8ae518dbb29595b3f79214be0b589066</td><td>SC553068</td><td>0.8093541</td><td>PC22192</td><td>Run4</td><td>NP0084-MB6</td><td>SC553068</td><td>Ext_Control</td><td>Ext_Blank</td><td>Water</td><td>...</td><td>0</td><td> 9.227570</td><td>Magnetic</td><td>Solution</td><td>D_0__Bacteria</td><td>D_1__Firmicutes   </td><td>D_2__Bacilli    </td><td>D_3__Bacillales   </td><td>D_4__Listeriaceae  </td><td>D_5__Listeria    </td></tr>\n",
       "\t<tr><th scope=row>2507</th><td>8ae518dbb29595b3f79214be0b589066</td><td>SC553048</td><td>0.7947894</td><td>PC22190</td><td>Run4</td><td>NP0084-MB6</td><td>SC553048</td><td>Ext_Control</td><td>Ext_Blank</td><td>Water</td><td>...</td><td>0</td><td> 7.888065</td><td>Magnetic</td><td>Solution</td><td>D_0__Bacteria</td><td>D_1__Firmicutes   </td><td>D_2__Bacilli    </td><td>D_3__Bacillales   </td><td>D_4__Listeriaceae  </td><td>D_5__Listeria    </td></tr>\n",
       "\t<tr><th scope=row>2505</th><td>8ae518dbb29595b3f79214be0b589066</td><td>SC552975</td><td>0.7427614</td><td>PC22190</td><td>Run4</td><td>NP0084-MB6</td><td>SC552975</td><td>Ext_Control</td><td>Ext_Blank</td><td>Water</td><td>...</td><td>0</td><td> 9.846112</td><td>Magnetic</td><td>Solution</td><td>D_0__Bacteria</td><td>D_1__Firmicutes   </td><td>D_2__Bacilli    </td><td>D_3__Bacillales   </td><td>D_4__Listeriaceae  </td><td>D_5__Listeria    </td></tr>\n",
       "\t<tr><th scope=row>2514</th><td>8ae518dbb29595b3f79214be0b589066</td><td>SC552961</td><td>0.7062563</td><td>PC22192</td><td>Run4</td><td>NP0084-MB6</td><td>SC552961</td><td>Ext_Control</td><td>Ext_Blank</td><td>Water</td><td>...</td><td>0</td><td>10.033773</td><td>Magnetic</td><td>Solution</td><td>D_0__Bacteria</td><td>D_1__Firmicutes   </td><td>D_2__Bacilli    </td><td>D_3__Bacillales   </td><td>D_4__Listeriaceae  </td><td>D_5__Listeria    </td></tr>\n",
       "\t<tr><th scope=row>822</th><td>333bbf9224442fc10cd497377d2b1d01</td><td>SC552952</td><td>0.4305763</td><td>PC22190</td><td>Run4</td><td>NP0084-MB6</td><td>SC552952</td><td>Ext_Control</td><td>Ext_Blank</td><td>Water</td><td>...</td><td>0</td><td>15.504106</td><td>Magnetic</td><td>Solution</td><td>D_0__Bacteria</td><td>D_1__Bacteroidetes</td><td>D_2__Bacteroidia</td><td>D_3__Bacteroidales</td><td>D_4__Bacteroidaceae</td><td>D_5__Bacteroides </td></tr>\n",
       "\t<tr><th scope=row>1811</th><td>675c847bccbc53942ebb7b8cbb4efc4d</td><td>SC552966</td><td>0.3654507</td><td>PC22192</td><td>Run4</td><td>NP0084-MB6</td><td>SC552966</td><td>Ext_Control</td><td>Ext_Blank</td><td>Water</td><td>...</td><td>0</td><td>18.981414</td><td>Magnetic</td><td>Solution</td><td>D_0__Bacteria</td><td>D_1__Bacteroidetes</td><td>D_2__Bacteroidia</td><td>D_3__Bacteroidales</td><td>D_4__Prevotellaceae</td><td>D_5__Prevotella 9</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6  36\n",
       "\\begin{tabular}{r|lllllllllllllllllllll}\n",
       "  & OTU & Sample & Abundance & Source.PCR.Plate & Run.ID & Project.ID & CGR.Sample.ID & Sample.Type & Sample.Des & Subject.ID & ... & Unassigned & frac\\_filtered & Mag.Col & IR & Domain & Phylum & Class & Order & Family & Genus\\\\\n",
       "  & <chr> & <chr> & <dbl> & <fct> & <fct> & <fct> & <fct> & <fct> & <fct> & <fct> & ... & <int> & <dbl> & <fct> & <fct> & <fct> & <fct> & <fct> & <fct> & <fct> & <fct>\\\\\n",
       "\\hline\n",
       "\t2512 & 8ae518dbb29595b3f79214be0b589066 & SC553068 & 0.8093541 & PC22192 & Run4 & NP0084-MB6 & SC553068 & Ext\\_Control & Ext\\_Blank & Water & ... & 0 &  9.227570 & Magnetic & Solution & D\\_0\\_\\_Bacteria & D\\_1\\_\\_Firmicutes    & D\\_2\\_\\_Bacilli     & D\\_3\\_\\_Bacillales    & D\\_4\\_\\_Listeriaceae   & D\\_5\\_\\_Listeria    \\\\\n",
       "\t2507 & 8ae518dbb29595b3f79214be0b589066 & SC553048 & 0.7947894 & PC22190 & Run4 & NP0084-MB6 & SC553048 & Ext\\_Control & Ext\\_Blank & Water & ... & 0 &  7.888065 & Magnetic & Solution & D\\_0\\_\\_Bacteria & D\\_1\\_\\_Firmicutes    & D\\_2\\_\\_Bacilli     & D\\_3\\_\\_Bacillales    & D\\_4\\_\\_Listeriaceae   & D\\_5\\_\\_Listeria    \\\\\n",
       "\t2505 & 8ae518dbb29595b3f79214be0b589066 & SC552975 & 0.7427614 & PC22190 & Run4 & NP0084-MB6 & SC552975 & Ext\\_Control & Ext\\_Blank & Water & ... & 0 &  9.846112 & Magnetic & Solution & D\\_0\\_\\_Bacteria & D\\_1\\_\\_Firmicutes    & D\\_2\\_\\_Bacilli     & D\\_3\\_\\_Bacillales    & D\\_4\\_\\_Listeriaceae   & D\\_5\\_\\_Listeria    \\\\\n",
       "\t2514 & 8ae518dbb29595b3f79214be0b589066 & SC552961 & 0.7062563 & PC22192 & Run4 & NP0084-MB6 & SC552961 & Ext\\_Control & Ext\\_Blank & Water & ... & 0 & 10.033773 & Magnetic & Solution & D\\_0\\_\\_Bacteria & D\\_1\\_\\_Firmicutes    & D\\_2\\_\\_Bacilli     & D\\_3\\_\\_Bacillales    & D\\_4\\_\\_Listeriaceae   & D\\_5\\_\\_Listeria    \\\\\n",
       "\t822 & 333bbf9224442fc10cd497377d2b1d01 & SC552952 & 0.4305763 & PC22190 & Run4 & NP0084-MB6 & SC552952 & Ext\\_Control & Ext\\_Blank & Water & ... & 0 & 15.504106 & Magnetic & Solution & D\\_0\\_\\_Bacteria & D\\_1\\_\\_Bacteroidetes & D\\_2\\_\\_Bacteroidia & D\\_3\\_\\_Bacteroidales & D\\_4\\_\\_Bacteroidaceae & D\\_5\\_\\_Bacteroides \\\\\n",
       "\t1811 & 675c847bccbc53942ebb7b8cbb4efc4d & SC552966 & 0.3654507 & PC22192 & Run4 & NP0084-MB6 & SC552966 & Ext\\_Control & Ext\\_Blank & Water & ... & 0 & 18.981414 & Magnetic & Solution & D\\_0\\_\\_Bacteria & D\\_1\\_\\_Bacteroidetes & D\\_2\\_\\_Bacteroidia & D\\_3\\_\\_Bacteroidales & D\\_4\\_\\_Prevotellaceae & D\\_5\\_\\_Prevotella 9\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6  36\n",
       "\n",
       "| <!--/--> | OTU &lt;chr&gt; | Sample &lt;chr&gt; | Abundance &lt;dbl&gt; | Source.PCR.Plate &lt;fct&gt; | Run.ID &lt;fct&gt; | Project.ID &lt;fct&gt; | CGR.Sample.ID &lt;fct&gt; | Sample.Type &lt;fct&gt; | Sample.Des &lt;fct&gt; | Subject.ID &lt;fct&gt; | ... ... | Unassigned &lt;int&gt; | frac_filtered &lt;dbl&gt; | Mag.Col &lt;fct&gt; | IR &lt;fct&gt; | Domain &lt;fct&gt; | Phylum &lt;fct&gt; | Class &lt;fct&gt; | Order &lt;fct&gt; | Family &lt;fct&gt; | Genus &lt;fct&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 2512 | 8ae518dbb29595b3f79214be0b589066 | SC553068 | 0.8093541 | PC22192 | Run4 | NP0084-MB6 | SC553068 | Ext_Control | Ext_Blank | Water | ... | 0 |  9.227570 | Magnetic | Solution | D_0__Bacteria | D_1__Firmicutes    | D_2__Bacilli     | D_3__Bacillales    | D_4__Listeriaceae   | D_5__Listeria     |\n",
       "| 2507 | 8ae518dbb29595b3f79214be0b589066 | SC553048 | 0.7947894 | PC22190 | Run4 | NP0084-MB6 | SC553048 | Ext_Control | Ext_Blank | Water | ... | 0 |  7.888065 | Magnetic | Solution | D_0__Bacteria | D_1__Firmicutes    | D_2__Bacilli     | D_3__Bacillales    | D_4__Listeriaceae   | D_5__Listeria     |\n",
       "| 2505 | 8ae518dbb29595b3f79214be0b589066 | SC552975 | 0.7427614 | PC22190 | Run4 | NP0084-MB6 | SC552975 | Ext_Control | Ext_Blank | Water | ... | 0 |  9.846112 | Magnetic | Solution | D_0__Bacteria | D_1__Firmicutes    | D_2__Bacilli     | D_3__Bacillales    | D_4__Listeriaceae   | D_5__Listeria     |\n",
       "| 2514 | 8ae518dbb29595b3f79214be0b589066 | SC552961 | 0.7062563 | PC22192 | Run4 | NP0084-MB6 | SC552961 | Ext_Control | Ext_Blank | Water | ... | 0 | 10.033773 | Magnetic | Solution | D_0__Bacteria | D_1__Firmicutes    | D_2__Bacilli     | D_3__Bacillales    | D_4__Listeriaceae   | D_5__Listeria     |\n",
       "| 822 | 333bbf9224442fc10cd497377d2b1d01 | SC552952 | 0.4305763 | PC22190 | Run4 | NP0084-MB6 | SC552952 | Ext_Control | Ext_Blank | Water | ... | 0 | 15.504106 | Magnetic | Solution | D_0__Bacteria | D_1__Bacteroidetes | D_2__Bacteroidia | D_3__Bacteroidales | D_4__Bacteroidaceae | D_5__Bacteroides  |\n",
       "| 1811 | 675c847bccbc53942ebb7b8cbb4efc4d | SC552966 | 0.3654507 | PC22192 | Run4 | NP0084-MB6 | SC552966 | Ext_Control | Ext_Blank | Water | ... | 0 | 18.981414 | Magnetic | Solution | D_0__Bacteria | D_1__Bacteroidetes | D_2__Bacteroidia | D_3__Bacteroidales | D_4__Prevotellaceae | D_5__Prevotella 9 |\n",
       "\n"
      ],
      "text/plain": [
       "     OTU                              Sample   Abundance Source.PCR.Plate\n",
       "2512 8ae518dbb29595b3f79214be0b589066 SC553068 0.8093541 PC22192         \n",
       "2507 8ae518dbb29595b3f79214be0b589066 SC553048 0.7947894 PC22190         \n",
       "2505 8ae518dbb29595b3f79214be0b589066 SC552975 0.7427614 PC22190         \n",
       "2514 8ae518dbb29595b3f79214be0b589066 SC552961 0.7062563 PC22192         \n",
       "822  333bbf9224442fc10cd497377d2b1d01 SC552952 0.4305763 PC22190         \n",
       "1811 675c847bccbc53942ebb7b8cbb4efc4d SC552966 0.3654507 PC22192         \n",
       "     Run.ID Project.ID CGR.Sample.ID Sample.Type Sample.Des Subject.ID ...\n",
       "2512 Run4   NP0084-MB6 SC553068      Ext_Control Ext_Blank  Water      ...\n",
       "2507 Run4   NP0084-MB6 SC553048      Ext_Control Ext_Blank  Water      ...\n",
       "2505 Run4   NP0084-MB6 SC552975      Ext_Control Ext_Blank  Water      ...\n",
       "2514 Run4   NP0084-MB6 SC552961      Ext_Control Ext_Blank  Water      ...\n",
       "822  Run4   NP0084-MB6 SC552952      Ext_Control Ext_Blank  Water      ...\n",
       "1811 Run4   NP0084-MB6 SC552966      Ext_Control Ext_Blank  Water      ...\n",
       "     Unassigned frac_filtered Mag.Col  IR       Domain       \n",
       "2512 0           9.227570     Magnetic Solution D_0__Bacteria\n",
       "2507 0           7.888065     Magnetic Solution D_0__Bacteria\n",
       "2505 0           9.846112     Magnetic Solution D_0__Bacteria\n",
       "2514 0          10.033773     Magnetic Solution D_0__Bacteria\n",
       "822  0          15.504106     Magnetic Solution D_0__Bacteria\n",
       "1811 0          18.981414     Magnetic Solution D_0__Bacteria\n",
       "     Phylum             Class            Order              Family             \n",
       "2512 D_1__Firmicutes    D_2__Bacilli     D_3__Bacillales    D_4__Listeriaceae  \n",
       "2507 D_1__Firmicutes    D_2__Bacilli     D_3__Bacillales    D_4__Listeriaceae  \n",
       "2505 D_1__Firmicutes    D_2__Bacilli     D_3__Bacillales    D_4__Listeriaceae  \n",
       "2514 D_1__Firmicutes    D_2__Bacilli     D_3__Bacillales    D_4__Listeriaceae  \n",
       "822  D_1__Bacteroidetes D_2__Bacteroidia D_3__Bacteroidales D_4__Bacteroidaceae\n",
       "1811 D_1__Bacteroidetes D_2__Bacteroidia D_3__Bacteroidales D_4__Prevotellaceae\n",
       "     Genus            \n",
       "2512 D_5__Listeria    \n",
       "2507 D_5__Listeria    \n",
       "2505 D_5__Listeria    \n",
       "2514 D_5__Listeria    \n",
       "822  D_5__Bacteroides \n",
       "1811 D_5__Prevotella 9"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#Calc rel abund\n",
    "tmp = transform_sample_counts(phy_blank, function(x) {x/sum(x)} )\n",
    "tmp = psmelt(tmp)\n",
    "head(tmp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#Create rel abun plots - summary\n",
    "title = \"Phyla Composition of Extraction Blanks \\n Organized by Homogenization Method\"\n",
    "file_name = \"relabun_extstd_blanks_summary\"\n",
    "rel_abun_plot(tmp,\"Ext.Kit\",\"Abundance\",\"Ext.Kit\",\"Phylum\",title,file_name)\n",
    "\n",
    "#Create rel abun plots - individual\n",
    "phy_tmp <- phy_blank %>%\n",
    "    tax_glom(taxrank = \"Phylum\") %>%                     # agglomerate at phylum level\n",
    "    transform_sample_counts(function(x) {x/sum(x)} ) %>% # Transform to rel. abundance\n",
    "    psmelt() %>%                                         # Melt to long format\n",
    "    group_by(Ext.Kit) %>%\n",
    "    arrange(Ext.Kit)\n",
    "file_name = \"relabun_extstd_blanks_indiv\"\n",
    "phy_tmp = phy_tmp[order(phy_tmp$Ext.Kit),]\n",
    "rel_abun_plot(phy_tmp,\"Sample\",\"Abundance\",\"Sample\",\"Phylum\",title,file_name)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h3> Homogenization, Automation Assessment <h3>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#List of all variables to test\n",
    "test_list = unique(sample_data(phy_r)$Ext.Kit)\n",
    "test_list = test_list[c(-4,-5,-6)] #two protocols are not included in assessment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\n",
       "Ext_Control Seq_Control       Study \n",
       "        155          36         160 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "\n",
       "      A00       A01       A02       A03       M16       M22       M98 NTC_Blank \n",
       "        5         5         5         5         5        24        49         2 \n",
       "PCR_Blank     Water       Z00       Z05       Z06       Z10       Z11   human_1 \n",
       "        2        50        55         5         5        21         2       111 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "\n",
       "Ext_Control Seq_Control       Study \n",
       "        117          30         135 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "\n",
       "    A00     A01     A02     A03     M16     M22     M98   Water     Z00     Z05 \n",
       "      4       5       5       5       4      23      47      21      51       5 \n",
       "    Z06     Z10     Z11 human_1 \n",
       "      5      18       1      88 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#Summarize data\n",
    "temp = as.data.frame(sample_data(phy_obj))\n",
    "table(temp$Sample.Type)\n",
    "table(temp$Subject.ID)\n",
    "write.csv(ddply(temp, .(temp$Ext.Kit, temp$Subject.ID, temp$Homo.Status), nrow),paste(stats_dir,\"counts_pre_hmo.csv\",sep=\"\"))\n",
    "\n",
    "#Summarize data\n",
    "temp = as.data.frame(sample_data(phy_r))\n",
    "table(temp$Sample.Type)\n",
    "table(temp$Subject.ID)\n",
    "write.csv(ddply(temp, .(temp$Ext.Kit, temp$Subject.ID, temp$Homo.Status), nrow),paste(stats_dir,\"counts_post_hmo.csv\",sep=\"\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Z MagBead\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in cor.test.default(my_data$obs, my_data$exp, method = \"spearman\"):\n",
      "\"Cannot compute exact p-value with ties\"\n",
      "Warning message in cor.test.default(my_data$obs, my_data$exp, method = \"spearman\"):\n",
      "\"Cannot compute exact p-value with ties\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Z MagBead\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in cor.test.default(my_data$obs, my_data$exp, method = \"spearman\"):\n",
      "\"Cannot compute exact p-value with ties\"\n",
      "Warning message in cor.test.default(my_data$obs, my_data$exp, method = \"spearman\"):\n",
      "\"Cannot compute exact p-value with ties\"\n",
      "`geom_smooth()` using formula 'y ~ x'\n",
      "\n",
      "`geom_smooth()` using formula 'y ~ x'\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Warning message:\n",
      "\"Removed 1 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 1 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 1 rows containing missing values (geom_segment).\"\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Warning message:\n",
      "\"Removed 1 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 1 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 1 rows containing missing values (geom_segment).\"\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Q PowerSoil\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in cor.test.default(my_data$obs, my_data$exp, method = \"spearman\"):\n",
      "\"Cannot compute exact p-value with ties\"\n",
      "Warning message in cor.test.default(my_data$obs, my_data$exp, method = \"spearman\"):\n",
      "\"Cannot compute exact p-value with ties\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Q PowerSoil\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in cor.test.default(my_data$obs, my_data$exp, method = \"spearman\"):\n",
      "\"Cannot compute exact p-value with ties\"\n",
      "Warning message in cor.test.default(my_data$obs, my_data$exp, method = \"spearman\"):\n",
      "\"Cannot compute exact p-value with ties\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Q PowerSoil\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in cor.test.default(my_data$obs, my_data$exp, method = \"spearman\"):\n",
      "\"Cannot compute exact p-value with ties\"\n",
      "Warning message in cor.test.default(my_data$obs, my_data$exp, method = \"spearman\"):\n",
      "\"Cannot compute exact p-value with ties\"\n",
      "`geom_smooth()` using formula 'y ~ x'\n",
      "\n",
      "`geom_smooth()` using formula 'y ~ x'\n",
      "\n",
      "`geom_smooth()` using formula 'y ~ x'\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message:\n",
      "\"Removed 1 rows containing missing values (geom_segment).\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message:\n",
      "\"Removed 1 rows containing missing values (geom_segment).\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message:\n",
      "\"Removed 1 rows containing missing values (geom_segment).\"\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message:\n",
      "\"Removed 1 rows containing missing values (geom_segment).\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message:\n",
      "\"Removed 1 rows containing missing values (geom_segment).\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message:\n",
      "\"Removed 1 rows containing missing values (geom_segment).\"\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Q PowerMicrobiome\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in cor.test.default(my_data$obs, my_data$exp, method = \"spearman\"):\n",
      "\"Cannot compute exact p-value with ties\"\n",
      "Warning message in cor.test.default(my_data$obs, my_data$exp, method = \"spearman\"):\n",
      "\"Cannot compute exact p-value with ties\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Q PowerMicrobiome\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in cor.test.default(my_data$obs, my_data$exp, method = \"spearman\"):\n",
      "\"Cannot compute exact p-value with ties\"\n",
      "Warning message in cor.test.default(my_data$obs, my_data$exp, method = \"spearman\"):\n",
      "\"Cannot compute exact p-value with ties\"\n",
      "`geom_smooth()` using formula 'y ~ x'\n",
      "\n",
      "`geom_smooth()` using formula 'y ~ x'\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Q PowerSoil Pro\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in cor.test.default(my_data$obs, my_data$exp, method = \"spearman\"):\n",
      "\"Cannot compute exact p-value with ties\"\n",
      "Warning message in cor.test.default(my_data$obs, my_data$exp, method = \"spearman\"):\n",
      "\"Cannot compute exact p-value with ties\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Q PowerSoil Pro\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in cor.test.default(my_data$obs, my_data$exp, method = \"spearman\"):\n",
      "\"Cannot compute exact p-value with ties\"\n",
      "Warning message in cor.test.default(my_data$obs, my_data$exp, method = \"spearman\"):\n",
      "\"Cannot compute exact p-value with ties\"\n",
      "`geom_smooth()` using formula 'y ~ x'\n",
      "\n",
      "`geom_smooth()` using formula 'y ~ x'\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"T MagMax\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in cor.test.default(my_data$obs, my_data$exp, method = \"spearman\"):\n",
      "\"Cannot compute exact p-value with ties\"\n",
      "Warning message in cor.test.default(my_data$obs, my_data$exp, method = \"spearman\"):\n",
      "\"Cannot compute exact p-value with ties\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"T MagMax\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in cor.test.default(my_data$obs, my_data$exp, method = \"spearman\"):\n",
      "\"Cannot compute exact p-value with ties\"\n",
      "Warning message in cor.test.default(my_data$obs, my_data$exp, method = \"spearman\"):\n",
      "\"Cannot compute exact p-value with ties\"\n",
      "`geom_smooth()` using formula 'y ~ x'\n",
      "\n",
      "`geom_smooth()` using formula 'y ~ x'\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for (variable in test_list){\n",
    "    \n",
    "    #Subset dataframe\n",
    "    OTUdf_clean2 = subset(OTUdf_clean,Ext.Kit==variable & Homo.Status!=\"Standard\" & Subject.ID!=\"Ext_Blank\")\n",
    "    \n",
    "    #Create TP,FP,FN df\n",
    "    CONTdf_use = CONTdf_s\n",
    "    F1df = bench_compare(OTUdf_clean2, CONTdf_use)\n",
    "    \n",
    "    #Create TP,FP,FN counts    \n",
    "    bench_key = \"hmo\"\n",
    "    F1df_counts = bench_counts(OTUdf_clean2,F1df,variable, bench_key)#swapped variable and benchkey from original\n",
    "    temp = F1df_counts\n",
    "    temp$sample_name = rownames(temp)\n",
    "    rownames(temp) = c()\n",
    "    mergedf_f1stats = rbind(mergedf_f1stats,temp)\n",
    "\n",
    "    #Create violin plots for each control - counts\n",
    "    bench_key = variable\n",
    "    x_list = c(\"FN\",\"TP\",\"FP\")\n",
    "    y_labs = c(\"False negative\",\"True positive\", \"False positive\")\n",
    "    plot_f1_violin(F1df_counts,x_list,y_labs,\"Subject.ID\",bench_key,\"counts_exthmo\")\n",
    "    \n",
    "    #Create violin plots for each control - summary\n",
    "    bench_key = variable\n",
    "    x_list = c(\"TAR\",\"TDR\",\"F1\")\n",
    "    y_labs = c(\"Taxonomic Accuracy Rate\",\"Taxonomic Detection Rate\", \"F-Measure\")\n",
    "    plot_f1_violin(F1df_counts,x_list,y_labs,\"Subject.ID\",bench_key,\"summary_exthmo\")\n",
    "    \n",
    "    #calculate relative abundance\n",
    "    RAdf = cal_relabun(OTUdf_clean2)\n",
    "    RAdf$Sample_name = rownames(RAdf)\n",
    "    RAdf_merged=rbind(RAdf_merged,RAdf)\n",
    "\n",
    "    #Create scatter plots\n",
    "    #Determine expected to observed scatter plots, normality, spearman coeff \n",
    "    title = paste(\"Scatterplots of controls homogenization variation: \\n\", variable,sep=\"\")\n",
    "    EOdf_stats = cal_EO_kit(RAdf,CONTdf_use,variable,\"hmo\",title)    \n",
    "    \n",
    "    #Create averaged OTU df\n",
    "    bench_key = \"hmo\"\n",
    "    dissdf = create_av_RAdf(RAdf,variable,bench_key,\"hmo\") #swap order of bench and variable, include hmo for 2nd option\n",
    "    dissdf_s = bind_rows(dissdf_s,dissdf)\n",
    "    \n",
    "    #save files\n",
    "    write.csv(F1df_counts,paste(stats_dir,\"F1_vals_exthmo_\",variable,\".csv\",sep=\"\"))#swap bench for variable\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in `[<-.factor`(`*tmp*`, thisvar, value = 0):\n",
      "\"invalid factor level, NA generated\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message:\n",
      "\"Removed 3 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 1 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 2 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 2 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 2 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 2 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 2 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 2 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 2 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 2 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 2 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 2 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 1 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 2 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 3 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 4 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 4 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 2 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 3 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 3 rows containing missing values (geom_segment).\"\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#Create distances matrix and plots\n",
    "Distdf_merged_hmo=data.frame()\n",
    "if(length(test_list)>1){\n",
    "    #add controls\n",
    "    RAdf_oe = RAdf_addcontrols(\"CONTdf_s\",\"dissdf_s\",\"hmo\")\n",
    "    RAdf_clean = calc_dist(RAdf_oe,CONTdf_s,\"hmo\")\n",
    "    plot_dist_cluster(RAdf_clean,\"exthmo\")\n",
    "    Distdf = plot_dist_violin_hmo(RAdf_clean,\"exthmo\")\n",
    "} \n",
    "Distdf_merged_hmo = bind_rows(Distdf_merged_hmo,Distdf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "'nperm' >= set of all permutations: complete enumeration.\n",
      "\n",
      "Set of permutations < 'minperm'. Generating entire set.\n",
      "\n",
      "'nperm' >= set of all permutations: complete enumeration.\n",
      "\n",
      "Set of permutations < 'minperm'. Generating entire set.\n",
      "\n",
      "'nperm' >= set of all permutations: complete enumeration.\n",
      "\n",
      "Set of permutations < 'minperm'. Generating entire set.\n",
      "\n",
      "'nperm' >= set of all permutations: complete enumeration.\n",
      "\n",
      "Set of permutations < 'minperm'. Generating entire set.\n",
      "\n",
      "'nperm' >= set of all permutations: complete enumeration.\n",
      "\n",
      "Set of permutations < 'minperm'. Generating entire set.\n",
      "\n",
      "'nperm' >= set of all permutations: complete enumeration.\n",
      "\n",
      "Set of permutations < 'minperm'. Generating entire set.\n",
      "\n",
      "Set of permutations < 'minperm'. Generating entire set.\n",
      "\n",
      "Set of permutations < 'minperm'. Generating entire set.\n",
      "\n",
      "'nperm' >= set of all permutations: complete enumeration.\n",
      "\n",
      "Set of permutations < 'minperm'. Generating entire set.\n",
      "\n",
      "'nperm' >= set of all permutations: complete enumeration.\n",
      "\n",
      "Set of permutations < 'minperm'. Generating entire set.\n",
      "\n",
      "'nperm' >= set of all permutations: complete enumeration.\n",
      "\n",
      "Set of permutations < 'minperm'. Generating entire set.\n",
      "\n",
      "Set of permutations < 'minperm'. Generating entire set.\n",
      "\n",
      "'nperm' >= set of all permutations: complete enumeration.\n",
      "\n",
      "Set of permutations < 'minperm'. Generating entire set.\n",
      "\n",
      "'nperm' >= set of all permutations: complete enumeration.\n",
      "\n",
      "Set of permutations < 'minperm'. Generating entire set.\n",
      "\n",
      "'nperm' >= set of all permutations: complete enumeration.\n",
      "\n",
      "Set of permutations < 'minperm'. Generating entire set.\n",
      "\n",
      "'nperm' >= set of all permutations: complete enumeration.\n",
      "\n",
      "Set of permutations < 'minperm'. Generating entire set.\n",
      "\n",
      "'nperm' >= set of all permutations: complete enumeration.\n",
      "\n",
      "Set of permutations < 'minperm'. Generating entire set.\n",
      "\n",
      "'nperm' >= set of all permutations: complete enumeration.\n",
      "\n",
      "Set of permutations < 'minperm'. Generating entire set.\n",
      "\n",
      "'nperm' >= set of all permutations: complete enumeration.\n",
      "\n",
      "Set of permutations < 'minperm'. Generating entire set.\n",
      "\n",
      "'nperm' >= set of all permutations: complete enumeration.\n",
      "\n",
      "Set of permutations < 'minperm'. Generating entire set.\n",
      "\n",
      "'nperm' >= set of all permutations: complete enumeration.\n",
      "\n",
      "Set of permutations < 'minperm'. Generating entire set.\n",
      "\n",
      "'nperm' >= set of all permutations: complete enumeration.\n",
      "\n",
      "Set of permutations < 'minperm'. Generating entire set.\n",
      "\n",
      "'nperm' >= set of all permutations: complete enumeration.\n",
      "\n",
      "Set of permutations < 'minperm'. Generating entire set.\n",
      "\n",
      "'nperm' >= set of all permutations: complete enumeration.\n",
      "\n",
      "Set of permutations < 'minperm'. Generating entire set.\n",
      "\n",
      "'nperm' >= set of all permutations: complete enumeration.\n",
      "\n",
      "Set of permutations < 'minperm'. Generating entire set.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#Run stats on distances of relative abundances\n",
    "ttestdf_hmo = cal_ttest_hmo(mergedf_f1stats)\n",
    "write.csv(ttestdf_hmo,paste(stats_dir,'ttest_hmo.csv',sep=\"\"))\n",
    "\n",
    "metadf_phy_r = as.data.frame(sample_data(phy_r))\n",
    "metadf_clean = subset(metadf_phy_r,select=c(\"Ext.Kit\",\"Ext.Robotics\",\"IR\",\"Mag.Col\"))\n",
    "metadf_clean= metadf_clean[!duplicated(metadf_clean$Ext.Kit), ]\n",
    "rownames(metadf_clean)=metadf_clean$Ext.Kit\n",
    "metadf_clean$Ext.Robotics <- as.character(metadf_clean$Ext.Robotics)\n",
    "metadf_clean$IR <- as.character(metadf_clean$IR)\n",
    "metadf_clean$Mag.Col <- as.character(metadf_clean$Mag.Col)\n",
    "\n",
    "\n",
    "ad_stat = adonis_bc_hmo(RAdf_oe)\n",
    "write.csv(ad_stat,paste(stats_dir,\"adtest_summary.csv\",sep=\"\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"Removed 1 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 1 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 1 rows containing missing values (geom_segment).\"\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Warning message:\n",
      "\"Removed 1 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 1 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 1 rows containing missing values (geom_segment).\"\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Warning message:\n",
      "\"Removed 2 rows containing non-finite values (stat_ydensity).\"\n",
      "Warning message:\n",
      "\"Removed 2 rows containing non-finite values (stat_summary).\"\n",
      "Warning message:\n",
      "\"Removed 1 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 1 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 1 rows containing missing values (geom_segment).\"\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message:\n",
      "\"Removed 3 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 1 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 2 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 2 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 2 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 2 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 2 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 2 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 2 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 2 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 2 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 1 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 1 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 1 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 2 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 3 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 1 rows containing missing values (geom_segment).\"\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message:\n",
      "\"Removed 3 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 1 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 2 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 2 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 2 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 2 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 2 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 2 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 2 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 2 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 2 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 1 rows containing missing values (geom_segment).\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"Removed 1 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 1 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 2 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 3 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 1 rows containing missing values (geom_segment).\"\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Warning message:\n",
      "\"Removed 2 rows containing non-finite values (stat_ydensity).\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message:\n",
      "\"Removed 2 rows containing non-finite values (stat_summary).\"\n",
      "Warning message:\n",
      "\"Removed 3 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 1 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 2 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 2 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 2 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 2 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 2 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 2 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 2 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 2 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 2 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 1 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 1 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 1 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 2 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 3 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 1 rows containing missing values (geom_segment).\"\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message:\n",
      "\"Removed 2 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 1 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 1 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 1 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 1 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 2 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 2 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 3 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 3 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 3 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 2 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 2 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 3 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 3 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 2 rows containing missing values (geom_segment).\"\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message:\n",
      "\"Removed 2 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 1 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 1 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 1 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 1 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 2 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 2 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 3 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 3 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 3 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 2 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 2 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 3 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 3 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 2 rows containing missing values (geom_segment).\"\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Warning message:\n",
      "\"Removed 2 rows containing non-finite values (stat_ydensity).\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message:\n",
      "\"Removed 2 rows containing non-finite values (stat_summary).\"\n",
      "Warning message:\n",
      "\"Removed 2 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 1 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 1 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 1 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 1 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 2 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 2 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 3 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 3 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 3 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 2 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 2 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 3 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 3 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 2 rows containing missing values (geom_segment).\"\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message:\n",
      "\"Removed 3 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 2 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 2 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 2 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 2 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 1 rows containing missing values (geom_segment).\"\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message:\n",
      "\"Removed 3 rows containing missing values (geom_segment).\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"Removed 2 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 2 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 2 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 2 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 1 rows containing missing values (geom_segment).\"\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Warning message:\n",
      "\"Removed 2 rows containing non-finite values (stat_ydensity).\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message:\n",
      "\"Removed 2 rows containing non-finite values (stat_summary).\"\n",
      "Warning message:\n",
      "\"Removed 3 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 2 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 2 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 2 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 2 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 1 rows containing missing values (geom_segment).\"\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message:\n",
      "\"Removed 3 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 2 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 2 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 2 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 1 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 1 rows containing missing values (geom_segment).\"\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message:\n",
      "\"Removed 3 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 2 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 2 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 2 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 1 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 1 rows containing missing values (geom_segment).\"\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Warning message:\n",
      "\"Removed 2 rows containing non-finite values (stat_ydensity).\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message:\n",
      "\"Removed 2 rows containing non-finite values (stat_summary).\"\n",
      "Warning message:\n",
      "\"Removed 3 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 2 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 2 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 2 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 1 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 1 rows containing missing values (geom_segment).\"\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#Overall violin plots\n",
    "x_list = c(\"TAR\",\"TDR\",\"F1\")\n",
    "y_labs = c(\"Taxonomic Accuracy Rate\",\"Taxonomic Detection Rate\", \"F-Measure\")\n",
    "\n",
    "#add homo.status\n",
    "mergedf_f1stats_hmo=mergedf_f1stats\n",
    "for(row in 1:nrow(mergedf_f1stats)){\n",
    "    sample = mergedf_f1stats[row,\"sample_name\"]\n",
    "    mergedf_f1stats_hmo[row,\"Homo.Status\"] = metadata[sample,\"Homo.Status\"]\n",
    "    mergedf_f1stats_hmo[row,\"Mag.Col\"] = metadata[sample,\"Mag.Col\"]\n",
    "    mergedf_f1stats_hmo[row,\"IR\"] = metadata[sample,\"IR\"]\n",
    "}\n",
    "#plot TAR,TDR,F1 individually, faceted  by homo status\n",
    "plot_f1_violin_hmo(mergedf_f1stats_hmo,x_list,y_labs,\"Homo.Status\",\"Ext Protocols Homogenization\",\"summary\",1)\n",
    "plot_f1_violin_hmo(mergedf_f1stats_hmo,x_list,y_labs,\"bmark\",\"Ext Protocols Homogenization\",\"summary_kit\")\n",
    "plot_f1_violin_hmo(mergedf_f1stats_hmo,x_list,y_labs,\"bmark\",\"Ext Protocols Homogenization\",\"summary_inverted\",3)\n",
    "plot_f1_violin_hmo(mergedf_f1stats_hmo,x_list,y_labs,\"Mag.Col\",\"Ext Protocols Homogenization\",\"summary_mc\")\n",
    "plot_f1_violin_hmo(mergedf_f1stats_hmo,x_list,y_labs,\"IR\",\"Ext Protocols Homogenization\",\"summary_ir\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h3> Human Assessment <h3>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "phyloseq-class experiment-level object\n",
       "otu_table()   OTU Table:         [ 1818 taxa and 111 samples ]\n",
       "sample_data() Sample Data:       [ 111 samples by 27 sample variables ]\n",
       "tax_table()   Taxonomy Table:    [ 1818 taxa by 7 taxonomic ranks ]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#Create object\n",
    "sub_list<-c(\"human_1\")\n",
    "phy_hum<-subset_samples(phy_obj, Sample.Des %in% sub_list)\n",
    "phy_hum"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"Removed 7 rows containing non-finite values (stat_boxplot).\"\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#Alpha div before rarefaction\n",
    "alpha_list = c(\"Observed\",\"ACE\",\"Shannon\",\"Simpson\",\"InvSimpson\")\n",
    "title = \"Alpha Diversity Plot of Unraryfied Data\"\n",
    "file_name = \"hum_unrare_extkit\"\n",
    "alpha_meas(phy_hum,alpha_list,\"Ext.Kit\",\"Ext.Kit\",title,file_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "phyloseq-class experiment-level object\n",
       "otu_table()   OTU Table:         [ 1206 taxa and 88 samples ]\n",
       "sample_data() Sample Data:       [ 88 samples by 27 sample variables ]\n",
       "tax_table()   Taxonomy Table:    [ 1206 taxa by 7 taxonomic ranks ]\n",
       "phy_tree()    Phylogenetic Tree: [ 1206 tips and 1205 internal nodes ]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "phyloseq-class experiment-level object\n",
       "otu_table()   OTU Table:         [ 1206 taxa and 69 samples ]\n",
       "sample_data() Sample Data:       [ 69 samples by 27 sample variables ]\n",
       "tax_table()   Taxonomy Table:    [ 1206 taxa by 7 taxonomic ranks ]\n",
       "phy_tree()    Phylogenetic Tree: [ 1206 tips and 1205 internal nodes ]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "phyloseq-class experiment-level object\n",
       "otu_table()   OTU Table:         [ 1206 taxa and 19 samples ]\n",
       "sample_data() Sample Data:       [ 19 samples by 27 sample variables ]\n",
       "tax_table()   Taxonomy Table:    [ 1206 taxa by 7 taxonomic ranks ]\n",
       "phy_tree()    Phylogenetic Tree: [ 1206 tips and 1205 internal nodes ]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "phyloseq-class experiment-level object\n",
       "otu_table()   OTU Table:         [ 1206 taxa and 5 samples ]\n",
       "sample_data() Sample Data:       [ 5 samples by 27 sample variables ]\n",
       "tax_table()   Taxonomy Table:    [ 1206 taxa by 7 taxonomic ranks ]\n",
       "phy_tree()    Phylogenetic Tree: [ 1206 tips and 1205 internal nodes ]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#subset samples\n",
    "phy_hum <-subset_samples(phy_r_t, Sample.Des %in% sub_list)\n",
    "phy_hum\n",
    "\n",
    "#subset for standard hmo\n",
    "phy_hum_std = subset_samples(phy_hum, Homo.Status==\"Standard\")\n",
    "phy_hum_std\n",
    "\n",
    "#susbet for hmo\n",
    "phy_hum_hmo = subset_samples(phy_hum, Homo.Status!=\"Standard\")\n",
    "phy_hum_hmo\n",
    "\n",
    "#subset for single samples\n",
    "#sample_list = c(\"SC553000\",\"SC553034\",\"SC553038\",\"SC553042\",\"SC553046\")#34 sucks\n",
    "#sample_list = c(\"SC553000\",\"SC553050\",\"SC553038\",\"SC553042\",\"SC553046\") #50 sucks\n",
    "sample_list = c(\"SC553000\",\"SC553084\",\"SC553038\",\"SC553042\",\"SC553046\") \n",
    "phy_hum_sub = subset_samples(phy_hum_hmo, CGR.Sample.ID %in% sample_list)\n",
    "phy_hum_sub"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\n",
       "Study \n",
       "   69 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "\n",
       "human_1 \n",
       "     69 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "\n",
       "Study \n",
       "   19 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "\n",
       "human_1 \n",
       "     19 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A data.frame: 4  3</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>temp$Ext.Kit</th><th scope=col>temp$Subject.ID</th><th scope=col>V1</th></tr>\n",
       "\t<tr><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>Q PowerMicrobiome</td><td>human_1</td><td>11</td></tr>\n",
       "\t<tr><td>Q PowerSoil      </td><td>human_1</td><td> 4</td></tr>\n",
       "\t<tr><td>Q PowerSoil Pro  </td><td>human_1</td><td> 3</td></tr>\n",
       "\t<tr><td>Z MagBead        </td><td>human_1</td><td> 1</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 4  3\n",
       "\\begin{tabular}{lll}\n",
       " temp\\$Ext.Kit & temp\\$Subject.ID & V1\\\\\n",
       " <fct> & <fct> & <int>\\\\\n",
       "\\hline\n",
       "\t Q PowerMicrobiome & human\\_1 & 11\\\\\n",
       "\t Q PowerSoil       & human\\_1 &  4\\\\\n",
       "\t Q PowerSoil Pro   & human\\_1 &  3\\\\\n",
       "\t Z MagBead         & human\\_1 &  1\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 4  3\n",
       "\n",
       "| temp$Ext.Kit &lt;fct&gt; | temp$Subject.ID &lt;fct&gt; | V1 &lt;int&gt; |\n",
       "|---|---|---|\n",
       "| Q PowerMicrobiome | human_1 | 11 |\n",
       "| Q PowerSoil       | human_1 |  4 |\n",
       "| Q PowerSoil Pro   | human_1 |  3 |\n",
       "| Z MagBead         | human_1 |  1 |\n",
       "\n"
      ],
      "text/plain": [
       "  temp$Ext.Kit      temp$Subject.ID V1\n",
       "1 Q PowerMicrobiome human_1         11\n",
       "2 Q PowerSoil       human_1          4\n",
       "3 Q PowerSoil Pro   human_1          3\n",
       "4 Z MagBead         human_1          1"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#Summarize data\n",
    "temp = as.data.frame(sample_data(phy_hum_std))\n",
    "table(temp$Sample.Type)\n",
    "table(temp$Subject.ID)\n",
    "write.csv(ddply(temp, .(temp$Ext.Kit, temp$Subject.ID), nrow),paste(stats_dir,\"counts_hum_std.csv\",sep=\"\"))\n",
    "\n",
    "#Summarize data\n",
    "temp = as.data.frame(sample_data(phy_hum_hmo))\n",
    "table(temp$Sample.Type)\n",
    "table(temp$Subject.ID)\n",
    "ddply(temp, .(temp$Ext.Kit, temp$Subject.ID), nrow)\n",
    "write.csv(ddply(temp, .(temp$Ext.Kit, temp$Subject.ID), nrow),paste(stats_dir,\"counts_hum_hmo.csv\",sep=\"\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#Alpha diversity, other metrics\n",
    "alpha_list = c(\"Observed\",\"ACE\",\"Shannon\",\"Simpson\",\"InvSimpson\")\n",
    "title = \"Alpha Diversity Plot of Unraryfied Data\"\n",
    "file_name = \"hum_rare_extkit\"\n",
    "alpha_meas(phy_hum,alpha_list,\"Ext.Kit\",\"Ext.Kit\",title,file_name)\n",
    "\n",
    "#std only\n",
    "alpha_list = c(\"Observed\",\"ACE\",\"Shannon\",\"Simpson\",\"InvSimpson\")\n",
    "title = \"Alpha Diversity Plot of Raryfied Data: \\n Std Processing\"\n",
    "file_name = \"hum_std_extkit\"\n",
    "alpha_meas(phy_hum_std,alpha_list,\"Ext.Kit\",\"Ext.Kit\",title,file_name)\n",
    "\n",
    "#hmo only\n",
    "alpha_list = c(\"Observed\",\"ACE\",\"Shannon\",\"Simpson\",\"InvSimpson\")\n",
    "title = \"Alpha Diversity Plot of Raryfied Data: \\n Non-Standard Processing\"\n",
    "file_name = \"hum_hmo_extkit\"\n",
    "alpha_meas(phy_hum_hmo,alpha_list,\"Ext.Kit\",\"Ext.Kit\",title,file_name)\n",
    "\n",
    "#break out each metric\n",
    "alpha_list = c(\"Observed\",\"ACE\",\"Shannon\",\"Simpson\",\"InvSimpson\")\n",
    "for (alpha in alpha_list){\n",
    "    title = paste(\"Alpha Diversity Plot of Raryfied Data: \\n\",alpha,\" - All Processing Methods\",sep=\"\")\n",
    "    file_name = paste(\"hum_rare_extkit_\",alpha,sep=\"\")\n",
    "    alpha_meas(phy_hum,alpha,\"Ext.Kit\",\"Ext.Kit\",title,file_name,\"Homo.Status\")\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Saving 6.67 x 6.67 in image\n",
      "\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA0gAAANICAMAAADKOT/pAAAATlBMVEUAAAAAsPYAv30aGhoz\nMzNNTU1oaGhycnJ8fHyBgYGMjIyampqjpQCnp6eysrK9vb3Hx8fQ0NDZ2dnh4eHna/Pp6enr\n6+vw8PD4dm3////j+RiAAAAACXBIWXMAABJ0AAASdAHeZh94AAAgAElEQVR4nO2di5qjKLtG\n2fOPmUwyqapUdTrt/d/ojmdA+DjLIe96nm6TAAriKgEVWQ8ACIblzgAALQCRAIgARAIgAhAJ\ngAhAJAAiAJEAiABE8uTGGLsuX16f9zGUP8oxRs63h20S3XqstnP5DNkKoMAu9WQ8Mp/bF2UM\nm3VM3C2T6NYzfbhqw2duxFY0iYEVEMmP+3hcfs3fwkVi33ZJdOsZFt+dJrUkrHIr2sTACuw9\nP86vv+6MneZv/iINi+f9xFgXniftBhfRXpm+6OKhvRcG9p4Xj0Gi1+H/M30NE+nFaWnchWAS\nacg108WDSGFg73nxwdjH8N9t+jofhcPi1rHT1/Lj8/Yy5Drb9nka+vvf3Fq2g/c+jVyMP3wv\noxjXqcH3uHasm4cjXhEep+G0Mq6ZXb629cxtt345T542M9btSCJ9vk5Rp485gEGlELDvvOiG\ngYbn2iDbRDoLffpuOj5Hk87CsAKf7MVzOvynHxYFptXf+XSv5Sv0s390y4Dfup7FhZd+g3QP\nxaDijyDSso7uAZHCwb7z4T51Ni7SaNtwVP70P92kzvDtexwnH47oT3Z+Dv9Ph37PJ9s+Tv9/\nTKMYX8NZb7R1XGf3nCIMqxmsfK36eR6kEk6IY94+ppXcpe0MfaTr9rUbN/PFywi8wd7z4Tod\npbNP/JE8/3zdvj2X08yDi9n30hdepPn0dBqH12/TWr4nPbgz072XTmSbIHNqfjsLP2u8r0nC\nwe0vOWPAGew9D9Y2XTdfStqO5OcWzNYW2ryTn98fZxuRhjPdY2ibDaeKrZ133rYw/DpcXxWv\n486L25z6ym9nhruOdJmagENM7VAesAZ7z4NP7k/80lfnFlKfnmtYSR0RTR9pOKXdllMRtyku\nxff804k/zS0bmgdC7vx2xs7Q9VuRPfkr8AJ7z4MTd3SPQ2QWIg1jBtevh0aku9B7eZ3puvFf\nrxOpf97moQLuCuuW+iS07GRJIFICsPfc+WE8w1/57UgezhDi6WUbivvppeN1+3IW7zm4jZd7\nx3aYYISQ/PE1yMQ1IrnUd75lpxYJTbuoYO+5c1uvHy2XkrYjeei/34Ufhb/46jPS94Wtp5/x\nh8eo6M+8sXlY4cxFOAmdM0mkx3qqkrYjfMVgQ1Sw99xh292qT/lI7r754e8lej8e+p/9z26w\nYeXec0nGi06nZQuvdT7PwqlvcPX8GGUQx7P7dSiCv+VIKZI0/L0mBl5AJGfWQe+B83g0biJN\nIwr8CMS8nIcHttuKekEk6e7v1+G93BG7XJAVb9y+zb8uV5eG365sbtB9sO2cKaQSvi4XZKcB\nizUx8AIiOXPmm0338Q8617b67NiZHxwTRu3Od+Ww9Pn27IUkwlnv5/o64i/ycxb34cg/iw8Y\nXWbFn3PXjdtOr/o63CI0r2JLDLyASBEppp9RTEbeB+zwiJRy/N7RSjucMmq+EUoR6Sy27MAB\nlFHzjVCGSGOvK3cm3o4Sar4ZyhBpeHopdx7ejxJqHoDqgUgARAAiARABiARABCASABGASABE\nACIBEAGIBEAEIBIAEYBIAEQAIgEQAYgEQAQgkhXCo6uxV34fHoE93X5UYcfcB1vG3bZVgx1o\nRUqRlukXppkepDfnQaRKwA60IqFI3LStw9N40pvzIFIlYAdakVCkE2PD3CcP9ev0cIhXAqrJ\nioQiLSsUp8jbbxgUDarJip1I4lxb4zxcJ3b67B9XtkzHxb+hb56oawnief16fXDbYfOqv06s\n+9hPcSy8EXB6ex//6r1+nhb8PM8Yxr/uT3jL3/6Vf/tcDlm4wWUrsI+sMIs0zQx56dbOjvCG\nvjXCflaSsY90uq3CzSJNQxC7mVnFNwKO07cKr97rx4le1zj86/6Et/wpXvm3y6UqC0AD9pEV\njGf+od+WbHyR3qDE6Wd+K5/4hr7hML+P05nupyWZj91OmOzxe335nyiS+EbA8e19u7mHX3Ge\n0yyWwuv+hLf86V75x+XyR5UFoAH7yAqzSN/CQn5DHxPe3ifx2e1PDts7AUWR9m8ElGbDn16Y\nOb8RQ3jdH+Pf8sc0r/zjcjln4Qsi2YB9ZIVZpH634N/QJ0bf8f0xvnKJOzl00+IpiyS+EXA8\nIYnvZ7nwk4sLr/sT3vJHvfJvm/b/ucsC0IB9ZMV2MFmLJLyhzyDSi+dV9T7Y/WDDbqNblPGD\nnGCzX3jLH/XKP936AQH2kRXuIolv6NMflN06W75qVTFFEt/yR73yDyK5g31khbtI4hv69Afl\nde7hSNeRlO0qpnwjoEXTbmV7y19PvPIPTTt3sI+sUIk0HGXfWpGmz+Yz0nDm+nyt6vskvDTs\npurpM+UbAXeDDUOgMNjwXPpI21v+yFf+MX6w4RMi2YB9ZMVOpMt4vH531BmJe0OfutE0snSl\n2HQqmd+cpxx7Zso3Au5fvfe9jGsLr/sT3vKnfOWfmMspC98Y/rYC+8iKnUhzZ/2sFUl8Qx8h\n0mN9R/o4mL28OY+8ICu8EVB69d4yjCBckB2/CG/5U73yTzpv4oKsA9hHVuxEGk82p0/DqN36\nhj5CpNdp4cK4t/Ytb85T3J8zfNm/EbCXXr33uG1v+BNe9ye85U/1yj+5ATrd9QSRbMA+yoHn\noZnnkIZINmAfZeDp+bbWXCLh3bJmIFIGPj1fqHesSGwconis4/OAAiJl4OL5YspjRfpYBkHw\n+j8LIFJFHNy0Gydl2QYxAAVEAiACEAmACEAkACIAkQCIAEQCIAIQCYAIQCQAIgCRAIgARAIg\nAsEi/V0B7ZREX5TcGbMh9FgrmXCRYuQiLbYipc1FFAiRDsyFJxVk0R+I5BotKxCpVCCSa7Ss\nQKRSgUiu0bICkUoFIrlGywpEKhWI5BotKxCpVCCSa7SsQKRSgUiu0bICkUoFIrlGywpEKhWI\n5BotKxCpVCCSa7SsQKRSySLS/2aGj9yvoVnRkVikuSTp8s8RXaQyMt8AeUSSlqpvEUkr0mHH\n4UBskdbMQ6RAIJJrNIn/CYvURBZpyzxECiS3SONfw6WVJy63Zoe0TJTFIJH6uTRjFulyrL97\nkEakfp9ZXWWkyXwDFCDS0rxY+kzrslcvU2XRU6T1wFqzbipHQFMqtkhb5uXM6gqRJPMNUMBg\ng9DQ0wnUe1ffAaN2/5MPN33+Q/4gkHn0zj036CO1FKL+NQvJYg3kOyP9b/moFWmVjf+rmS6L\nQfX8v14r0pLvQkXq5ZxPrVPpd4hkIGPTzvQXm6uuMJVyi8Q3iIoXSdceleN5AJEoDhEprH13\n0KgdIdIBx2KMUTuHpQ8QiSJgsMHyj19wHSYebBj/7xUiGcrnQ/TBhiUvfObQtPOggFG7bWS1\nF+uqiuFvLnu7M5Ju+LtPcCzGuLNh3ulKcZJmvgFwr51rNGs8DzcS3GtXKhDJNZo1EEmigiz6\nA5Fco1kDkSQqyKI/EMk1WlYgUqlAJNdoWYFIpQKRXKNlBSKVCkRyjZYViFQqEMk1WlYgUqlA\nJNdoWYFIpYL3I1VVEkKkCgg91komxRnpN53CEByYXBHsf0ZKm9WoRYleD0dmvgEgEhUNIsVM\nDpEoIJJ3MERqCIhERYNIMZNDJAqI5B0MkRoCIlHRIFLM5BCJAiJ5B0OkhoBIVDSIFDM5RKKA\nSN7BEKkhIBIVDSLFTA6RKCCSdzBEagiIREWDSDGTQySK+CIZsmRIrkidTSTTznUvKZfH7oU6\nhEjvEJw08+0BkahoBYvUrf/JIVR6h2CI5AREoqJBJO9wiOQIRHLKi0u4WSRlCJXeIRgiOQGR\nqGhFi7T2kaan5n5HhgWG74FIFBDJKS8u4bRIg0Vo2hUDRKKilSzS+p8cQqV3CIZITtQj0vwz\n+01FKlqkXdZqEilq5hukFpEut9PtMi6ul/WbZeoyRFLkuR6RIme+QWoR6Xljt8e4uD7Wb5ap\nyxBJked6RIqc+QapRaT+fDtPi3+4b3apyxBJkeeK7myInPn2qEWk57W/PMfFv8/1m2XqMkRS\n5DnmsZj4jJQ28w1QokhhWGUxTkmMx6J7Xg1rP1KkAzPfACWKFJS8KJEMycsWiQ7GGUkEIlHR\nIJJ3OERyBCI55cUlHCJVBESiokEk73CI5AhEcsqLSzhEqgiIREWDSN7hEMkRiOSUF5dwiFQR\nEImKBpG8wyGSIxDJKS8u4RCpIiASFQ0ieYdDJEcgklNeXMIhUkUUKFJYcv8KjC9S/OQHinRk\n5hsAIlHRIFLM5BCJAiJ5B0OkhoBIVDSIFDM5RKKASN7BEKkhIBIVDSLFTA6RKCCSdzBEagiI\nREWDSDGTQyQKiOQdDJEaAiJR0SBSzOQQiQIieQdDpIaASFQ0iBQzOUSigEjewRCpISASFQ0i\nxUwOkSggkncwRGoIiERFg0gxk0MkCojkHQyRGgIiUdEgUszkEIkCInkHQ6SGqE8kQ45LmrMB\nIglAJAqI5B0MkRoCIlHRIFLM5BCJAiJ5B0OkhoBIVDSIFDM5RKKASN7BEKkhINIW7Xf5QKRS\nqVckplrgjBS0eodgiCRSq0iX2+l2kRdbsCmLKiCSSzBEEqlVpOeN3R7yYgs2ZVEFRHIJhkgi\ntYrUn2/n/aKHSEGrdwiGSCK1ivS89penvNiCTVlUAZFcgiGSSLhIRw9cMQMuQ11iSfY/QaSY\nySESRUPD385ZKelYhEh5gUhUNIgUMzlEooBI3sEQqSEgEhUNIsVMDpEoIJJ3MERqCIhERYNI\nMZNDJAqI5B0MkRoCIlHRIFLM5BCJAiJ5B0OkhoBIVDSIFDM5RKKASN7BEKkh6hMpWQVCJJdg\niCQCkahoEClmcohEAZG8gyFSQ0AkKhpEipkcIlFAJO9giNQQEImKBpFiJodIFBDJOxgiNQRE\noqLVLZKhbsOu10EkEYhERYNI3skhkiNNixR4MB15LEKkvEAkKhpE8k4OkRyBSHogkgBEooBI\neiCSAESigEh6IJIARKKASHqKFokVlPkGgEhUtHZFutxO1wu5dYjkBESioun2zvJ2s99krJJF\net7Y9UFuHSI5AZGoaOq9s73d7Cq+5Mwu+UbOpt359g+9dYjkRIEiuT9Lblh7bJG2t5tdxZec\n2SXfyHlGuvb/PsmtQyQnIBIVTbOt9e1m/3DfFOQWKYwDM98AEImKpjkjrW83+1d8yZld8g0M\nf7cDRKKiBf5Nh0gCEImibZEMSQo6FiFSXiASFQ0ieSeHSI5AJD0QSQAiUSQTSbnoIZIGiJSX\nUkXaLnrepIueEEkJRMpLqSJtFz1v0kVPiKQEIuWlVJG2i578Ygt2WTtE8tk6RHKiVJG2i54X\n6aInRFICkfJSokhBVznfVqTfO9j+JwfCUquASBQY/taT+YwU/Tb6wOQQiQIi6YFIAhCJommR\n4h9MgckhUqlAJCoaRIqZHCJRQCTvYIjUEBCJigaRYiaHSBQQyTsYIjUERKKiQaSYySESBUTy\nDoZIDQGRqGgQKWZyiEQBkbyDIVJDFChS9PqHSJG2HpgcIlFAJO9giNQQEImKBpFiJodIFBDJ\nOxgiNQREoqJBpJjJIRIFRPIOhkgNAZGoaBApZnKIRAGRvIMhUkNAJCoaRIqZHCJRQCTvYIjU\nEBCJigaRYiaHSBQQyTsYIjUERKKiQaSYySESBUTyDoZIDfGGInWaLxDJJRgiibyfSF2n+QKR\nXIIhksjbidTxJ6EOZyTfYIgk8nYi8a25Dk0772CIJAKRpigD0V+/EB+IVCopRAp7L09gcgeR\nuh5nJO9giCTyxiJ1/Jd9NIttlXQsQqS8vLNIE7poFtsq6ViESHl5Y5HkLxDJJRgiiUAkXTSL\nbZV0LEKkvLytSB3/RRXNYlslHYsQKS9vKJIOiOQSDJFEDhRp/pn9NmwYIumDIVKpHCbS5Xa6\nXcbF9bJ+88kSRFICkfJymEjPG7s9xsX1sX7zyRJEUgKR8nJc0+58O0+Lf7hvHlmCSEogUl6O\nOyNd+8tzXPz7XL/5ZAkiKYFIeUkiUiD0BiGSEoiUFwx/U9EgUszkEIkCInkHQ6SGgEhUNIgU\nMzlEooBI3sEQqSEgEhUNIsVMDpEoIJJ3MERqCIhERYNIMZNDJAqI5B2cWqToOzJw7e8u0vI8\ntrycgEjewRCpIcwiLXOEyMsZiOQdDJEaAiJR0Ux7p6BjESLlxbKP5CTS4X+nDcEQyXb1DsEQ\nSSREpL8rmZ8UItmu3iEYIonYiDQNLuCMFCMvLquHSBWBph0V7S1EYqoFRHIEIlHR3kCkbS6N\nmziXBkRyIsWoHUSyCy5CpG0ujZs4lwZEcgIiUdEsRdI0jqoQaZtLg1+ot+6xdp43F8n9zoY3\nE0nbOKpCpG0ujYs4lwZEciLFvXZvJpK2cZRdpLREzHwDQCQqml3TTtc4yi6S8+odgnFGEoFI\nVDS7M5KucQSRRCASRdsi1dM6gkh5gUhUtDcY/vYNhkgiEImKBpH0QCQBiERFg0h6IJIARKKi\nQSQ9EEkAIlHRIJIeiCQAkahoEEkPRBKASFQ0iKQHIgkUOB2XITxZBUIkl2CIJAKRqGh1ixT9\n1B6YHCJRQCTvYIjUEBCJigaRYiaHSBRNi5R2XCTqsQiR8gKRqGgQKWZyiEQBkbyDIVJDZBTJ\nc6YDiKQEIuUlm0jeMx1AJCUQKS/ZRPKe6QAiKYFIecnXtPOd6QAiKYFIecl3RvKd6QAiKYFI\neUkiUjUTHZiiQaSYySESBYa/vYMhUkNAJCoaRIqZnKwHVYvjs9tCx8WZsfv4+bPbrSAzEImK\nVrJIXeKpowsQafs8fRo9Gj+bGvjHA5GoaEWLpA3xX71DcHyRqN/GT2fWPYjYeYFIVDSIFDO5\nrUgX9tP3P+zMnZ6GD4tHr8/mMafDgUhUtIJFkjsJDYn0ZKdBmx9RpDM7r58h0gBE8k/OibR1\nkZp4KTbfRfpg9y92E5t2HWNfy2c07UYgkn9y6YzU4mDD8GX+I8GLNJj0mD9DpBGI5J9cymNT\nInFfvqbTDy/S7dW2O60xIVIPkUKSv69Iff/o2HX5DJF6iBSSvOWmHfelO53kpt3rv+9Nr/cQ\n6fAaMgQ3K1KnDAlYvUNwSpE+2P3OPvYXZD/HbhJEsguGSESwdGeDOsR/9Q7ByQYb2DT8fWLP\nYayuF8S5DkPg0/fS7hGCSFS0kkUyh9Qq0nxB9vI6A0ki9Wd2m+61g0gQKSR5uyLVjizScFvt\nN+s+rFcAkbyDIVJDSCK9unPDMCNj1iZBJO9giNQQkkgn9v369/lj3wSFSN7BEKkhJJFeJ6T7\nMGZiP7wIkbyDIVJDSMJ07HFlP0MvyXYFEMk7GCI1hCTSBxtH6IdbmyyBSN7BEKkh5CbcjXX3\n14nJ2qOWRMr9mIEFEKlUcB2JioYzUszkEIkCInkHQ6SG2In0eRke6/2xXgFE8g6GSA0hifQ8\njXc7MfZtuwKI5B0MkRpCEuk63RT4xc7q6HsgkncwRGqI/QVZx+c9IJJ3METS8cc6ZjFAJCoa\nRIqZ/J1Empt2t+npeBvcK9DkaNiT6hApYPUOwRBJRB5s6KaHq9a5YY1AJO9giKSjfpH6/uPE\n2On2VMVVApG8gyGSwB81hmyUQoYLshDJP3nTIjn8Wh4QiYoGkWImfyeRbF8wuQGRvIMhkgBE\nkoFIyZJDJPFNFPYHK9uWuuhMWjqhTPQ4p5z8RMqvuNBlyXrtEClg9Q7BKUXihxgkkSZpmPoY\nopiTUNIFdXPUiZ8pJz+ZNnm5nW4XeUFkyXbtEClg9Q7BCUUSButkkYRF8SIlvbNhWvXzxm4P\neUFlyXLtEMl29Yb9bDoCkokkDnuLIjHpA7eYmnhs/ZWtrUA2vpusnz/ysacg/gvrd+FThH77\nUYk64CvlnA3LpJm3836hz5Lt2iGS7eqLE8niOpJepFkeJv3fL/e7iSJJQeKXXTgTNqFGN9iQ\ncM6G+Yx07S9PeaHKkuPaIZLt6osTaVm6nJHWwYb10FfYMi13IvW8G0xaRS+54ydS0jkbmAk6\nOURS0o5IVB9Jc0bqJQvY2rJjghWsDxKJOjqDOlgDGP72DoZIAjajdjYiCX2hXVDoGUkHRKKi\nmUwIy6rHjnoPkYhf5z6LIML2qRCRHNpXGxDJKS8u4RBp/6v6OlLPHfKsFz4xKYhsuhnCxS1K\nGRMyCZGcttWOSLuWUqki9fwwt5DJ5aDdRNrGsHtZJCGIHP4W0jsPfzsAkZzy4hJ+mEiKq+I5\nRMJjFBIQyTJ5KSIpropnEImjFns45N11Q9POflutiKS4Kg6RHJF21w19JIdttSKS4qo4RHJE\n2l0d+zmzx/OccoLItxVpFzuzSEHXxSGSiLS7Xrvvg937Z8oJIt9UJLErb7P6Nxv+rpu9SHf2\nyQ3UG4FIlnkRu/I2q4dIFSHtrgv7erBT/40JIm225ea80JW3WT1Eqghpdw0GnYcGcsIJIt9U\nJLErb7N6iFQR8u66n4bpVu2fonhzkdLiUhSIlJcMF2RbEsk5Ly7h7yvSf9Yxi4HfXdbTFPNA\nJKe8uIRDpIoQblM8f9lPVbwAkZzy4hIOkSqC313D2/qcXYJITnlxCYdIFSHsrscwgz673F1W\nAJGc8uIS/mYi/afGkI1SkHfX5NLV3iWI5JQXl/B3E8nh1/JQ7K7JJdsVQCSnvLiEJxbp8Gp6\nN5Fe3CGSzbYgklNwHJGEC2x299cqEvbrs7KR0J2REk4QCZEsw99WJL5nJInkPfc399S49Fsc\n0EeiokGkmMltRRLGGGSRhIWDSMqYqURytujF37+LByL5bD2bSOJonSgSkz5wi6nltp13WM9P\nasLvzPXnVCK5j333b35GKuhYbEAki+FvvUizPEz6f/Kp53tIbEuRSCTc2eC4rZKOxQZEWpYu\nZ6R1sEHQg/FucYn2MaPBr8rdoh4iBQRDJAGbPpLmjNSrRWJcc2+Lll4kLyCSdzBEErAZtbMR\naW3XMTlRD5GCgiFSpK0HJo9xHWnq7igcMYm0603VL5Jpk2HJIVLA6h2Cc4mkvI7Uc7awXvi0\nmsMUP0Mkp9QQKdLWA5PHvLNhf1loGZbbRBIn+N7G7ZIPf3sBkbyDIZJAO/faGeYLUAORvIMh\nkkA7j1FAJMeslHQs1i8SRy32cMiH5eX86PvH+aKMrAIieQdDJB31i3Rh01VZZm0SRPIOhkg6\n6hdpbtI9czTtlsGV32QsiKQGIuVFOizPbGraHX9G2t4ad929PM4iObVxiBRp64HJ7UWqEOmw\nfHTTUENnPcddLJG2t8Zddy+Ps0hObRwiRdp6YPJ3Eul1BJ8YO33Y378arWm3vjXuH+6bfXJi\n4xAp0tYDk7+VSM5EOyOtb437d/fyOIvk1MYhUqStByaHSBQeIgVCrx0iBazeIRgiiewOy8/L\n8GaXH+sVYPjbOxgi6fjLOmYxSIfl8zT+zWc53yELkfTBEKlUpMPyym7DtaSvnO+QhUj6YIhU\nKooLsss/OyCSdzBE0gGRRiBSsuQti/SXGkM2SkHdtLulfIcsRPJP3rRIDr+WhzzYkO3OhljJ\nIVLA6h2CIZLI7rAcJls93TLc2RApOUQKWL1DcDaRhEuJLg/PyRG5B9bdHsFTr138an/9aAEi\neQdDJIFNGb5nJImUZBJ9h7XQq+e+ntJPWQyR/JO/hUjCGIMskrBwUEAZM51IwwtdHO5Y7WsU\naXtjTfdCG80iKyUdi+2IJI7WiSIx6QO34GYXWifcWqYLMkyivy3YPGbt3s6TEzxuHWMX6/sa\nElRg/PoXs7i5063/KaLZbKukY7EBkSyGv/UizfIw6f9+uZazycG2FPu1sJ6X0QFF/O8bY6cv\n2xXUJlLXQ6QYwYWckdZhAkEPxrvFJdJMor9fSxSRXqelpLMIZT4j9Z3mG0RyCc7TR9KckXq1\nSIxr7m3RtGekPrJI39fXGenTdgVtiPT3QO4XollAiJQ7a2YijNrZiLS265icqD9QpLGPdG25\njySKhMEGz+BM15Gm7o7CEZNIu95UWpGGq7GfiUftIJJ/cojEtv97pQKsFz6t5jDFz+lEeodX\nX3aazxDJJTiTSH0v9HuY/DsvkvUk+pJIMYa/T9Y3qy5ULVJHRLPYVknHYgMitXT3d+dsYs0i\ndVQ0i22VdCzWLxJHLfZwSOL8nHXTyemoVaRuurGhw/C3XzBEEpH7SJZz9WzUJ5IOiOQSDJFE\nIBIVDSLFTP5OIrkDkbyDIZIOiDQCkZIlfxORKmQnUvoJIiGSf3KIVCqSSEdMEAmR/JNDpFKR\nRDpigkiI5J/cSSRDs90QDJGc2I3abf/sgEjewRCpISASFQ0ieYcHifR/1jGLQd20SzpBJETy\nTw6RSkUebDhggkiI5J8cIpXKbm+mnyASIvknh0ilgguyVDRTVgOPRffkLYv0f2oM2SgFiERF\na0Ik5cJr6y7hHiI5/Foe8t787Pr+m3Uf1iuASE55CUzuLtLldrpd5IXn1l3C31ykT8b6xzDg\nYG0SRHLKS2Byd5GeNzY8YiYuPLfuEp5MpN1c+OsT5Nz/x7Ob/OT79e/zh3Xq6HsgklNeApN7\nNO3Ot/N+4bd1l/AQkfiekSSSNPkJ/32esSGTSfsLsnd2wgVZu23VINLz2l+e8sJz6y7hASIJ\nYwyySMJC+F6USB17XNnP0EuyXcHRIpnGcSDStnoDzlt3CfcXSRytE2tbJdDyQRCJmz9/vFGH\n+5IGacUfw8XYYcM32xUcLJJxRLQgkVhukWJv3SU8zfC3SST+X7/c6rZOvsV6Y6m9kdd7Y939\ndWKy9uhgkczXFooR6XI7XS9EuCG5eutNi7QsA85ITPHbZpWQIjKVXUeqSKTnjV0NN1pBJB6b\nPpJBJG7gTvW6JCFFZCASFS2obXa+/ROS/H1FIkbtTCIJ72fJKtLzNtxr5/DSPvSRNDyv/b+G\n3QiReKyuI819Hqb4zp+R9u4cLNKj+Lu/yxq1C4ReO0Ta/0peR1p+ZYtfkkgHDjac2fml0OPM\nDN3kjfcWiQ7GqJ1AzDsb2P678LIXxpjijHTc8NOukKYAABrfSURBVPe8oWexF2TradolSd60\nSKnv/k4mkWLlFzY160s9I9U02ACRJHI/j3SkSP1latpZewSR3PISmBwiBXCYSA794A2I5JSX\nwOQQqVQqEwl9JB3R6wEiOYELslQ0iBQzOUSigEhOeQlM/iYiVYhUl/fLq1V3+XJYAURyyktg\ncohUKkJdPs5zB+lk//5L9JGc8hKYHCKVilCXJ3a+vxbf5+EhWUtwZ4NTXgKTQ6RS4evyc30H\nxZl92q7g8EfNDR3RQ0Uq6FhsSqRf1jGLgRfpvL4V6bvg17pAJCUQKS/idSTVRwMQyTsYIumA\nSCPv00cq6FiESHmprWlX1qhdQcciRMoLL9JX+YMNhV1HKuhYrF+kX2oM2SgFoQl3ZufhnFTe\n8LfmURWVURApYPUOwQlEcvi1PMS+0HJB9lzYBVlxghntNJya1BAp0tYDk7+RSPMtQneHFRwr\nktRHgkhUSIMi8Y8mcA+WyxieYEjxYFIdN61ywgjtOYhEhVQrEt8zUui1TMXF+IUyhmZLEAln\nJG3khkQSxhj2Iq1nIojkHI4+0sw7iCSO1hEi9dskXEtDjp88qJdD+BjRgUhUNIgUM3mc4e9t\nSrtNJH4KyG2yrnUaO0WM6EAkKhpEipk8yhmJu/uGOyPNIasu62AD49Mw6XtMKhMJfSQd7YhE\n95F4kfjTEWPSeafnlNrHiE5lIuGMpKMhkahROyZ8XEVivXhG4mJwCSESRJp5D5GIX9UiqXpC\nwqddjOhAJCoaRIqZPMYFWekzd55h0rLfqcXHiI7FOrsXquUE+kjewRBJwEIkJt/ZsP5jjF8y\nuWu0ixEd8zq7+T95OYM7G7yDIZJArLu/U2gSY6tliYQzkob6ReLwvVM1WQ/IctNmcouExygm\nIBKNy2TbcQkR6e+R3wegFckqNUSKtPXA5G/zhKyerscZSUVBxyJEykslIq0fMPytoSmRKsRK\npI7/DyJZh0MkgbcXqdv+h0gu4RBJ4N1F6rgFRHIJh0gCby5S1823MpRwZwNE0gGR8oJ77aho\nEClmcohEsd87f/78IVNAJP/kEKlU4ov054/BpBCRcIuQDoiUl+gi/fljMilIJNy0qgEi5aU6\nkUy/Q6SA1TsEQyQRiERFg0gxk0MkioP7SMbfA0RKfOdtDCBSqRx/Rvrvv//oNaKPpA+GSKVy\nuEj//WcyCaN2+mCIVCpHi/Tff0aTcB1JHwyRSuXoPpKnSHgeaQIilcrRdzb4ibSi9YdKDZEi\nbT0wOUSiUOwdcszOq4+0ApF0QKS8HC6S16jdCkTSAZHyUtkZqTd5BJFCVu8QDJFEjhYpsI8k\nXUfao5pRECJF2npgcohEcbRI2rscRpRzc0KkSFsPTA6RKIoSST3LbbMiddoQz9U7BEMkkdr6\nSBBpo4NI5YAzEhWtaJE6nJEKoi2R3qmP1KFpVxKNifRGo3abSMfNwR4GRKIoqo/UK6eNblKk\ndRbpXUjA6h2CcUYSae2MFFck094xhCc8FqVpOj1ECvx7BpFEIBIVrWCR1nk75RDL1YdWA0QS\nqezu7x4icQSckcz1AJGcOHzOhqr6SBDJOxwiOXLInA08EGkFIpVDhum4DB5BJCI42r12qUUK\nuAxRJzgjUdEgkmd4yIXxOkEfiYrGxKW46CGSjqBbteqktVG7uHc2THvncjvdLvKCC9cCkQQg\nEoXjGamue+2mvfO8sdtDXnDhWsoT6T8dzpmDSCJ1zGvHc+Td3/PeOd/O+wUXrqNAkdYPYj1E\nFgl9JHdcBxuS9JF+aVBnUYP+jHTtL095wYVrKVgkqR5ii4RRO2eKGLVbKk06I4WLZIJeY8Ei\nJT4jhQz61Ekskf5oUKRIcB1pU0doUkQ6I+l5Y5EMk9BAJFcWkZbvokAQKSC8ZJFM06KhaeeK\nLJI4ard8cBgtKq+PpOdtRTJN1InBBmcsRdKkhkhuwZWIhOFvd+yadlxFCachd5GU1QeRyhAp\nSj3UyeEiGYddLWpw9yv6SEdfkKXrIWj0tE6OFsn8l5CsQd1fwu0T36CASDIHiRT0B61Oju4j\nJRYJZyQ6fSyRUv5Bq5P4Igk3rR4sUvQLsoYk9Ym0fkjTR1LHhkgW7EXiid1HStnJhUjGzJle\nQQqR/HEVKWzULuWwK0QyZs70UmyI5M+xZyTTFfWQC4GOT/RYhBcskt9Nq/YioY/kiqNIgX0k\nvGhMh6tIxpZBkEgYtXPmaJGiHn2maC2LZPqdFElqGawi4YKsN8eO2nkFQ6SVaCKJLQOIVNt1\nJK9giLQSTyTl7xDJn2PvbPAKhkgrqUVafkAfyRmI5B2cWCTVY5Wpm3brLxi1cwUieQenFUn5\ngHKsUTvdYIM6bxDJAtfnkQKvI0EkDXZzZxx2HUkAIlngOGoX/DwSRFLjKBJuEYpMbXc2QCQN\nECkvR4lkf/s+RCKCo/WRIFJkIBIVrWaR6FE73LQaGYhERataJFWdQKRU4F47Klq7InmO2qFp\npwP32lHR2hPJvmVAZk46Iy1AJH9wHck7OINIG7qG9cJff/1FJVd6BJECcLyzAdeRgpJHE0nf\nRZ346y+zSfsfIZI/ziIJQCS34NBRuxWDSH/9ZTZJ8RtE8gcieQdnuI60ApEiA5GoaBWLpH+t\nzoRFy44UCU07EfSRqGgNi2R44ZuFRxhs4MGoHRWtZZHo91QZRMLwt8xBb+zDdaQ4yY8SydS0\ng0gyEImK9nYi/aVBjgeRZCASFa1lkZR9pMUYyZ/9qQl9JIna3o8EkTTEGbXb1BHOQ4o2Hkbt\nRCASFa1ikfyuI9mLhOtIIpj8hIpWs0hedzbY9pF6iCQBkahoVYvkc68dRPLlKJHsb9+HSERw\n6ru/IZIvB4m0YrrruIdIVHDyM9KyNI3aQSQJiERFez+RcEby5GiRjA+UpRep2z690Eaz2FZ7\nIi3o/VnA8LfI4SIZbpbsk4u0udOt/ymi2WyrYpFC7/7GBVmJ40Wib5bsU4vU9RCpDz4j4RYh\nmbcTqYdIAxApMgfda8dTokh/D/wun6MeNcdjFK7EEmnF72ZJHpyR9MFHPWpumkUIIskcLlJB\no3ZNi2SoB/NlCDpzGGyQOFqkAq4jQaSB0GrA8LdIdJFCmxQQiQo2imTfVw1sGCjDIZI/kTu5\nPUSigs0iLd/No6cJMg+R/FGMdf2hxp1Gj1INaqlRijT83+CdDRApE0efkQoYtdMBkVyCIZLI\n0X2kPv91JB0QySXYINKvX792P0Mkijiz1/BAJH1wLSL9+sWbBJEsgEjewe2K9OuXYBJEsgAi\neQc3KJIGIYttApGoaBWLpJ46Omzr1uHSGWkGIlFAJO/gdkXqVR5BJBKM2nkHt9e021B4BJFI\nnK8jQaSA5NWIpLqkBJEoXB9xhkghySFSqUAkKlrFIuXtI0EkZyCSdzBEagiIREWrWCQ07Y4F\nIlHRIJJ3OERyBKN23sEQqSFwHYmKVrFI6CMdyxvOIqTDXaQUT2vTwQ4iCS0DiJSaN5xFSIez\nSKEPzUOkhnjDWYR0uIoUOqMVmnYt8YazCOloQyQNUbbuEg6RHHEctatfpP802OXFIdztvk+I\nlJfDh7+r7yNpXoerKBM9669mjsUVxycRIFJe0LQjoikLYn1GsnhVF2GS67Nx6CPlBYMN+mjq\ngtiK5PlmlAVnkaY5/HSjdsfOJPji1/4niETRrkiakpQq0pxpIf/bZ0UKnJFi0tp1pIAnM21F\nWj8o+0i61xnvjEotEpp2h3L8LUImj4IqOGSugCJE0s3DY5yIB/fa5SXDTasJb2ALmr3Gto+0\nfvAS6f80LCXQFAUiFU6KxygOv5VzJapIfmekFd2JaP0gCLQTyXWyUoiUF4ikjZZmsMEkUqym\nHfpIx9KWSEf0kVZMA5DqQTuI1CaNiXTAqN2GaQBSOfht6iNhsKFOWhMp/XUkDp+BfNvBBvSR\n6gIi6aNZeOR+adlaJMdXDEGkvEAkIpqFR/6PUWiuI2liQ6TCgUj6aDZdJIikASI50q5IoaN2\neUXCqN2xQCQ52h8duwRht99CpKaASHI0TeNofyymPiNhsKEmIJIcrRSRMPxdFRBJjnaUSOon\nZHHTaqWUJ5JppoNiRAqd104tEm4RqpPiRDLNdFCOSEmadiuu99/ijJSX0kQyPaD9NiL5ziK0\nfIdIxwKR5GjWw9+JRfKc1275DpGOBSLJ0Y66jmSYjkud3CySbeYhUlxKE6mAPpL9sRg4jYvB\nIy+RVvQCEat3CIZIIqWJVMAZaflu6iN55SUwOUQqlUJEsp16x2vtECnS1l3CIZIjkURaPwgC\nQSQJe5HMb06ESDEpTSSxj5RFpGL662EihU6LBpGcKEWkcpp2K+bGUdkimd4uD5FiApH00Ywe\nQSQ9EMmRdkXCGSkkHCI50lwfaQUihYRDJEdiiyTc/Q2RJDDYUCrliWT6GSIpwfB3XsoTCWck\nIhgXZEulOJHQR6KCIVKplCZS/jsbViBSSDhEcgTD397B6CM1RCEiregF8l/7W4qEM9KxlCZS\n/ueRNqq+swEiHUtxImWfRYjD5BFE0gORHIkuku6akv/avUVKYEJgcohUKhCJilaxSBhsOBaI\nREWrWSTcInQoEImKVrVIgbmDSE5AJCoaRPIOh0iOQCTvYIjUEBCJigaRvMMhkiMNifS7fCBS\nqZQnUkEXZPOekbzm/nbYfEGZb4DiRCroFqG8Ivm9jcJh8wVlvgFKE6mgm1bN20p5LHq+H8lh\n8wVlvgEgEhUNIvmFQyR3IJJ3MERqiNJEQh9pAX2kqihOJIzaLWDUribKEyl+/VcqUt0XZP3r\noU4gEhUNIsVMDpEoIJJ3MERqCIhERYNIMZNDJAqI5B0MkRoCIlHRIFLM5BCJAiJ5B0OkhoBI\nVDSIFDM5RKKASN7BEKkhIBIVDSLFTA6RKCCSdzBEagiIREWDSDGTQyQKiOQdDJEaAiJR0SBS\nzOQQqZv+f8EvJyCSdzBEaggbkWZ/5v+2LyMQyTsYIjWEhUhdD5Gi5SUwOUQqFfumHUSKkZfA\n5BCpVEJE+nsk9+SjZiBSpK0HJodIOCNFy0tgcohUKhCJigaRYiaHSBApWl4Ck0OkUgkXqQLs\ndkXuXFpRde7t6qFOgkVSYNhhpv0ZljxqbaXNauqipN2RR9ZDBQTf2aAAIh2T3AREOpDge+0U\nQKRjkpuASAeSQiQA3g6IBEAEIBIAEYBIAEQAIgEQgXQijYPk6nHyboNK7hlss3YniJLYbMyU\nlcRFCauHzJmvh2QizbXnuw87YeEabHHV2CkrCUuSuihhuc+c+YpoUyRT/bpmJaNIoUXJKlLU\neiibLCJ123mfSM4tIge7QR6K5qIULFLxma+ISs9IhjZDTWekxEVJe0Y6sh7KJulgQ2Af2dgH\nNnV0vTcedWVWwybpihKWPnPm6yHb8HdDu7jqolSd+YLIJZJl47sG7PoJhVJ15ksiddMusBfr\nuXZTsBtRxkX8Vx+lZRk2auO79rj1UDQpBxssgg292IC1d9H+nBqdNkSz2xNEcFBRAushb+Yr\nolSR7IZlibWHDLQ5bcv4R93uhEYEBxUlrB4yZ74icokUOphkDD5OpLKL0k49lE3CPlKyNZvX\nPtfeQW27pKsPLko79VA0uQYbzGmD1j7WYNzrSKlKkrooAbnPn/l6yNS0m3Yw2QMiV3Bk3Vj0\nMrRFMZekgPNduZmviDwiddw/XdpKRKKLkv9YDKiH/JmviCwideJ/RLjP2sdw4hEiV8wZ0ec3\nwrEYWJSAesif+YrIMtgQLpI5NOJoUdpjMXVR/OuhgMzXQ7ozEtEN3Xau9xnJeJkpXgWS2zIU\nxeZPQtqiBNRD/sxXRJZ77cJFMm/goAoMPxbNG0hWlKozXxh5blrt6FrqNoLWf0QF0kUJLkna\nolSd+bLI9DzSFJLsVswjn0cqvSjt1EPRpB61S7QTu+2vaXLSliR1Udqph7LJJVLgvj+yzWA6\nFMsuSjv1UDYtijRdvIjWpMgpUnhR2qmHssnWtAvbu3PtaQebuvARJ2F95MqKLko79VA2uSY/\nCRwO4mpJGRRh6JZbIZ3TwovSTj0UTZFzf4eNu5ZUgYEjyHmLUnXmj6ZIkcJoqAKrLkrVmXcm\njUiduc3QRbidUX/jS7QKtChJwUVppx6KJ+N0XMRwj03bXRt4dAVSRbEqSc6iVJ35kihSJJv0\nnX5EN6xp756VmotSdeZLItuoXUgFTus9rHpMLbuyi9JOPRRNxguydCUQ9W9YdWxMmzMVhTqS\n0xclsB7yZr4i8l2QNf6Z1yc3/iWM2qgwHi/Gv/lE8uRFCauHzJmviFJvWjXXP7nqtY8boQYT\nlyRxUcJynznzFZH4UXNKBHNq0187i9TRTknksWRMTWckaVGC6iF35ush0XUki3O6YXBbWGhX\nYUwdWoFWQ0/+jVRuFcbkHkUJroecma+LvHc2GO9g813vujisAlM93HZIUarOfBnkFClhF3Sp\n/6Pqr+qiVJ35Usj26suQv3MHDwZZtOzCVp20KO3UQ9EU+mDfONBTSP0YBxCNqXOWJKgecme+\nInJeR6ITBw2axvxDaTEETPfWw4Z/A4sSUg/ZM18ROa8j0YM95LCtoendxezg2oxcBdy4mbgo\nAfWQP/MVkfeCrOezlXOyRHeQKXPjOQRsPBaTFyWgHvJnviKyDTZQ1y+2nU+Ipg3tY1egsXmi\nDzaV5ICi+NdDAZmvh2yPUZAHZk/W4FZ/BhOPqUCbzp75T0KmolSd+ZLI9g5ZU2KqZbcuqT+F\n0S6PJCzJAUUJyX32zFdEoS9jnoLoxygsNhIF40bIopAlOaAoQfWQO/MVketlzHY9eNOaD2q7\nWQWXWpR26qFosrwfqTdXIFk1xk5uVExX7w1FsTqQE3oUVA+ZM18RGeds4BbacE1gxy9yE3Ys\nZi5K1ZkviVJFMlYh2bA/9lY8YaEN1wZnLUrVmS+JXA/2WdxL6V0FsSvOUBLzuIn/wRShKEH1\nkDvz9VDqo+aBG49Zh1lLElyUduqhbNoUqY9Zh7lLElaU3Ll/F5fyiGTRWiBvhrTMQpwqpA9F\nc1EilCSgKIH1kDfzFZFFJFOnYwrtjJHIFRxzRjIXJbgk6c5IxWe+IrLctJpcpLi1R/W3kx+L\nwUUJqofcma+HLMPf4SLNB4euwRKSOTfCj8WMRak684VRpUj0hcBDr1+EHotZi1J15gsjy/uR\nLEQyt6fMq4iBoSRWfxPIq5bSMi7h9ZAx83VR6hnJmHxc5q/BqotSdeYLI49I5pO+6YGzaZm/\nAs1FMT7COC2ziFRz5guj3HfImlt2tbQpLBpH5Ral6swfSLkiDRBNe+lT6ZTxjKInVWf+IPJN\nom+zFl2AtExI4pKkHmyIkftcma+LRGek8BFPqu4Nw66KTyH5CF8D1d9LW5Tg3OfMfFWkfEI2\nKLlx7RYd5DgVmLYkqYsSeCo1r/2weiiblH2kwAaFb+JurcJoFZipJJGKEpL77JmvhcSDDfTj\nk4Zd7Ps8Wb+sN2YFGh5+ozfmezRGK0pIPWTPfBXkOSPZPuzvVYfdmvKIM5JlUTyf9V2Tpjkj\nVZD5WsjTR7IZ7vEealraE7GuE9L5sCiK/6BZhKKE1kPWzFdEllG7rf5Mf+i9th1hHdYZMRcl\nwsNEAc2rwHrImvmqyHIdyUKkONmIsArDlZiyi9JOPRRPvptWpU+7OKEXQY/BoijllqTqzBdG\nlrm/bSrQvJajsGod6aNlLklYPRRUDYWTZxJ9cye3oBq07STREUoUqfjMV0SeSfTNw64l1SB9\nLJqKkrskQfWQO/P1kG0SfUPTu6AaNN32aShK7jNSUD0UVA2FU+xjFO30cqsuSdWZP5JiRQKg\nJpLOa9dIm6DuktSd+2pIOWrnf5tVUc2JuksSkvvsma+IIkXq1v9KoO6SBOQ+f+YrAiIZqbsk\nEOkYIJKRuksCkY4hyyT6xqTrf0VQd0n8c19A5uuhyOHvdmqw6pJUnfmjSXhnQ0Da9b8SqLsk\noQ92FVMNhZPwjBTwbKXhIaCjqbsk/u3SAjJfDbkmP6mOuktSd+5rIKlI7VRf3SWpO/d1kPgW\noSaouyR1574a8jzYVxV1l6Tu3NcDzkhm6i5J3bmvBvSRrKi7JHXnvg5KHbWbB13LOQLqLknQ\n8H3uzFdBkdeRrKc0Poi6SxJyFYxfAIoi72ywmGXoQOouSYy5UouohsIp9167cVl7FVZdkqoz\nfzTl3v3dF1ODdZck8O7vvphqKJvU15GCarCMRkXdJQnIff7MVwREMlJ3SSDSMRQpUlm93LpL\nEuPPQBHVUDiFilTSuGvdJQkSKXfmK6LIwYYldczsBFB3SYJynzvz9VDk8DcAtQGRAIhAsnfI\nttIiqLskdee+JhK9Q3b9r3bqLkndua8KiERTd0nqzn1VQCSauktSd+6rAiLR1F2SunNfFRCJ\npu6S1J37qoBINHWXpO7cV0UikZqZo7PuktSd+6rABVkAIgCRAIgARAIgAhAJgAhAJAAiAJEA\niEDCee1amaOz7pLUnft6SPYYBb+ombpLUnfuayKVSNKyXuouSd25r4mEtwj1LUwtWHdJ6s59\nVUAkmrpLUnfuqyKtSPW3KeouSd25rwqIRFN3SerOfVVgsMFA3SWpO/c1geFvA3WXpO7c10Tq\nC7INUHdJ6s59PeAWIQAiAJEAiABEAiACEAmACEAkACIAkQCIAEQCIAIQCYAIQCQAIgCR/Pm6\nMNZdv3NnA5QARPLmzCZuuTMCCgAi+XJm59fJ6PnVsZ/cWQH5gUie3Nlp+XB97Ub20537/nFl\n7Proh+/9/P/r34Wdh9/6j46dPrNlGCQFInlyZff507MfdDm/fHp2Q1Ove4oiXeffbmNDECa1\nCUTypBP23NRTurHz0OS7iSKdn8tvj/6b4ZGGNoFIniyqDEyS9P1p+P8xtPl4kX7m3zp2vZOr\nBBUDkTyRReJ/U3+6vxp+p0eu/IK0QCRPLmsfyVakvv85sQ6XndoEInmyjtpxukhNu8fa6HsM\nnaeBT4Yd3iaoV1/O7DSck+6XTaRtsKFjX/3zPA82DJ8+ht+++x8MNjQKRPJmubNhaK1NIm3D\n3+NQ98cs0vBbvwx/f2TONUgDRPLnfn2Jc/kaPs4ttvWCbH/rXsrMTbvz9lsHjxoFIiUGnaL3\nANWcGIj0HqCaEwOR3gNUc2Ig0nuAagYgAhAJgAhAJAAiAJEAiABEAiACEAmACEAkACIAkQCI\nwP8DGmbBWcSw+soAAAAASUVORK5CYII=",
      "text/plain": [
       "plot without title"
      ]
     },
     "metadata": {
      "image/png": {
       "height": 420,
       "width": 420
      },
      "text/plain": {
       "height": 420,
       "width": 420
      }
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "fig = plot_anova_diversity(phy_hum_std,method=c(\"richness\",\"simpson\",\"shannon\"),grouping_column=\"Ext.Kit\")\n",
    "fname_in = \"alpha_plot_stats_hum_extkit.jpg\"\n",
    "title_in = \"Alpha Diversity Plot \\n Human, Std processing\"\n",
    "annotate_figure(fig, #https://rpkgs.datanovia.com/ggpubr/reference/annotate_figure.html\n",
    "                    top = text_grob(title_in, color = \"black\", face = \"bold\", size = 14))\n",
    "ggsave(paste(img_dir,\"\",fname_in,sep=\"\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#Beta diveresity\n",
    "\n",
    "#plot all samples\n",
    "title = \"Beta Diversity Plot of Raryfied Data: \\n All samples\"\n",
    "file_name = \"beta_all_subj&kit\"\n",
    "plot_beta(phy_r_t,\"PCoA\",\"bray\",\"Subject.ID\",title,file_name,\"Ext.Kit\")\n",
    "\n",
    "#Plot extracted material only\n",
    "phy_sub = subset_samples(phy_r_t,Sample.Type!=\"Seq_Control\")\n",
    "\n",
    "title = \"Beta Diversity Plot of Raryfied Data: \\n Extacted samples\"\n",
    "file_name = \"beta_all_strat_st\"\n",
    "plot_beta(phy_sub,\"PCoA\",\"bray\",\"Homo.Method\",title,file_name,\"Ext.Kit\",\"Subject.ID\")\n",
    "\n",
    "title = \"Beta Diversity Plot of Raryfied Data: \\n Extacted samples\"\n",
    "file_name = \"beta_all_strat_hm\"\n",
    "plot_beta(phy_sub,\"PCoA\",\"bray\",\"Subject.ID\",title,file_name,\"Ext.Kit\",\"Homo.Method\")\n",
    "\n",
    "title = \"Beta Diversity Plot of Raryfied Data: \\n Extacted samples\"\n",
    "file_name = \"beta_all_strat_ek\"\n",
    "plot_beta(phy_sub,\"PCoA\",\"bray\",\"Subject.ID\",title,file_name,\"Homo.Method\",\"Ext.Kit\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [],
   "source": [
    "beta_nmds <- function(title_in,phy_in,group_in,filename_in){\n",
    "    \n",
    "    ord.res = ordination_MS(phy_in,method=\"NMDS\",grouping_column=group_in)\n",
    "    \n",
    "    fname = paste(img_dir,\"\",filename_in,sep=\"\")\n",
    "    fig = plot.ordination(ord.res,method=\"NMDS\",show.pvalues=T)\n",
    "    annotate_figure(fig, #https://rpkgs.datanovia.com/ggpubr/reference/annotate_figure.html\n",
    "                    top = text_grob(title_in, color = \"black\", face = \"bold\", size = 14))\n",
    "    ggsave(fname)   \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Square root transformation\n",
      "Wisconsin double standardization\n",
      "Run 0 stress 0.1325511 \n",
      "Run 1 stress 0.1417247 \n",
      "Run 2 stress 0.1500223 \n",
      "Run 3 stress 0.1432668 \n",
      "Run 4 stress 0.1438513 \n",
      "Run 5 stress 0.1375304 \n",
      "Run 6 stress 0.1454245 \n",
      "Run 7 stress 0.1411488 \n",
      "Run 8 stress 0.1496079 \n",
      "Run 9 stress 0.1420763 \n",
      "Run 10 stress 0.1446658 \n",
      "Run 11 stress 0.1380095 \n",
      "Run 12 stress 0.1411047 \n",
      "Run 13 stress 0.145693 \n",
      "Run 14 stress 0.1434878 \n",
      "Run 15 stress 0.1433467 \n",
      "Run 16 stress 0.1406994 \n",
      "Run 17 stress 0.1424287 \n",
      "Run 18 stress 0.1500032 \n",
      "Run 19 stress 0.1477479 \n",
      "Run 20 stress 0.1379481 \n",
      "*** No convergence -- monoMDS stopping criteria:\n",
      "    20: stress ratio > sratmax\n",
      "                     groups      p_value label\n",
      "Water-human_1 Water-human_1 3.001837e-18   ***\n",
      "M98-Water         M98-Water 2.812441e-17   ***\n",
      "A01-human_1     A01-human_1 9.101305e-14   ***\n",
      "Water-Z00         Water-Z00 3.513261e-13   ***\n",
      "Z05-human_1     Z05-human_1 1.822713e-12   ***\n",
      "Z06-human_1     Z06-human_1 1.841093e-12   ***\n",
      "A02-human_1     A02-human_1 1.619827e-11   ***\n",
      "A03-human_1     A03-human_1 3.190190e-11   ***\n",
      "M22-Water         M22-Water 6.226036e-11   ***\n",
      "M98-human_1     M98-human_1 1.758494e-08   ***\n",
      "A01-M98             A01-M98 8.883932e-08   ***\n",
      "A01-Water         A01-Water 1.817676e-07   ***\n",
      "Water-Z05         Water-Z05 3.468526e-07   ***\n",
      "Water-Z06         Water-Z06 3.483850e-07   ***\n",
      "M22-human_1     M22-human_1 5.388754e-07   ***\n",
      "A02-Water         A02-Water 5.634567e-07   ***\n",
      "Z00-human_1     Z00-human_1 6.077573e-07   ***\n",
      "A03-Water         A03-Water 6.521672e-07   ***\n",
      "M98-Z05             M98-Z05 7.968779e-07   ***\n",
      "M98-Z06             M98-Z06 8.053219e-07   ***\n",
      "A00-A01             A00-A01 2.336565e-06   ***\n",
      "A02-M98             A02-M98 3.917518e-06   ***\n",
      "A03-M98             A03-M98 6.354765e-06   ***\n",
      "A00-human_1     A00-human_1 6.891808e-06   ***\n",
      "A00-Z05             A00-Z05 1.043433e-05   ***\n",
      "A00-Z06             A00-Z06 2.705741e-05   ***\n",
      "A00-Water         A00-Water 3.724496e-05   ***\n",
      "A00-A03             A00-A03 3.890974e-05   ***\n",
      "M16-human_1     M16-human_1 5.093217e-05   ***\n",
      "M16-Water         M16-Water 6.742572e-05   ***\n",
      "A01-M22             A01-M22 1.096049e-04   ***\n",
      "A01-M16             A01-M16 1.174750e-04   ***\n",
      "A00-A02             A00-A02 1.666523e-04   ***\n",
      "A01-A03             A01-A03 2.559229e-04   ***\n",
      "M16-Z05             M16-Z05 3.483973e-04   ***\n",
      "M22-Z05             M22-Z05 3.944499e-04   ***\n",
      "M22-Z06             M22-Z06 3.975773e-04   ***\n",
      "M16-Z06             M16-Z06 4.367651e-04   ***\n",
      "A02-M22             A02-M22 1.004246e-03    **\n",
      "A03-M16             A03-M16 1.060424e-03    **\n",
      "A02-M16             A02-M16 1.242317e-03    **\n",
      "A03-M22             A03-M22 1.328732e-03    **\n",
      "Water-Z10         Water-Z10 1.636265e-03    **\n",
      "A01-Z00             A01-Z00 3.665156e-03    **\n",
      "A01-A02             A01-A02 4.387793e-03    **\n",
      "Z00-Z06             Z00-Z06 8.863702e-03    **\n",
      "Z00-Z05             Z00-Z05 8.864886e-03    **\n",
      "A00-M98             A00-M98 1.339058e-02     *\n",
      "A01-Z10             A01-Z10 1.540736e-02     *\n",
      "A02-Z00             A02-Z00 1.625364e-02     *\n",
      "A01-Z05             A01-Z05 1.679783e-02     *\n",
      "A03-Z05             A03-Z05 1.780932e-02     *\n",
      "A03-Z00             A03-Z00 1.965426e-02     *\n",
      "M98-Z10             M98-Z10 2.015038e-02     *\n",
      "Z05-Z10             Z05-Z10 2.263439e-02     *\n",
      "Z06-Z10             Z06-Z10 2.263940e-02     *\n",
      "Z00-Z10             Z00-Z10 2.350752e-02     *\n",
      "A02-Z10             A02-Z10 2.978709e-02     *\n",
      "A03-Z10             A03-Z10 3.250411e-02     *\n",
      "A01-Z06             A01-Z06 3.402573e-02     *\n",
      "A03-Z06             A03-Z06 3.617802e-02     *\n",
      "M16-M98             M16-M98 4.510179e-02     *\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Saving 6.67 x 6.67 in image\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Square root transformation\n",
      "Wisconsin double standardization\n",
      "Run 0 stress 0.0001568025 \n",
      "Run 1 stress 0.0001210542 \n",
      "... New best solution\n",
      "... Procrustes: rmse 0.05571643  max resid 0.5429925 \n",
      "Run 2 stress 0.0001372732 \n",
      "... Procrustes: rmse 0.04947462  max resid 0.5647796 \n",
      "Run 3 stress 0.0001172647 \n",
      "... New best solution\n",
      "... Procrustes: rmse 0.04587223  max resid 0.6204615 \n",
      "Run 4 stress 0.0001499195 \n",
      "... Procrustes: rmse 0.0403846  max resid 0.5479715 \n",
      "Run 5 stress 0.0001648667 \n",
      "... Procrustes: rmse 0.01686309  max resid 0.2432275 \n",
      "Run 6 stress 0.0001168312 \n",
      "... New best solution\n",
      "... Procrustes: rmse 0.03890701  max resid 0.458699 \n",
      "Run 7 stress 9.10918e-05 \n",
      "... New best solution\n",
      "... Procrustes: rmse 0.05444176  max resid 0.643114 \n",
      "Run 8 stress 0.0001305017 \n",
      "... Procrustes: rmse 0.04307668  max resid 0.6366981 \n",
      "Run 9 stress 0.0001492449 \n",
      "... Procrustes: rmse 0.05056978  max resid 0.6911863 \n",
      "Run 10 stress 0.0001092772 \n",
      "... Procrustes: rmse 0.05012605  max resid 0.6250108 \n",
      "Run 11 stress 0.0001479446 \n",
      "... Procrustes: rmse 0.03435794  max resid 0.408576 \n",
      "Run 12 stress 0.0001560136 \n",
      "... Procrustes: rmse 0.04153129  max resid 0.5030922 \n",
      "Run 13 stress 0.0001128929 \n",
      "... Procrustes: rmse 0.04270751  max resid 0.4298667 \n",
      "Run 14 stress 0.0001073101 \n",
      "... Procrustes: rmse 0.0389415  max resid 0.4876429 \n",
      "Run 15 stress 0.0001524888 \n",
      "... Procrustes: rmse 0.04991826  max resid 0.6839078 \n",
      "Run 16 stress 0.0001482704 \n",
      "... Procrustes: rmse 0.04559382  max resid 0.4795854 \n",
      "Run 17 stress 0.0001244946 \n",
      "... Procrustes: rmse 0.04835497  max resid 0.5627126 \n",
      "Run 18 stress 0.0001397167 \n",
      "... Procrustes: rmse 0.03470242  max resid 0.3525909 \n",
      "Run 19 stress 0.0001633643 \n",
      "... Procrustes: rmse 0.03972311  max resid 0.3973971 \n",
      "Run 20 stress 0.0001562946 \n",
      "... Procrustes: rmse 0.04630554  max resid 0.5032129 \n",
      "*** No convergence -- monoMDS stopping criteria:\n",
      "     1: stress < smin\n",
      "    19: scale factor of the gradient < sfgrmin\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in metaMDS(veganifyOTU(physeq), distance, ...):\n",
      "\"stress is (nearly) zero: you may have insufficient data\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                                             groups\n",
      "Q DSP Virus-Q QIAamp                           Q DSP Virus-Q QIAamp\n",
      "Q PowerSoil Pro-Q QIAamp                   Q PowerSoil Pro-Q QIAamp\n",
      "Q DSP Virus-Q PowerMicrobiome         Q DSP Virus-Q PowerMicrobiome\n",
      "Q QIAamp-Z MagBead                               Q QIAamp-Z MagBead\n",
      "Q PowerMicrobiome-Q PowerSoil Pro Q PowerMicrobiome-Q PowerSoil Pro\n",
      "Q PowerMicrobiome-Z MagBead             Q PowerMicrobiome-Z MagBead\n",
      "Q PowerSoil-Q QIAamp                           Q PowerSoil-Q QIAamp\n",
      "Q PowerMicrobiome-Q PowerSoil         Q PowerMicrobiome-Q PowerSoil\n",
      "Q QIAamp-T MagMax                                 Q QIAamp-T MagMax\n",
      "Q DSP Virus-Q PowerSoil                     Q DSP Virus-Q PowerSoil\n",
      "Q PowerMicrobiome-T MagMax               Q PowerMicrobiome-T MagMax\n",
      "                                       p_value label\n",
      "Q DSP Virus-Q QIAamp              6.379709e-05   ***\n",
      "Q PowerSoil Pro-Q QIAamp          1.128763e-04   ***\n",
      "Q DSP Virus-Q PowerMicrobiome     2.238348e-04   ***\n",
      "Q QIAamp-Z MagBead                3.322825e-04   ***\n",
      "Q PowerMicrobiome-Q PowerSoil Pro 4.723169e-04   ***\n",
      "Q PowerMicrobiome-Z MagBead       1.180764e-03    **\n",
      "Q PowerSoil-Q QIAamp              1.370785e-03    **\n",
      "Q PowerMicrobiome-Q PowerSoil     3.713200e-03    **\n",
      "Q QIAamp-T MagMax                 1.094652e-02     *\n",
      "Q DSP Virus-Q PowerSoil           2.253821e-02     *\n",
      "Q PowerMicrobiome-T MagMax        2.687811e-02     *\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Saving 6.67 x 6.67 in image\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Square root transformation\n",
      "Wisconsin double standardization\n",
      "Run 0 stress 0.1275733 \n",
      "Run 1 stress 0.1275736 \n",
      "... Procrustes: rmse 8.1857e-05  max resid 0.0003478253 \n",
      "... Similar to previous best\n",
      "Run 2 stress 0.1275889 \n",
      "... Procrustes: rmse 0.0007819736  max resid 0.005052197 \n",
      "... Similar to previous best\n",
      "Run 3 stress 0.2005745 \n",
      "Run 4 stress 0.1275736 \n",
      "... Procrustes: rmse 5.035162e-05  max resid 0.0002241239 \n",
      "... Similar to previous best\n",
      "Run 5 stress 0.1275733 \n",
      "... New best solution\n",
      "... Procrustes: rmse 9.509514e-06  max resid 4.877762e-05 \n",
      "... Similar to previous best\n",
      "Run 6 stress 0.1896089 \n",
      "Run 7 stress 0.1275734 \n",
      "... Procrustes: rmse 4.049996e-05  max resid 0.0002107608 \n",
      "... Similar to previous best\n",
      "Run 8 stress 0.1275733 \n",
      "... Procrustes: rmse 8.128788e-06  max resid 3.255237e-05 \n",
      "... Similar to previous best\n",
      "Run 9 stress 0.1736069 \n",
      "Run 10 stress 0.1425107 \n",
      "Run 11 stress 0.1456629 \n",
      "Run 12 stress 0.1470012 \n",
      "Run 13 stress 0.144289 \n",
      "Run 14 stress 0.1458219 \n",
      "Run 15 stress 0.1606936 \n",
      "Run 16 stress 0.1275733 \n",
      "... Procrustes: rmse 1.164362e-05  max resid 6.418584e-05 \n",
      "... Similar to previous best\n",
      "Run 17 stress 0.1275733 \n",
      "... Procrustes: rmse 1.827862e-05  max resid 9.349912e-05 \n",
      "... Similar to previous best\n",
      "Run 18 stress 0.142617 \n",
      "Run 19 stress 0.1849389 \n",
      "Run 20 stress 0.1425108 \n",
      "*** Solution reached\n",
      "                                                             groups\n",
      "Q PowerSoil Pro-Q QIAamp                   Q PowerSoil Pro-Q QIAamp\n",
      "Q PowerSoil-Q PowerSoil Pro             Q PowerSoil-Q PowerSoil Pro\n",
      "Q PowerMicrobiome-Q PowerSoil Pro Q PowerMicrobiome-Q PowerSoil Pro\n",
      "Q PowerSoil Pro-Z MagBead                 Q PowerSoil Pro-Z MagBead\n",
      "                                       p_value label\n",
      "Q PowerSoil Pro-Q QIAamp          0.0007349795   ***\n",
      "Q PowerSoil-Q PowerSoil Pro       0.0166654003     *\n",
      "Q PowerMicrobiome-Q PowerSoil Pro 0.0209200318     *\n",
      "Q PowerSoil Pro-Z MagBead         0.0393378323     *\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Saving 6.67 x 6.67 in image\n",
      "\n"
     ]
    }
   ],
   "source": [
    "title = \"Beta Diversity Plot of Raryfied Data: \\nAll samples\"\n",
    "filename = \"beta_nmds_all_subid.png\"\n",
    "beta_nmds(title,phy_r,\"Subject.ID\",filename)\n",
    "\n",
    "sub_list<-c(\"Ext_Control\",\"Study\")\n",
    "phy_sub<-subset_samples(phy_obj, Sample.Type %in% sub_list)\n",
    "title = \"Beta Diversity Plot of Raryfied Data: \\nExtracted samples\"\n",
    "filename = \"beta_nmds_all_extkit.png\"\n",
    "beta_nmds(title,phy_sub,\"Ext.Kit\",filename)\n",
    "\n",
    "title = \"Beta Diversity Plot of Raryfied Data: \\nHuman, std processing\"\n",
    "filename = \"beta_nmds_human_extkit.png\"\n",
    "beta_nmds(title,phy_hum_std,\"Ext.Kit\",filename)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h2> Comparative Analysis <h2>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [],
   "source": [
    "subset_df <- function(tmp_in,max_in=1){\n",
    "    df_out=data.frame()\n",
    "    max_count=max_in\n",
    "    #for each kit add rows until max\n",
    "    for (kit in unique(tmp_in$Ext.Kit)){\n",
    "        for(cont in unique(tmp_in$Subject.ID)){\n",
    "            tmp2 = subset(tmp_in,Ext.Kit==kit&Subject.ID==cont)\n",
    "            for(row in 1:max_count){\n",
    "                df_out = rbind(df_out,tmp2[row,])   \n",
    "            } \n",
    "        }    \n",
    "    }    \n",
    "    return(df_out)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [],
   "source": [
    "adonis_stat <- function(dist_in,test_in,test_name){    \n",
    "    distdf_sub = dist_in[,1:(ncol(dist_in)-5)]\n",
    "    distdf = vegdist(distdf_sub,method=\"bray\")\n",
    "\n",
    "    metadata$CGR.ID = rownames(metadata)\n",
    "    metadata_sub = subset(metadata,CGR.ID %in% rownames(as.data.frame(as.matrix(distdf))))\n",
    "\n",
    "    #perform comparisons\n",
    "    #print(adonis(distdf ~ get(test_in), data=metadata_sub, permutations = 99)$aov.tab$\"Pr(>F)\"[1])\n",
    "    df_out = data.frame()\n",
    "    df_out[nrow(df_out)+1,\"pval\"]=adonis(distdf ~ get(test_in), data=metadata_sub, permutations = 99)$aov.tab$\"Pr(>F)\"[1]\n",
    "    if(df_out[nrow(df_out),\"pval\"]>.05) {df_out[nrow(df_out),\"sig\"]=\"ns\"} else{df_out[nrow(df_out),\"sig\"]=\"sig\"}\n",
    "    df_out[nrow(df_out),\"testing\"]=test_in\n",
    "    df_out[nrow(df_out),\"description\"]=test_name\n",
    "    return(df_out)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Prepare samples\n",
    "#Create OTUdf\n",
    "OTUdf = otu_extract(phy_r)\n",
    "\n",
    "#Create TAXdf\n",
    "TAXdf = tax_extract(phy_r)\n",
    "\n",
    "#Clean df's\n",
    "OTUdf_clean = otu_tax_clean(OTUdf,TAXdf)\n",
    "RAdf = cal_relabun(OTUdf_clean)\n",
    "RAdf_sub = subset(RAdf,Subject.ID!=\"Water\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [],
   "source": [
    "#stats\n",
    "stats_df=data.frame()\n",
    "\n",
    "#group 1\n",
    "#all kits, D6300, Ext,Kit\n",
    "Subject.List = c(\"Z00\",\"M98\")\n",
    "tmp = subset(RAdf_sub, Subject.ID %in% Subject.List&Homo.Status==\"Standard\")\n",
    "distdf = subset_df(tmp)\n",
    "stats_df = (adonis_stat(distdf,\"Ext.Kit\",\"Z00,M98 by Ext.Kit for all methods\"))\n",
    "stats_df = rbind(stats_df,adonis_stat(distdf,\"Subject.ID\",\"Z00,M98 by Subject.ID for all methods\"))\n",
    "\n",
    "\n",
    "#group 2\n",
    "#all kits, D6300, Ext,Kit\n",
    "Ext.List = c(\"Q DSP Virus\",\"Q PowerMicrobiome\", \"Q PowerSoil Pro\",\"Q PowerSoil\",\"Q QIAamp\")\n",
    "Subject.List = c(\"Z00\",\"M22\",\"M98\")\n",
    "tmp = subset(RAdf_sub, Ext.Kit %in% Ext.List & Subject.ID %in% Subject.List&Homo.Status==\"Standard\")\n",
    "distdf = subset_df(tmp)\n",
    "stats_df = rbind(stats_df,adonis_stat(distdf,\"Ext.Kit\",\"Z00,M22,M98 by Ext.Kit for all qiagen methods\"))\n",
    "stats_df = rbind(stats_df,adonis_stat(distdf,\"Subject.ID\",\"Z00,M22,M98 by Subject.ID for all qiagen methods\"))\n",
    "\n",
    "#group 3\n",
    "#select methods, Z00/human_1, Ext.Kit/Subject.ID,Homo.Status\n",
    "Ext.List = c(\"Q PowerMicrobiome\", \"Q PowerSoil Pro\",\"Q PowerSoil\",\"Z MagBead\")\n",
    "Subject.List = c(\"Z00\",\"human_1\")\n",
    "tmp = subset(RAdf_sub, Ext.Kit %in% Ext.List & Subject.ID %in% Subject.List&Homo.Status==\"Standard\")\n",
    "distdf = subset_df(tmp)\n",
    "tmp = subset(RAdf_sub, Ext.Kit %in% Ext.List & Subject.ID %in% Subject.List&Homo.Status!=\"Standard\")\n",
    "distdf2 = subset_df(tmp)\n",
    "distdf=rbind(distdf,distdf2)\n",
    "distdf$Homo.Status = as.character(distdf$Homo.Status)\n",
    "for(row in 1:nrow(distdf)){\n",
    "    if(distdf[row,\"Homo.Status\"]!=\"Standard\"){\n",
    "        distdf[row,\"Homo.Status\"]=\"Variation\"\n",
    "    }\n",
    "}\n",
    "stats_df = rbind(stats_df,adonis_stat(distdf,\"Ext.Kit\",\"Z00&human_1 by Ext.Kit for group3 methods\"))\n",
    "stats_df = rbind(stats_df,adonis_stat(distdf,\"Subject.ID\",\"Z00&human_1 by Subject.ID for group3 methods\"))\n",
    "stats_df = rbind(stats_df,adonis_stat(distdf,\"Homo.Status\",\"Z00&human_1 by Subject.ID for group3 methods\"))\n",
    "\n",
    "\n",
    "#group 4\n",
    "Ext.List = c(\"Q PowerMicrobiome\", \"Q PowerSoil Pro\",\"Q PowerSoil\")\n",
    "Subject.List = c(\"Z00\",\"Z10\",\"M22\",\"M98\")\n",
    "tmp = subset(RAdf_sub, Ext.Kit %in% Ext.List & Subject.ID %in% Subject.List&Homo.Status==\"Standard\")\n",
    "distdf = subset_df(tmp)\n",
    "stats_df = rbind(stats_df,adonis_stat(distdf,\"Ext.Kit\",\"Z00,Z10,M22,M98 by Ext.Kit for group4 methods\"))\n",
    "stats_df = rbind(stats_df,adonis_stat(distdf,\"Subject.ID\",\"Z00,Z10,M22,M98 by Subject.ID for group4 methods\"))\n",
    "\n",
    "#group 5\n",
    "Ext.List = c(\"Q PowerMicrobiome\", \"Q PowerSoil Pro\",\"Q PowerSoil\", \"Z MagBead\",\"Q QIAamp\")\n",
    "Subject.List = c(\"human_1\")\n",
    "tmp = subset(RAdf_sub, Ext.Kit %in% Ext.List & Subject.ID %in% Subject.List&Homo.Status==\"Standard\")\n",
    "distdf = subset_df(tmp,5)\n",
    "stats_df = rbind(stats_df,adonis_stat(distdf,\"Ext.Kit\",\"human by Ext.Kit for group5 methods\"))\n",
    "stats_df = rbind(stats_df,adonis_stat(distdf,\"IR\",\"human by IR for group5 methods\"))\n",
    "stats_df = rbind(stats_df,adonis_stat(distdf,\"Mag.Col\",\"human by Mag.Col for group5 methods\"))\n",
    "\n",
    "write.csv(stats_df,paste(stats_dir,\"adtest_summary_groups.csv\",sep=\"\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [],
   "source": [
    "network_plotting <- function(phy_in,color_in,title_in,fname_in,shape_in=NULL){ \n",
    "    #https://joey711.github.io/phyloseq/plot_network-examples.html\n",
    "    ig = make_network(phy_in, dist.fun=\"bray\", max.dist=0.3)\n",
    "    fig = plot_network(ig, physeq=phy_in, color=color_in, line_weight=0.4, label=NULL,shape=shape_in) +\n",
    "        scale_shape_manual(values = c(15,16,17,18,3,8,0,1,2))\n",
    "    #plot_net(phy_hum_std, color=\"Ext.Kit\", distance=\"bray\") #alt, not used\n",
    "\n",
    "    annotate_figure(fig, #https://rpkgs.datanovia.com/ggpubr/reference/annotate_figure.html\n",
    "                       top = text_grob(title_in, color = \"black\", face = \"bold\", size = 14))\n",
    "    ggsave(paste(img_dir,\"\",fname_in,\".jpg\",sep=\"\"))\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"attributes are not identical across measure variables; they will be dropped\"\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Warning message:\n",
      "\"attributes are not identical across measure variables; they will be dropped\"\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Warning message:\n",
      "\"attributes are not identical across measure variables; they will be dropped\"\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Warning message:\n",
      "\"attributes are not identical across measure variables; they will be dropped\"\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#Create network tree\n",
    "title=\"Network created using Bray-Curtis dissimilarity matrix: \\n Human Samples, Std Processing\"\n",
    "network_plotting(phy_hum_std,\"Run.ID\",title,\"net_human_std_runid.jpg\")\n",
    "\n",
    "title=\"Network created using Bray-Curtis dissimilarity matrix: \\n Human Samples, Std Processing\"\n",
    "network_plotting(phy_hum_std,\"Ext.Kit\",title,\"net_human_std_extkit.jpg\",\"Run.ID\")\n",
    "\n",
    "title=\"Network created using Bray-Curtis dissimilarity matrix: \\n All Samples, All Processing\"\n",
    "network_plotting(phy_r_t,\"Subject.ID\",title,\"net_all_subjid.jpg\",\"Ext.Kit\")\n",
    "\n",
    "title=\"Network created using Bray-Curtis dissimilarity matrix: \\n All Samples, All Processing\"\n",
    "network_plotting(phy_r_t,\"Ext.Kit\",title,\"net_all_extkit.jpg\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.5.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
