{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"package 'ape' was built under R version 3.5.3\"\n",
      "Warning message:\n",
      "\"package 'cowplot' was built under R version 3.5.1\"\n",
      "Loading required package: ggplot2\n",
      "\n",
      "Warning message:\n",
      "\"package 'ggplot2' was built under R version 3.5.1\"\n",
      "\n",
      "Attaching package: 'cowplot'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:ggplot2':\n",
      "\n",
      "    ggsave\n",
      "\n",
      "\n",
      "Warning message:\n",
      "\"package 'data.table' was built under R version 3.5.1\"\n",
      "Warning message:\n",
      "\"package 'dplyr' was built under R version 3.5.3\"\n",
      "\n",
      "Attaching package: 'dplyr'\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:data.table':\n",
      "\n",
      "    between, first, last\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:stats':\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:base':\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "Warning message:\n",
      "\"package 'yaml' was built under R version 3.5.3\"\n"
     ]
    }
   ],
   "source": [
    "source(\"sources/ggrare.R\") #github library: https://rdrr.io/github/gauravsk/ranacapa/\n",
    "source(\"sources/miseqR.R\")\n",
    "source(\"sources/qiime2R.R\") #github https://github.com/jbisanz/qiime2R\n",
    "source('sources/abundances.R') #github https://github.com/microbiome/microbiome\n",
    "\n",
    "library(biomformat)\n",
    "library(ape)\n",
    "library(cowplot)\n",
    "library(cluster)\n",
    "library(data.table)\n",
    "library(dplyr)\n",
    "#library(\"expss\")\n",
    "#library(ggplot2)\n",
    "#library(\"ggpubr\")\n",
    "#library(Hmisc)\n",
    "#library(otuSummary)\n",
    "#library(phyloseq)\n",
    "#library(scales)\n",
    "#library(tibble)\n",
    "#library(vegan)\n",
    "library(yaml)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "#read in config\n",
    "configfile = read_yaml('config.yaml')\n",
    "data_dir=configfile$raw_data_dir\n",
    "analysis=configfile$analysis_dir\n",
    "output_dir=configfile$output_dir\n",
    "manifest_file=configfile$clean_manifest\n",
    "variable_file=configfile$variable_manifest\n",
    "binf_file=configfile$binf_manifest\n",
    "ref_dir=configfile$ref_dir\n",
    "qza_dir=configfile$qza_dir"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A data.frame: 24 × 7</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>tax</th><th scope=col>group</th><th scope=col>group.ref</th><th scope=col>tax.ref</th><th scope=col>file.name</th><th scope=col>include</th><th scope=col>classifier</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>blast_cl_gg_gg.qza</th><td>blast  </td><td>closed</td><td>greengenes</td><td>greengenes</td><td>blast_cl_gg_gg.qza        </td><td>Y</td><td>refdb_tax_gg.qza                               </td></tr>\n",
       "\t<tr><th scope=row>blast_dada_gg.qza</th><td>blast  </td><td>dada2 </td><td>greengenes</td><td>greengenes</td><td>blast_dada_gg.qza         </td><td>Y</td><td>refdb_tax_gg.qza                               </td></tr>\n",
       "\t<tr><th scope=row>blast_dn_gg.qza</th><td>blast  </td><td>denovo</td><td>NA        </td><td>greengenes</td><td>blast_dn_gg.qza           </td><td>Y</td><td>refdb_tax_gg.qza                               </td></tr>\n",
       "\t<tr><th scope=row>blast_op_gg_gg.qza</th><td>blast  </td><td>open  </td><td>greengenes</td><td>greengenes</td><td>blast_op_gg_gg.qza        </td><td>Y</td><td>refdb_tax_gg.qza                               </td></tr>\n",
       "\t<tr><th scope=row>scikit_clgg_gg.qza</th><td>scikit </td><td>closed</td><td>greengenes</td><td>greengenes</td><td>scikit_clgg_gg.qza        </td><td>Y</td><td>classify-sklearn_gg-13-8-99-nb-classifier.qza  </td></tr>\n",
       "\t<tr><th scope=row>scikit_dada_gg.qza</th><td>scikit </td><td>dada2 </td><td>greengenes</td><td>greengenes</td><td>scikit_dada_gg.qza        </td><td>Y</td><td>classify-sklearn_gg-13-8-99-nb-classifier.qza  </td></tr>\n",
       "\t<tr><th scope=row>scikit_dn_gg.qza</th><td>scikit </td><td>denovo</td><td>NA        </td><td>greengenes</td><td>scikit_dn_gg.qza          </td><td>Y</td><td>classify-sklearn_gg-13-8-99-nb-classifier.qza  </td></tr>\n",
       "\t<tr><th scope=row>scikit_opgg_gg.qza</th><td>scikit </td><td>open  </td><td>greengenes</td><td>greengenes</td><td>scikit_opgg_gg.qza        </td><td>Y</td><td>classify-sklearn_gg-13-8-99-nb-classifier.qza  </td></tr>\n",
       "\t<tr><th scope=row>vsearch_cl_gg_gg.qza</th><td>vsearch</td><td>closed</td><td>greengenes</td><td>greengenes</td><td>vsearch_cl_gg_gg.qza      </td><td>Y</td><td>refdb_tax_gg.qza                               </td></tr>\n",
       "\t<tr><th scope=row>vsearch_dada_gg.qza</th><td>vsearch</td><td>dada2 </td><td>greengenes</td><td>greengenes</td><td>vsearch_dada_gg.qza       </td><td>Y</td><td>refdb_tax_gg.qza                               </td></tr>\n",
       "\t<tr><th scope=row>vsearch_dn_gg.qza</th><td>vsearch</td><td>denovo</td><td>NA        </td><td>greengenes</td><td>vsearch_dn_gg.qza         </td><td>Y</td><td>refdb_tax_gg.qza                               </td></tr>\n",
       "\t<tr><th scope=row>vsearch_op_gg_gg.qza</th><td>vsearch</td><td>open  </td><td>greengenes</td><td>greengenes</td><td>vsearch_op_gg_gg.qza      </td><td>Y</td><td>refdb_tax_gg.qza                               </td></tr>\n",
       "\t<tr><th scope=row>blast_cl_SILVA_silva.qza</th><td>blast  </td><td>closed</td><td>silva     </td><td>silva     </td><td>blast_cl_SILVA_silva.qza  </td><td>Y</td><td>refdb_tax_silva.qza                            </td></tr>\n",
       "\t<tr><th scope=row>blast_dada_silva.qza</th><td>blast  </td><td>dada2 </td><td>silva     </td><td>silva     </td><td>blast_dada_silva.qza      </td><td>Y</td><td>refdb_tax_silva.qza                            </td></tr>\n",
       "\t<tr><th scope=row>blast_dn_silva.qza</th><td>blast  </td><td>denovo</td><td>NA        </td><td>silva     </td><td>blast_dn_silva.qza        </td><td>Y</td><td>refdb_tax_silva.qza                            </td></tr>\n",
       "\t<tr><th scope=row>blast_op_SILVA_silva.qza</th><td>blast  </td><td>open  </td><td>silva     </td><td>silva     </td><td>blast_op_SILVA_silva.qza  </td><td>Y</td><td>refdb_tax_silva.qza                            </td></tr>\n",
       "\t<tr><th scope=row>scikit_clSILVA_SILVA.qza</th><td>scikit </td><td>closed</td><td>silva     </td><td>silva     </td><td>scikit_clSILVA_SILVA.qza  </td><td>Y</td><td>classify-sklearn_silva-132-99-nb-classifier.qza</td></tr>\n",
       "\t<tr><th scope=row>scikit_dada_SILVA.qza</th><td>scikit </td><td>dada2 </td><td>silva     </td><td>silva     </td><td>scikit_dada_SILVA.qza     </td><td>Y</td><td>classify-sklearn_silva-132-99-nb-classifier.qza</td></tr>\n",
       "\t<tr><th scope=row>scikit_dn_SILVA.qza</th><td>scikit </td><td>denovo</td><td>NA        </td><td>silva     </td><td>scikit_dn_SILVA.qza       </td><td>Y</td><td>classify-sklearn_silva-132-99-nb-classifier.qza</td></tr>\n",
       "\t<tr><th scope=row>scikit_opSILVA_SILVA.qza</th><td>scikit </td><td>open  </td><td>silva     </td><td>silva     </td><td>scikit_opSILVA_SILVA.qza  </td><td>Y</td><td>classify-sklearn_silva-132-99-nb-classifier.qza</td></tr>\n",
       "\t<tr><th scope=row>vsearch_cl_SILVA_silva.qza</th><td>vsearch</td><td>closed</td><td>silva     </td><td>silva     </td><td>vsearch_cl_SILVA_silva.qza</td><td>Y</td><td>refdb_tax_silva.qza                            </td></tr>\n",
       "\t<tr><th scope=row>vsearch_dada_silva.qza</th><td>vsearch</td><td>dada2 </td><td>silva     </td><td>silva     </td><td>vsearch_dada_silva.qza    </td><td>Y</td><td>refdb_tax_silva.qza                            </td></tr>\n",
       "\t<tr><th scope=row>vsearch_dn_silva.qza</th><td>vsearch</td><td>denovo</td><td>NA        </td><td>silva     </td><td>vsearch_dn_silva.qza      </td><td>Y</td><td>refdb_tax_silva.qza                            </td></tr>\n",
       "\t<tr><th scope=row>vsearch_op_SILVA_gg.qza</th><td>vsearch</td><td>open  </td><td>silva     </td><td>silva     </td><td>vsearch_op_SILVA_gg.qza   </td><td>Y</td><td>refdb_tax_silva.qza                            </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 24 × 7\n",
       "\\begin{tabular}{r|lllllll}\n",
       "  & tax & group & group.ref & tax.ref & file.name & include & classifier\\\\\n",
       "  & <fct> & <fct> & <fct> & <fct> & <fct> & <fct> & <fct>\\\\\n",
       "\\hline\n",
       "\tblast\\_cl\\_gg\\_gg.qza & blast   & closed & greengenes & greengenes & blast\\_cl\\_gg\\_gg.qza         & Y & refdb\\_tax\\_gg.qza                               \\\\\n",
       "\tblast\\_dada\\_gg.qza & blast   & dada2  & greengenes & greengenes & blast\\_dada\\_gg.qza          & Y & refdb\\_tax\\_gg.qza                               \\\\\n",
       "\tblast\\_dn\\_gg.qza & blast   & denovo & NA         & greengenes & blast\\_dn\\_gg.qza            & Y & refdb\\_tax\\_gg.qza                               \\\\\n",
       "\tblast\\_op\\_gg\\_gg.qza & blast   & open   & greengenes & greengenes & blast\\_op\\_gg\\_gg.qza         & Y & refdb\\_tax\\_gg.qza                               \\\\\n",
       "\tscikit\\_clgg\\_gg.qza & scikit  & closed & greengenes & greengenes & scikit\\_clgg\\_gg.qza         & Y & classify-sklearn\\_gg-13-8-99-nb-classifier.qza  \\\\\n",
       "\tscikit\\_dada\\_gg.qza & scikit  & dada2  & greengenes & greengenes & scikit\\_dada\\_gg.qza         & Y & classify-sklearn\\_gg-13-8-99-nb-classifier.qza  \\\\\n",
       "\tscikit\\_dn\\_gg.qza & scikit  & denovo & NA         & greengenes & scikit\\_dn\\_gg.qza           & Y & classify-sklearn\\_gg-13-8-99-nb-classifier.qza  \\\\\n",
       "\tscikit\\_opgg\\_gg.qza & scikit  & open   & greengenes & greengenes & scikit\\_opgg\\_gg.qza         & Y & classify-sklearn\\_gg-13-8-99-nb-classifier.qza  \\\\\n",
       "\tvsearch\\_cl\\_gg\\_gg.qza & vsearch & closed & greengenes & greengenes & vsearch\\_cl\\_gg\\_gg.qza       & Y & refdb\\_tax\\_gg.qza                               \\\\\n",
       "\tvsearch\\_dada\\_gg.qza & vsearch & dada2  & greengenes & greengenes & vsearch\\_dada\\_gg.qza        & Y & refdb\\_tax\\_gg.qza                               \\\\\n",
       "\tvsearch\\_dn\\_gg.qza & vsearch & denovo & NA         & greengenes & vsearch\\_dn\\_gg.qza          & Y & refdb\\_tax\\_gg.qza                               \\\\\n",
       "\tvsearch\\_op\\_gg\\_gg.qza & vsearch & open   & greengenes & greengenes & vsearch\\_op\\_gg\\_gg.qza       & Y & refdb\\_tax\\_gg.qza                               \\\\\n",
       "\tblast\\_cl\\_SILVA\\_silva.qza & blast   & closed & silva      & silva      & blast\\_cl\\_SILVA\\_silva.qza   & Y & refdb\\_tax\\_silva.qza                            \\\\\n",
       "\tblast\\_dada\\_silva.qza & blast   & dada2  & silva      & silva      & blast\\_dada\\_silva.qza       & Y & refdb\\_tax\\_silva.qza                            \\\\\n",
       "\tblast\\_dn\\_silva.qza & blast   & denovo & NA         & silva      & blast\\_dn\\_silva.qza         & Y & refdb\\_tax\\_silva.qza                            \\\\\n",
       "\tblast\\_op\\_SILVA\\_silva.qza & blast   & open   & silva      & silva      & blast\\_op\\_SILVA\\_silva.qza   & Y & refdb\\_tax\\_silva.qza                            \\\\\n",
       "\tscikit\\_clSILVA\\_SILVA.qza & scikit  & closed & silva      & silva      & scikit\\_clSILVA\\_SILVA.qza   & Y & classify-sklearn\\_silva-132-99-nb-classifier.qza\\\\\n",
       "\tscikit\\_dada\\_SILVA.qza & scikit  & dada2  & silva      & silva      & scikit\\_dada\\_SILVA.qza      & Y & classify-sklearn\\_silva-132-99-nb-classifier.qza\\\\\n",
       "\tscikit\\_dn\\_SILVA.qza & scikit  & denovo & NA         & silva      & scikit\\_dn\\_SILVA.qza        & Y & classify-sklearn\\_silva-132-99-nb-classifier.qza\\\\\n",
       "\tscikit\\_opSILVA\\_SILVA.qza & scikit  & open   & silva      & silva      & scikit\\_opSILVA\\_SILVA.qza   & Y & classify-sklearn\\_silva-132-99-nb-classifier.qza\\\\\n",
       "\tvsearch\\_cl\\_SILVA\\_silva.qza & vsearch & closed & silva      & silva      & vsearch\\_cl\\_SILVA\\_silva.qza & Y & refdb\\_tax\\_silva.qza                            \\\\\n",
       "\tvsearch\\_dada\\_silva.qza & vsearch & dada2  & silva      & silva      & vsearch\\_dada\\_silva.qza     & Y & refdb\\_tax\\_silva.qza                            \\\\\n",
       "\tvsearch\\_dn\\_silva.qza & vsearch & denovo & NA         & silva      & vsearch\\_dn\\_silva.qza       & Y & refdb\\_tax\\_silva.qza                            \\\\\n",
       "\tvsearch\\_op\\_SILVA\\_gg.qza & vsearch & open   & silva      & silva      & vsearch\\_op\\_SILVA\\_gg.qza    & Y & refdb\\_tax\\_silva.qza                            \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 24 × 7\n",
       "\n",
       "| <!--/--> | tax &lt;fct&gt; | group &lt;fct&gt; | group.ref &lt;fct&gt; | tax.ref &lt;fct&gt; | file.name &lt;fct&gt; | include &lt;fct&gt; | classifier &lt;fct&gt; |\n",
       "|---|---|---|---|---|---|---|---|\n",
       "| blast_cl_gg_gg.qza | blast   | closed | greengenes | greengenes | blast_cl_gg_gg.qza         | Y | refdb_tax_gg.qza                                |\n",
       "| blast_dada_gg.qza | blast   | dada2  | greengenes | greengenes | blast_dada_gg.qza          | Y | refdb_tax_gg.qza                                |\n",
       "| blast_dn_gg.qza | blast   | denovo | NA         | greengenes | blast_dn_gg.qza            | Y | refdb_tax_gg.qza                                |\n",
       "| blast_op_gg_gg.qza | blast   | open   | greengenes | greengenes | blast_op_gg_gg.qza         | Y | refdb_tax_gg.qza                                |\n",
       "| scikit_clgg_gg.qza | scikit  | closed | greengenes | greengenes | scikit_clgg_gg.qza         | Y | classify-sklearn_gg-13-8-99-nb-classifier.qza   |\n",
       "| scikit_dada_gg.qza | scikit  | dada2  | greengenes | greengenes | scikit_dada_gg.qza         | Y | classify-sklearn_gg-13-8-99-nb-classifier.qza   |\n",
       "| scikit_dn_gg.qza | scikit  | denovo | NA         | greengenes | scikit_dn_gg.qza           | Y | classify-sklearn_gg-13-8-99-nb-classifier.qza   |\n",
       "| scikit_opgg_gg.qza | scikit  | open   | greengenes | greengenes | scikit_opgg_gg.qza         | Y | classify-sklearn_gg-13-8-99-nb-classifier.qza   |\n",
       "| vsearch_cl_gg_gg.qza | vsearch | closed | greengenes | greengenes | vsearch_cl_gg_gg.qza       | Y | refdb_tax_gg.qza                                |\n",
       "| vsearch_dada_gg.qza | vsearch | dada2  | greengenes | greengenes | vsearch_dada_gg.qza        | Y | refdb_tax_gg.qza                                |\n",
       "| vsearch_dn_gg.qza | vsearch | denovo | NA         | greengenes | vsearch_dn_gg.qza          | Y | refdb_tax_gg.qza                                |\n",
       "| vsearch_op_gg_gg.qza | vsearch | open   | greengenes | greengenes | vsearch_op_gg_gg.qza       | Y | refdb_tax_gg.qza                                |\n",
       "| blast_cl_SILVA_silva.qza | blast   | closed | silva      | silva      | blast_cl_SILVA_silva.qza   | Y | refdb_tax_silva.qza                             |\n",
       "| blast_dada_silva.qza | blast   | dada2  | silva      | silva      | blast_dada_silva.qza       | Y | refdb_tax_silva.qza                             |\n",
       "| blast_dn_silva.qza | blast   | denovo | NA         | silva      | blast_dn_silva.qza         | Y | refdb_tax_silva.qza                             |\n",
       "| blast_op_SILVA_silva.qza | blast   | open   | silva      | silva      | blast_op_SILVA_silva.qza   | Y | refdb_tax_silva.qza                             |\n",
       "| scikit_clSILVA_SILVA.qza | scikit  | closed | silva      | silva      | scikit_clSILVA_SILVA.qza   | Y | classify-sklearn_silva-132-99-nb-classifier.qza |\n",
       "| scikit_dada_SILVA.qza | scikit  | dada2  | silva      | silva      | scikit_dada_SILVA.qza      | Y | classify-sklearn_silva-132-99-nb-classifier.qza |\n",
       "| scikit_dn_SILVA.qza | scikit  | denovo | NA         | silva      | scikit_dn_SILVA.qza        | Y | classify-sklearn_silva-132-99-nb-classifier.qza |\n",
       "| scikit_opSILVA_SILVA.qza | scikit  | open   | silva      | silva      | scikit_opSILVA_SILVA.qza   | Y | classify-sklearn_silva-132-99-nb-classifier.qza |\n",
       "| vsearch_cl_SILVA_silva.qza | vsearch | closed | silva      | silva      | vsearch_cl_SILVA_silva.qza | Y | refdb_tax_silva.qza                             |\n",
       "| vsearch_dada_silva.qza | vsearch | dada2  | silva      | silva      | vsearch_dada_silva.qza     | Y | refdb_tax_silva.qza                             |\n",
       "| vsearch_dn_silva.qza | vsearch | denovo | NA         | silva      | vsearch_dn_silva.qza       | Y | refdb_tax_silva.qza                             |\n",
       "| vsearch_op_SILVA_gg.qza | vsearch | open   | silva      | silva      | vsearch_op_SILVA_gg.qza    | Y | refdb_tax_silva.qza                             |\n",
       "\n"
      ],
      "text/plain": [
       "                           tax     group  group.ref  tax.ref   \n",
       "blast_cl_gg_gg.qza         blast   closed greengenes greengenes\n",
       "blast_dada_gg.qza          blast   dada2  greengenes greengenes\n",
       "blast_dn_gg.qza            blast   denovo NA         greengenes\n",
       "blast_op_gg_gg.qza         blast   open   greengenes greengenes\n",
       "scikit_clgg_gg.qza         scikit  closed greengenes greengenes\n",
       "scikit_dada_gg.qza         scikit  dada2  greengenes greengenes\n",
       "scikit_dn_gg.qza           scikit  denovo NA         greengenes\n",
       "scikit_opgg_gg.qza         scikit  open   greengenes greengenes\n",
       "vsearch_cl_gg_gg.qza       vsearch closed greengenes greengenes\n",
       "vsearch_dada_gg.qza        vsearch dada2  greengenes greengenes\n",
       "vsearch_dn_gg.qza          vsearch denovo NA         greengenes\n",
       "vsearch_op_gg_gg.qza       vsearch open   greengenes greengenes\n",
       "blast_cl_SILVA_silva.qza   blast   closed silva      silva     \n",
       "blast_dada_silva.qza       blast   dada2  silva      silva     \n",
       "blast_dn_silva.qza         blast   denovo NA         silva     \n",
       "blast_op_SILVA_silva.qza   blast   open   silva      silva     \n",
       "scikit_clSILVA_SILVA.qza   scikit  closed silva      silva     \n",
       "scikit_dada_SILVA.qza      scikit  dada2  silva      silva     \n",
       "scikit_dn_SILVA.qza        scikit  denovo NA         silva     \n",
       "scikit_opSILVA_SILVA.qza   scikit  open   silva      silva     \n",
       "vsearch_cl_SILVA_silva.qza vsearch closed silva      silva     \n",
       "vsearch_dada_silva.qza     vsearch dada2  silva      silva     \n",
       "vsearch_dn_silva.qza       vsearch denovo NA         silva     \n",
       "vsearch_op_SILVA_gg.qza    vsearch open   silva      silva     \n",
       "                           file.name                  include\n",
       "blast_cl_gg_gg.qza         blast_cl_gg_gg.qza         Y      \n",
       "blast_dada_gg.qza          blast_dada_gg.qza          Y      \n",
       "blast_dn_gg.qza            blast_dn_gg.qza            Y      \n",
       "blast_op_gg_gg.qza         blast_op_gg_gg.qza         Y      \n",
       "scikit_clgg_gg.qza         scikit_clgg_gg.qza         Y      \n",
       "scikit_dada_gg.qza         scikit_dada_gg.qza         Y      \n",
       "scikit_dn_gg.qza           scikit_dn_gg.qza           Y      \n",
       "scikit_opgg_gg.qza         scikit_opgg_gg.qza         Y      \n",
       "vsearch_cl_gg_gg.qza       vsearch_cl_gg_gg.qza       Y      \n",
       "vsearch_dada_gg.qza        vsearch_dada_gg.qza        Y      \n",
       "vsearch_dn_gg.qza          vsearch_dn_gg.qza          Y      \n",
       "vsearch_op_gg_gg.qza       vsearch_op_gg_gg.qza       Y      \n",
       "blast_cl_SILVA_silva.qza   blast_cl_SILVA_silva.qza   Y      \n",
       "blast_dada_silva.qza       blast_dada_silva.qza       Y      \n",
       "blast_dn_silva.qza         blast_dn_silva.qza         Y      \n",
       "blast_op_SILVA_silva.qza   blast_op_SILVA_silva.qza   Y      \n",
       "scikit_clSILVA_SILVA.qza   scikit_clSILVA_SILVA.qza   Y      \n",
       "scikit_dada_SILVA.qza      scikit_dada_SILVA.qza      Y      \n",
       "scikit_dn_SILVA.qza        scikit_dn_SILVA.qza        Y      \n",
       "scikit_opSILVA_SILVA.qza   scikit_opSILVA_SILVA.qza   Y      \n",
       "vsearch_cl_SILVA_silva.qza vsearch_cl_SILVA_silva.qza Y      \n",
       "vsearch_dada_silva.qza     vsearch_dada_silva.qza     Y      \n",
       "vsearch_dn_silva.qza       vsearch_dn_silva.qza       Y      \n",
       "vsearch_op_SILVA_gg.qza    vsearch_op_SILVA_gg.qza    Y      \n",
       "                           classifier                                     \n",
       "blast_cl_gg_gg.qza         refdb_tax_gg.qza                               \n",
       "blast_dada_gg.qza          refdb_tax_gg.qza                               \n",
       "blast_dn_gg.qza            refdb_tax_gg.qza                               \n",
       "blast_op_gg_gg.qza         refdb_tax_gg.qza                               \n",
       "scikit_clgg_gg.qza         classify-sklearn_gg-13-8-99-nb-classifier.qza  \n",
       "scikit_dada_gg.qza         classify-sklearn_gg-13-8-99-nb-classifier.qza  \n",
       "scikit_dn_gg.qza           classify-sklearn_gg-13-8-99-nb-classifier.qza  \n",
       "scikit_opgg_gg.qza         classify-sklearn_gg-13-8-99-nb-classifier.qza  \n",
       "vsearch_cl_gg_gg.qza       refdb_tax_gg.qza                               \n",
       "vsearch_dada_gg.qza        refdb_tax_gg.qza                               \n",
       "vsearch_dn_gg.qza          refdb_tax_gg.qza                               \n",
       "vsearch_op_gg_gg.qza       refdb_tax_gg.qza                               \n",
       "blast_cl_SILVA_silva.qza   refdb_tax_silva.qza                            \n",
       "blast_dada_silva.qza       refdb_tax_silva.qza                            \n",
       "blast_dn_silva.qza         refdb_tax_silva.qza                            \n",
       "blast_op_SILVA_silva.qza   refdb_tax_silva.qza                            \n",
       "scikit_clSILVA_SILVA.qza   classify-sklearn_silva-132-99-nb-classifier.qza\n",
       "scikit_dada_SILVA.qza      classify-sklearn_silva-132-99-nb-classifier.qza\n",
       "scikit_dn_SILVA.qza        classify-sklearn_silva-132-99-nb-classifier.qza\n",
       "scikit_opSILVA_SILVA.qza   classify-sklearn_silva-132-99-nb-classifier.qza\n",
       "vsearch_cl_SILVA_silva.qza refdb_tax_silva.qza                            \n",
       "vsearch_dada_silva.qza     refdb_tax_silva.qza                            \n",
       "vsearch_dn_silva.qza       refdb_tax_silva.qza                            \n",
       "vsearch_op_SILVA_gg.qza    refdb_tax_silva.qza                            "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#decide which files to include, add Y or N flag\n",
    "benchmarkdf = read.csv(binf_file,sep=\"\\t\")\n",
    "benchmarkdf = subset(benchmarkdf,include==\"Y\")\n",
    "rownames(benchmarkdf) = benchmarkdf$file.name\n",
    "benchmarkdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "################################################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "'blast_cl_gg_gg.qza'"
      ],
      "text/latex": [
       "'blast\\_cl\\_gg\\_gg.qza'"
      ],
      "text/markdown": [
       "'blast_cl_gg_gg.qza'"
      ],
      "text/plain": [
       "[1] \"blast_cl_gg_gg.qza\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#List of all variables to test\n",
    "test_list = rownames(benchmarkdf)\n",
    "test_list = test_list[1]\n",
    "test_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "phylo_create <- function(tax_in,tab_in){\n",
    "    #Read OTUS\n",
    "    otus<-read_qza(tab_in)\n",
    "\n",
    "    #Read Greengenes taxonomy file\n",
    "    taxonomy<-read_qza(tax_in)\n",
    "\n",
    "    #Create taxonomy table\n",
    "    tax_table = data.frame()\n",
    "    col = 1\n",
    "    row = 1\n",
    "\n",
    "    for (i in taxonomy$data$Taxon){\n",
    "        tax_list = as.list(strsplit(as.character(i),\";\"))\n",
    "        for (i in 1:lengths(tax_list)){\n",
    "            tax_table[row,col] = tax_list[[1]][i]\n",
    "            col = col + 1\n",
    "        }\n",
    "        col = 1\n",
    "        row = row + 1\n",
    "    }\n",
    "    rownames(tax_table)<-taxonomy$data$Feature.ID\n",
    "    colnames(tax_table)<-c(\"Domain\", \"Phylum\", \"Class\", \"Order\", \"Family\", \"Genus\", \"Species\")\n",
    "    tax_convert<-as.matrix(tax_table)  \n",
    "\n",
    "    #read metadata\n",
    "    metadata<-read.table('../manifest/m_cleaned.txt',sep='\\t', header=T, row.names=1, comment=\"\")\n",
    "\n",
    "    #Create phylo object\n",
    "    phy_obj<-phyloseq(otu_table(otus$data, taxa_are_rows = T), tax_table(tax_convert), sample_data(metadata))\n",
    "    print(phy_obj) \n",
    "    return(phy_obj)\n",
    "} "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"\\n****************************************************************************\"\n",
      "[1] \"Analysis for  blast_cl_gg_gg.qza\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in unzip(file, exdir = tmp):\n",
      "\"error 1 in extracting from zip file\"\n"
     ]
    },
    {
     "ename": "ERROR",
     "evalue": "Error in unzip(file, exdir = tmp, list = TRUE): zip file '/Users/slsevilla/Google Drive/MyDocuments_Current/Education/George Mason University/Dissertation/data/aim1/ref/refdb_tax_gg.qza' cannot be opened\n",
     "output_type": "error",
     "traceback": [
      "Error in unzip(file, exdir = tmp, list = TRUE): zip file '/Users/slsevilla/Google Drive/MyDocuments_Current/Education/George Mason University/Dissertation/data/aim1/ref/refdb_tax_gg.qza' cannot be opened\nTraceback:\n",
      "1. phylo_create(tax_file, table_file)",
      "2. read_qza(tax_in)   # at line 6 of file <text>",
      "3. unzip(file, exdir = tmp, list = TRUE)"
     ]
    }
   ],
   "source": [
    "for (variable in test_list){\n",
    "    print(\"\\n****************************************************************************\")\n",
    "    print(paste(\"Analysis for \",variable))\n",
    "    \n",
    "    #Create the phyloseq object\n",
    "    tax_file = paste(ref_dir,benchmarkdf[variable,\"classifier\"],sep=\"\")\n",
    "    table_file = paste(qza_dir,variable,sep=\"\")\n",
    "    phy_obj = phylo_create(tax_file,table_file)  \n",
    "\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "    \n",
    "    #Create glomed, subsetted obj\n",
    "    seq_obj = phy_filt(phy_obj)\n",
    "    remove(phy_obj)\n",
    "    \n",
    "    #Create OTUdf\n",
    "    OTUdf = otu_extract(seq_obj)\n",
    "    \n",
    "    #Create TAXdf\n",
    "    TAXdf = tax_extract(seq_obj)\n",
    "    \n",
    "    #Clean df's\n",
    "    OTUdf_clean = otu_tax_clean(OTUdf,TAXdf)\n",
    "    \n",
    "    #Create TP,FP,FN df\n",
    "    tax_ref = as.character(benchmarkdf[variable,\"tax.ref\"])\n",
    "    if(tax_ref=='silva'){\n",
    "        CONTdf_use = CONTdf_s\n",
    "        F1df = bench_compare(OTUdf_clean, CONTdf_use)\n",
    "    } else{\n",
    "        CONTdf_use = CONTdf_g\n",
    "        F1df = bench_compare(OTUdf_clean, CONTdf_use)\n",
    "    }\n",
    "    \n",
    "    #Create TP,FP,FN counts    \n",
    "    if(benchmarkdf[variable,\"group.type\"]==\"dada\" | benchmarkdf[variable,\"group.type\"]==\"dn\"){\n",
    "        bench_key = paste(benchmarkdf[variable,\"group.type\"],benchmarkdf[variable,\"tax.type\"],\n",
    "                         benchmarkdf[variable,\"tax.ref\"],sep=\"_\")\n",
    "    } else{\n",
    "        bench_key = paste(benchmarkdf[variable,\"group.type\"],benchmarkdf[variable,\"group.ref\"],\n",
    "                         benchmarkdf[variable,\"tax.type\"],benchmarkdf[variable,\"tax.ref\"],sep=\"_\")\n",
    "    }         \n",
    "    F1df_counts = bench_counts(OTUdf_clean,F1df,bench_key,variable)\n",
    "    temp = F1df_counts\n",
    "    temp$sample_name = rownames(temp)\n",
    "    mergedf_f1stats = rbind(mergedf_f1stats,temp)\n",
    "    \n",
    "    #Create violin plots for each control - counts\n",
    "    x_list = c(\"FN\",\"TP\",\"FP\")\n",
    "    y_labs = c(\"False negative\",\"True positive\", \"False positive\")\n",
    "    plot_f1_violin(F1df_counts,x_list,y_labs,\"Subject.ID\",bench_key,\"counts\")\n",
    "    \n",
    "    #Create violin plots for each control - summary\n",
    "    x_list = c(\"TAR\",\"TDR\",\"F1\")\n",
    "    y_labs = c(\"Taxonmic Accuracy Rate\",\"Taxonmic Detection Rate\", \"F-Measure\")\n",
    "    plot_f1_violin(F1df_counts,x_list,y_labs,\"Subject.ID\",bench_key,\"summary\")\n",
    "\n",
    "    #calculate relative abundance\n",
    "    RAdf = cal_relabun(OTUdf_clean)\n",
    "    RAdf$Sample_name = rownames(RAdf)\n",
    "    \n",
    "    #Create scatter plots\n",
    "    #Determine expected to observed scatter plots, normality, spearman coeff \n",
    "    title = paste(\"Scatterplots of controls: \", variable,sep=\"\")\n",
    "    EOdf_stats = cal_EO(RAdf,CONTdf_use,variable,title)    \n",
    "    EOdf_stats_merged = rbind(EOdf_stats_merged,EOdf_stats)   \n",
    "    \n",
    "    #Create averaged OTU df\n",
    "    dissdf = create_av_RAdf(RAdf,bench_key,variable)\n",
    "    if(benchmarkdf[variable,\"tax.ref\"]==\"silva\"){\n",
    "        dissdf_s = bind_rows(dissdf_s,dissdf)\n",
    "    } else{\n",
    "        dissdf_g = bind_rows(dissdf_g,dissdf)\n",
    "    }\n",
    "    \n",
    "    #save files\n",
    "    #write.csv(OTUdf_clean,paste(stat_dir,\"OTUcounts_\",bench_key,\".csv\",sep=\"\"))\n",
    "    #write.csv(F1df,paste(stat_dir,\"F1counts_\",bench_key,\".csv\",sep=\"\"))\n",
    "    write.csv(F1df_counts,paste(stats_dir,\"F1_values_binf_\",bench_key,\".csv\",sep=\"\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "#Create a merged dfs for all variables\n",
    "mergedf_f1stats = data.frame()\n",
    "EOdf_stats_merged = data.frame()\n",
    "dissdf_s = data.frame()\n",
    "dissdf_g = data.frame()\n",
    "\n",
    "#List of all variables to test\n",
    "test_list = unique(benchmarkdf$x)\n",
    "#test_list = test_list[1]\n",
    "#test_list\n",
    "\n",
    "for (variable in test_list){\n",
    "    print(\"\\n****************************************************************************\")\n",
    "    print(paste(\"Analysis for \",variable))\n",
    "    \n",
    "    #Create the phyloseq object\n",
    "    tax_file = paste(data_dir,'taxonomic_classification/',variable,sep=\"\")\n",
    "    table_file = paste(data_dir,'taxonomic_classification/',benchmarkdf[variable,\"table.file\"],sep=\"\")\n",
    "    phy_obj = phylo_create(tax_file,table_file)  \n",
    "    \n",
    "    #Create glomed, subsetted obj\n",
    "    seq_obj = phy_filt(phy_obj)\n",
    "    remove(phy_obj)\n",
    "    \n",
    "    #Create OTUdf\n",
    "    OTUdf = otu_extract(seq_obj)\n",
    "    \n",
    "    #Create TAXdf\n",
    "    TAXdf = tax_extract(seq_obj)\n",
    "    \n",
    "    #Clean df's\n",
    "    OTUdf_clean = otu_tax_clean(OTUdf,TAXdf)\n",
    "    \n",
    "    #Create TP,FP,FN df\n",
    "    tax_ref = as.character(benchmarkdf[variable,\"tax.ref\"])\n",
    "    if(tax_ref=='silva'){\n",
    "        CONTdf_use = CONTdf_s\n",
    "        F1df = bench_compare(OTUdf_clean, CONTdf_use)\n",
    "    } else{\n",
    "        CONTdf_use = CONTdf_g\n",
    "        F1df = bench_compare(OTUdf_clean, CONTdf_use)\n",
    "    }\n",
    "    \n",
    "    #Create TP,FP,FN counts    \n",
    "    if(benchmarkdf[variable,\"group.type\"]==\"dada\" | benchmarkdf[variable,\"group.type\"]==\"dn\"){\n",
    "        bench_key = paste(benchmarkdf[variable,\"group.type\"],benchmarkdf[variable,\"tax.type\"],\n",
    "                         benchmarkdf[variable,\"tax.ref\"],sep=\"_\")\n",
    "    } else{\n",
    "        bench_key = paste(benchmarkdf[variable,\"group.type\"],benchmarkdf[variable,\"group.ref\"],\n",
    "                         benchmarkdf[variable,\"tax.type\"],benchmarkdf[variable,\"tax.ref\"],sep=\"_\")\n",
    "    }         \n",
    "    F1df_counts = bench_counts(OTUdf_clean,F1df,bench_key,variable)\n",
    "    temp = F1df_counts\n",
    "    temp$sample_name = rownames(temp)\n",
    "    mergedf_f1stats = rbind(mergedf_f1stats,temp)\n",
    "    \n",
    "    #Create violin plots for each control - counts\n",
    "    x_list = c(\"FN\",\"TP\",\"FP\")\n",
    "    y_labs = c(\"False negative\",\"True positive\", \"False positive\")\n",
    "    plot_f1_violin(F1df_counts,x_list,y_labs,\"Subject.ID\",bench_key,\"counts\")\n",
    "    \n",
    "    #Create violin plots for each control - summary\n",
    "    x_list = c(\"TAR\",\"TDR\",\"F1\")\n",
    "    y_labs = c(\"Taxonmic Accuracy Rate\",\"Taxonmic Detection Rate\", \"F-Measure\")\n",
    "    plot_f1_violin(F1df_counts,x_list,y_labs,\"Subject.ID\",bench_key,\"summary\")\n",
    "\n",
    "    #calculate relative abundance\n",
    "    RAdf = cal_relabun(OTUdf_clean)\n",
    "    RAdf$Sample_name = rownames(RAdf)\n",
    "    \n",
    "    #Create scatter plots\n",
    "    #Determine expected to observed scatter plots, normality, spearman coeff \n",
    "    title = paste(\"Scatterplots of controls: \", variable,sep=\"\")\n",
    "    EOdf_stats = cal_EO(RAdf,CONTdf_use,variable,title)    \n",
    "    EOdf_stats_merged = rbind(EOdf_stats_merged,EOdf_stats)   \n",
    "    \n",
    "    #Create averaged OTU df\n",
    "    dissdf = create_av_RAdf(RAdf,bench_key,variable)\n",
    "    if(benchmarkdf[variable,\"tax.ref\"]==\"silva\"){\n",
    "        dissdf_s = bind_rows(dissdf_s,dissdf)\n",
    "    } else{\n",
    "        dissdf_g = bind_rows(dissdf_g,dissdf)\n",
    "    }\n",
    "    \n",
    "    #save files\n",
    "    #write.csv(OTUdf_clean,paste(stat_dir,\"OTUcounts_\",bench_key,\".csv\",sep=\"\"))\n",
    "    #write.csv(F1df,paste(stat_dir,\"F1counts_\",bench_key,\".csv\",sep=\"\"))\n",
    "    write.csv(F1df_counts,paste(stats_dir,\"F1_values_binf_\",bench_key,\".csv\",sep=\"\"))\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "#Read in taxonomy file\n",
    "CONTdf = read.csv('../taxonomy/Taxonomy_Complete.csv',stringsAsFactors = FALSE)\n",
    "\n",
    "#Convert the tax file to SILVA naming with \"D_5__\"\n",
    "CONTdf_s = CONTdf\n",
    "for (i in 1:nrow(CONTdf)){\n",
    "    old = as.character(CONTdf[i,\"Genus\"])\n",
    "    new = paste(\"D_5__\",old,sep=\"\")\n",
    "    new = gsub(\" \", \".\", new, fixed = TRUE)\n",
    "    CONTdf_s[i,\"Genus\"] = new\n",
    "    \n",
    "    old = as.character(CONTdf[i,\"Family\"])\n",
    "    new = paste(\"D_4__\",old,sep=\"\")\n",
    "    new = gsub(\" \", \".\", new, fixed = TRUE)\n",
    "    CONTdf_s[i,\"Family\"] = new\n",
    "}\n",
    "\n",
    "#Convert the tax file to gg naming with \"f\"\n",
    "CONTdf_g = CONTdf\n",
    "for (i in 1:nrow(CONTdf)){\n",
    "    old = as.character(CONTdf[i,\"Genus\"])\n",
    "    new = paste(\"g__\",old,sep=\"\")\n",
    "    new = gsub(\" \", \".\", new, fixed = TRUE)\n",
    "    CONTdf_g[i,\"Genus\"] = new\n",
    "    \n",
    "    old = as.character(CONTdf[i,\"Family\"])\n",
    "    new = paste(\"f__\",old,sep=\"\")\n",
    "    new = gsub(\" \", \".\", new, fixed = TRUE)\n",
    "    CONTdf_g[i,\"Family\"] = new\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "phy_filt <- function(phy_in){\n",
    "    #glom to tax level, subset for only seq controls\n",
    "    phy_obj = tax_glom(phy_in,taxrank=\"Genus\") #only merges genera that have the same upper taxonomy\n",
    "    phy_obj = subset_samples(phy_obj,Sample.Type==\"Seq_Control\" & Sample.Des!=\"Seq_Blank\")\n",
    "    return(phy_obj)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "otu_extract <- function(phy_in){\n",
    "    # Extract abundance matrix from the phyloseq object\n",
    "    OTU1 = as(otu_table(phy_in), \"matrix\")\n",
    "    if(taxa_are_rows(phy_in)){OTU1 <- t(OTU1)} # transpose\n",
    "    OTUdf = as.data.frame(OTU1)# Coerce to data.frame\n",
    "    return(OTUdf)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "tax_extract <- function(phy_in){\n",
    "    #extract taxonomy information into dataframe\n",
    "    TAXdf = tax_table(phy_in)\n",
    "    return(TAXdf)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "otu_tax_clean <- function(OTU_in,TAX_in){\n",
    "    #Remove zero cols\n",
    "    OTUdf_clean = OTU_in[, which(colSums(OTU_in) != 0)]\n",
    "\n",
    "    #rename col from features to genus\n",
    "    for (i in colnames(OTU_in)){\n",
    "        newname = paste(as.character(TAX_in[i,\"Family\"]),as.character(TAX_in[i,\"Genus\"]),sep=\"\")\n",
    "        \n",
    "        #standardize gg naming\n",
    "        newname = trimws(newname) #remove first space in gg naming\n",
    "        newname = sub(\" g__\",\"g__\",newname)\n",
    "        newname = sub(\"\\\\[\",\"\",newname)\n",
    "        newname = sub(\"\\\\]\",\"\",newname)\n",
    "        \n",
    "        if(is.na(newname)){\n",
    "            next;\n",
    "        } else{\n",
    "            names(OTUdf_clean)[names(OTUdf_clean) == i] <- newname\n",
    "        }\n",
    "    }\n",
    "    \n",
    "    #add metadata col\n",
    "    metadata<-read.table('../manifest/m_cleaned.txt',sep='\\t', header=T, row.names=1, comment=\"\")\n",
    "    for (a in rownames(OTUdf_clean)){\n",
    "        OTUdf_clean[a,\"Subject.ID\"] = metadata[a,\"Subject.ID\"]\n",
    "    }\n",
    "    OTUdf_clean = data.frame(OTUdf_clean,stringsAsFactors=FALSE)\n",
    "    \n",
    "    return(OTUdf_clean)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Create TP,FP,FN dataframe\n",
    "bench_compare <- function(OTUdf_in,CONTdf_in){\n",
    "    \n",
    "    F1df = OTUdf_in\n",
    "    F1df$Sample_names = rownames(F1df) #create col of sample_names\n",
    "    \n",
    "    #For each of the controls in the cleaned df\n",
    "    for (cont in unique(OTUdf_in$Subject.ID)){\n",
    "        temp=filter(CONTdf_in, get(cont) !=0) #create df of taxa related to control\n",
    "        \n",
    "        #Create taxa list, merging family and genus - this will help avoid id errors from taxonomic differences\n",
    "        #IE one taxa labeled at the genus level may be from family1 and the same taxa may also be from family2\n",
    "        taxa_list=list()\n",
    "        for (x in rownames(temp)){taxa_list = c(taxa_list,paste(temp[x,\"Family\"],temp[x,\"Genus\"],sep=\"\"))}  \n",
    "        \n",
    "        #Create list of samples that match the control\n",
    "        sample_list = filter(F1df, Subject.ID == cont)$Sample_names\n",
    "        \n",
    "        #If the data is not in the df (FN), then add the taxa as a new col\n",
    "        for (taxa in taxa_list){ \n",
    "            if(!(taxa %in% unique(colnames(F1df)))){\n",
    "                F1df[taxa[!(taxa %in% colnames(F1df))]] = 0\n",
    "            } \n",
    "        }\n",
    "\n",
    "        for (sample in sample_list){\n",
    "            for (col in unique(colnames(F1df))[unique(colnames(F1df)) != \"Subject.ID\" & unique(colnames(F1df)) != \"Sample_names\"]){ \n",
    "                #removes subjectID, sample_names cols from loop\n",
    "                value_check = F1df[sample,col]\n",
    "\n",
    "                if(col %in% taxa_list && value_check > 0){ #column matches expected taxa, has features\n",
    "                    F1df[sample,col] = \"TP\"\n",
    "                } else if (col %in% taxa_list && value_check == 0){ #column matches expected taxa, no features\n",
    "                    F1df[sample,col] = \"FN\"\n",
    "                } else if ((!col %in% taxa_list) && value_check > 0){ #column doesnt match expected taxa, has features\n",
    "                    F1df[sample,col] = \"FP\"\n",
    "                } else{ #column doesnt match expected taxa, has no features\n",
    "                    next\n",
    "                }\n",
    "            }\n",
    "        }\n",
    "    }\n",
    "    return(F1df)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Create counts of FP, FN, TP\n",
    "bench_counts <- function(OTUdf_in,F1_in,bmark_in,var_in){\n",
    "    F1df_counts = subset(OTUdf_in, select=c(\"Subject.ID\"))\n",
    "\n",
    "    for (sample in rownames(F1df)){\n",
    "        FN = rowSums(F1_in==\"FN\")[sample][[1]]\n",
    "        FP = rowSums(F1_in==\"FP\")[sample][[1]]\n",
    "        TP = rowSums(F1_in==\"TP\")[sample][[1]]\n",
    "\n",
    "        F1df_counts[sample,\"check\"]= FN + TP\n",
    "        F1df_counts[sample,\"FN\"] = FN\n",
    "        F1df_counts[sample,\"FP\"] = FP\n",
    "        F1df_counts[sample,\"TP\"] = TP\n",
    "        F1df_counts[sample,\"TAR\"] = TP / (TP + FP)\n",
    "        F1df_counts[sample,\"TDR\"] = TP / (TP + FN)\n",
    "\n",
    "        num = F1df_counts[sample,\"TAR\"]*F1df_counts[sample,\"TDR\"]\n",
    "        den = F1df_counts[sample,\"TAR\"]+F1df_counts[sample,\"TDR\"] \n",
    "        F1df_counts[sample,\"F1\"] = 2*(num/den)\n",
    "        F1df_counts[sample,\"Subject.ID\"] = OTUdf_in[sample,\"Subject.ID\"]\n",
    "        \n",
    "        F1df_counts[sample,\"bmark\"] = bmark_in\n",
    "        F1df_counts[sample,\"group.type\"] = benchmarkdf[variable,\"group.type\"]  \n",
    "        F1df_counts[sample,\"group.ref\"] = benchmarkdf[variable,\"group.ref\"]\n",
    "        F1df_counts[sample,\"tax.type\"] = benchmarkdf[variable,\"tax.type\"]\n",
    "        F1df_counts[sample,\"tax.ref\"] = benchmarkdf[variable,\"tax.ref\"]\n",
    "    }\n",
    "    return(F1df_counts)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "#F1_in=F1df_counts\n",
    "bench_means <- function(F1_in){\n",
    "    df = with(F1_in, tapply(X = FN,       \n",
    "                    INDEX = Subject.ID,   \n",
    "                    FUN = mean,     \n",
    "                    na.rm = TRUE))\n",
    "    return(df)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "plot_violin <- function(df_in,x_lab,y_ax,y_lab,col_in){\n",
    "    p = ggplot(df_in, aes(x=get(x_lab), y=get(y_ax), color=get(col_in))) + \n",
    "        geom_violin(trim=FALSE) + \n",
    "        labs(color = col_in) +\n",
    "        theme(axis.text.x = element_text(angle = 90, hjust = 1),\n",
    "              axis.title.x = element_blank(), \n",
    "             legend.position = \"none\") + \n",
    "        ylab(y_lab) +\n",
    "        stat_summary(fun.data=mean_sdl, color=\"red\")\n",
    "    return (p)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_f1_violin <- function(F1_in,x_list,y_list,id_in,title,ftype){\n",
    "    \n",
    "    p_1 = plot_violin(F1_in,id_in,x_list[1],y_list[1],id_in)\n",
    "    p_2 = plot_violin(F1_in,id_in,x_list[2],y_list[2],id_in)\n",
    "    p_3 = plot_violin(F1_in,id_in,x_list[3],y_list[3],id_in)\n",
    "\n",
    "    #Create count violin plots\n",
    "    fig = ggarrange(p_1, p_2, p_3, \n",
    "              labels = c(\"A\", \"B\", \"C\"),\n",
    "              title = title,\n",
    "              #common.legend = TRUE,\n",
    "              ncol = 2, nrow = 2)\n",
    "    annotate_figure(fig, #https://rpkgs.datanovia.com/ggpubr/reference/annotate_figure.html\n",
    "                   top = text_grob(title, color = \"blue\", face = \"bold\", size = 14)\n",
    "    )\n",
    "    \n",
    "    ggsave(paste(img_dir,\"violin_f1_binf_\",ftype,\"_\",title,\".jpg\",sep=\"\"))\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "cal_relabun <- function(df_in){\n",
    "    temp = df_in\n",
    "    temp$Total = rowSums(df_in[,c(-(ncol(df_in)))])\n",
    "    col_list = subset(colnames(temp),colnames(temp)!=\"Subject.ID\" & colnames(temp)!=\"Total\")\n",
    "\n",
    "    #cal_relabund <- function(x){}\n",
    "    for (row in rownames(temp)){\n",
    "        for (col in col_list){\n",
    "            val = temp[row,col]\n",
    "            if(val==0){\n",
    "                next\n",
    "            } else{\n",
    "                temp[row,col] = val/temp[row,\"Total\"]\n",
    "            }\n",
    "        }\n",
    "    }\n",
    "    return(temp)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h2> Taxa Database <h2>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "create_taxadf <- function(cont,contdf_in){\n",
    "    taxdf=filter(contdf_in, get(cont) !=0) #create df of taxa related to control\n",
    "    taxdf$taxa.id=paste(taxdf$Family,taxdf$Genus,sep=\"\") #create fam/genus col\n",
    "    taxdf = taxdf[,c(\"taxa.id\",cont),drop=FALSE] #remove all cols but the merged and control col\n",
    "    taxdf = aggregate(get(cont) ~ taxa.id, data = taxdf, sum) #collapse df for any dups    \n",
    "    names(taxdf)[2] = cont\n",
    "    rownames(taxdf) = taxdf$taxa.id\n",
    "    return(taxdf)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h2> Expected to Observed Correlation <h2>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "cal_EO <- function(df_in,contdf_in,bench_in,title_in){\n",
    "    count=1\n",
    "    plot_list = list()\n",
    "    EOdf_merged = data.frame()\n",
    "    eo_out = data.frame()\n",
    "    \n",
    "    #For each of the controls in the cleaned df\n",
    "    for (cont in unique(df_in$Subject.ID)){\n",
    "\n",
    "        taxdf = create_taxadf(cont,contdf_in)\n",
    "        taxa_list=taxdf$taxa.id\n",
    "        \n",
    "        #Create list of samples that match the control\n",
    "        sample_list = filter(RAdf, Subject.ID == cont)$Sample_name\n",
    "        EOdf= data.frame()\n",
    "        \n",
    "        for (taxa in taxa_list){\n",
    "            for(sample in sample_list){\n",
    "                EOdf[nrow(EOdf)+1,\"Sample_name\"] = sample\n",
    "                EOdf[nrow(EOdf),\"taxa.id\"] = taxa\n",
    "                EOdf[nrow(EOdf),\"Subject.ID\"] = cont\n",
    "                EOdf[nrow(EOdf),\"exp\"] = taxdf[taxa,cont]\n",
    "                EOdf[nrow(EOdf),\"bench\"] = bench_in\n",
    "\n",
    "                if(is.null(df_in[sample,taxa])){\n",
    "                    next\n",
    "                } else{\n",
    "                    EOdf[nrow(EOdf),\"obs\"] = df_in[sample,taxa]\n",
    "                }\n",
    "            }\n",
    "        }\n",
    "        EOdf_merged = rbind(EOdf_merged,EOdf) #Merge dfs\n",
    "        \n",
    "        #Create scatter plots\n",
    "        plot_list[[count]] = plot_EO(EOdf,cont)\n",
    "        count=count+1\n",
    "        \n",
    "        #Calc stats\n",
    "        EOdf[is.na(EOdf)] = 0\n",
    "        print(EOdf)\n",
    "        temp = cal_EO_stattest(EOdf,cont,bench_in)\n",
    "        eo_out = rbind(eo_out,temp) #merge df       \n",
    "    }\n",
    "    \n",
    "    #print plots\n",
    "    fig = ggarrange(plot_list[[1]], plot_list[[2]], plot_list[[3]], plot_list[[4]], plot_list[[5]], plot_list[[6]], \n",
    "              labels = c(\"A\", \"B\", \"C\", \"D\", \"E\", \"F\"),\n",
    "              #top = as.character(title_in),\n",
    "              #common.legend = TRUE#,\n",
    "              ncol = 2, nrow = 3)\n",
    "    annotate_figure(fig, top = text_grob(title_in, color = \"blue\", face = \"bold\", size = 14)\n",
    "    )\n",
    "    ggsave(paste(img_dir,\"scatter_binf_\",bench_in,\".jpg\",sep=\"\"))\n",
    "\n",
    "    return(eo_out)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_EO <- function(df_in,cont_in){\n",
    "    my_data=subset(df_in,Subject.ID==cont_in)\n",
    "\n",
    "    p = ggscatter(my_data, x = \"obs\", y = \"exp\", \n",
    "              add = \"reg.line\", conf.int = TRUE, \n",
    "              cor.coef = TRUE, cor.method = \"pearson\",\n",
    "              xlab = \"Expected Rel Abunance (%)\", ylab = \"Observed Rel Abun (%)\") +\n",
    "        ggtitle(paste(\"Scatter plot of Rel Abun:\",cont_in,sep=\"\"))\n",
    "    return(p)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "cal_EO_stattest <- function(df_in,cont_in,bench_in){\n",
    "    my_data=subset(df_in,Subject.ID==cont_in)\n",
    "    df_out=data.frame()\n",
    "    \n",
    "    df_out[nrow(df_out)+1,\"bmark\"] = bench_in\n",
    "    df_out[nrow(df_out),\"cont\"] = cont_in\n",
    "    df_out[nrow(df_out),\"shapiro\"] = shapiro.test(my_data$obs)$p.value\n",
    "    df_out[nrow(df_out),\"spear.p\"] = cor.test(my_data$obs,my_data$exp,method = \"spearman\")$p.value\n",
    "    df_out[nrow(df_out),\"spear.rho\"] = cor.test(my_data$obs,my_data$exp,method = \"spearman\")$estimate[[1]]\n",
    "\n",
    "    #-1 indicates a strong negative correlation : this means that every time x increases, y decreases (left panel figure)\n",
    "    #0 means that there is no association between the two variables (x and y) (middle panel figure)\n",
    "    #1 indicates a strong positive correlation : this means that y increases with x (right panel figure)\n",
    "    return(df_out)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h2>Dissimilarity <h2>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Create averaged relative abundance df\n",
    "create_av_RAdf <- function(RAdf_in,bench_in,var_in){\n",
    "    tmp = aggregate(RAdf_in[,1:(ncol(RAdf_in)-3)], list(RAdf_in$Subject.ID), mean)\n",
    "    tmp$bmark=bench_in\n",
    "    tmp$db = benchmarkdf[var_in,\"tax.ref\"]\n",
    "    return(tmp)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "#Add expected values to RA average dfs\n",
    "RAdf_addcontrols<-function(cont_in,df_in){\n",
    "    \n",
    "    tmp=get(df_in)\n",
    "    cont_leng = length((unique(OTUdf_clean$Subject.ID)))\n",
    "    \n",
    "    for(i in 1:cont_leng){\n",
    "        cont = as.character(tmp[i,\"Group.1\"])\n",
    "        taxadf = create_taxadf(cont,get(cont_in))\n",
    "        tmp_count = nrow(tmp)+1\n",
    "        \n",
    "        for(tx_row in 1:nrow(taxadf)){\n",
    "            tmp[tmp_count,\"Group.1\"]=cont\n",
    "            tmp[tmp_count,taxadf[tx_row,\"taxa.id\"]] = taxadf[tx_row,cont]\n",
    "            tmp[tmp_count,\"bmark\"]=\"expected\"\n",
    "            tmp[tmp_count,\"db\"]=tmp[i,\"db\"]\n",
    "        }\n",
    "        \n",
    "    }\n",
    "    tmp[is.na(tmp)] = 0 \n",
    "    return(tmp)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "calc_dist <- function(RA_in,CONTdf_in){\n",
    "    rownames(RA_in)=paste(RA_in$Group.1,RA_in$bmark,sep=\"-\")\n",
    "    RA_in = subset(RA_in, select = -c(db,Group.1,bmark))\n",
    "\n",
    "    taxa_list<-unique(paste(CONTdf_in$Family,CONTdf_in$Genus,sep=\"\"))\n",
    "    df_clean = RA_in[,colnames(RA_in) %in% taxa_list] #remove cols not expected\n",
    "\n",
    "    df_out = vegdist(df_clean,method=\"bray\")   \n",
    "    return(df_out)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_dist_cluster <- function(df_in,ref_in){\n",
    "    fname = paste(img_dir,\"hclust_summary_bc_binf_allgroups_\",ref_in,\".png\",sep=\"\")\n",
    "    png(fname,width = 900, height = 600)\n",
    "    p = plot(hclust(df_in), cex = .9, main = \"All Groups - Bray-Curtis Dissimilarity\",\n",
    "             xlab=\"Bioinformatic Variables\", sub=\"\")\n",
    "    dev.off()\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_dist_violin <- function(df_in,ref_in){\n",
    "    df_in = as.data.frame(as.matrix(df_in))\n",
    "    \n",
    "    df_plot=data.frame()\n",
    "    for (row in rownames(df_in)){\n",
    "        splits = strsplit(row,\"-\")\n",
    "        if(splits[[1]][2]==\"expected\"){\n",
    "            next\n",
    "        }else{\n",
    "            df_plot[nrow(df_plot)+1,\"dist\"]=df_in[row,paste(splits[[1]][1],\"expected\",sep=\"-\")]\n",
    "            df_plot[nrow(df_plot),\"name\"]=splits[[1]][2]\n",
    "        }\n",
    "    }\n",
    "\n",
    "    p = ggplot(df_plot, aes(x=name, y=dist, color=name)) + \n",
    "            geom_violin(trim=FALSE) + \n",
    "            labs(color = df_plot$name) +\n",
    "            theme(axis.text.x = element_text(angle = 90, hjust = 1),\n",
    "                  axis.title.x = element_blank(), \n",
    "                 legend.position = \"none\") + \n",
    "            xlab(\"Benchmark ID\") +\n",
    "            ylab(\"Bray-Curtis Dissimilarity\") +\n",
    "            stat_summary(fun.data=mean_sdl, color=\"red\")\n",
    "    \n",
    "    \n",
    "    fig = ggarrange(p, \n",
    "              #labels = c(\"A\"),\n",
    "              #top = as.character(title_in),\n",
    "              #common.legend = TRUE#,\n",
    "              ncol = 1, nrow = 1)\n",
    "    annotate_figure(fig, top = text_grob(\"Bray-Curtis Dissimilarity\", color = \"blue\", face = \"bold\", size = 14)\n",
    "    )\n",
    "    ggsave(paste(img_dir,\"violin_bc_summary_binf_allgroups_\",ref_in,\".jpg\",sep=\"\"))\n",
    "    return(df_plot)\n",
    "\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h2> Statistics <h2>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "tstat_paired_f1 <- function(f1_in,subgroup_in){\n",
    "    \n",
    "    df_out = data.frame()\n",
    "    \n",
    "    cal_list = c(\"TAR\",\"TDR\",\"F1\")\n",
    "    \n",
    "    for (cal in cal_list){\n",
    "    \n",
    "        comb_list = combn(unique(f1_in[,subgroup_in]),2)\n",
    "        count=1\n",
    "        \n",
    "        for (i in 1:(length(comb_list)/2)){\n",
    "            \n",
    "            tstatdf=data.frame()\n",
    "            var1 = as.character(comb_list[count])\n",
    "            var2 = as.character(comb_list[count+1])\n",
    "            count=count+2\n",
    "\n",
    "            df = f1_in %>%\n",
    "                filter(f1_in[,subgroup_in] == var1 | f1_in[,subgroup_in] == var2) %>%\n",
    "                select(subgroup_in, cal)\n",
    "            \n",
    "            tmp = t.test(get(cal) ~ get(subgroup_in), data = df)\n",
    "            tstatdf[nrow(tstatdf)+1,\"level\"] = cal\n",
    "            tstatdf[nrow(tstatdf),\"comparison\"] = paste(var1,var2,sep=\"_\")\n",
    "            tstatdf[nrow(tstatdf),\"pval\"] = tmp$p.value\n",
    "            if(tmp$p.value<0.05){\n",
    "                tstatdf[nrow(tstatdf),\"sig\"] = \"sig\"\n",
    "            } else{\n",
    "                tstatdf[nrow(tstatdf),\"sig\"] = \"ns\"\n",
    "            }\n",
    "            \n",
    "\n",
    "            df_out = bind_rows(df_out,tstatdf)\n",
    "        }\n",
    "    }\n",
    "        \n",
    "    return(df_out)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "adonis_bc <- function(df_in){\n",
    "    \n",
    "    for(i in 1:nrow(df_in)){\n",
    "        split = strsplit(df_in[i,\"name\"],\"_\")\n",
    "        df_in[i,\"tax.ref\"]=split[[1]][length(split[[1]])] #add col for ref db\n",
    "        \n",
    "        if(length(split[[1]])==4){ #dada or dn do not have add' tax ref so they are then 2, otherwise 3\n",
    "            df_in[i,\"tax.type\"]=split[[1]][3]\n",
    "        } else{\n",
    "            df_in[i,\"tax.type\"]=split[[1]][2]\n",
    "        }\n",
    "        \n",
    "        df_in[i,\"group.type\"]=split[[1]][1]  \n",
    "    }\n",
    "\n",
    "    cal_list = c(\"tax.ref\",\"tax.type\",\"group.type\")\n",
    "    df_out=data.frame()\n",
    "    \n",
    "    for (cal in cal_list){\n",
    "        st = adonis(df_in[,1] ~ df_in[,cal],\n",
    "                    data=df_in, permutations = 99, method=\"bray\")\n",
    "        df_out[nrow(df_out)+1,\"level\"] = cal\n",
    "        df_out[nrow(df_out),\"comparison\"] = \"bray-curtis\"\n",
    "        df_out[nrow(df_out),\"pval\"] = as.numeric(st$aov.tab$\"Pr(>F)\"[1])\n",
    "        \n",
    "        if(df_out[nrow(df_out),\"pval\"]<0.05){\n",
    "            df_out[nrow(df_out),\"sig\"] = \"sig\"\n",
    "        } else{\n",
    "            df_out[nrow(df_out),\"sig\"] = \"ns\"\n",
    "        }   \n",
    "    }\n",
    "    \n",
    "    dist_s = vegdist(subset(dissdf_s,select = -c(bmark,db,Group.1)),na.rm=TRUE)\n",
    "    dist_g = vegdist(subset(dissdf_g,select = -c(bmark,db,Group.1)),na.rm=TRUE)\n",
    "    \n",
    "    df_out[nrow(df_out)+1,\"homo_s\"] = anova(betadisper(dist_s,dissdf_s$bmark))$\"Pr(>F)\"[1]\n",
    "    df_out[nrow(df_out)+1,\"homo_g\"] = anova(betadisper(dist_g,dissdf_g$bmark))$\"Pr(>F)\"[1]\n",
    "    \n",
    "    return(df_out)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h2> Run Production Code <h2>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "write.csv(EOdf_stats_merged,paste(stats_dir,\"EO_stats_binf_summary.csv\",sep=\"\"))\n",
    "write.csv(mergedf_f1stats,paste(stats_dir,\"F1_values_summary_binf.csv\",sep=\"\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h2> Benchmarked Assessment<h2>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Saving 6.67 x 6.67 in image\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Saving 6.67 x 6.67 in image\n",
      "\n"
     ]
    }
   ],
   "source": [
    "Distdf_merged=data.frame()\n",
    "#Silva\n",
    "if(length(test_list)>1){\n",
    "    RAdf_oe = RAdf_addcontrols(\"CONTdf_s\",\"dissdf_s\")\n",
    "    RAdf_clean = calc_dist(RAdf_oe,CONTdf_s)\n",
    "    plot_dist_cluster(RAdf_clean,\"s\")\n",
    "    Distdf = plot_dist_violin(RAdf_clean,\"s\")\n",
    "} \n",
    "Distdf_merged = bind_rows(Distdf_merged,Distdf)\n",
    "\n",
    "#GreenGenes\n",
    "RAdf_oe = RAdf_addcontrols(\"CONTdf_g\",\"dissdf_g\")\n",
    "RAdf_clean = calc_dist(RAdf_oe,CONTdf_g)\n",
    "plot_dist_cluster(RAdf_clean,\"g\")\n",
    "Distdf = plot_dist_violin(RAdf_clean,\"g\")\n",
    "Distdf_merged = bind_rows(Distdf_merged,Distdf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A data.frame: 5 × 6</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>level</th><th scope=col>comparison</th><th scope=col>pval</th><th scope=col>sig</th><th scope=col>homo_s</th><th scope=col>homo_g</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>tax.ref   </td><td>bray-curtis</td><td>0.01</td><td>sig</td><td>NA</td><td>       NA</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>tax.type  </td><td>bray-curtis</td><td>0.55</td><td>ns </td><td>NA</td><td>       NA</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>group.type</td><td>bray-curtis</td><td>1.00</td><td>ns </td><td>NA</td><td>       NA</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>NA        </td><td>NA         </td><td>  NA</td><td>NA </td><td> 1</td><td>       NA</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>NA        </td><td>NA         </td><td>  NA</td><td>NA </td><td>NA</td><td>0.9999574</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 5 × 6\n",
       "\\begin{tabular}{r|llllll}\n",
       "  & level & comparison & pval & sig & homo\\_s & homo\\_g\\\\\n",
       "  & <chr> & <chr> & <dbl> & <chr> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t1 & tax.ref    & bray-curtis & 0.01 & sig & NA &        NA\\\\\n",
       "\t2 & tax.type   & bray-curtis & 0.55 & ns  & NA &        NA\\\\\n",
       "\t3 & group.type & bray-curtis & 1.00 & ns  & NA &        NA\\\\\n",
       "\t4 & NA         & NA          &   NA & NA  &  1 &        NA\\\\\n",
       "\t5 & NA         & NA          &   NA & NA  & NA & 0.9999574\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 5 × 6\n",
       "\n",
       "| <!--/--> | level &lt;chr&gt; | comparison &lt;chr&gt; | pval &lt;dbl&gt; | sig &lt;chr&gt; | homo_s &lt;dbl&gt; | homo_g &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|\n",
       "| 1 | tax.ref    | bray-curtis | 0.01 | sig | NA |        NA |\n",
       "| 2 | tax.type   | bray-curtis | 0.55 | ns  | NA |        NA |\n",
       "| 3 | group.type | bray-curtis | 1.00 | ns  | NA |        NA |\n",
       "| 4 | NA         | NA          |   NA | NA  |  1 |        NA |\n",
       "| 5 | NA         | NA          |   NA | NA  | NA | 0.9999574 |\n",
       "\n"
      ],
      "text/plain": [
       "  level      comparison  pval sig homo_s homo_g   \n",
       "1 tax.ref    bray-curtis 0.01 sig NA            NA\n",
       "2 tax.type   bray-curtis 0.55 ns  NA            NA\n",
       "3 group.type bray-curtis 1.00 ns  NA            NA\n",
       "4 NA         NA            NA NA   1            NA\n",
       "5 NA         NA            NA NA  NA     0.9999574"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "adonis_stats = adonis_bc(Distdf_merged)\n",
    "adonis_stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in as_grob.default(plot):\n",
      "\"Cannot convert object of class character into a grob.\"\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Warning message in as_grob.default(plot):\n",
      "\"Cannot convert object of class character into a grob.\"\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Warning message in as_grob.default(plot):\n",
      "\"Cannot convert object of class character into a grob.\"\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Warning message in as_grob.default(plot):\n",
      "\"Cannot convert object of class character into a grob.\"\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#Overall violin plots\n",
    "x_list = c(\"TAR\",\"TDR\",\"F1\")\n",
    "y_labs = c(\"Taxonmic Accuracy Rate\",\"Taxonmic Detection Rate\", \"F-Measure\")\n",
    "\n",
    "#Create violin plots collapsed to the group.type (merges all ref dbs, and tax class methods)\n",
    "plot_f1_violin(mergedf_f1stats,x_list,y_labs,\"group.type\",\"Group Types\",\"summary\")\n",
    "\n",
    "#Create violin plots collapsed to the tax.type (merges all ref dbs, and grouping methods)\n",
    "plot_f1_violin(mergedf_f1stats,x_list,y_labs,\"tax.type\",\"Tax Types\",\"summary\")\n",
    "\n",
    "#Create violin plots collapsed to the tax.ref (merges all tax class methods and grouping methods)\n",
    "plot_f1_violin(mergedf_f1stats,x_list,y_labs,\"tax.ref\",\"Tax Ref Types\",\"summary\")\n",
    "\n",
    "#Create violin plots for all binf variations (merges all variables)\n",
    "plot_f1_violin(mergedf_f1stats,x_list,y_labs,\"bmark\",\"All Grouping and Assignment Variables\",\"summary\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Note: Using an external vector in selections is ambiguous.\n",
      "\u001b[34mi\u001b[39m Use `all_of(subgroup_in)` instead of `subgroup_in` to silence this message.\n",
      "\u001b[34mi\u001b[39m See <https://tidyselect.r-lib.org/reference/faq-external-vector.html>.\n",
      "\u001b[90mThis message is displayed once per session.\u001b[39m\n",
      "\n",
      "Note: Using an external vector in selections is ambiguous.\n",
      "\u001b[34mi\u001b[39m Use `all_of(cal)` instead of `cal` to silence this message.\n",
      "\u001b[34mi\u001b[39m See <https://tidyselect.r-lib.org/reference/faq-external-vector.html>.\n",
      "\u001b[90mThis message is displayed once per session.\u001b[39m\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#T-tests\n",
    "tstatdf_merged=data.frame()\n",
    "variable_list = c(\"group.type\",\"tax.type\",\"tax.ref\")\n",
    "\n",
    "for (variable in variable_list){\n",
    "    tstatdf = tstat_paired_f1(mergedf_f1stats,variable)\n",
    "    tstatdf_merged=bind_rows(tstatdf_merged,tstatdf)  \n",
    "}\n",
    "write.csv(tstatdf_merged,paste(stats_dir,\"ttest_stats_binf.csv\",sep=\"\"))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.5.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
