---
title: "R Notebook"
output: html_notebook
editor_options: 
  chunk_output_type: console
---

##################################################
# 
##################################################
Prepare Environment
# set dir
```{r}
git_base="~/../../Volumes/sevillas2/git/gmu/dissertation/"
data_base="~/../../Volumes/data/gmu/"
```

#load libraries
```{r}
source(paste0(git_base,"sources/ggrare.R"))
source(paste0(git_base,"sources/miseqR.R"))
source(paste0(git_base,'sources/abundances.R')) #github https://github.com/microbiome/microbiome
source(paste0(git_base,"sources/qiime2R.R")) #github https://github.com/jbisanz/qiime2R

library(ape)
library(biomformat)
library(corrplot)
library(cowplot)
library(cluster)
library(data.table)
library(dplyr)
library(ggplot2)
library(matrixStats)
library(microbiome)
library(phyloseq)
library(ggpubr)
library(plyr)
library(RColorBrewer)
library(Rmisc)
library(stringr)
library(tidyr)
library(vegan)
library(yaml)
```

# Pre-Processing
## Read manifest
```{r}
#system
system_opt = "m" #swapping between mac and PC

#read in config
config_file = read_yaml(paste0(git_base,"aim1/analysis/config.yaml"))

#set dirs
git_dir=str_replace(config_file$git_dir[[system_opt]],'git_base/',git_base)
data_dir=str_replace(config_file$data_dir[[system_opt]],'data_base/',data_base)
analysis_dir=str_replace(config_file$analysis_dir[[system_opt]],'data_base/',data_base)
ref_dir=str_replace(config_file$ref_dir[[system_opt]],'data_base/',data_base)
manifest_dir=str_replace(config_file$manifest_dir[[system_opt]],'git_base/',git_base)

#set manifests
manifest_file=paste(manifest_dir, config_file$clean_manifest,sep="")
variable_file=paste(manifest_dir, config_file$variable_manifest,sep="")
binf_file=paste(manifest_dir, config_file$binf_manifest,sep="")
```

## Select BINF variables
```{r}
#decide which files to include, add Y or N flag
benchmarkdf = read.csv(binf_file,sep="\t")
benchmarkdf = subset(benchmarkdf,include=="Y")
rownames(benchmarkdf) = benchmarkdf$tax_file

benchmark_files = subset(benchmarkdf,include=="Y")$tax_file
benchmark_files
```

## Create PhyloSeq object
```{r}
# test information
#benchmark_files = benchmark_files[1:2]
binfdf=data.frame()
variable_df = read.csv(variable_file,sep="\t",row.names = 1)

#start variable counter
i = 1
phy_list=list()

for (binf_variable in benchmark_files){
  #Read OTUS
  otus<-read_qza(paste0(data_dir,'04_filtered/tmp/',benchmarkdf[binf_variable,"otu_file"])) #files must be unzipped before reading
  colnames(otus$data) <- paste0(colnames(otus$data),"-",variable_df[binf_variable,"Identifier"])
  
  #Read taxonomy reference file
  taxonomy<-read_qza(paste(data_dir,'05_class/tmp/',binf_variable,sep=""))
  
  #convert tax to matrix for phyloseq input
  GenData <- function(x){
      tax_list = unlist(strsplit(as.character(x),";"))
      
      for (i in (length(tax_list)+1):17){
        tax_list = append(tax_list,"")
      }
      
      return(tax_list)
  }
  
  taxonomy_list <- as.list(taxonomy$data$Taxon)
  taxonomy_table <- do.call("rbind", lapply(taxonomy_list, GenData))
  taxonomy_table <- taxonomy_table[,c(1:7)]
  colnames(taxonomy_table) <- c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species")
  taxonomy_table <- as.data.frame(taxonomy_table)
  taxonomy_table <-taxonomy_table %>% 
    mutate_all(funs(str_replace(., "D_.__", "")))
  taxonomy_table <-taxonomy_table %>% 
    mutate_all(funs(str_replace(., "]", "")))
  taxonomy_table <-taxonomy_table %>% 
    mutate_all(funs(str_replace(., "-", "_")))
  taxonomy_table <-taxonomy_table %>% 
    mutate_all(funs(str_replace(., " ", "")))
  
  rownames(taxonomy_table)<-taxonomy$data$Feature.ID
  tax_convert<-as.matrix(taxonomy_table)

  #read metadata
  metadata<-read.table(manifest_file,sep='\t', header=T, row.names=1, comment="")
  
  #update metadata for analysis
  metadata$Sample.Des[metadata$Sample.Des=="art_col"] = "Chemostat"
  metadata$Sample.Des[metadata$Sample.Des=="Ext_Blank"] = "Blank"
  metadata$Sample.Des[metadata$Sample.Des=="ZymoBiomics" & metadata$Sample.Type=="Ext_Control"] = "Artificial-U"
  metadata$Sample.Des[metadata$Sample.Des=="H01"] = "Human"
  metadata$BINFID = variable_df[binf_variable,"Identifier"]
  metadata$UniqueID = paste0(rownames(metadata),"-",metadata$BINFID)
  rownames(metadata) = metadata$UniqueID
  
  #create phyloseq object
  phy_obj<-phyloseq(otu_table(otus$data, taxa_are_rows = T), tax_table(tax_convert), sample_data(metadata))

  #Create phy tree
  #random_tree = rtree(ntaxa(phy_obj), rooted=TRUE, tip.label=taxa_names(phy_obj))
  #phy_t = merge_phyloseq(phy_obj, random_tree)
  phy_t = phy_obj
  
  #assign variable
  nam <- paste("phy_t", i, sep = "")
  assign(nam, phy_t)
  phy_list[i] = nam
  i = i+1
  
  #create binf df
  binfdf[nam,"tax_name"] = binf_variable
  
  #remove(phy_obj,phy_t,random_tree,tax_convert,taxonomy_list,taxonomy_table)
  remove(phy_obj,phy_t,tax_convert,taxonomy_list,taxonomy_table)
}

#df of phy object name : completed methods
binfdf
```

## merge phyloseq obj
```{r}
i = 1
remove(phy_merge)
for (phyid in phy_list){
  if (i==1){
    phy_merge = get(phyid)
  } else{
    phy_merge = merge_phyloseq(get(phyid),phy_merge)
  }
  i=i+1
}
phy_merge
remove(phy_t2,phy_t3,phy_t4,phy_t5,phy_t6,phy_t7,phy_t8,phy_t9,phy_t10,phy_t11,phy_t12)
```

## Define groups for extraction and binf, subset phy object
```{r}
#determine counts for grouping
determine_groups <- function(){
  control_list = c("EX-1","EX-2","EX-3","EX-4","EX-5","EX-6","EX-7","Seq")
  
  count_df = data.frame(control_list)
  rownames(count_df) = count_df$control_list
  
  #subset ext
  for (extid in unique(subset(sample_data(phy_merge),Sample.Type!="Seq_Control")$Ext.Kit)){
    tmp_df = subset(sample_data(phy_merge),Ext.Kit==extid & BINFID=="BINF-01")
    
    #subset by subjectid (H01)
    for (sampleid in unique(tmp_df$Subject.ID)){
      count_df[extid,sampleid] = nrow(subset(tmp_df,Subject.ID==sampleid))
    }
  }
  
  #subset seq
  for (seqid in unique(subset(sample_data(phy_merge),Sample.Type=="Seq_Control")$Subject.ID)){
    tmp_df = subset(sample_data(phy_merge),Subject.ID==seqid & BINFID=="BINF-01")
    count_df["Seq",seqid] = nrow(subset(tmp_df,Subject.ID==seqid))
  }
  return(count_df)
}
count_df = determine_groups()
count_df

#assign extraction groups
assign_ext_groups <-function(){
  group_df=data.frame(c("EX-1","EX-2","EX-3","EX-4","EX-5","EX-6","EX-7"),c("EX-G1","EX-G2","EX-G1","EX-G2","EX-G2","EX-G1","EX-G2"))
  colnames(group_df)=c("Ext.Kit","Ext.Group")
  rownames(group_df)=group_df$Ext.Kit
  
  #create list of extraction ids
  group_list = sample_data(phy_merge)$Ext.Kit
  
  #for each kit assign group
  for (gid in rownames(group_df)){
    group_list = gsub(gid,group_df[gid,"Ext.Group"],group_list)
  }
    
  #add col to sample data with group
  sample_data(phy_merge)$Ext.Group <- group_list
  return(phy_merge)
}
phy_merge = assign_ext_groups()
unique(sample_data(phy_merge)$Ext.Group)

#assign binf groups
assign_binf_groups <-function(){
  group_df=data.frame(c("BINF-01","BINF-02","BINF-03","BINF-04","BINF-05","BINF-06",
                        "BINF-07","BINF-08","BINF-09","BINF-10","BINF-11","BINF-12"),
                      c("BINF-G1","BINF-G2","BINF-G1","BINF-G2","BINF-G1","BINF-G2",
                        "BINF-G3","BINF-G4","BINF-G3","BINF-G4","BINF-G3","BINF-G4"))
  colnames(group_df)=c("BINFID","BINF.Group")
  rownames(group_df)=group_df$BINFID
  
  #create list of extraction ids
  group_list = sample_data(phy_merge)$BINFID
  
  #for each kit assign group
  for (gid in rownames(group_df)){
    group_list = gsub(gid,group_df[gid,"BINF.Group"],group_list)
  }
  
  #add col to sample data with group
  sample_data(phy_merge)$BINF.Group <- group_list
  return(phy_merge)
}
phy_merge = assign_binf_groups()
unique(sample_data(phy_merge)$BINF.Group)

#remove Z10 from phy_merge
phy_merge = subset_samples(phy_merge,Subject.ID!="Z10")

#create subset sample lists
create_sample_lists<-function(){
  #sub phylo object for only one binf variable
  phy_sub = subset_samples(phy_merge,BINFID == "BINF-01")
  
  #list groups
  ext_group1 = c("EX-2","EX-4","EX-5","EX-7")
  ext_group2 = c("EX-1","EX-3","EX-6")
  
  sample_list=""
  
  #extraction group 1
  for (extid in ext_group1){
    #robo = 7
    sample_list = append(sample_list,
                          (subset(sample_data(phy_sub),Sample.Des=="Robogut" & Ext.Kit == extid)$UniqueID)[1:3])
      
    #chemo = 5
    sample_list = append(sample_list,
                          (subset(sample_data(phy_sub),Sample.Des=="Chemostat" & Ext.Kit == extid)$UniqueID)[1:3])
    
    #Artifical-U = 6
    sample_list = append(sample_list,
                          (subset(sample_data(phy_sub),Sample.Des=="Artificial-U" & Ext.Kit == extid)$UniqueID)[1:3])
    #blank = 3
    sample_list = append(sample_list,
                          (subset(sample_data(phy_sub),Sample.Des=="Blank" & Ext.Kit == extid)$UniqueID)[1:3])
    #human = 5
    sample_list = append(sample_list,
                          (subset(sample_data(phy_sub),Sample.Des=="Human" & Ext.Kit == extid)$UniqueID)[1:3])  
  }
  
  #extraction group 2
  for (extid in ext_group2){
    #robo = 1
    sample_list = append(sample_list,
                          (subset(sample_data(phy_sub),Sample.Des=="Robogut" & Ext.Kit == extid)$UniqueID)[1])
      
    #chemo = 1
    sample_list = append(sample_list,
                          (subset(sample_data(phy_sub),Sample.Des=="Chemostat" & Ext.Kit == extid)$UniqueID)[1])
    
    #Artifical-U = 1
    sample_list = append(sample_list,
                          (subset(sample_data(phy_sub),Sample.Des=="Artificial-U" & Ext.Kit == extid)$UniqueID)[1])
  }

  #seq controls
  for (seqid in unique(subset(sample_data(phy_sub),Sample.Type=="Seq_Control")$Subject.ID)){
    sample_list = append(sample_list,
                          (subset(sample_data(phy_sub),Subject.ID==seqid)$UniqueID)[1:2])
  }

  #clean first element
  sample_list = sample_list[2:length(sample_list)]
  
  #add all binf ids
  sample_list_complete = ""
  
  for (binfid in unique(sample_data(phy_merge)$BINFID)){
    sub_list = gsub("BINF-01",binfid,sample_list)
    sample_list_complete = append(sample_list_complete,sub_list)
  }
  return(sample_list_complete)
}
sample_list = create_sample_lists()

#remove first blank in sample_list
sample_list = sample_list[2:length(sample_list)]
length(sample_list)

#subset phy_merge
phy_merge_sub = subset_samples(phy_merge,UniqueID %in% sample_list)
remove(phy_merge)

#check subset merge
variable_counting <- function(){
  
  #for each group list the kits and sample types
  for (group_id in unique(sample_data(phy_merge_sub)$Ext.Group)){
    #count n times extraction kit is used
    print(paste0("Groupid:", group_id))
    
    sub_df = subset(sample_data(phy_merge_sub),Ext.Group==group_id)
    
    #for each kit
    for (kit_id in unique(sub_df$Ext.Kit)){
      print(paste0("--Ext Kit:", kit_id))
      sub_df2 = subset(sub_df,Ext.Kit==kit_id)
      
      for (sample_id in unique(sub_df$Sample.Des)){
        print(paste0("----Sample type ", sample_id, ": ", nrow(subset(sub_df2,Sample.Des==sample_id))))
      }
    }
  }
  
  print("*********************************************")
  #for each seq control
  tmp_df = subset(sample_data(phy_merge_sub),Sample.Type == "Seq_Control")
  for (sample_id in unique(tmp_df$Subject.ID)){
    print(paste0("Seq ID ", sample_id, ": ", nrow(subset(tmp_df,Subject.ID==sample_id))))
  }
}
variable_counting()

#prune taxa
phy_merge_sub_p = prune_taxa(taxa_sums(phy_merge_sub) > 0, phy_merge_sub)
```

# In Text
## General Extraction
```{r}
calc_quant_means<-function(phy.in, type.in){
  #create dataframe
  quant_means = data.frame()
  
  #Read in quant report
  quant_df = read.csv(paste0(git_dir,"manifest/quant.txt"),sep="\t")
  rownames(quant_df)=quant_df$Sample.ID

  #for each extraction kit
  for (ext_kit in sort(unique(subset(sample_data(phy.in),Ext.Kit !="Seq")$Ext.Kit))){
  
    #create list of ids for ext.kit
    sample_list = gsub("-BINF-01","",rownames(subset(sample_data(phy.in),Ext.Kit==ext_kit)))
    
    for (sampleid in sample_list){
      quant_means[nrow(quant_means)+1,"Ext.Kit"] = ext_kit
      quant_means[nrow(quant_means),"qdna"] = quant_df[sampleid,"QDNA"]
      quant_means[nrow(quant_means),"sample"] = sampleid
    }
  }
  
  sumdf = summarySE(quant_means, measurevar="qdna", groupvars=c(type.in))
  print(sumdf)
  
  #plot
  p = ggplot(sumdf, aes(x=get(type.in), y=qdna, fill=get(type.in))) + 
      geom_bar(position=position_dodge(), stat="identity",
               colour="black", # Use black outlines,
               size=.3) +      # Thinner lines
      geom_errorbar(aes(ymin=qdna-se, ymax=qdna+se),
                    size=.3,    # Thinner lines
                    width=.2,
                    position=position_dodge(.9)) +
      xlab("Extraction method") +
      ylab("Quant values ng/uL") +
      scale_fill_hue(name=type.in) + # Legend label, use darker colors
      ggtitle("Concentration by extraction method") +
      theme_bw()
  print(p)
  file_save = paste0(analysis_dir,"concentration.png")
  ggsave(file_save,p)
}

#only need one binf variable
phy_tmp = subset_samples(phy_merge_sub_p,BINFID=="BINF-01")

#run function
calc_quant_means(phy_tmp,"Ext.Kit")
```

## General sequencing
```{r}
#### BEFORE SEQ
#total samples sequenced
nrow(subset(metadata,Subject.ID!="Z10"))

# negative controls
nrow(subset(metadata,Subject.ID %in% c("BP","BN")))
nrow(subset(metadata,Subject.ID=="BW"))

#non negative controls
nrow(subset(metadata,(Subject.ID %in% c("A00","A01","A02","A03","Z05","Z06","Z11"))))
nrow(subset(metadata,(Subject.ID %in% c("Z00"))))
nrow(subset(metadata,(Subject.ID %in% c("H01"))))
nrow(subset(metadata,(Subject.ID %in% c("M98","M16","M22"))))
nrow(subset(metadata,(Subject.ID %in% c("A00","A01","A02","A03","Z05","Z06","Z11")))) + nrow(subset(metadata,(Subject.ID %in% c("Z00")))) + nrow(subset(metadata,(Subject.ID %in% c("H01"))))+nrow(subset(metadata,(Subject.ID %in% c("M98","M16","M22"))))

#verify
nrow(subset(metadata,Subject.ID!="Z10")) - nrow(subset(metadata,Subject.ID %in% c("BP","BN"))) - nrow(subset(metadata,Subject.ID=="BW")) - nrow(subset(metadata,(Subject.ID %in% c("A00","A01","A02","A03","Z05","Z06","Z11")))) - nrow(subset(metadata,(Subject.ID %in% c("Z00")))) - nrow(subset(metadata,(Subject.ID %in% c("H01")))) -nrow(subset(metadata,(Subject.ID %in% c("M98","M16","M22"))))

#### AFTER SEQ
#total samples sequenced
nrow(subset(sample_data(phy_t1),Subject.ID!="Z10"))

# negative controls
nrow(subset(sample_data(phy_t1),Subject.ID %in% c("BP","BN")))
nrow(subset(metadata,Subject.ID=="BW"))-nrow(subset(sample_data(phy_t1),Subject.ID=="BW"))

#non negative controls
nrow(subset(metadata,(Subject.ID %in% c("A00","A01","A02","A03","Z05","Z06","Z11"))))-nrow(subset(sample_data(phy_t1),(Subject.ID %in% c("A00","A01","A02","A03","Z05","Z06","Z11"))))
nrow(subset(metadata,(Subject.ID %in% c("Z00"))))-nrow(subset(sample_data(phy_t1),(Subject.ID %in% c("Z00"))))
nrow(subset(metadata,(Subject.ID %in% c("H01"))))-nrow(subset(sample_data(phy_t1),(Subject.ID %in% c("H01"))))
nrow(subset(metadata,(Subject.ID %in% c("M98","M16","M22"))))-nrow(subset(sample_data(phy_t1),(Subject.ID %in% c("M98","M16","M22"))))
nrow(subset(sample_data(phy_t1),(Subject.ID %in% c("A00","A01","A02","A03","Z05","Z06","Z11")))) + nrow(subset(sample_data(phy_t1),(Subject.ID %in% c("Z00")))) + nrow(subset(sample_data(phy_t1),(Subject.ID %in% c("H01"))))+nrow(subset(sample_data(phy_t1),(Subject.ID %in% c("M98","M16","M22"))))

#verify
nrow(subset(sample_data(phy_t1),Subject.ID!="Z10")) - (nrow(subset(sample_data(phy_t1),(Subject.ID %in% c("A00","A01","A02","A03","Z05","Z06","Z11")))) + nrow(subset(sample_data(phy_t1),(Subject.ID %in% c("Z00")))) + nrow(subset(sample_data(phy_t1),(Subject.ID %in% c("H01"))))+nrow(subset(sample_data(phy_t1),(Subject.ID %in% c("M98","M16","M22")))))-nrow(subset(sample_data(phy_t1),Subject.ID=="BW"))

### After subsampling
nrow(subset(sample_data(phy_merge_sub_p),BINFID=="BINF-01"))

#determine num of reads by Sample.Desc
plot_read_depth<-function(phy.in){
  sampledf = data.frame()
  
  #for each extraction kit
  for(typeid in unique(sample_data(phy.in)$Sample.Des)){
  
    #create list of ids for ext.kit
    sample_list = gsub("-BINF-01","",rownames(subset(sample_data(phy.in),Sample.Des==typeid)))
    
    for (sampleid in sample_list){
      sampledf[nrow(sampledf)+1,"Sample.Des"] = typeid
      sampledf[nrow(sampledf),"reads"] = sample_sums(phy.in)[sampleid][[1]]
      sampledf[nrow(sampledf),"sample"] = sampleid
    }
  }
  sampledf$reads = as.numeric(sampledf$reads)
  
  #combine MSA and zymo
  sampledf$Sample.Des=gsub("MSA","Artificial-E",sampledf$Sample.Des)
  sampledf$Sample.Des=gsub("ZymoBiomics","Artificial-E",sampledf$Sample.Des)

  sumdf = summarySE(sampledf, measurevar="reads", groupvars=c("Sample.Des"))
  print(sumdf)
  
  #plot
  p = ggplot(sumdf, aes(x=Sample.Des, y=reads, fill=Sample.Des)) + 
        geom_bar(position=position_dodge(), stat="identity",
                 colour="black", # Use black outlines,
                 size=.3) +      # Thinner lines
        geom_errorbar(aes(ymin=reads-se, ymax=reads+se),
                      size=.3,    # Thinner lines
                      width=.2,
                      position=position_dodge(.9)) +
        xlab("Sample Description") +
        ylab("Average Read Length") +
        scale_fill_hue(name="Sample.Des") + # Legend label, use darker colors
        ggtitle(paste0("Average Numer of Reads \n By Sample Description")) +
        theme_bw()
    print(p)
    file_save = paste0(analysis_dir,"readlength_sampledescription.png")
    ggsave(file_save,p)
}
plot_read_depth(phy_t1)
```

## Overview of OTU analysis
counts
```{r}
##################################################
# counts
##################################################
nrow(subset(sample_data(phy_merge_sub_p),BINFID=="BINF-01"))
phy_merge_sub_p

nrow(subset(sample_data(phy_merge_sub_p),Subject.ID %in% c("H01","M98","M22","M16")))
nrow(subset(sample_data(phy_merge_sub_p),Subject.ID %in% c("Z00","BW")))
nrow(subset(sample_data(phy_merge_sub_p),Sample.Type=="Seq_Control"))

nrow(subset(sample_data(phy_merge_sub_p),Subject.ID %in% c("H01","M98","M22","M16"))) + nrow(subset(sample_data(phy_merge_sub_p),Subject.ID %in% c("Z00","BW"))) + nrow(subset(sample_data(phy_merge_sub_p),Sample.Type=="Seq_Control"))

##################################################
# otu summary
##################################################
calc_binf_counts<-function(phy.in){
  #average reads / non-blank sample
  sampledf = data.frame()
  
  #for each extraction kit
  for(binfid in sort(unique(sample_data(phy.in)$BINFID))){
  
    #create list of ids for ext.kit
    sample_list = rownames(subset(sample_data(phy.in),BINFID==binfid))
    
    sampledf[nrow(sampledf)+1,"BINF ID"] = binfid
    sampledf[nrow(sampledf),"mean OTU/ASV"] = mean(colSums(otu_table(phy.in)[,sample_list]))
    sampledf[nrow(sampledf),"median OTU/ASV"] = median(colSums(otu_table(phy.in)[,sample_list]))
    sampledf[nrow(sampledf),"std dev"] = sd(colSums(otu_table(phy.in)[,sample_list]))
  }

  print(sampledf)
  
  file_save = paste0(analysis_dir,'binf_otu.csv')
  write.csv(sampledf,file_save)
}
calc_binf_counts(phy_merge_sub_p)
calc_binf_counts(phy_merge_sub)

```

alpha table
```{r}
##################################################
# richness summary
##################################################
richness_binf_variables<-function(phy.in){
  #calculate alpha at binf level by sample type
  compare_binf = data.frame()
  sampledf = data.frame()
    
  # Calculate richness
  df_ob = phyloseq::estimate_richness(phy.in, measures = "Observed")
  df_ch = phyloseq::estimate_richness(phy.in, measures = "Chao1")
  df_inv = phyloseq::estimate_richness(phy.in, measures = "InvSimpson")
  
  # Store inv distances within type - all binfs are already present in df
  for (rowid in rownames(df_ob)){
    lookupval = gsub("[.]","-",rowid)
    compare_binf[nrow(compare_binf)+1,"type"] = as.character(sample_data(phy.in)[lookupval,"Sample.Des"])
    compare_binf[nrow(compare_binf),"binf"] = paste0("BINF-", strsplit(lookupval,"-")[[1]][length(strsplit(lookupval,"-")[[1]])])
    compare_binf[nrow(compare_binf),"BINF.Group"] = as.character(subset(sample_data(phy.in),
                                                                                  BINFID==compare_binf[nrow(compare_binf),
                                                                                                       "binf"])$BINF.Group)[1]
    compare_binf[nrow(compare_binf),"INV"] = df_inv[lookupval,"InvSimpson"]  
    compare_binf[nrow(compare_binf),"CH"] = df_ch[rowid,"Chao1"]  
    compare_binf[nrow(compare_binf),"OBS"] = df_ob[rowid,"Observed"]  
  }
  #combine MSA and zymo
  compare_binf$type=gsub("MSA","Artificial-E",compare_binf$type)
  compare_binf$type=gsub("ZymoBiomics","Artificial-E",compare_binf$type)
  
  #merge at grouup level
  for(binfid in sort(unique(compare_binf$BINF.Group))){
    for (typeid in sort(unique(compare_binf$type))){
      sub_df = subset(compare_binf, BINF.Group==binfid & type==typeid)
      sampledf[nrow(sampledf)+1,"Specimen Type"] = typeid
      sampledf[nrow(sampledf),"Bioinformatics Group"] = binfid
      sampledf[nrow(sampledf),"Inverse Simpson"] = paste0(round(mean(sub_df$INV),2)," (", round(sd(sub_df$INV),2),")")
      sampledf[nrow(sampledf),"Chao1"] = paste0(round(mean(sub_df$CH),2)," (", round(sd(sub_df$CH),2),")")
      sampledf[nrow(sampledf),"Observed Species"] = paste0(round(mean(sub_df$OBS),2)," (", round(sd(sub_df$OBS),2),")")
    }
  }
  
  print(sampledf)
  
  file_save = paste0(analysis_dir,'binf_richness.csv')
  write.csv(sampledf,file_save)
}
richness_binf_variables(phy_merge_sub_p)
```

wetvdry individual
```{r}
##################################################
# alpha plots
##################################################
#calculate alpha at binf level by sample type
create_alpha_plots<-function(phy.in){
  
  #calculate richness
  df_obs = phyloseq::estimate_richness(phy.in, measures = "Observed")
  df_ch = phyloseq::estimate_richness(phy.in, measures = "Chao1")
  df_inv = phyloseq::estimate_richness(phy.in, measures = "InvSimpson")
  df_sh = phyloseq::estimate_richness(phy.in, measures = "Shannon")
  
  #create df
  sampledf=data.frame()
  
  for (rowid in rownames(df_obs)){
    lookupval = gsub("[.]","-",rowid)
    sampledf[nrow(sampledf)+1,"Extraction Method"] = as.character(sample_data(phy.in)[lookupval,"Ext.Kit"])
    sampledf[nrow(sampledf),"Bioinformatic Group"] = as.character(sample_data(phy.in)[lookupval,"BINF.Group"])
    sampledf[nrow(sampledf),"Observed"] = log10(df_obs[rowid,"Observed"])
    sampledf[nrow(sampledf),"Chao1"] = log10(df_ch[rowid,"Chao1"])
    sampledf[nrow(sampledf),"Inverse Simpson"] = log10(df_inv[lookupval,])
    sampledf[nrow(sampledf),"Shannon"] = log10(df_sh[lookupval,"Shannon"])
  }
  
  plot_values<-function(x.in,y.in,ylab.in){
    p <- sampledf %>% 
      ggplot(aes(x=get(x.in), y=get(y.in))) +
      theme(axis.title=element_text(size=8),
        axis.title.x = element_blank()) +
      labs(fill = "") +
      ylab(ylab.in) + 
      geom_boxplot()
    return(p)
  }
  
  p1 = plot_values("Extraction Method","Observed","log10(Observed)")
  p2 = plot_values("Extraction Method","Chao1","log10(Chao1)")
  p3 = plot_values("Extraction Method","Inverse Simpson","log10(Inverse Simpson)")
  p4 = plot_values("Extraction Method","Shannon","log10(Shannon)")
  p5 = plot_values("Bioinformatic Group","Observed","log10(Observed)")
  p6 = plot_values("Bioinformatic Group","Chao1","log10(Chao1)")
  p7 = plot_values("Bioinformatic Group","Inverse Simpson","log10(Inverse Simpson)")
  p8 = plot_values("Bioinformatic Group","Shannon","log10(Shannon)")
  
  file_name = paste0(analysis_dir,"alpha_diversity_wetvdry.png")
  pf = ggarrange(p1, p2, p3, p4 , p5, p6, p7, p8,
            labels = c("A", "", "", "", "B", "", "", ""),
            ncol = 2, nrow = 4)
  
  pf = annotate_figure(pf, top = text_grob("Alpha-Diversity Analysis", 
                face = "bold", size = 14))
  print(pf)
  ggsave(file_name,pf)
}
create_alpha_plots(subset_samples(phy_merge_sub_p,Ext.Group=="EX-G2"))

##################################################
# beta plots
##################################################
create_ordin_plots<-function(phy.in,color.in,title.in,file.in,shape.in=""){
  #set colors
  allGroupsColors<- c("grey0")
  allGroupsColors<-append(allGroupsColors,brewer.pal(n = 8, name = "Dark2"))
  allGroupsColors<-append(allGroupsColors,brewer.pal(n = 8, name = "Set2"))
  
  #perform ordination
  ordu = phyloseq::ordinate(phy.in, method="PCoA", distance="bray")
  
  #create plot to pull axis infor
  p_data = plot_ordination(phy.in,ordu)
  
  if(shape.in==""){
    p <- plot_ordination(phy.in, ordu, color=color.in)+ 
      theme(legend.position = "bottom") + 
      ylab(gsub("Axis.2   ","PCo2 ",p_data$labels$y)) +
      xlab(gsub("Axis.1   ","PCo1 ",p_data$labels$x)) +
      geom_point(size = 3) +
      scale_color_manual(values = allGroupsColors) +
      guides(fill=guide_legend(title=color.in))
  } else{
    p <- plot_ordination(phy.in, ordu, color=color.in,shape=shape.in)+ 
      theme(legend.position = "bottom") + 
      ylab(gsub("Axis.2   ","PCo2 ",p_data$labels$y)) +
      xlab(gsub("Axis.1   ","PCo1 ",p_data$labels$x)) +
      geom_point(size = 3) +
      scale_shape_manual(values = c(15,16,17,18,3,7,8,5)) +
      scale_color_manual(values = allGroupsColors)
  }
  
  addSmallLegend <- function(myPlot, pointSize = 1, textSize = 6, spaceLegend = 0.1) {
    myPlot +
        guides(shape = guide_legend(override.aes = list(size = pointSize)),
               color = guide_legend(override.aes = list(size = pointSize))) +
        theme(legend.title = element_text(size = textSize), 
              legend.text  = element_text(size = textSize),
              legend.key.size = unit(spaceLegend, "lines"))
  }
  
  pf = addSmallLegend(p)
  pf = pf+labs(title=title.in) + geom_jitter(position=position_jitter(0.01))
  print(pf)
  file_name = paste0(analysis_dir,file.in)
  ggsave(file_name,pf)
  return(pf)

}
create_ordin_plots(subset_samples(phy_merge_sub_p,Ext.Group=="EX-G2"),"BINF.Group","Beta-Diversity Analysis","beta_diversity_wetvdry.png","Ext.Kit")

subset_samples(phy_merge_sub_p,Ext.Group=="EX-G2") #total 
subset_samples(phy_merge_sub_p,Ext.Group=="EX-G2" & BINFID=="BINF-01") #replicates
unique(sample_data(subset_samples(phy_merge_sub_p,Ext.Group=="EX-G2" & BINFID=="BINF-01"))$Subject.ID) #number of sub
```

filtering
```{r}
##################################################
# otu summary filtered
##################################################
calc_binf_filtered<-function(phy.in){
  #create tree
  random_tree = rtree(ntaxa(phy.in), rooted=TRUE, tip.label=taxa_names(phy.in))
  phy.in.t = merge_phyloseq(phy.in, random_tree)
  
  #List OTUs that do not show appear more than x times in more than x% the samples
  otu_filt_list = genefilter_sample(phy.in.t, filterfun_sample(function(x) x > 1), A=0.10*nsamples(phy.in.t))
  print(count(otu_filt_list==TRUE))
  
  #remove OTUs
  phy_c = prune_taxa(otu_filt_list, phy.in.t)
  
  #Transform to even sampling depth.
  phy_c2 = transform_sample_counts(phy_c, function(x) 1E6 * x/sum(x))
  
  #average reads / non-blank sample
  sampledf = data.frame()

  #for each extraction kit
  for (typeid in sort(unique(sample_data(phy_c2)$Sample.Des))){
    sampledf[nrow(sampledf)+1,"Sample Type"] = typeid
    
    for(binfid in sort(unique(sample_data(phy_c2)$BINF.Group))){
    #create list of ids for ext.kit
      sample_list = rownames(subset(sample_data(phy_c),BINF.Group==binfid & Sample.Des == typeid))
      sampledf[nrow(sampledf),binfid] = paste0(round(mean(colSums(otu_table(phy_c)[,sample_list]>0)),2),
                                               " (",
                                               round(sd(colSums(otu_table(phy_c)[,sample_list]>0)),2),")")
    }
  }

  rownames(sampledf)=sampledf$`Sample Type`
  sampledf=sampledf[-1]
  print(sampledf)
  
  file_save = paste0(analysis_dir,'binf_otu_filt.csv')
  write.csv(sampledf,file_save)
  
  return(phy_c)
}
phy_clean = calc_binf_filtered(phy_merge_sub_p)

#otus before filtering
phy_merge_sub

#no values
phy_merge_sub_p

#after 10%
phy_clean

```

wetvsdry comparison
```{r}
##################################################
# diversity within and between
##################################################
binf_ext_diversity<-function(phy.in){
  #subset for group 2
  phy_tmp = subset_samples(phy.in,Ext.Group=="EX-G2")
  compare_binf=data.frame()
  
  # Calculate bray curtis distances
  df_bray <-  as.data.frame(as.matrix(phyloseq::distance(phy_tmp, method = "bray")))
  
  # Calculate inv simpson
  df_inv = phyloseq::estimate_richness(phy_tmp, measures = "InvSimpson")
  
  #for each binf variable
  for (binfid in unique(sample_data(phy_tmp)$BINFID)){
    print(paste0("Working on ",binfid))
    
    for (extid in unique(sample_data(phy_tmp)$Ext.Kit)){
      #must match binf method, type, extraction method
      sample_list = subset(sample_data(phy_tmp), BINFID == binfid & Ext.Kit == extid)$UniqueID
        
      #for each sample determine within (type, binf, ext) variation
      for (sampleid in sample_list){
        df_sub = df_bray[sampleid,sample_list]
          
        for(colid in colnames(df_sub)){
            if(sampleid==colid){ next
              } else{
              compare_binf[nrow(compare_binf)+1,"binf"] = binfid
              compare_binf[nrow(compare_binf),"ext.kit"] = subset(sample_data(phy_tmp), UniqueID == sampleid)$Ext.Kit
              compare_binf[nrow(compare_binf),"value"] = df_sub[sampleid,colid]
              compare_binf[nrow(compare_binf),"cat"] = "Bray-Curtis, \n within"
            }
          }
      }
    }
  }
  
  # Store inv distances within type - all binfs are already present in df
  for (rowid in rownames(df_inv)){
    compare_binf[nrow(compare_binf)+1,"binf"] = paste0("BINF-",
                                                       strsplit(rowid,"-")[[1]][length(strsplit(rowid,"-")[[1]])])
    compare_binf[nrow(compare_binf),"ext.kit"] = subset(sample_data(phy_tmp), UniqueID == rowid)$Ext.Kit
    compare_binf[nrow(compare_binf),"value"] = df_inv[rowid,"InvSimpson"]  
    compare_binf[nrow(compare_binf),"cat"] = "Inv. Simpson, \nwithin"
  }
  
  plot_values<-function(df.in, x.in,y.in,ylab.in){
    p <- df.in %>% 
      ggplot(aes(x=get(x.in), y=get(y.in))) +
      facet_grid(ext.kit ~ ., scales = "free_y") +
      theme(axis.title=element_text(size=8),
        axis.title.x = element_blank(),
        axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
      labs(fill = "") +
      ylab(ylab.in) + 
      geom_boxplot()
    print(p)
    return(p)
  }
  
  p1 = plot_values(subset(compare_binf,cat=="Bray-Curtis, \n within"),"binf","value","Bray-Curtis")
  
  p2 = plot_values(subset(compare_binf,cat=="Inv. Simpson, \nwithin"),"binf","value","Inverse Simpson")
  
  file_name = paste0(analysis_dir,"diversity_wetvsdry.png")
  pf = ggarrange(p1, p2,
            labels = c("A", "B"),
            ncol = 2, nrow = 1)
  
  pf = annotate_figure(pf, top = text_grob("Bioinformatic and Extraction Protocols", 
                face = "bold", size = 14))
  print(pf)
  ggsave(file_name,pf)
}
binf_ext_diversity(phy_clean)
 
##################################################
# calculate sig tests
##################################################
beta_sig_type<-function(phy.in){
  #Calculate distance and save as a matrix
  print("Calculating distance")
  BC.dist=as.data.frame(as.matrix(phyloseq::distance(phy.in, method = "bray")))

  #Run PERMANOVA on distances.
  print("Sample.Type")
  print(adonis(BC.dist ~ Sample.Type, data = data.frame(sample_data(phy.in)), permutations = 1000))
  
  print("Ext.Kit")
  print(adonis(BC.dist ~ Ext.Kit, data = data.frame(sample_data(phy.in)), permutations = 1000))
  
  print("Ext.Group")
  print(adonis(BC.dist ~ Ext.Group, data = data.frame(sample_data(phy.in)), permutations = 1000))
  
  print("BINF ID")
  print(adonis(BC.dist ~ BINFID, data = data.frame(sample_data(phy.in)), permutations = 1000))
  
  print("BINF GROUP")
  print(adonis(BC.dist ~ BINF.Group, data = data.frame(sample_data(phy.in)), permutations = 1000))
}
beta_sig_type(phy_clean)

##################################################
# correlation matrix
##################################################
create_alpha_matrix<-function(){
  
  #melt and format correlation plots
  create_melted_corr<-function(df.in){
    # Get lower triangle of the correlation matrix
    get_lower_tri<-function(cormat){
      cormat[upper.tri(cormat)] <- NA
      return(cormat)
    }
    # Get upper triangle of the correlation matrix
    get_upper_tri <- function(cormat){
      cormat[lower.tri(cormat)]<- NA
      return(cormat)
    }
    
    M = cor(df.in)
    upper_tri <- get_upper_tri(M)
    melted_M <- melt(upper_tri, na.rm = TRUE)
    melted_M$value <- round(melted_M$value,2)
    return(melted_M)
  }
  
  #plot correlation
  create_corr_plots <- function(M.in,title.in){
    p <- ggplot(M.in, aes(Var2, Var1, fill = value))+
      geom_tile(color = "white")+
      scale_fill_gradient2(low = "blue", high = "red", mid = "white", 
                           midpoint = 0, limit = c(-1,1), space = "Lab", 
                           name="Pearson\nCorrelation") +
      theme_minimal()+ # minimal theme
      theme(axis.text.x = element_text(angle = 45, vjust = 1, 
                                       size = 12, hjust = 1))+
      coord_fixed() + 
      geom_text(aes(Var2, Var1, label = value), color = "black", size = 4) +
      theme(
        axis.title.x = element_blank(),
        axis.title.y = element_blank(),
        panel.grid.major = element_blank(),
        panel.border = element_blank(),
        panel.background = element_blank(),
        axis.ticks = element_blank(),
        legend.justification = c(1, 0),
        legend.position = c(0.4, 0.8),
        legend.direction = "horizontal")+
      guides(fill = guide_colorbar(barwidth = 7, barheight = 1, title.position = "top", title.hjust = 0.5)) + 
      ggtitle (title.in)
    return(p)
  }
  
  #estimate inv simp
  df_inv = phyloseq::estimate_richness(phy_clean, measures = "InvSimpson")
  
  #for binf variables
  sample_df=data.frame()
  for (binfid in sort(unique(sample_data(phy_clean)$BINFID))){
    sample_list = subset(sample_data(phy_clean),BINFID==binfid)$UniqueID
    
    i=1
    for (sampleid in sample_list){
      sample_df[i,binfid] = log10(df_inv[sampleid,"InvSimpson"])
      i=i+1
    }
  }
  M = create_melted_corr(sample_df)
  p1 = create_corr_plots(M, "Bioinformatics Protocol")
  
  #run for ext kit
  sample_df=data.frame()
  for (extid in sort(unique(subset(sample_data(phy_clean),Ext.Group=="EX-G2")$Ext.Kit))){
    sample_list = subset(sample_data(phy_clean),Ext.Kit==extid)$UniqueID
    
    i=1
    for (sampleid in sample_list){
      sample_df[i,extid] = log10(df_inv[sampleid,"InvSimpson"])
      i=i+1
    }
  }
  M = create_melted_corr(sample_df)
  p2 = create_corr_plots(M,"Extraction Protocol")
  
  file_name = paste0(analysis_dir,"alpha_spearman.png")
  pf = ggarrange(p1, p2,
            labels = c("A", "B"),
            ncol = 2, nrow = 1)
  
  pf = annotate_figure(pf, top = text_grob("Bioinformatic and Extraction Protocols", 
                face = "bold", size = 14))
  print(pf)
  ggsave(file_name,pf)
}
create_alpha_matrix()
```

between, within, inv + replicates
```{r}
##################################################
# bray-curtis within, beteween & invsimpson within
##################################################
alpha_beta_diversity_boxplots<-function(phy.in){
  #calculate alpha at binf level by sample type
  compare_binf = data.frame()
    
  # Calculate bray curtis distances
  df_bray <-  as.data.frame(as.matrix(phyloseq::distance(phy.in, method = "bray")))
  
  # Calculate inv simpson
  df_inv = phyloseq::estimate_richness(phy.in, measures = "InvSimpson")
  
  ###### BINF
  #beta diversity
  for (binfid in unique(sample_data(phy.in)$BINFID)){
    print(paste0("Working on ",binfid))
    
    # Store distances within type
    for (typeid in unique(sample_data(phy.in)$Sample.Des)){
      for (extid in unique(sample_data(phy.in)$Ext.Kit)){
        #must match binf method, type, extraction method
        sample_list = subset(sample_data(phy.in),Sample.Des==typeid & BINFID == binfid & Ext.Kit == extid)$UniqueID
        
        #for each sample determine within (type, binf, ext) variation
        for (sampleid in sample_list){
          df_sub = df_bray[sampleid,sample_list]
          
          for(colid in colnames(df_sub)){
            if(sampleid==colid){ next
              } else{
              compare_binf[nrow(compare_binf)+1,"type"] = typeid
              compare_binf[nrow(compare_binf),"sampleid"] = sampleid
              compare_binf[nrow(compare_binf),"binf"] = binfid
              compare_binf[nrow(compare_binf),"ext"] = extid
              compare_binf[nrow(compare_binf),"value"] = df_sub[sampleid,colid]
              compare_binf[nrow(compare_binf),"cat"] = "Bray-Curtis, \n within"
            }
          }
        }
      }
    }
  
    # Store distances between type
    for (typeid in unique(sample_data(phy.in)$Sample.Des)){
      for (extid in unique(sample_data(phy.in)$Ext.Kit)){
        #create list of samples within group, and outside of group
        sample_list = subset(sample_data(phy.in),Sample.Des==typeid & BINFID == binfid & Ext.Kit == extid)$UniqueID
        sample_list_compare = subset(sample_data(phy.in),Sample.Des!=typeid & BINFID == binfid & Ext.Kit == extid)$UniqueID
        
        #for each sample (row) compare to all other samples outside of group (col)
        for (sampleid in sample_list){
          df_sub = df_bray[sampleid,sample_list_compare]
          
          for(colid in colnames(df_sub)){
            if(sampleid==colid){ next
              } else{
              compare_binf[nrow(compare_binf)+1,"type"] = typeid
              compare_binf[nrow(compare_binf),"sampleid"] = sampleid
              compare_binf[nrow(compare_binf),"binf"] = binfid 
              compare_binf[nrow(compare_binf),"ext"] = extid
              compare_binf[nrow(compare_binf),"value"] = df_sub[sampleid,colid]
              compare_binf[nrow(compare_binf),"cat"] = "Bray-Curtis, \n between"
            }
          }
      }
    }
  }
      
  }    
    
  #alpha diversity
  print("Working on Inv")
  for (rowid in rownames(df_inv)){
    compare_binf[nrow(compare_binf)+1,"type"] = as.character(sample_data(phy.in)[rowid,"Sample.Des"])
    compare_binf[nrow(compare_binf),"ext"] = as.character(sample_data(phy.in)[rowid,"Ext.Kit"])
    compare_binf[nrow(compare_binf),"binf"] = paste0("BINF-", strsplit(rowid,"-")[[1]][length(strsplit(rowid,"-")[[1]])])
    compare_binf[nrow(compare_binf),"sampleid"] = rowid
    compare_binf[nrow(compare_binf),"value"] = df_inv[rowid,"InvSimpson"]  
    compare_binf[nrow(compare_binf),"cat"] = "Inv. Simpson, \nwithin"
  }

  #save df
  file_save = paste0(analysis_dir,'compare_binf.csv')
  write.csv(compare_binf,file_save)
  
  #create boxplot
  p1 <- compare_binf %>% 
    ggplot(aes(x=binf, y=value, fill=factor(type))) +
    facet_grid(cat ~., scales = "free_y") +
    ggtitle ("Bioinformatics Protocols") + 
    theme(
      axis.title.x = element_blank(),
      axis.title.y = element_blank(),
      axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
    labs(fill = "") +
    geom_boxplot() 

  p2 <- compare_binf %>% 
    ggplot(aes(x=ext, y=value, fill=factor(type))) +
    facet_grid(cat ~., scales = "free_y") +
    ggtitle ("Extraction Protocols") + 
    theme(
      axis.title.x = element_blank(),
      axis.title.y = element_blank(),
      axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
    labs(fill = "") +
    geom_boxplot() 
  
  file_save = paste0(analysis_dir,'bray_simp_boxplots.png')
  pf = ggarrange(p1, p2,
                 labels = c("A", "B"),
                 ncol = 2, nrow = 1,
                 common.legend=TRUE,
                 legend="bottom")
    
  pf = annotate_figure(pf, top = text_grob("Bioinformatic and Extraction Protocols", 
                                           face = "bold", size = 14))
  print(pf)
  ggsave(file_name,pf)
}
alpha_beta_diversity_boxplots(subset_samples(phy_merge_sub_p,Sample.Type!="Seq_Control" & Ext.Group=="EX-G2"))

##################################################
# bray-curtis variations
##################################################
impact_of_variables <-function(){
  
  #run bray-curtis distances
  df_tmp <- as.data.frame(as.matrix(phyloseq::distance(phy_merge_sub_p, method = "bray")))

  #for each row and col of distance df
  compare_df = data.frame()
  for (rowid in rownames(df_tmp)){
    #pull variables
    binfid = as.character(sample_data(phy_merge_sub_p)[rowid,"BINFID"])
    extid = as.character(sample_data(phy_merge_sub_p)[rowid,"Ext.Kit"])
    subid = as.character(sample_data(phy_merge_sub_p)[rowid,"Subject.ID"])
    uniqid = as.character(sample_data(phy_merge_sub_p)[rowid,"UniqueID"])
    typeid = as.character(sample_data(phy_merge_sub_p)[rowid,"Sample.Des"])
    
    # "bioinformatic variability"
    sample_list = subset(sample_data(phy_merge_sub_p),
                         BINFID!=binfid, Ext.Kit==extid, Subject.ID == subid)
    df_sub = df_tmp[rowid,sample_list]
    for(colid in colnames(df_sub)){
      compare_df[nrow(compare_df)+1,"type"] = typeid
      compare_df[nrow(compare_df),"sampleid"] = rowid
      compare_df[nrow(compare_df),"value"] = df_sub[rowid,colid]
      compare_df[nrow(compare_df),"cat"] = "bioinformatic variability"
    }
      
    # technical
    sample_list = sample_list = subset(sample_data(phy_merge_sub_p),
                         BINFID==binfid, Ext.Kit==extid, Subject.ID == subid,
                         UniqueID != uniqid)
    df_sub = df_tmp[rowid,sample_list]
    for(colid in colnames(df_sub)){
      compare_df[nrow(compare_df)+1,"type"] = typeid
      compare_df[nrow(compare_df),"sampleid"] = rowid
      compare_df[nrow(compare_df),"value"] = df_sub[rowid,colid]
      compare_df[nrow(compare_df),"cat"] = "technical replicates"
    }
    
    #"extraction variability"
    sample_list = subset(sample_data(phy_merge_sub_p),
                         BINFID==binfid, Ext.Kit!=extid, Subject.ID == subid)
    df_sub = df_tmp[rowid,sample_list]
    for(colid in colnames(df_sub)){
      compare_df[nrow(compare_df)+1,"type"] = typeid
      compare_df[nrow(compare_df),"sampleid"] = rowid
      compare_df[nrow(compare_df),"value"] = df_sub[rowid,colid]
      compare_df[nrow(compare_df),"cat"] = "extraction variability"
    }
    
    #sample type
    sample_list = subset(sample_data(phy_merge_sub_p),
                         BINFID==binfid, Ext.Kit==extid, Subject.ID != subid,
                         Sample.Des==typeid)
    df_sub = df_tmp[rowid,sample_list]
    for(colid in colnames(df_sub)){
      compare_df[nrow(compare_df)+1,"type"] = typeid
      compare_df[nrow(compare_df),"sampleid"] = rowid
      compare_df[nrow(compare_df),"value"] = df_sub[rowid,colid]
      compare_df[nrow(compare_df),"cat"] =  "sample type variability"
    }
  }
  
  file_save = paste0(analysis_dir,'compare_variables.csv')
  write.csv(compare_df,file_save)
  
  p <- compare_df %>% 
    ggplot(aes(x=cat, y=value, fill=factor(type))) +
    ggtitle ("Impact of Variables on Microbiome Analysis") + 
    theme(
      axis.title.x = element_blank(),
      axis.title.y = element_blank(),
      axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
    labs(fill = "") +
    geom_boxplot() 
  
  p
  file_save = paste0(analysis_dir,'compare_variables.png')
  ggsave(file_save,p)
}
impact_of_variables
```

beta plots sub by 4 variables
```{r}
##################################################
# beta plots
##################################################
create_beta_subplots<-function(phy.in){
  print("Creating plots p1,2")
  p1 = create_ordin_plots(phy.in,"BINFID","Bioinformatic Protocol",
                     "beta_diversity_binfid.png")
  p2 = create_ordin_plots(phy.in,"Sample.Des","Sample Type",
                     "beta_diversity_sampletype.png")
  print("Creating plots p3,4")
  p3 = create_ordin_plots(phy.in,"Ext.Kit","Extraction Method",
                     "beta_diversity_extkit.png")
  p4 = create_ordin_plots(phy.in,"Subject.ID","Subject",
                     "beta_diversity_subjectid.png")
  
  file_name = paste0(analysis_dir,"beta_diversity_combo.png")
  pf = ggarrange(p1, p2, p3, p4,
                 labels = c("A", "B","C","D"),
                 ncol = 2, nrow = 2)
    
  pf = annotate_figure(pf, top = text_grob("Beta Diversity", 
                                           face = "bold", size = 14))
  print(pf)
  ggsave(file_name,pf)
}
create_beta_subplots(subset_samples(phy_merge_sub_p,Ext.Group=="EX-G2"))
```


Diversity measurements
```{r}



```






# Results


paragraph 2 - counts and alpha diversity
```{r}


#determine number of OTU's per sample
get_otu_counts <- function(phy_in,df.in,binfid){
  #List OTUs that do not show appear more than x times in more than x% the samples
  otu_filt_list = genefilter_sample(phy_in, filterfun_sample(function(x) x > 2), A=0.15*nsamples(phy_in))
  
  #remove OTUs
  phy_clean = prune_taxa(otu_filt_list, phy_in)
  
  #Transform to even sampling depth.
  phy_clean = rarefy_even_depth(phy_clean, rngseed=1, sample.size=10000, replace=F) 
  phy_clean = transform_sample_counts(phy_clean, function(x) 1E6 * x/sum(x))
  
  print(binfid)
  print(phy_clean)
  
  p <-plot_richness(phy_clean, x = "Sample.Des", measures = "Observed" )
  OTU_counts <- data.frame("OTUs" = p$data)
  for(typeid in unique(OTU_counts$OTUs.Sample.Des)){
    df.in[typeid,binfid] = paste(round(mean(subset(OTU_counts,OTUs.Sample.Des==typeid)$OTUs.value),2)," (", round(sd(subset(OTU_counts,OTUs.Sample.Des==typeid)$OTUs.value),2),")", sep="")
  }
  return(df.in)
}
#Create df
otu_df=data.frame(1:7)
otu_df$type=unique(sample_data(phy_t1)$Sample.Des)
rownames(otu_df)=otu_df$type
otu_df=otu_df[-1]

#Run counts
otu_df = get_otu_counts(phy_t6,otu_df,"BINF-A") # BINFA
otu_df = get_otu_counts(phy_t9,otu_df,"BINF-B") # BINFB
otu_df = get_otu_counts(phy_t1,otu_df,"BINF-C") # BINFC
otu_df = get_otu_counts(phy_t10,otu_df,"BINF-D") # BINFD

otu_df=otu_df[-1]
print(otu_df)
file_save = paste0(analysis_dir,'otu_counts_binfgroups.csv')
write.csv(binf_alpha,file_save)
```

paragraph 3 - beta diversity
```{r}


#merge phylo objects
phy_binf_groups = merge_phyloseq(phy_t6,phy_t9,phy_t1,phy_t10)
phy_binf_groups

#clean 
phy_clean = phy_cleanup(phy_merge)
phy_clean

#run ordination
ordu = phyloseq::ordinate(phy_clean, method="PCoA", distance="bray")


    


#add group info
group_list = sample_data(phy_clean)$BINFID
group_list = gsub("BINF-08","BINF-D",group_list)
group_list = gsub("BINF-07","BINF-C",group_list)
group_list = gsub("BINF-03","BINF-A",group_list)
group_list = gsub("BINF-06","BINF-B",group_list)
sample_data(phy_clean)$`Binf.Variable` <- group_list

group_list = sample_data(phy_clean)$Sample.Des
group_list = gsub("Blank","BW",group_list)
group_list = gsub("Robogut","M98",group_list)
group_list = gsub("Chemostat","M12,M16",group_list)
group_list = gsub("Human","H01",group_list)
group_list = gsub("ZymoBiomics","Z00,Z10",group_list)
group_list = gsub("Artificial","Z05,Z06,Z11",group_list)
group_list = gsub("MSA","A00-A03",group_list)
sample_data(phy_clean)$`Sample.Description` <- group_list

#create plots
p1 = create_plots_ordin(phy_clean,"Binf.Variable")
p2 = create_plots_ordin(phy_clean,"Sample.Description")
p3 = create_plots_ordin(phy_clean,"Sample.Type")
p4 = create_plots_ordin(phy_clean,"Ext.Kit")

#plot
library(ggpubr)
file_name = paste0(analysis_dir,"beta_diversity_4plots.png")
pf = ggarrange(p1, p2, p3, p4 , 
          labels = c("A", "B", "C", "D"),
          ncol = 2, nrow = 2)

pf = annotate_figure(pf, top = text_grob("Beta-Diversity Analysis", 
              face = "bold", size = 14))
print(pf)
ggsave(file_name,pf)




#total number of samples
phy_clean

#total number of replicated
nrow(subset(sample_data(phy_clean),BINFID=="BINF-08"))
nrow(subset(sample_data(subset_samples(phy_clean,Sample.Type!="Seq_Control")),BINFID=="BINF-03"))

#total number of physical

```


```{r}
#https://www.biostars.org/p/451706/
```


